// shaders/dist/react/utils/generatePresetCode.js
var DEFAULT_TRANSFORM = {
	offsetX: 0,
	offsetY: 0,
	rotation: 0,
	scale: 1,
	anchorX: .5,
	anchorY: .5,
	edges: "transparent"
};
function toJSLiteral(value) {
	if (typeof value === "string") return JSON.stringify(value);
	if (typeof value === "boolean" || typeof value === "number" || value === null || value === void 0) return String(value);
	if (Array.isArray(value)) return `[${value.map((v) => toJSLiteral(v)).join(", ")}]`;
	if (typeof value === "object") return `{ ${Object.entries(value).map(([k, v]) => `${k}: ${toJSLiteral(v)}`).join(", ")} }`;
	return String(value);
}
function toObjectLiteral(obj) {
	return `{ ${Object.entries(obj).map(([key, value]) => {
		return `${key}: ${toJSLiteral(value)}`;
	}).join(", ")} }`;
}
var shaderMetadata = {
	"AngularBlur": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 20,
		"center": {
			"x": .5,
			"y": .5
		}
	},
	"Ascii": {
		"opacity": 1,
		"blendMode": "normal",
		"characters": "@%#*+=-:.",
		"cellSize": 30,
		"fontFamily": "JetBrains Mono",
		"spacing": 1,
		"gamma": 1,
		"alphaThreshold": 0,
		"preserveAlpha": true
	},
	"Aurora": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#a533f8",
		"colorB": "#22ee88",
		"colorC": "#1694e8",
		"colorSpace": "linear",
		"balance": 50,
		"intensity": 80,
		"curtainCount": 4,
		"speed": 5,
		"waviness": 50,
		"rayDensity": 20,
		"height": 120,
		"center": {
			"x": .5,
			"y": 0
		},
		"seed": 0
	},
	"BarShift": {
		"opacity": 1,
		"blendMode": "normal",
		"count": 6,
		"angle": 0,
		"intensity": .15,
		"seed": 0,
		"speed": 0,
		"edges": "mirror"
	},
	"Beam": {
		"opacity": 1,
		"blendMode": "normal",
		"startPosition": {
			"x": .2,
			"y": .5
		},
		"endPosition": {
			"x": .8,
			"y": .5
		},
		"startThickness": .2,
		"endThickness": .2,
		"startSoftness": .5,
		"endSoftness": .5,
		"insideColor": "#FF0000",
		"outsideColor": "#0000FF",
		"colorSpace": "linear"
	},
	"Blob": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#ff6b35",
		"colorB": "#e91e63",
		"size": .5,
		"deformation": .5,
		"softness": .5,
		"highlightIntensity": .5,
		"highlightX": .3,
		"highlightY": -.3,
		"highlightZ": .4,
		"highlightColor": "#ffe11a",
		"speed": .5,
		"seed": 1,
		"center": {
			"x": .5,
			"y": .5
		},
		"colorSpace": "linear"
	},
	"Blur": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 50
	},
	"BrickPattern": {
		"opacity": 1,
		"blendMode": "normal",
		"colorBrick": "#000000",
		"colorMortar": "#ffffff",
		"cellsX": 8,
		"cellsY": 10,
		"mortar": .05,
		"angle": 0,
		"speed": 0,
		"offset": 0,
		"speedVariance": 0,
		"seed": 0,
		"colorSpace": "linear"
	},
	"BrightnessContrast": {
		"opacity": 1,
		"blendMode": "normal",
		"brightness": 0,
		"contrast": 0
	},
	"Bulge": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"strength": 1,
		"radius": 1,
		"falloff": .5,
		"edges": "stretch"
	},
	"ChannelBlur": {
		"opacity": 1,
		"blendMode": "normal",
		"redIntensity": 0,
		"greenIntensity": 20,
		"blueIntensity": 40
	},
	"Checkerboard": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#cccccc",
		"colorB": "#999999",
		"cells": 8,
		"softness": 0,
		"colorSpace": "linear"
	},
	"Chevron": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#000000",
		"colorB": "#ffffff",
		"count": 5,
		"angle": 0,
		"balance": .5,
		"softness": 0,
		"speed": 0,
		"offset": 0,
		"colorSpace": "linear"
	},
	"ChromaFlow": {
		"opacity": 1,
		"blendMode": "normal",
		"baseColor": "#0066ff",
		"upColor": "#00ff00",
		"downColor": "#ff0000",
		"leftColor": "#0000ff",
		"rightColor": "#ffff00",
		"intensity": 1,
		"radius": 3,
		"momentum": 30
	},
	"ChromaticAberration": {
		"opacity": 1,
		"blendMode": "normal",
		"strength": .2,
		"angle": 0,
		"redOffset": -1,
		"greenOffset": 0,
		"blueOffset": 1
	},
	"Circle": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"radius": 1,
		"softness": 0,
		"center": {
			"x": .5,
			"y": .5
		},
		"strokeThickness": 0,
		"strokeColor": "#000000",
		"strokePosition": "center",
		"colorSpace": "linear"
	},
	"ColorWheel": {
		"opacity": 1,
		"blendMode": "normal",
		"mode": "rainbow",
		"colorA": "#ff0000",
		"colorB": "#00ff88",
		"colorC": "#0066ff",
		"scale": 1,
		"angle": 0,
		"speed": .05,
		"colorSpace": "oklch"
	},
	"ConcentricSpin": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 20,
		"rings": 8,
		"smoothness": .03,
		"seed": 0,
		"speed": .1,
		"speedRandomness": .5,
		"edges": "mirror",
		"center": {
			"x": .5,
			"y": .5
		}
	},
	"ConicGradient": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#FF0080",
		"colorB": "#00BFFF",
		"center": {
			"x": .5,
			"y": .5
		},
		"rotation": 0,
		"repeat": 1,
		"colorSpace": "oklch"
	},
	"ContourLines": {
		"opacity": 1,
		"blendMode": "normal",
		"levels": 5,
		"lineWidth": 2,
		"softness": 0,
		"gamma": .5,
		"invert": false,
		"source": "luminance",
		"colorMode": "source",
		"lineColor": "#000000",
		"backgroundColor": "transparent"
	},
	"Crescent": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"center": {
			"x": .5,
			"y": .5
		},
		"radius": .3,
		"innerRatio": .8,
		"offset": .2,
		"rotation": 0,
		"softness": 0,
		"strokeThickness": 0,
		"strokeColor": "#000000",
		"strokePosition": "center",
		"colorSpace": "linear"
	},
	"Cross": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"center": {
			"x": .5,
			"y": .5
		},
		"radius": .35,
		"thickness": .08,
		"rounding": 0,
		"rotation": 0,
		"softness": 0,
		"strokeThickness": 0,
		"strokeColor": "#000000",
		"strokePosition": "center",
		"colorSpace": "linear"
	},
	"CRTScreen": {
		"opacity": 1,
		"blendMode": "normal",
		"pixelSize": 128,
		"colorShift": 1,
		"scanlineIntensity": .3,
		"scanlineFrequency": 200,
		"brightness": 1,
		"contrast": 1,
		"vignetteIntensity": 1,
		"vignetteRadius": .5
	},
	"Crystal": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"scale": 1,
		"cutout": false,
		"refraction": .5,
		"dispersion": .5,
		"facets": 5,
		"fresnel": .05,
		"fresnelSoftness": 1,
		"fresnelColor": "#ffffff",
		"edgeSoftness": 0,
		"innerZoom": 1.5,
		"lightAngle": 270,
		"highlights": .5,
		"shadows": .3,
		"brightness": 1.2,
		"tintColor": "#e8e0ff",
		"tintIntensity": 0,
		"tintPreserveLuminosity": true,
		"shape": "{\"type\":\"polygonSDF\",\"radius\":0.35,\"sides\":10}"
	},
	"CursorRipples": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 10,
		"decay": 10,
		"radius": .5,
		"chromaticSplit": 1,
		"edges": "stretch"
	},
	"CursorTrail": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#00aaff",
		"colorB": "#ff00aa",
		"radius": .5,
		"length": .5,
		"shrink": 1,
		"colorSpace": "linear"
	},
	"DiamondGradient": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#4ffb4a",
		"colorB": "#4f1238",
		"center": {
			"x": .5,
			"y": .5
		},
		"size": .7,
		"rotation": 0,
		"repeat": 1,
		"roundness": 0,
		"colorSpace": "oklch"
	},
	"DiffuseBlur": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 30,
		"edges": "stretch"
	},
	"Dither": {
		"opacity": 1,
		"blendMode": "normal",
		"pattern": "bayer4",
		"pixelSize": 4,
		"threshold": .5,
		"spread": 1,
		"colorMode": "custom",
		"colorA": "transparent",
		"colorB": "#ffffff"
	},
	"DOMTexture": {
		"opacity": 1,
		"blendMode": "normal"
	},
	"DotGrid": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"density": 30,
		"dotSize": .3,
		"twinkle": 0
	},
	"DropShadow": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#000000",
		"distance": .1,
		"angle": 135,
		"blur": 5,
		"intensity": .5,
		"cutout": false
	},
	"Duotone": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#ff0000",
		"colorB": "#023af4",
		"blend": .5,
		"colorSpace": "linear"
	},
	"Ellipse": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"center": {
			"x": .5,
			"y": .5
		},
		"radiusX": .35,
		"radiusY": .2,
		"rotation": 0,
		"softness": 0,
		"strokeThickness": 0,
		"strokeColor": "#000000",
		"strokePosition": "center",
		"colorSpace": "linear"
	},
	"Emboss": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"scale": 1,
		"depth": -.5,
		"lightAngle": 260,
		"lightIntensity": .6,
		"shadowIntensity": .3,
		"shape": "{\"type\":\"circleSDF\",\"radius\":0.35}"
	},
	"FallingLines": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#ffffff",
		"colorB": "#ffffff00",
		"colorSpace": "linear",
		"angle": 90,
		"speed": .5,
		"speedVariance": .3,
		"density": 15,
		"trailLength": .35,
		"balance": .5,
		"strokeWidth": .15,
		"rounding": 1
	},
	"FilmGrain": {
		"opacity": 1,
		"blendMode": "normal",
		"strength": .5,
		"bias": 2,
		"animated": false
	},
	"FloatingParticles": {
		"opacity": 1,
		"blendMode": "normal",
		"randomness": .25,
		"speed": .25,
		"angle": 90,
		"particleSize": 2,
		"particleSoftness": 0,
		"twinkle": .5,
		"count": 5,
		"particleColor": "#ffffff",
		"speedVariance": .3,
		"angleVariance": 30,
		"particleDensity": 3
	},
	"Flower": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"center": {
			"x": .5,
			"y": .5
		},
		"radius": .4,
		"sides": 5,
		"innerRatio": .4,
		"rotation": 0,
		"softness": 0,
		"strokeThickness": 0,
		"strokeColor": "#000000",
		"strokePosition": "center",
		"colorSpace": "linear"
	},
	"FlowField": {
		"opacity": 1,
		"blendMode": "normal",
		"strength": .15,
		"detail": 2,
		"speed": 0,
		"evolutionSpeed": 0,
		"edges": "mirror"
	},
	"FlowingGradient": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#0a0015",
		"colorB": "#6b17e6",
		"colorC": "#ff4d6a",
		"colorD": "#ff6b35",
		"colorSpace": "oklch",
		"speed": 1,
		"distortion": .5,
		"seed": 0
	},
	"FlutedGlass": {
		"opacity": 1,
		"blendMode": "normal",
		"shape": "bars",
		"angle": 0,
		"frequency": 10,
		"softness": .5,
		"waveAmplitude": .06,
		"waveFrequency": 1.5,
		"speed": 0,
		"refraction": 1.5,
		"aberration": .2,
		"lightAngle": 30,
		"highlight": .2,
		"highlightSoftness": .3,
		"highlightColor": "#ffffff",
		"edges": "mirror"
	},
	"Fog": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#e0e0e0",
		"colorB": "#888888",
		"seed": 0,
		"speed": 1,
		"turbulence": 1,
		"detail": 15,
		"blending": .3,
		"mouseInfluence": .1,
		"mouseRadius": .1,
		"colorSpace": "linear"
	},
	"Form3D": {
		"opacity": 1,
		"blendMode": "normal",
		"shape3d": "{\"type\":\"ribbon\",\"angle\":0,\"twist\":50,\"width\":40,\"thickness\":20,\"seed\":0}",
		"center": {
			"x": .5,
			"y": .5
		},
		"zoom": 50,
		"glossiness": 50,
		"lighting": 50,
		"uvMode": "stretch",
		"speed": 1
	},
	"FractalNoise": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#000000",
		"colorB": "#ffffff",
		"octaves": 4,
		"detail": 2,
		"contrast": .5,
		"speed": .15,
		"angle": 0,
		"seed": 0,
		"colorSpace": "linear"
	},
	"Glass": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"scale": 1,
		"cutout": false,
		"refraction": 1,
		"edgeSoftness": .1,
		"blur": 0,
		"thickness": .2,
		"aberration": .5,
		"innerZoom": 1,
		"lightAngle": 300,
		"highlight": .05,
		"highlightColor": "#ffffff",
		"highlightSoftness": .5,
		"fresnel": .1,
		"fresnelSoftness": .1,
		"fresnelColor": "#ffffff",
		"tintColor": "#ffffff",
		"tintIntensity": 0,
		"tintPreserveLuminosity": true,
		"shape": "{\"type\":\"circleSDF\",\"radius\":0.35}"
	},
	"GlassTiles": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 2,
		"tileCount": 20,
		"rotation": 0,
		"roundness": 0
	},
	"Glitch": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": .5,
		"speed": 1,
		"rgbShift": 5,
		"blockDensity": 10,
		"colorBarIntensity": .2,
		"mirrorAmount": .3,
		"scanlineIntensity": .2
	},
	"Glow": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 1,
		"threshold": .5,
		"size": 25
	},
	"Godrays": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": 0,
			"y": 0
		},
		"density": .3,
		"intensity": .8,
		"spotty": 1,
		"speed": .5,
		"rayColor": "#4283fb",
		"backgroundColor": "transparent"
	},
	"Grayscale": {
		"opacity": 1,
		"blendMode": "normal"
	},
	"Grid": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"cells": 10,
		"thickness": 1,
		"rotation": 0
	},
	"GridDistortion": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 1,
		"decay": 3,
		"radius": 1,
		"gridSize": 20,
		"edges": "stretch"
	},
	"Group": {
		"opacity": 1,
		"blendMode": "normal"
	},
	"Halftone": {
		"opacity": 1,
		"blendMode": "normal",
		"style": "classic",
		"frequency": 100,
		"angle": 45,
		"cyanAngle": 15,
		"magentaAngle": 75,
		"yellowAngle": 0,
		"blackAngle": 45,
		"misprint": 0,
		"misprintAngle": 0,
		"paperColor": "#ffffff",
		"cyanColor": "#00ffff",
		"magentaColor": "#ff00ff",
		"yellowColor": "#ffff00",
		"blackColor": "#000000"
	},
	"HexGrid": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#000000",
		"colorB": "#ffffff",
		"cells": 8,
		"thickness": 1,
		"colorSpace": "linear"
	},
	"HueShift": {
		"opacity": 1,
		"blendMode": "normal",
		"shift": 0
	},
	"ImageTexture": {
		"opacity": 1,
		"blendMode": "normal",
		"url": "https://shaders.com/sample.jpg",
		"objectFit": "cover"
	},
	"Invert": {
		"opacity": 1,
		"blendMode": "normal"
	},
	"Kaleidoscope": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"segments": 6,
		"angle": 0,
		"edges": "mirror"
	},
	"LensFlare": {
		"opacity": 1,
		"blendMode": "normal",
		"lightPosition": {
			"x": .3,
			"y": .3
		},
		"intensity": .5,
		"ghostIntensity": .4,
		"ghostSpread": .7,
		"ghostChroma": .3,
		"haloIntensity": .4,
		"haloRadius": .6,
		"haloChroma": .6,
		"haloSoftness": .8,
		"starburstIntensity": .3,
		"starburstPoints": 6,
		"streakIntensity": .15,
		"streakLength": .5,
		"glareIntensity": .2,
		"glareSize": .5,
		"edgeFade": .2,
		"speed": .5
	},
	"LinearBlur": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 30,
		"angle": 0
	},
	"LinearGradient": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#1aff00",
		"colorB": "#0000ff",
		"start": {
			"x": 0,
			"y": .5
		},
		"end": {
			"x": 1,
			"y": .5
		},
		"angle": 0,
		"edges": "stretch",
		"colorSpace": "linear"
	},
	"Liquify": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 10,
		"stiffness": 3,
		"damping": 3,
		"radius": 1,
		"edges": "stretch"
	},
	"Marble": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#ffffff",
		"colorB": "#3a2d54",
		"colorC": "#0f0f0f",
		"scale": 2,
		"turbulence": 10,
		"speed": .05,
		"seed": 0,
		"colorSpace": "linear"
	},
	"Mirror": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"angle": 0,
		"edges": "mirror"
	},
	"MultiPointGradient": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#4776E6",
		"positionA": {
			"x": .2,
			"y": .2
		},
		"colorB": "#C44DFF",
		"positionB": {
			"x": .8,
			"y": .2
		},
		"colorC": "#1ABC9C",
		"positionC": {
			"x": .2,
			"y": .8
		},
		"colorD": "#F8BBD9",
		"positionD": {
			"x": .8,
			"y": .8
		},
		"colorE": "#FF8C42",
		"positionE": {
			"x": .5,
			"y": .5
		},
		"smoothness": 2
	},
	"Neon": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"scale": 1,
		"color": "#00ddff",
		"secondaryColor": "#ff00aa",
		"secondaryBlend": .5,
		"glowColor": "#00ddff",
		"tubeThickness": .2,
		"intensity": 1.5,
		"hotCoreIntensity": .6,
		"glowIntensity": .6,
		"glowRadius": .25,
		"lightAngle": 300,
		"specularIntensity": .5,
		"specularSize": .5,
		"cornerSmoothing": .15,
		"flickerSpeed": 0,
		"flickerAmount": .2,
		"flowSpeed": 0,
		"flowAmount": .3,
		"shape": "{\"type\":\"circleSDF\",\"radius\":0.35}"
	},
	"Paper": {
		"opacity": 1,
		"blendMode": "normal",
		"roughness": .3,
		"grainScale": 1,
		"displacement": .15,
		"seed": 0
	},
	"Perspective": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"pan": 0,
		"tilt": 0,
		"fov": 60,
		"zoom": 1,
		"offset": {
			"x": .5,
			"y": .5
		},
		"edges": "transparent"
	},
	"Pixelate": {
		"opacity": 1,
		"blendMode": "normal",
		"scale": 50,
		"gap": 0,
		"roundness": 0
	},
	"Plasma": {
		"opacity": 1,
		"blendMode": "normal",
		"density": 2,
		"speed": 2,
		"intensity": 1.5,
		"warp": .4,
		"contrast": 1,
		"balance": 50,
		"colorA": "#7018be",
		"colorB": "#000000",
		"colorSpace": "linear"
	},
	"PolarCoordinates": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"wrap": 1,
		"radius": 1,
		"intensity": 1,
		"edges": "transparent"
	},
	"Polygon": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"center": {
			"x": .5,
			"y": .5
		},
		"radius": .4,
		"sides": 6,
		"rounding": 0,
		"rotation": 0,
		"softness": 0,
		"strokeThickness": 0,
		"strokeColor": "#000000",
		"strokePosition": "center",
		"colorSpace": "linear"
	},
	"Posterize": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 5
	},
	"ProgressiveBlur": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 50,
		"angle": 0,
		"center": {
			"x": 0,
			"y": .5
		},
		"falloff": 1
	},
	"RadialGradient": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#ff0000",
		"colorB": "#0000ff",
		"center": {
			"x": .5,
			"y": .5
		},
		"radius": 1,
		"repeat": 1,
		"aspect": 1,
		"skewAngle": 0,
		"colorSpace": "linear"
	},
	"RectangularCoordinates": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"scale": 1,
		"intensity": 1,
		"edges": "transparent"
	},
	"ReflectivePlane": {
		"opacity": 1,
		"blendMode": "normal",
		"height": .7,
		"distance": .5,
		"falloff": .5,
		"blur": 3,
		"blurDistance": .3,
		"edges": "stretch"
	},
	"Ring": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"center": {
			"x": .5,
			"y": .5
		},
		"radius": .3,
		"thickness": .07,
		"softness": 0,
		"strokeThickness": 0,
		"strokeColor": "#000000",
		"strokePosition": "center",
		"colorSpace": "linear"
	},
	"Ripples": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"colorA": "#ffffff",
		"colorB": "#000000",
		"speed": 1,
		"frequency": 20,
		"softness": 0,
		"thickness": .5,
		"phase": 0
	},
	"RoundedRect": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"center": {
			"x": .5,
			"y": .5
		},
		"width": .35,
		"height": .25,
		"rounding": .05,
		"rotation": 0,
		"softness": 0,
		"strokeThickness": 0,
		"strokeColor": "#000000",
		"strokePosition": "center",
		"colorSpace": "linear"
	},
	"Saturation": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 1
	},
	"Sharpness": {
		"opacity": 1,
		"blendMode": "normal",
		"sharpness": 0
	},
	"Shatter": {
		"opacity": 1,
		"blendMode": "normal",
		"crackWidth": 1,
		"intensity": 4,
		"radius": .4,
		"decay": 1,
		"seed": 2,
		"chromaticSplit": 1,
		"refractionStrength": 5,
		"shardLighting": .1,
		"edges": "mirror"
	},
	"SimplexNoise": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#ffffff",
		"colorB": "#000000",
		"colorSpace": "linear",
		"scale": 2,
		"balance": 0,
		"contrast": 0,
		"seed": 0,
		"speed": 1
	},
	"SineWave": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"amplitude": .15,
		"frequency": 1,
		"speed": 1,
		"angle": 0,
		"position": {
			"x": .5,
			"y": .5
		},
		"thickness": .2,
		"softness": .4
	},
	"Smoke": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#fc83f9",
		"colorB": "#c21c79",
		"emitFrom": {
			"x": .5,
			"y": 1
		},
		"direction": 0,
		"speed": 20,
		"spread": 60,
		"emitRadius": .08,
		"intensity": 1,
		"dissipation": .2,
		"detail": 25,
		"gravity": .5,
		"colorDecay": .4,
		"mouseInfluence": .1,
		"mouseRadius": .1,
		"colorSpace": "linear"
	},
	"SmokeFill": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#8cf3ff",
		"colorB": "#04a0d6",
		"center": {
			"x": .5,
			"y": .5
		},
		"scale": 1,
		"emitFrom": {
			"x": .5,
			"y": .5
		},
		"direction": 0,
		"speed": 10,
		"spread": 60,
		"emitRadius": .03,
		"intensity": 1,
		"dissipation": .3,
		"detail": 25,
		"gravity": .5,
		"colorDecay": .4,
		"mouseInfluence": .1,
		"mouseRadius": .1,
		"colorSpace": "linear",
		"shape": "{\"type\":\"circleSDF\",\"radius\":0.35}"
	},
	"Solarize": {
		"opacity": 1,
		"blendMode": "normal",
		"threshold": .5,
		"strength": 1
	},
	"SolidColor": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#5b18ca"
	},
	"Spherize": {
		"opacity": 1,
		"blendMode": "normal",
		"radius": 1,
		"depth": 1,
		"center": {
			"x": .5,
			"y": .5
		},
		"lightPosition": {
			"x": .3,
			"y": .3
		},
		"lightIntensity": .5,
		"lightSoftness": .5,
		"lightColor": "#ffffff"
	},
	"Spiral": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#000000",
		"colorB": "#ffffff",
		"strokeWidth": .5,
		"strokeFalloff": 0,
		"softness": 0,
		"speed": 1,
		"center": {
			"x": .5,
			"y": .5
		},
		"scale": 1,
		"colorSpace": "linear"
	},
	"Star": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"center": {
			"x": .5,
			"y": .5
		},
		"radius": .4,
		"sides": 5,
		"innerRatio": .4,
		"rotation": 0,
		"softness": 0,
		"strokeThickness": 0,
		"strokeColor": "#000000",
		"strokePosition": "center",
		"colorSpace": "linear"
	},
	"Strands": {
		"opacity": 1,
		"blendMode": "normal",
		"speed": .5,
		"amplitude": 1,
		"frequency": 1,
		"lineCount": 12,
		"lineWidth": .1,
		"waveColor": "#f1c907",
		"pinEdges": true,
		"start": {
			"x": 0,
			"y": .5
		},
		"end": {
			"x": 1,
			"y": .5
		}
	},
	"Stretch": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"strength": 1,
		"angle": 0,
		"falloff": 0,
		"edges": "stretch"
	},
	"Stripes": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#000000",
		"colorB": "#ffffff",
		"angle": 45,
		"density": 5,
		"balance": .5,
		"softness": 0,
		"speed": .2,
		"offset": 0,
		"colorSpace": "linear"
	},
	"StudioBackground": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#d8dbec",
		"keyColor": "#d5e4ea",
		"keyIntensity": 40,
		"keySoftness": 50,
		"fillColor": "#d5e4ea",
		"fillIntensity": 10,
		"fillSoftness": 70,
		"fillAngle": 70,
		"backColor": "#c8d4e8",
		"backIntensity": 20,
		"backSoftness": 80,
		"brightness": 20,
		"vignette": 0,
		"center": {
			"x": .5,
			"y": .8
		},
		"lightTarget": 100,
		"wallCurvature": 10,
		"ambientIntensity": 50,
		"ambientSpeed": 2,
		"seed": 0
	},
	"SunBurst": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffdd88",
		"background": "#000000",
		"center": {
			"x": .5,
			"y": .5
		},
		"rayCount": 12,
		"softness": .3,
		"radius": .8,
		"feather": .5,
		"speed": .2
	},
	"Swirl": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#1275d8",
		"colorB": "#e19136",
		"speed": 1,
		"detail": 1,
		"blend": 50,
		"colorSpace": "linear"
	},
	"TiltShift": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 50,
		"width": .3,
		"falloff": .3,
		"angle": 0,
		"center": {
			"x": .5,
			"y": .5
		}
	},
	"Tint": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ff8800",
		"amount": .5,
		"preserveLuminosity": true
	},
	"Trapezoid": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"center": {
			"x": .5,
			"y": .5
		},
		"bottomWidth": .35,
		"topWidth": .2,
		"height": .25,
		"rotation": 0,
		"softness": 0,
		"strokeThickness": 0,
		"strokeColor": "#000000",
		"strokePosition": "center",
		"colorSpace": "linear"
	},
	"Tritone": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#ce1bea",
		"colorB": "#2fff00",
		"colorC": "#ffff00",
		"blendMid": .5,
		"colorSpace": "linear"
	},
	"Truchet": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#000000",
		"colorB": "#ffffff",
		"cells": 10,
		"thickness": 2,
		"seed": 0,
		"colorSpace": "linear"
	},
	"Twirl": {
		"opacity": 1,
		"blendMode": "normal",
		"center": {
			"x": .5,
			"y": .5
		},
		"intensity": 1,
		"edges": "stretch"
	},
	"Vesica": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#ffffff",
		"center": {
			"x": .5,
			"y": .5
		},
		"radius": .35,
		"spread": .5,
		"rotation": 0,
		"softness": 0,
		"strokeThickness": 0,
		"strokeColor": "#000000",
		"strokePosition": "center",
		"colorSpace": "linear"
	},
	"VHS": {
		"opacity": 1,
		"blendMode": "normal",
		"wobble": 1,
		"scanlineNoise": .6,
		"smear": .2,
		"speed": 1
	},
	"Vibrance": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 0
	},
	"VideoTexture": {
		"opacity": 1,
		"blendMode": "normal",
		"url": "https://shaders.com/sample.mp4",
		"objectFit": "cover",
		"loop": true
	},
	"Vignette": {
		"opacity": 1,
		"blendMode": "normal",
		"color": "#000000",
		"center": {
			"x": .5,
			"y": .5
		},
		"radius": .5,
		"falloff": .5,
		"intensity": 1
	},
	"Voronoi": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#3186cf",
		"colorB": "#fc02dd",
		"colorBorder": "#000000",
		"scale": 6,
		"speed": .5,
		"seed": 0,
		"edgeIntensity": .5,
		"edgeSoftness": .05,
		"colorSpace": "oklch"
	},
	"WaveDistortion": {
		"opacity": 1,
		"blendMode": "normal",
		"strength": .3,
		"frequency": 1,
		"speed": 1,
		"angle": 0,
		"waveType": "sine",
		"edges": "stretch"
	},
	"Weave": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#c4c4c4",
		"colorB": "#4d4d4d",
		"cells": 10,
		"gap": .25,
		"rotation": 0
	},
	"WebcamTexture": {
		"opacity": 1,
		"blendMode": "normal",
		"objectFit": "cover",
		"mirror": true
	},
	"WorleyNoise": {
		"opacity": 1,
		"blendMode": "normal",
		"colorA": "#ffffff",
		"colorB": "#000000",
		"colorSpace": "linear",
		"scale": 6,
		"mode": "f1",
		"distance": "euclidean",
		"octaves": 1,
		"lacunarity": 2,
		"persistence": .5,
		"jitter": 1,
		"contrast": 1,
		"balance": 0,
		"seed": 0,
		"speed": .5
	},
	"ZoomBlur": {
		"opacity": 1,
		"blendMode": "normal",
		"intensity": 30,
		"center": {
			"x": .5,
			"y": .5
		}
	}
};
function isPropMapValue(value) {
	return typeof value === "object" && value !== null && "type" in value && value.type === "map";
}
function generatePropString(props, componentType, indent = "  ") {
	return Object.entries(props).sort(([a], [b]) => a.localeCompare(b)).filter(([key, value]) => {
		if (key === "maskType" && value === "alpha") return false;
		if ((key === "shape" || key === "shapeType") && props.shapeSdfUrl) return false;
		if (key === "transform" && typeof value === "object") {
			if (Object.keys(DEFAULT_TRANSFORM).every((k) => value[k] === DEFAULT_TRANSFORM[k])) return false;
		}
		const componentDefaults = shaderMetadata[componentType] || {};
		if (componentDefaults.hasOwnProperty(key)) {
			const defaultValue = componentDefaults[key];
			if (value != null && defaultValue != null && typeof value === "object" && typeof defaultValue === "object") return JSON.stringify(value) !== JSON.stringify(defaultValue);
			return value !== defaultValue;
		}
		if (key === "opacity" && value === 1) return false;
		if (key === "blendMode" && value === "normal") return false;
		return true;
	}).map(([key, value]) => {
		if (key === "transform" && typeof value === "object") {
			const nonDefaultKeys = {};
			for (const k in value) if (value[k] !== DEFAULT_TRANSFORM[k]) nonDefaultKeys[k] = value[k];
			const keys = Object.keys(nonDefaultKeys).sort();
			if (keys.length > 1) return `${key}={{\n${indent}    ${keys.map((k) => `${k}: ${typeof nonDefaultKeys[k] === "string" ? `"${nonDefaultKeys[k]}"` : nonDefaultKeys[k]}`).join(`,\n${indent}    `)}\n${indent}  }}`;
			else if (keys.length === 1) {
				const k = keys[0];
				const v = nonDefaultKeys[k];
				return `${key}={{ ${k}: ${typeof v === "string" ? `"${v}"` : v} }}`;
			}
		}
		if (isPropMapValue(value)) return `${key}={{\n${indent}    ${Object.keys(value).map((k) => {
			const v = value[k];
			return `${k}: ${typeof v === "string" ? `"${v}"` : v}`;
		}).join(`,\n${indent}    `)}\n${indent}  }}`;
		if (typeof value === "string") {
			if (value.startsWith("{") && value.endsWith("}")) try {
				const parsed = JSON.parse(value);
				if (typeof parsed === "object" && parsed !== null) {
					const keys = Object.keys(parsed);
					if (keys.length > 1) return `${key}={{\n${indent}    ${keys.map((k) => `${k}: ${toJSLiteral(parsed[k])}`).join(`,\n${indent}    `)}\n${indent}  }}`;
					return `${key}={${toObjectLiteral(parsed)}}`;
				}
			} catch {}
			return `${key}="${value}"`;
		} else if (value !== null && typeof value === "object") {
			const roundedValue = { ...value };
			if ("x" in roundedValue && typeof roundedValue.x === "number") roundedValue.x = Math.round(roundedValue.x * 100) / 100;
			if ("y" in roundedValue && typeof roundedValue.y === "number") roundedValue.y = Math.round(roundedValue.y * 100) / 100;
			const keys = Object.keys(roundedValue);
			if (keys.length > 1) return `${key}={{\n${indent}    ${keys.map((k) => `${k}: ${typeof roundedValue[k] === "string" ? `"${roundedValue[k]}"` : roundedValue[k]}`).join(`,\n${indent}    `)}\n${indent}  }}`;
			return `${key}={${toObjectLiteral(roundedValue)}}`;
		} else return `${key}={${value}}`;
	}).join("\n" + indent + "  ");
}
function isIdReferenced(id, allComponents) {
	const flatComponents = [];
	function flattenComponents(components) {
		for (const component of components) {
			flatComponents.push(component);
			if (component.children) flattenComponents(component.children);
		}
	}
	flattenComponents(allComponents);
	return flatComponents.some((component) => {
		if (component.props?.maskSource === id) return true;
		if (component.props) {
			for (const value of Object.values(component.props)) if (isPropMapValue(value) && value.source === id) return true;
		}
		return false;
	});
}
function generatePresetCode(preset, colorSpace) {
	const componentTypes = /* @__PURE__ */ new Set();
	function collectTypes(components) {
		for (const config of components) {
			componentTypes.add(config.type);
			if (config.children) collectTypes(config.children);
		}
	}
	collectTypes(preset.components);
	const generateComponentString = (config, indent = "      ") => {
		const propString = config.props ? generatePropString(config.props, config.type, indent) : "";
		let idString = "";
		if (config.id) {
			if (isIdReferenced(config.id, preset.components)) idString = `id="${config.id}"`;
		}
		const attributes = [idString, propString].filter(Boolean).join("\n" + indent + "  ");
		const sampleChildren = config.type === "DOMTexture" && (!config.children || config.children.length === 0) ? `${indent}  <h1>Your HTML content here</h1>` : null;
		if (sampleChildren) if (attributes) return `${indent}<${config.type}\n${indent}  ${attributes}>\n${sampleChildren}\n${indent}</${config.type}>`;
		else return `${indent}<${config.type}>\n${sampleChildren}\n${indent}</${config.type}>`;
		else if (config.children && config.children.length > 0) {
			const childrenString = config.children.map((child) => generateComponentString(child, indent + "  ")).join("\n");
			if (attributes) return `${indent}<${config.type}\n${indent}  ${attributes}>\n${childrenString}\n${indent}</${config.type}>`;
			else return `${indent}<${config.type}>\n${childrenString}\n${indent}</${config.type}>`;
		} else if (attributes) return `${indent}<${config.type}\n${indent}  ${attributes} />`;
		else return `${indent}<${config.type} />`;
	};
	const componentStrings = preset.components.map((config) => generateComponentString(config)).join("\n");
	const shaderProps = colorSpace && colorSpace !== "p3-linear" ? ` colorSpace="${colorSpace}"` : "";
	const allImports = ["Shader", ...Array.from(componentTypes).sort()];
	return `${allImports.length > 3 ? `import {\n  ${allImports.join(",\n  ")},\n} from 'shaders/react'` : `import { ${allImports.join(", ")} } from 'shaders/react'`}

export default function ShaderEffect() {
  return (
    <Shader${shaderProps}>
${componentStrings}
    </Shader>
  )
}`;
}
const availableComponents = [
	"AngularBlur",
	"Ascii",
	"Aurora",
	"BarShift",
	"Beam",
	"Blob",
	"Blur",
	"BrickPattern",
	"BrightnessContrast",
	"Bulge",
	"CRTScreen",
	"ChannelBlur",
	"Checkerboard",
	"Chevron",
	"ChromaFlow",
	"ChromaticAberration",
	"Circle",
	"ColorWheel",
	"ConcentricSpin",
	"ConicGradient",
	"ContourLines",
	"Crescent",
	"Cross",
	"Crystal",
	"CursorRipples",
	"CursorTrail",
	"DOMTexture",
	"DiamondGradient",
	"DiffuseBlur",
	"Dither",
	"DotGrid",
	"DropShadow",
	"Duotone",
	"Ellipse",
	"Emboss",
	"FallingLines",
	"FilmGrain",
	"FloatingParticles",
	"FlowField",
	"Flower",
	"FlowingGradient",
	"FlutedGlass",
	"Fog",
	"Form3D",
	"FractalNoise",
	"Glass",
	"GlassTiles",
	"Glitch",
	"Glow",
	"Godrays",
	"Grayscale",
	"Grid",
	"GridDistortion",
	"Group",
	"Halftone",
	"HexGrid",
	"HueShift",
	"ImageTexture",
	"Invert",
	"Kaleidoscope",
	"LensFlare",
	"LinearBlur",
	"LinearGradient",
	"Liquify",
	"Marble",
	"Mirror",
	"MultiPointGradient",
	"Neon",
	"Paper",
	"Perspective",
	"Pixelate",
	"Plasma",
	"PolarCoordinates",
	"Polygon",
	"Posterize",
	"ProgressiveBlur",
	"RadialGradient",
	"RectangularCoordinates",
	"ReflectivePlane",
	"Ring",
	"Ripples",
	"RoundedRect",
	"Saturation",
	"Sharpness",
	"Shatter",
	"SimplexNoise",
	"SineWave",
	"Smoke",
	"SmokeFill",
	"Solarize",
	"SolidColor",
	"Spherize",
	"Spiral",
	"Star",
	"Strands",
	"Stretch",
	"Stripes",
	"StudioBackground",
	"SunBurst",
	"Swirl",
	"TiltShift",
	"Tint",
	"Trapezoid",
	"Tritone",
	"Truchet",
	"Twirl",
	"VHS",
	"Vesica",
	"Vibrance",
	"VideoTexture",
	"Vignette",
	"Voronoi",
	"WaveDistortion",
	"Weave",
	"WebcamTexture",
	"WorleyNoise",
	"ZoomBlur"
];
export { availableComponents, generatePresetCode };
