// shaders/dist/core/AngularBlur-C1oRJ8Y9.js
import { l as transformPosition } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { Fn, convertToTexture, cos, float, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
var rawWeights = Array.from({ length: 32 }, (_, i) => {
	const t = (i / 31 - .5) * 2;
	return Math.exp(-.5 * t * t / .64);
});
var wSum = rawWeights.reduce((a, b) => a + b);
var blurWeights = rawWeights.map((w) => w / wSum);
const componentDefinition = {
	name: "AngularBlur",
	category: "Blurs",
	description: "Radial motion blur rotating around a center point",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 20,
			description: "Intensity of the angular blur effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Blur Intensity",
				group: "Effect"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the rotation",
			ui: {
				type: "position",
				label: "Center Position",
				group: "Position"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const center = vec2(uniforms.center.uniform.x, uniforms.center.uniform.y.oneMinus());
		const angleStep = uniforms.intensity.uniform.mul(.005).div(float(31));
		const aspect = viewportSize.x.div(viewportSize.y);
		const cosStep = cos(angleStep);
		const sinStep = sin(angleStep);
		return Fn(() => {
			const initialCoord = screenUV.sub(center);
			const aspectCorrectedCoord = vec2(initialCoord.x.mul(aspect), initialCoord.y);
			const total = vec4(0).toVar();
			const curCos = float(1).toVar();
			const curSin = float(0).toVar();
			for (let i = 0; i < 32; i++) {
				const rotatedX = aspectCorrectedCoord.x.mul(curCos).sub(aspectCorrectedCoord.y.mul(curSin));
				const rotatedY = aspectCorrectedCoord.x.mul(curSin).add(aspectCorrectedCoord.y.mul(curCos));
				const sampleCoord = vec2(rotatedX.div(aspect), rotatedY).add(center);
				total.assign(total.add(childTexture.sample(sampleCoord).mul(float(blurWeights[i]))));
				const nextCos = curCos.mul(cosStep).sub(curSin.mul(sinStep));
				const nextSin = curSin.mul(cosStep).add(curCos.mul(sinStep));
				curCos.assign(nextCos);
				curSin.assign(nextSin);
			}
			return unpremultiplyAlpha(total);
		})();
	}
};
var AngularBlur_default = componentDefinition;
export { componentDefinition as n, AngularBlur_default as t };
