// shaders/dist/core/Ascii-BwSMmD6Z.js
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { DataTexture, NearestFilter, RGBAFormat, UnsignedByteType } from "three";
import { abs, any, clamp, convertToTexture, dot, float, floor, fract, greaterThan, or, pow, screenUV, select, texture, uniform, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Ascii",
	category: "Stylize",
	description: "Convert imagery to ASCII character art",
	requiresRTT: true,
	requiresChild: true,
	props: {
		characters: {
			default: "@%#*+=-:.",
			description: "Characters ordered from dense to sparse. First character is used for bright areas, last for dark areas.",
			ui: {
				type: "text",
				label: "Characters",
				group: "Effect"
			}
		},
		cellSize: {
			default: 30,
			description: "Size of each ASCII character cell (normalized to 1080p reference, scales proportionally at other resolutions)",
			ui: {
				type: "range",
				min: 8,
				max: 100,
				step: 1,
				label: "Cell Size",
				group: "Effect"
			}
		},
		fontFamily: {
			default: "JetBrains Mono",
			description: "Font family for characters",
			ui: {
				type: "select",
				options: [
					{
						label: "Azeret Mono",
						value: "Azeret Mono"
					},
					{
						label: "Courier Prime",
						value: "Courier Prime"
					},
					{
						label: "Cutive Mono",
						value: "Cutive Mono"
					},
					{
						label: "Fira Code",
						value: "Fira Code"
					},
					{
						label: "Geist Mono",
						value: "Geist Mono"
					},
					{
						label: "IBM Plex Mono",
						value: "IBM Plex Mono"
					},
					{
						label: "JetBrains Mono",
						value: "JetBrains Mono"
					},
					{
						label: "Major Mono Display",
						value: "Major Mono Display"
					},
					{
						label: "Martian Mono",
						value: "Martian Mono"
					},
					{
						label: "Nova Mono",
						value: "Nova Mono"
					},
					{
						label: "Press Start 2P",
						value: "Press Start 2P"
					},
					{
						label: "Roboto Mono",
						value: "Roboto Mono"
					},
					{
						label: "Share Tech Mono",
						value: "Share Tech Mono"
					},
					{
						label: "Silkscreen",
						value: "Silkscreen"
					},
					{
						label: "Source Code Pro",
						value: "Source Code Pro"
					},
					{
						label: "Space Mono",
						value: "Space Mono"
					},
					{
						label: "Syne Mono",
						value: "Syne Mono"
					},
					{
						label: "VT323",
						value: "VT323"
					},
					{
						label: "Xanh Mono",
						value: "Xanh Mono"
					}
				],
				label: "Font Family",
				group: "Appearance"
			}
		},
		spacing: {
			default: 1,
			description: "Character size within each cell (1.0 = optimal size, 0.0 = smallest)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Character Size",
				group: "Effect"
			},
			transform: (value) => .1 + value * 1.4
		},
		gamma: {
			default: 1,
			description: "Brightness curve adjustment. <1 brightens darks (more light characters), >1 darkens midtones (more dark characters). Use to better fit characters to image brightness range.",
			ui: {
				type: ["range", "map"],
				min: .25,
				max: 3,
				step: .05,
				label: "Gamma",
				group: "Effect"
			}
		},
		alphaThreshold: {
			default: 0,
			description: "Pixels with alpha below this threshold become fully transparent.",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Alpha Threshold",
				group: "Effect"
			}
		},
		preserveAlpha: {
			default: true,
			description: "When enabled, output alpha matches input alpha. When disabled, pixels above the alpha threshold become fully opaque.",
			ui: {
				type: "checkbox",
				label: "Preserve Alpha",
				group: "Effect"
			},
			transform: (value) => value ? 1 : 0
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup, onBeforeRender }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const ATLAS_TEXTURE_SIZE = 2048;
		const GOOGLE_FONTS_MAP = {
			"JetBrains Mono": "JetBrains+Mono",
			"Fira Code": "Fira+Code",
			"Source Code Pro": "Source+Code+Pro",
			"IBM Plex Mono": "IBM+Plex+Mono",
			"Space Mono": "Space+Mono",
			"Roboto Mono": "Roboto+Mono",
			"Courier Prime": "Courier+Prime",
			"Geist Mono": "Geist+Mono",
			"VT323": "VT323",
			"Press Start 2P": "Press+Start+2P",
			"Silkscreen": "Silkscreen",
			"Major Mono Display": "Major+Mono+Display",
			"Syne Mono": "Syne+Mono",
			"Nova Mono": "Nova+Mono",
			"Xanh Mono": "Xanh+Mono",
			"Cutive Mono": "Cutive+Mono",
			"Share Tech Mono": "Share+Tech+Mono",
			"Martian Mono": "Martian+Mono",
			"Azeret Mono": "Azeret+Mono"
		};
		const loadedFonts = /* @__PURE__ */ new Set();
		async function loadGoogleFont(fontFamily) {
			if (loadedFonts.has(fontFamily)) return;
			if (!GOOGLE_FONTS_MAP[fontFamily]) return;
			const googleFontName = GOOGLE_FONTS_MAP[fontFamily];
			if (!document.querySelector(`link[href*="${googleFontName}"]`)) {
				const link = document.createElement("link");
				link.rel = "stylesheet";
				link.href = `https://fonts.googleapis.com/css2?family=${googleFontName}:wght@400&display=swap`;
				document.head.appendChild(link);
			}
			try {
				if (document.fonts && document.fonts.load) {
					const specs = [
						`400 12px "${fontFamily}"`,
						`12px "${fontFamily}"`,
						`400 12px ${fontFamily}`,
						`12px ${fontFamily}`
					];
					for (const spec of specs) try {
						await document.fonts.load(spec);
						break;
					} catch (e) {}
					await new Promise((resolve) => setTimeout(resolve, 500));
					loadedFonts.add(fontFamily);
				} else {
					await new Promise((resolve) => setTimeout(resolve, 1e3));
					loadedFonts.add(fontFamily);
				}
			} catch (error) {}
		}
		let reusableCanvas = null;
		let reusableCtx = null;
		let reusableDataBuffer = null;
		const createCharacterAtlas = async (characters, cellSize, fontFamily, spacing) => {
			if (!characters || characters.length === 0) throw new Error("Characters string cannot be empty");
			const charCount$1 = characters.length;
			const atlasSize$1 = Math.max(2, Math.ceil(Math.sqrt(charCount$1)));
			const baseAtlasCellSize = 128;
			const spacingMultiplier = Math.max(1, 2 / spacing);
			const actualCellSize = Math.min(baseAtlasCellSize * spacingMultiplier, ATLAS_TEXTURE_SIZE / atlasSize$1);
			const fontSize = actualCellSize * .75;
			if (!reusableCanvas) {
				reusableCanvas = document.createElement("canvas");
				reusableCanvas.width = ATLAS_TEXTURE_SIZE;
				reusableCanvas.height = ATLAS_TEXTURE_SIZE;
				reusableCtx = reusableCanvas.getContext("2d", { willReadFrequently: true });
				if (!reusableCtx) {
					console.error("[Ascii] Failed to create 2D canvas context for character atlas");
					throw new Error("Ascii shader: Canvas 2D context unavailable");
				}
			}
			if (!reusableCtx) {
				console.error("[Ascii] Canvas context unexpectedly null during atlas creation");
				throw new Error("Ascii shader: Canvas 2D context unavailable");
			}
			const ctx = reusableCtx;
			ctx.clearRect(0, 0, ATLAS_TEXTURE_SIZE, ATLAS_TEXTURE_SIZE);
			ctx.fillStyle = "#ffffff";
			ctx.font = `${fontSize}px "${fontFamily}", ${fontFamily}, monospace`;
			if (!ctx.font.includes(fontFamily) && !ctx.font.includes("IBM") && !ctx.font.includes("Plex")) ctx.font = `${fontSize}px ${fontFamily}, monospace`;
			ctx.textAlign = "center";
			ctx.textBaseline = "middle";
			for (let i = 0; i < charCount$1; i++) {
				const char = characters[i];
				const row = Math.floor(i / atlasSize$1);
				const x = i % atlasSize$1 * actualCellSize + actualCellSize / 2;
				const y = row * actualCellSize + actualCellSize / 2;
				ctx.fillText(char, x, y);
			}
			const imageData = ctx.getImageData(0, 0, ATLAS_TEXTURE_SIZE, ATLAS_TEXTURE_SIZE);
			if (!reusableDataBuffer) reusableDataBuffer = new Uint8Array(ATLAS_TEXTURE_SIZE * ATLAS_TEXTURE_SIZE * 4);
			reusableDataBuffer.set(imageData.data);
			return {
				data: reusableDataBuffer,
				width: ATLAS_TEXTURE_SIZE,
				height: ATLAS_TEXTURE_SIZE,
				atlasSize: atlasSize$1,
				cellSize,
				charCount: charCount$1,
				characters,
				fontFamily,
				uvScale: atlasSize$1 * actualCellSize / ATLAS_TEXTURE_SIZE
			};
		};
		const defaultCharacters = componentDefinition.props.characters.default;
		const defaultCellSize = componentDefinition.props.cellSize.default;
		const defaultFontFamily = componentDefinition.props.fontFamily.default;
		const defaultSpacing = componentDefinition.props.spacing.transform ? componentDefinition.props.spacing.transform(componentDefinition.props.spacing.default) : componentDefinition.props.spacing.default;
		const initialCharacters = uniforms?.characters?.uniform?.value ?? defaultCharacters;
		const initialCellSize = uniforms?.cellSize?.uniform?.value ?? defaultCellSize;
		const initialFontFamily = uniforms?.fontFamily?.uniform?.value ?? defaultFontFamily;
		const initialSpacing = uniforms?.spacing?.uniform?.value ?? defaultSpacing;
		let atlasData = {
			data: new Uint8Array(ATLAS_TEXTURE_SIZE * ATLAS_TEXTURE_SIZE * 4),
			width: ATLAS_TEXTURE_SIZE,
			height: ATLAS_TEXTURE_SIZE,
			atlasSize: 3,
			cellSize: initialCellSize,
			charCount: initialCharacters.length,
			characters: initialCharacters,
			fontFamily: initialFontFamily,
			uvScale: 1
		};
		const atlasTexture = new DataTexture(atlasData.data, ATLAS_TEXTURE_SIZE, ATLAS_TEXTURE_SIZE, RGBAFormat, UnsignedByteType);
		atlasTexture.minFilter = NearestFilter;
		atlasTexture.magFilter = NearestFilter;
		atlasTexture.generateMipmaps = false;
		atlasTexture.needsUpdate = true;
		const atlasField = texture(atlasTexture);
		const charCountUniform = uniform(atlasData.charCount);
		const cellSizeUniform = uniform(atlasData.cellSize);
		const atlasScaleUniform = uniform(atlasData.uvScale);
		const atlasSizeUniform = uniform(atlasData.atlasSize);
		let lastCharacters = initialCharacters;
		let lastCellSize = initialCellSize;
		let lastFontFamily = initialFontFamily;
		let lastSpacing = initialSpacing;
		const initializeAtlas = async () => {
			const currentCharacters = uniforms?.characters?.uniform?.value ?? lastCharacters;
			const currentCellSize = uniforms?.cellSize?.uniform?.value ?? lastCellSize;
			const currentFontFamily = uniforms?.fontFamily?.uniform?.value ?? lastFontFamily;
			const currentSpacing = uniforms?.spacing?.uniform?.value ?? lastSpacing;
			atlasData = await createCharacterAtlas(currentCharacters, currentCellSize, currentFontFamily, currentSpacing);
			atlasTexture.image.data = atlasData.data;
			atlasTexture.needsUpdate = true;
			charCountUniform.value = atlasData.charCount;
			cellSizeUniform.value = atlasData.cellSize;
			atlasScaleUniform.value = atlasData.uvScale;
			atlasSizeUniform.value = atlasData.atlasSize;
			lastCharacters = currentCharacters;
			lastCellSize = currentCellSize;
			lastFontFamily = currentFontFamily;
			lastSpacing = currentSpacing;
		};
		setTimeout(() => {
			initializeAtlas().catch((error) => {
				console.error("Failed to initialize atlas:", error);
			});
			loadGoogleFont(uniforms?.fontFamily?.uniform?.value ?? initialFontFamily).then(() => {
				initializeAtlas().catch((error) => {
					console.error("Failed to update atlas with custom font:", error);
				});
			}).catch((error) => {
				console.error("Failed to load font:", error);
			});
		}, 0);
		let lastUpdateTime = 0;
		let isUpdating = false;
		onBeforeRender(() => {
			const now = Date.now();
			if (now - lastUpdateTime < 16 || isUpdating) return;
			const currentCharacters = uniforms.characters?.uniform?.value ?? lastCharacters;
			const currentCellSize = uniforms.cellSize?.uniform?.value ?? lastCellSize;
			const currentFontFamily = uniforms.fontFamily?.uniform?.value ?? lastFontFamily;
			const currentSpacing = uniforms.spacing?.uniform?.value ?? lastSpacing;
			if (!currentCharacters || currentCharacters.length === 0) return;
			if (currentCharacters !== lastCharacters || currentCellSize !== lastCellSize || currentFontFamily !== lastFontFamily || currentSpacing !== lastSpacing) {
				lastUpdateTime = now;
				isUpdating = true;
				createCharacterAtlas(currentCharacters, currentCellSize, currentFontFamily, currentSpacing).then((newAtlasData) => {
					atlasData = newAtlasData;
					atlasTexture.image.data = atlasData.data;
					atlasTexture.needsUpdate = true;
					charCountUniform.value = atlasData.charCount;
					cellSizeUniform.value = atlasData.cellSize;
					atlasScaleUniform.value = atlasData.uvScale;
					atlasSizeUniform.value = atlasData.atlasSize;
					lastCharacters = currentCharacters;
					lastCellSize = currentCellSize;
					lastFontFamily = currentFontFamily;
					lastSpacing = currentSpacing;
					isUpdating = false;
				}).catch((error) => {
					console.error("Failed to update atlas:", error);
					isUpdating = false;
				});
			}
		});
		onCleanup(() => {
			atlasTexture.dispose();
			if (reusableCanvas) {
				reusableCanvas.width = 0;
				reusableCanvas.height = 0;
				reusableCanvas = null;
				reusableCtx = null;
			}
			reusableDataBuffer = null;
		});
		const defaultGamma = componentDefinition.props.gamma.default;
		const defaultAlphaThreshold = componentDefinition.props.alphaThreshold.default;
		const defaultPreserveAlpha = componentDefinition.props.preserveAlpha.transform ? componentDefinition.props.preserveAlpha.transform(componentDefinition.props.preserveAlpha.default) : componentDefinition.props.preserveAlpha.default ? 1 : 0;
		const gammaUniform = uniforms.gamma?.uniform ?? uniform(defaultGamma);
		const alphaThresholdUniform = uniforms.alphaThreshold?.uniform ?? uniform(defaultAlphaThreshold);
		const preserveAlphaUniform = uniforms.preserveAlpha?.uniform ?? uniform(defaultPreserveAlpha);
		const scaleFactor = viewportSize.y.div(1080);
		const effectiveCellSize = cellSizeUniform.mul(scaleFactor);
		const mainGridSize = viewportSize.div(effectiveCellSize);
		const gridCoords = screenUV.mul(mainGridSize);
		const cellCoords = floor(gridCoords);
		const rawCellUV = fract(gridCoords);
		const characterSize = uniforms.spacing?.uniform ?? uniform(initialSpacing);
		const center = vec2(.5);
		const centeredUV = rawCellUV.sub(center);
		const cellUV = centeredUV.div(characterSize).add(center);
		const isOutsideBounds = any(greaterThan(abs(centeredUV), vec2(characterSize.mul(.5))));
		const cellCenter = cellCoords.add(vec2(.5)).div(mainGridSize);
		const cellColor = unpremultiplyAlpha(childTexture.sample(cellCenter));
		const adjustedBrightness = pow(dot(cellColor.rgb, vec3(.299, .587, .114)), gammaUniform);
		const charCount = charCountUniform;
		const charIndex = clamp(floor(float(1).sub(adjustedBrightness).mul(charCount)), 0, charCount.sub(1));
		const atlasSize = atlasSizeUniform;
		const atlasCol = charIndex.mod(atlasSize);
		const atlasRow = floor(charIndex.div(atlasSize));
		const atlasCellSize = float(1).div(atlasSize).mul(atlasScaleUniform);
		const atlasSampleUV = vec2(atlasCol, atlasRow).mul(atlasCellSize).add(cellUV.mul(atlasCellSize));
		const asciiChar = atlasField.sample(atlasSampleUV);
		const isBackground = dot(asciiChar.rgb, vec3(.299, .587, .114)).lessThan(.1);
		const sourceAlpha = cellColor.a;
		const isBelowThreshold = sourceAlpha.lessThan(alphaThresholdUniform);
		const outputAlpha = select(preserveAlphaUniform.greaterThan(.5), sourceAlpha, float(1));
		const finalAlpha = select(or(or(isBackground, isOutsideBounds), isBelowThreshold), float(0), outputAlpha);
		return vec4(asciiChar.rgb.mul(cellColor.rgb), finalAlpha);
	}
};
var Ascii_default = componentDefinition;
export { componentDefinition as n, Ascii_default as t };
