// shaders/dist/core/Aurora-D5Pjk4ve.js
import { a as transformColor, l as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { Fn, clamp, exp, float, fract, mix, mx_noise_float, pow, screenUV, sin, smoothstep, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Aurora",
	category: "Textures",
	description: "Mesmerizing aurora borealis with layered curtains, vertical rays, and flowing light.",
	props: {
		colorA: {
			default: "#a533f8",
			transform: transformColor,
			description: "Edge color at the curtain base",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#22ee88",
			transform: transformColor,
			description: "Core color in the bright center",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		colorC: {
			default: "#1694e8",
			transform: transformColor,
			description: "Tip color at the ray ends",
			ui: {
				type: "color",
				label: "Color C",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		balance: {
			default: 50,
			description: "Shifts color distribution across the curtain height",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Balance",
				group: "Colors"
			}
		},
		intensity: {
			default: 80,
			description: "Overall aurora brightness",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Aurora"
			}
		},
		curtainCount: {
			default: 4,
			compileTime: true,
			description: "Number of aurora curtain layers",
			ui: {
				type: "range",
				min: 1,
				max: 4,
				step: 1,
				label: "Curtains",
				group: "Aurora"
			}
		},
		speed: {
			default: 5,
			description: "Animation speed",
			ui: {
				type: "range",
				min: -10,
				max: 10,
				step: .1,
				label: "Speed",
				group: "Aurora"
			}
		},
		waviness: {
			default: 50,
			description: "How much the curtains undulate",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 200,
				step: 1,
				label: "Waviness",
				group: "Aurora"
			}
		},
		rayDensity: {
			default: 20,
			description: "Density of vertical ray structures",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Ray Detail",
				group: "Aurora"
			}
		},
		height: {
			default: 120,
			description: "How tall the aurora extends",
			ui: {
				type: ["range", "map"],
				min: 10,
				max: 200,
				step: 1,
				label: "Height",
				group: "Scene"
			}
		},
		center: {
			default: {
				x: .5,
				y: 0
			},
			transform: transformPosition,
			description: "Center position of the aurora",
			ui: {
				type: "position",
				label: "Center",
				group: "Scene"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for variation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Aurora"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const colorA = uniforms.colorA.uniform;
		const colorB = uniforms.colorB.uniform;
		const colorC = uniforms.colorC.uniform;
		const csMode = uniforms.colorSpace.uniform.value;
		const balanceExp = mix(float(2.5), float(.4), uniforms.balance.uniform.mul(.01));
		const intensity = uniforms.intensity.uniform.mul(.018);
		const curtainCount = uniforms.curtainCount.uniform.value | 0;
		const waviness = uniforms.waviness.uniform.mul(.01);
		const rayAmount = uniforms.rayDensity.uniform.mul(.01);
		const auroraHeight = uniforms.height.uniform.mul(.01);
		const centerX = uniforms.center.uniform.x;
		const centerY = uniforms.center.uniform.y.oneMinus();
		const seed = uniforms.seed.uniform;
		const aspect = viewportSize.x.div(viewportSize.y);
		const t = createAnimatedTime(params, uniforms.speed);
		const auroraCurtain = Fn(([uv$1, layerIdx, layerTime]) => {
			const tSlow = t.mul(.06).add(layerTime);
			const lSeed = seed.add(layerIdx.mul(31.7));
			const wx = uv$1.x.sub(centerX).mul(aspect).mul(2);
			const ph1 = fract(sin(lSeed.mul(12.9)).mul(4758.5)).mul(Math.PI * 2);
			const ph2 = fract(sin(lSeed.mul(78.2)).mul(2847.1)).mul(Math.PI * 2);
			const ph3 = fract(sin(lSeed.mul(41.6)).mul(1593.7)).mul(Math.PI * 2);
			const arc = sin(wx.mul(.7).add(tSlow.mul(.4)).add(ph1)).mul(.15).add(sin(wx.mul(1.6).add(tSlow.mul(.65)).add(ph2)).mul(.08)).add(sin(wx.mul(2.9).add(tSlow.mul(.3)).add(ph3)).mul(.04));
			const noiseWarp = mx_noise_float(vec3(wx.mul(.5), tSlow.mul(.25), lSeed.mul(.1)));
			const pathY = centerY.add(arc.mul(waviness)).add(noiseWarp.mul(.14).mul(waviness));
			const warpedX = wx.add(noiseWarp.mul(2.5));
			const rayFreq = mix(float(6), float(20), rayAmount);
			const rayPhase = fract(sin(lSeed.mul(53.7)).mul(3847.2)).mul(Math.PI * 2);
			const rays = pow(sin(warpedX.mul(rayFreq.mul(.15)).add(tSlow.mul(.1)).add(rayPhase)).mul(.5).add(.5), float(1.5));
			const localHeight = auroraHeight.mul(mix(float(.4), float(1), rays));
			const distFromPath = uv$1.y.sub(pathY);
			const bottomFade = smoothstep(float(-.04), float(.015), distFromPath);
			const topExtent = localHeight.mul(rays.mul(.4).add(.6));
			const topFade = smoothstep(topExtent, float(0), distFromPath);
			const envelope = bottomFade.mul(topFade);
			const sectionBright = mix(float(.4), float(1), noiseWarp.mul(.5).add(.5));
			const colorT = clamp(distFromPath.div(topExtent.add(.001)), 0, 1);
			const coreD = colorT.sub(.2);
			const coreBright = exp(coreD.mul(coreD).mul(-6)).mul(.3).add(1);
			return vec2(colorT, envelope.mul(rays).mul(sectionBright).mul(coreBright));
		});
		return Fn(() => {
			const uv$1 = screenUV;
			const weightedT = float(0).toVar();
			const totalAlpha = float(0).toVar();
			const a1 = auroraCurtain(uv$1, float(0), float(0));
			weightedT.assign(weightedT.add(a1.x.mul(a1.y)));
			totalAlpha.assign(totalAlpha.add(a1.y));
			if (curtainCount >= 2) {
				const a2 = auroraCurtain(uv$1, float(1), float(3.1));
				const w2 = a2.y.mul(.65);
				weightedT.assign(weightedT.add(a2.x.mul(w2)));
				totalAlpha.assign(totalAlpha.add(w2));
			}
			if (curtainCount >= 3) {
				const a3 = auroraCurtain(uv$1, float(2), float(6.7));
				const w3 = a3.y.mul(.45);
				weightedT.assign(weightedT.add(a3.x.mul(w3)));
				totalAlpha.assign(totalAlpha.add(w3));
			}
			if (curtainCount >= 4) {
				const a4 = auroraCurtain(uv$1, float(3), float(9.4));
				const w4 = a4.y.mul(.3);
				weightedT.assign(weightedT.add(a4.x.mul(w4)));
				totalAlpha.assign(totalAlpha.add(w4));
			}
			const avgT = pow(weightedT.div(totalAlpha.add(.001)), balanceExp);
			const lowToMid = smoothstep(float(0), float(.2), avgT);
			const midToHigh = smoothstep(float(.4), float(.85), avgT);
			const col = mixColors(mixColors(colorA, colorB, lowToMid, csMode), colorC, midToHigh, csMode).rgb.mul(totalAlpha).mul(intensity);
			const alpha = clamp(totalAlpha.mul(intensity).mul(.5), 0, 1);
			return vec4(clamp(col, 0, 1), alpha);
		})();
	}
};
var Aurora_default = componentDefinition;
export { componentDefinition as n, Aurora_default as t };
