// shaders/dist/core/BarShift-2XBhva1A.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges, r as transformAngle } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { Fn, convertToTexture, cos, float, floor, fract, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "BarShift",
	category: "Distortions",
	description: "Slices content into parallel bars, each offset independently for a fractured or glitch-like effect",
	requiresRTT: true,
	requiresChild: true,
	props: {
		count: {
			default: 6,
			description: "Number of bars across the longest viewport dimension",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Count",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Angle of bar orientation in degrees (0 = vertical bars, 90 = horizontal bars)",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		intensity: {
			default: .15,
			description: "Maximum displacement per bar",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Intensity",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Randomization seed for per-bar offset variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		speed: {
			default: 0,
			description: "Animation speed — each bar drifts at its own rate and direction",
			ui: {
				type: "range",
				min: -2,
				max: 2,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms, childNode, onCleanup } = params;
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const aspect = viewportSize.x.div(viewportSize.y);
		const animTime = createAnimatedTime(params, uniforms.speed);
		const hash = (x) => fract(sin(x.mul(12.9898)).mul(43758.5453)).sub(.5).mul(2);
		const angleRad = uniforms.angle.uniform.mul(.0174533);
		const cosA = cos(angleRad);
		const sinA = sin(angleRad);
		const longestAxis = aspect.max(float(1));
		return Fn(() => {
			const uv$1 = screenUV;
			const au = uv$1.x.mul(aspect);
			const av = uv$1.y;
			const cu = au.sub(aspect.mul(.5));
			const cv = av.sub(.5);
			const projectedU = cu.mul(cosA).add(cv.mul(sinA));
			const projectedV = cu.mul(sinA).negate().add(cv.mul(cosA));
			const barIndex = floor(projectedU.mul(uniforms.count.uniform).div(longestAxis));
			const seedOffset = uniforms.seed.uniform.mul(.137);
			const staticOffset = hash(barIndex.add(seedOffset));
			const barSpeed = hash(barIndex.add(seedOffset).add(float(42.7)));
			const animOffset = animTime.mul(barSpeed).mul(.2);
			const offset = staticOffset.add(animOffset).mul(uniforms.intensity.uniform);
			const distPV = projectedV.add(offset);
			const dcu = projectedU.mul(cosA).sub(distPV.mul(sinA));
			const dcv = projectedU.mul(sinA).add(distPV.mul(cosA));
			const sampleUV = vec2(dcu.div(aspect).add(.5), dcv.add(.5));
			return unpremultiplyAlpha(applyEdgeHandling(sampleUV, childTexture.sample(sampleUV), childTexture, uniforms.edges.uniform.value));
		})();
	}
};
var BarShift_default = componentDefinition;
export { componentDefinition as n, BarShift_default as t };
