// shaders/dist/core/Blur-5_t3tR_g.js
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { i as createVariableGaussianBlurCompute, r as createGaussianBlurCompute } from "./computeBlur-CprKe0xa.js";
import { Fn, If, clamp, convertToTexture, dot, float, instanceIndex, int, mix, pow, screenUV, texture, textureLoad, textureStore, uint, uniform, uvec2, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Blur",
	category: "Blurs",
	description: "A simple Gaussian blur effect",
	requiresRTT: true,
	requiresChild: true,
	props: { intensity: {
		default: 50,
		description: "Intensity of the blur effect",
		ui: {
			type: ["range", "map"],
			min: 0,
			max: 200,
			step: 1,
			label: "Intensity",
			group: "Effect"
		}
	} },
	computeNode: ({ childNode, onCleanup, dimensions, renderer, uniforms, onResize }) => {
		if (!childNode) return null;
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const dpr = renderer?.getPixelRatio?.() ?? 1;
		const texWidth = Math.round(dimensions.width * dpr);
		const texHeight = Math.round(dimensions.height * dpr);
		const mapInfo = uniforms.intensity.mapComputeInfo;
		if (mapInfo) {
			const blur$1 = createVariableGaussianBlurCompute(childTexture, texWidth, texHeight, onCleanup);
			const { computeNodes: blurNodes, outputTexture: outputTexture$1, blurMapWriteNode, computeWidth, computeHeight } = blur$1;
			const inputWidthU = uniform(texWidth);
			const inputHeightU = uniform(texHeight);
			const cw = uint(computeWidth);
			const ch = uint(computeHeight);
			const { sourceTexture, channel, inputMin, inputMax, outputMin, outputMax, curve } = mapInfo;
			const fillBlurMap = Fn(() => {
				const idx = instanceIndex;
				const cx = idx.mod(cw);
				const cy = idx.div(cw);
				If(cy.lessThan(ch), () => {
					const u = cx.toFloat().add(float(.5)).div(float(computeWidth));
					const v = cy.toFloat().add(float(.5)).div(float(computeHeight));
					const sample = textureLoad(sourceTexture, uvec2(u.mul(inputWidthU).round().toInt().clamp(int(0), int(inputWidthU.sub(1))).toUint(), v.mul(inputHeightU).round().toInt().clamp(int(0), int(inputHeightU.sub(1))).toUint()), int(0));
					let raw;
					if (channel === "alpha") raw = sample.a;
					else if (channel === "alphaInverted") raw = sample.a.oneMinus();
					else if (channel === "luminanceInverted") raw = dot(sample.rgb, vec3(.2126, .7152, .0722)).oneMinus();
					else raw = dot(sample.rgb, vec3(.2126, .7152, .0722));
					const inputRange = inputMax.sub(inputMin).max(float(1e-4));
					const radius = mix(outputMin, outputMax, pow(clamp(raw.sub(inputMin).div(inputRange), float(0), float(1)), pow(float(2), curve.negate().mul(float(2))))).mul(float(.36));
					textureStore(blurMapWriteNode, uvec2(cx, cy), vec4(radius, float(0), float(0), float(1))).toWriteOnly();
				});
			})().compute(computeWidth * computeHeight, [256]);
			onResize(({ width, height }) => {
				const newW = Math.round(width * dpr);
				const newH = Math.round(height * dpr);
				inputWidthU.value = newW;
				inputHeightU.value = newH;
				blur$1.setInputDimensions(newW, newH);
			});
			return {
				outputs: {
					childTexture,
					blurredTexture: outputTexture$1
				},
				getComputeNodes: () => [fillBlurMap, ...blurNodes]
			};
		}
		const blur = createGaussianBlurCompute(childTexture, texWidth, texHeight, onCleanup);
		const { computeNodes, outputTexture, updateRadius } = blur;
		onResize(({ width, height }) => {
			blur.setInputDimensions(Math.round(width * dpr), Math.round(height * dpr));
		});
		return {
			outputs: {
				childTexture,
				blurredTexture: outputTexture
			},
			getComputeNodes: () => {
				const intensityEntry = uniforms.intensity;
				const intensity = intensityEntry.getCpuValue?.() ?? (typeof intensityEntry.uniform.value === "number" ? intensityEntry.uniform.value : null);
				if (intensity !== null) updateRadius(intensity * .36);
				return computeNodes;
			}
		};
	},
	fragmentNode: ({ uniforms, childNode, onCleanup, computeOutputs }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		if (computeOutputs?.blurredTexture) {
			const sharp = computeOutputs.childTexture.sample(screenUV);
			const blurred = texture(computeOutputs.blurredTexture).sample(screenUV);
			const intensityPin = uniforms.intensity.uniform.mul(float(0));
			return unpremultiplyAlpha(vec4(blurred.rgb.add(intensityPin), sharp.a));
		}
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const weights = [
			.056,
			.135,
			.278,
			.487,
			.726,
			.923,
			1,
			.923,
			.726,
			.487,
			.278,
			.135,
			.056
		];
		const weightSum = 6.214;
		const scaledIntensity = uniforms.intensity.uniform.mul(.06);
		const horizontalTexture = convertToTexture(Fn(() => {
			const uv$1 = screenUV;
			const blurRadius = scaledIntensity;
			const pixelSize = vec2(1).div(viewportSize);
			const total = vec4(0).toVar();
			for (let x = -6; x <= 6; x++) {
				const weight = float(weights[x + 6] / weightSum);
				const offset = vec2(float(x), float(0)).mul(blurRadius).mul(pixelSize);
				const sampleCoord = uv$1.add(offset);
				const sample = childTexture.sample(sampleCoord).mul(weight);
				total.assign(total.add(sample));
			}
			return total;
		})());
		onCleanup(() => {
			if (horizontalTexture?.renderTarget?.dispose) horizontalTexture.renderTarget.dispose();
		});
		return Fn(() => {
			const uv$1 = screenUV;
			const blurRadius = scaledIntensity;
			const pixelSize = vec2(1).div(viewportSize);
			const total = vec4(0).toVar();
			for (let y = -6; y <= 6; y++) {
				const weight = float(weights[y + 6] / weightSum);
				const offset = vec2(float(0), float(y)).mul(blurRadius).mul(pixelSize);
				const sampleCoord = uv$1.add(offset);
				const sample = horizontalTexture.sample(sampleCoord).mul(weight);
				total.assign(total.add(sample));
			}
			return unpremultiplyAlpha(total);
		})();
	}
};
var Blur_default = componentDefinition;
export { componentDefinition as n, Blur_default as t };
