// shaders/dist/core/BrickPattern-BfTsjp4I.js
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { cos, float, floor, fract, mix, mod, screenUV, sin, step, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "BrickPattern",
	category: "Textures",
	description: "Classic brick wall pattern with alternating rows and mortar gaps",
	props: {
		colorBrick: {
			default: "#000000",
			transform: transformColor,
			description: "Brick color",
			ui: {
				type: "color",
				label: "Brick Color",
				group: "Colors"
			}
		},
		colorMortar: {
			default: "#ffffff",
			transform: transformColor,
			description: "Mortar / gap color",
			ui: {
				type: "color",
				label: "Mortar Color",
				group: "Colors"
			}
		},
		cellsX: {
			default: 8,
			description: "Number of bricks per row",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Bricks per Row",
				group: "Effect"
			}
		},
		cellsY: {
			default: 10,
			description: "Number of brick rows",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Rows",
				group: "Effect"
			}
		},
		mortar: {
			default: .05,
			description: "Width of mortar gaps — equal pixel thickness in both directions",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Mortar Gap",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Rotation angle in degrees",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		speed: {
			default: 0,
			description: "Animation speed",
			ui: {
				type: "range",
				min: -2,
				max: 2,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		offset: {
			default: 0,
			description: "Static horizontal offset — shifts the brick pattern without animating",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Offset",
				group: "Animation"
			}
		},
		speedVariance: {
			default: 0,
			description: "How much each row's speed varies — at high values rows move at different speeds and directions",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Speed Variance",
				group: "Animation"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for per-row speed variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Animation"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const au = uv$1.x.mul(aspect);
		const av = uv$1.y;
		const cu = au.sub(aspect.mul(.5));
		const cv = av.sub(float(.5));
		const rotRad = uniforms.angle.uniform.mul(float(.0174533));
		const cosR = cos(rotRad);
		const sinR = sin(rotRad);
		const ru = cu.mul(cosR).add(cv.mul(sinR));
		const rv = cu.mul(sinR).negate().add(cv.mul(cosR));
		const u = ru.div(aspect).add(float(.5));
		const v = rv.add(float(.5));
		const cellsX = uniforms.cellsX.uniform;
		const cellsY = uniforms.cellsY.uniform;
		const animTime = createAnimatedTime(params, uniforms.speed);
		const mortarU = uniforms.mortar.uniform.mul(float(.3));
		const mortarV = mortarU.mul(aspect).mul(cellsY).div(cellsX);
		const row = floor(v.mul(cellsY));
		const rowOffset = mod(row, float(2)).mul(.5);
		const rowDir = mod(row, float(2)).mul(2).sub(1);
		const seedVal = uniforms.seed.uniform;
		const rowSpeedMult = mix(rowDir, fract(sin(row.add(seedVal.mul(.137))).mul(43758.5453)).sub(.5).mul(2), uniforms.speedVariance.uniform);
		const cellU = fract(u.mul(cellsX).add(rowOffset).add(animTime.mul(rowSpeedMult)).add(uniforms.offset.uniform));
		const cellV = fract(v.mul(cellsY));
		const brickU = step(mortarU, cellU).mul(step(mortarU, float(1).sub(cellU)));
		const brickV = step(mortarV, cellV).mul(step(mortarV, float(1).sub(cellV)));
		const brickMask = brickU.mul(brickV);
		const result = mixColors(uniforms.colorMortar.uniform, uniforms.colorBrick.uniform, brickMask, uniforms.colorSpace.uniform.value);
		return vec4(result.rgb, result.a);
	}
};
var BrickPattern_default = componentDefinition;
export { componentDefinition as n, BrickPattern_default as t };
