// shaders/dist/core/BrightnessContrast-GppIqtkO.js
import { vec3, vec4 } from "three/tsl";
const componentDefinition = {
	name: "BrightnessContrast",
	category: "Adjustments",
	description: "Adjust brightness and contrast of the image",
	requiresChild: true,
	props: {
		brightness: {
			default: 0,
			description: "Brightness adjustment (-1 to 1)",
			ui: {
				type: ["range", "map"],
				min: -1,
				max: 1,
				step: .05,
				label: "Brightness",
				group: "Appearance"
			}
		},
		contrast: {
			default: 0,
			description: "Contrast adjustment (-1 to 1)",
			ui: {
				type: ["range", "map"],
				min: -1,
				max: 1,
				step: .05,
				label: "Contrast",
				group: "Appearance"
			},
			transform: (value) => value + 1
		}
	},
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Brightness Contrast shader.");
			return vec4(0);
		}
		const brightness = uniforms.brightness.uniform;
		const contrast = uniforms.contrast.uniform;
		return vec4(vec3(childNode.r, childNode.g, childNode.b).sub(.5).mul(contrast).add(.5).add(brightness), childNode.a);
	}
};
var BrightnessContrast_default = componentDefinition;
export { componentDefinition as n, BrightnessContrast_default as t };
