// shaders/dist/core/Checkerboard-BVBGTvd-.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { float, floor, fract, min, mod, screenUV, smoothstep, vec2, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Checkerboard",
	category: "Textures",
	description: "Classic checkerboard pattern with two alternating colors",
	props: {
		colorA: {
			default: "#cccccc",
			transform: transformColor,
			description: "First color of the checkerboard pattern",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#999999",
			transform: transformColor,
			description: "Second color of the checkerboard pattern",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		cells: {
			default: 8,
			description: "Number of cells along the shortest canvas edge (creates square cells)",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 50,
				step: 1,
				label: "Cells",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Smoothness of the transition between colors (0 = hard edges, 1 = very soft)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Softness",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const correctedUV = vec2(uv$1.x.mul(aspect), uv$1.y.oneMinus());
		const cells = uniforms.cells.uniform;
		const softness = uniforms.softness.uniform;
		const gridUV = correctedUV.mul(cells);
		const gridCoords = floor(gridUV);
		const checkerValue = mod(gridCoords.x.add(gridCoords.y), 2);
		const cellUV = fract(gridUV);
		const distToEdge = min(min(cellUV.x, float(1).sub(cellUV.x)), min(cellUV.y, float(1).sub(cellUV.y)));
		const softRange = softness.mul(.5).add(1e-4);
		const edgeBlend = smoothstep(float(0), softRange, distToEdge);
		const blendFactor = checkerValue.mul(edgeBlend).add(float(.5).mul(float(1).sub(edgeBlend)));
		return mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, blendFactor, uniforms.colorSpace.uniform.value);
	}
};
var Checkerboard_default = componentDefinition;
export { componentDefinition as n, Checkerboard_default as t };
