// shaders/dist/core/Chevron-BKZjAXR_.js
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { cos, fract, fwidth, mix, screenUV, sin, smoothstep, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Chevron",
	category: "Textures",
	description: "Animated chevron / zigzag stripe pattern",
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "First color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Second color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		count: {
			default: 5,
			description: "Number of chevron pairs visible",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Count",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Rotation angle of the chevrons",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		balance: {
			default: .5,
			description: "Ratio of the two colors",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Balance",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Edge softness",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		speed: {
			default: 0,
			description: "Animation speed",
			ui: {
				type: "range",
				min: -2,
				max: 2,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		offset: {
			default: 0,
			description: "Phase offset for pattern positioning",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Offset",
				group: "Animation"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const au = uv$1.x.mul(aspect);
		const av = uv$1.y;
		const angleRad = uniforms.angle.uniform.mul(.0174533);
		const cosA = cos(angleRad);
		const sinA = sin(angleRad);
		const cu = au.sub(aspect.mul(.5));
		const cv = av.sub(.5);
		const rotU = cu.mul(cosA).add(cv.mul(sinA));
		const rotV = cu.mul(sinA).negate().add(cv.mul(cosA));
		const count = uniforms.count.uniform;
		const animTime = createAnimatedTime(params, uniforms.speed);
		const zigzag = fract(rotV.mul(count)).sub(.5).abs().mul(2);
		const stripeCoord = rotU.mul(count).add(zigzag.mul(.5)).add(animTime).add(uniforms.offset.uniform);
		const pattern = fract(stripeCoord);
		const fw = fwidth(stripeCoord);
		const balance = uniforms.balance.uniform;
		const soft = uniforms.softness.uniform;
		const mask = smoothstep(balance.sub(soft).sub(fw), balance.add(soft).add(fw), pattern);
		return vec4(mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, mask, uniforms.colorSpace.uniform.value).rgb, mix(uniforms.colorA.uniform.a, uniforms.colorB.uniform.a, mask));
	}
};
var Chevron_default = componentDefinition;
export { componentDefinition as n, Chevron_default as t };
