// shaders/dist/core/ChromaticAberration-vygkvgz-.js
import { r as transformAngle } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { convertToTexture, cos, radians, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "ChromaticAberration",
	category: "Stylize",
	description: "Separate RGB channels for a prismatic distortion effect",
	requiresRTT: true,
	requiresChild: true,
	props: {
		strength: {
			default: .2,
			description: "Overall strength of the chromatic aberration effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Strength",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Direction of the chromatic aberration in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		redOffset: {
			default: -1,
			description: "Red channel offset multiplier",
			ui: {
				type: ["range", "map"],
				min: -2,
				max: 2,
				step: .1,
				label: "Red Offset",
				group: "Effect"
			}
		},
		greenOffset: {
			default: 0,
			description: "Green channel offset multiplier",
			ui: {
				type: ["range", "map"],
				min: -2,
				max: 2,
				step: .1,
				label: "Green Offset",
				group: "Effect"
			}
		},
		blueOffset: {
			default: 1,
			description: "Blue channel offset multiplier",
			ui: {
				type: ["range", "map"],
				min: -2,
				max: 2,
				step: .1,
				label: "Blue Offset",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const angleRad = radians(uniforms.angle.uniform);
		const aspect = viewportSize.x.div(viewportSize.y);
		const direction = vec2(cos(angleRad).div(aspect), sin(angleRad));
		const scaledStrength = uniforms.strength.uniform.mul(.1);
		const redUV = screenUV.add(direction.mul(scaledStrength).mul(uniforms.redOffset.uniform));
		const greenUV = screenUV.add(direction.mul(scaledStrength).mul(uniforms.greenOffset.uniform));
		const blueUV = screenUV.add(direction.mul(scaledStrength).mul(uniforms.blueOffset.uniform));
		const redChannel = childTexture.sample(redUV).r;
		const greenChannel = childTexture.sample(greenUV).g;
		const blueChannel = childTexture.sample(blueUV).b;
		const alpha = childTexture.sample(screenUV).a;
		return unpremultiplyAlpha(vec4(redChannel, greenChannel, blueChannel, alpha));
	}
};
var ChromaticAberration_default = componentDefinition;
export { componentDefinition as n, ChromaticAberration_default as t };
