// shaders/dist/core/Circle-BycqVCMt.js
import { a as transformColor, l as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { float, length, screenUV, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
var transformStrokePosition = (value) => {
	const mode = {
		"outside": 0,
		"center": 1,
		"inside": 2
	}[value.toLowerCase()];
	if (mode === void 0) {
		console.warn(`Invalid stroke position: ${value}. Defaulting to 'outside'.`);
		return 0;
	}
	return mode;
};
const componentDefinition = {
	name: "Circle",
	category: "Shapes",
	description: "Generate a circle with adjustable size and softness",
	props: {
		color: {
			default: "#ffffff",
			transform: transformColor,
			description: "The color of the circle",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		radius: {
			default: 1,
			description: "The radius of the circle. A value of one (1) is sets the circle to fit the canvas.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .05,
				label: "Radius",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Edge softness. Lower values like zero (0) are sharp, higher values like one (1) are softer.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the circle",
			ui: {
				type: "position",
				label: "Center Position",
				group: "Position"
			}
		},
		strokeThickness: {
			default: 0,
			description: "The thickness of the stroke outline. Zero (0) means no stroke.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .5,
				step: .01,
				label: "Stroke Thickness",
				group: "Stroke"
			}
		},
		strokeColor: {
			default: "#000000",
			transform: transformColor,
			description: "The color of the stroke outline",
			ui: {
				type: "color",
				label: "Stroke Color",
				group: "Colors"
			}
		},
		strokePosition: {
			default: "center",
			transform: transformStrokePosition,
			description: "Position of the stroke relative to the circle edge",
			ui: {
				type: "select",
				options: [
					{
						label: "Outside",
						value: "outside"
					},
					{
						label: "Center",
						value: "center"
					},
					{
						label: "Inside",
						value: "inside"
					}
				],
				label: "Stroke Position",
				group: "Stroke"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for blending fill and stroke colors in soft edges",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Blending",
				group: "Colors"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(uv$1.x.mul(aspect), uv$1.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus());
		const distanceFromCenter = length(aspectCorrectedUV.sub(centerPos));
		const edgeSoftness = uniforms.softness.uniform;
		const circleEdge = uniforms.radius.uniform.mul(.5);
		const strokeThickness = uniforms.strokeThickness.uniform;
		const strokePositionMode = uniforms.strokePosition.uniform;
		const isCenter = strokePositionMode.greaterThanEqual(float(.5)).and(strokePositionMode.lessThan(float(1.5)));
		const isInside = strokePositionMode.greaterThanEqual(float(1.5));
		const halfThickness = strokeThickness.mul(.5);
		const strokeInner = isInside.select(circleEdge.sub(strokeThickness), isCenter.select(circleEdge.sub(halfThickness), circleEdge));
		const strokeOuter = isInside.select(circleEdge, isCenter.select(circleEdge.add(halfThickness), circleEdge.add(strokeThickness)));
		const overallMask = smoothstep(strokeOuter.sub(edgeSoftness), strokeOuter, distanceFromCenter).oneMinus();
		const strokeBlend = strokeThickness.greaterThan(float(0)).select(smoothstep(strokeInner.sub(edgeSoftness), strokeInner, distanceFromCenter), float(0));
		const blendedColor = mixColors(vec4(uniforms.color.uniform.rgb, uniforms.color.uniform.a), vec4(uniforms.strokeColor.uniform.rgb, uniforms.strokeColor.uniform.a), strokeBlend, uniforms.colorSpace.uniform.value);
		return vec4(blendedColor.xyz, blendedColor.w.mul(overallMask));
	}
};
var Circle_default = componentDefinition;
export { componentDefinition as n, Circle_default as t };
