// shaders/dist/core/ColorWheel-BQUKQPM-.js
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { clamp, cos, float, floor, fract, mix, screenUV, sin, step, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "ColorWheel",
	category: "Textures",
	description: "A directional gradient that smoothly cycles through rainbow colors or a custom set of three colors",
	props: {
		mode: {
			default: "rainbow",
			compileTime: true,
			description: "Rainbow cycles through the full spectrum; Custom loops through your three chosen colors",
			ui: {
				type: "select",
				options: [{
					label: "Rainbow",
					value: "rainbow"
				}, {
					label: "Custom",
					value: "custom"
				}],
				label: "Mode",
				group: "Colors"
			}
		},
		colorA: {
			default: "#ff0000",
			transform: transformColor,
			description: "First color in the cycle",
			ui: {
				type: "color",
				label: "Color 1",
				group: "Colors",
				condition: { mode: "custom" }
			}
		},
		colorB: {
			default: "#00ff88",
			transform: transformColor,
			description: "Second color in the cycle",
			ui: {
				type: "color",
				label: "Color 2",
				group: "Colors",
				condition: { mode: "custom" }
			}
		},
		colorC: {
			default: "#0066ff",
			transform: transformColor,
			description: "Third color in the cycle",
			ui: {
				type: "color",
				label: "Color 3",
				group: "Colors",
				condition: { mode: "custom" }
			}
		},
		scale: {
			default: 1,
			description: "Number of color cycles across the viewport",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 10,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Direction the gradient flows",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		speed: {
			default: .05,
			description: "Speed at which the gradient cycles",
			ui: {
				type: "range",
				min: -1,
				max: 1,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		colorSpace: {
			default: "oklch",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for blending between custom colors",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors",
				condition: { mode: "custom" }
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const angleRad = uniforms.angle.uniform.mul(.0174533);
		const cosA = cos(angleRad);
		const sinA = sin(angleRad);
		const au = uv$1.x.mul(aspect);
		const av = uv$1.y;
		const projectedCoord = au.mul(cosA).add(av.mul(sinA));
		const animTime = createAnimatedTime(params, uniforms.speed);
		const t = fract(projectedCoord.mul(uniforms.scale.uniform).mul(float(.2)).add(animTime));
		if (uniforms.mode.uniform.value === "rainbow") {
			const h6 = t.mul(float(6));
			return vec4(clamp(h6.sub(float(3)).abs().sub(float(1)), float(0), float(1)), clamp(float(2).sub(h6.sub(float(2)).abs()), float(0), float(1)), clamp(float(2).sub(h6.sub(float(4)).abs()), float(0), float(1)), float(1));
		} else {
			const t3 = t.mul(float(3));
			const si = fract(t3);
			const floorT3 = floor(t3);
			const c01 = mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, si, uniforms.colorSpace.uniform.value);
			const c12 = mixColors(uniforms.colorB.uniform, uniforms.colorC.uniform, si, uniforms.colorSpace.uniform.value);
			let result = mixColors(uniforms.colorC.uniform, uniforms.colorA.uniform, si, uniforms.colorSpace.uniform.value);
			result = mix(result, c12, step(floorT3, float(1.5)));
			result = mix(result, c01, step(floorT3, float(.5)));
			return vec4(result.rgb, result.a);
		}
	}
};
var ColorWheel_default = componentDefinition;
export { componentDefinition as n, ColorWheel_default as t };
