// shaders/dist/core/computeBlur-CprKe0xa.js
import { StorageTexture } from "three/webgpu";
import { FloatType } from "three";
import { Fn, If, float, instanceIndex, int, storageTexture, textureLoad, textureStore, uint, uniform, uniformArray, uvec2, vec4 } from "three/tsl";
var DEFAULT_HALF_KERNEL = 24;
const DEFAULT_COMPUTE_WIDTH = 1024;
const DEFAULT_COMPUTE_HEIGHT = 640;
function buildFixedWeights(halfKernel) {
	const sigma = halfKernel / 3;
	const size = halfKernel * 2 + 1;
	const weights = [];
	let sum = 0;
	for (let i = -halfKernel; i <= halfKernel; i++) {
		const w = Math.exp(-(i * i) / (2 * sigma * sigma));
		weights.push(w);
		sum += w;
	}
	for (let i = 0; i < size; i++) weights[i] /= sum;
	return weights;
}
var PASSTHROUGH_THRESHOLD = .5;
function createGaussianBlurCompute(inputTexture, inputWidth, inputHeight, onCleanup, halfKernel = DEFAULT_HALF_KERNEL, computeWidth = DEFAULT_COMPUTE_WIDTH, computeHeight = 640) {
	const KERNEL_SIZE = halfKernel * 2 + 1;
	const HALF_KERNEL = halfKernel;
	const intermediateTex = new StorageTexture(computeWidth, computeHeight);
	intermediateTex.type = FloatType;
	const outputTex = new StorageTexture(computeWidth, computeHeight);
	outputTex.type = FloatType;
	onCleanup(() => {
		intermediateTex.dispose();
		outputTex.dispose();
	});
	const intermediateWriteNode = storageTexture(intermediateTex);
	const intermediateReadNode = storageTexture(intermediateTex).toReadOnly();
	const outputNode = storageTexture(outputTex);
	const weightsH = uniformArray(new Array(KERNEL_SIZE).fill(0), "float");
	const weightsV = uniformArray(new Array(KERNEL_SIZE).fill(0), "float");
	let lastRadius = -1;
	const inputWidthU = uniform(inputWidth);
	const inputHeightU = uniform(inputHeight);
	let scaleY = inputHeight / computeHeight;
	const cw = uint(computeWidth);
	const ch = uint(computeHeight);
	const computeHm1 = int(computeHeight - 1);
	const inputWm1 = int(inputWidthU.sub(1));
	const inputHm1 = int(inputHeightU.sub(1));
	const horizontalPass = Fn(() => {
		const idx = instanceIndex;
		const cx = idx.mod(cw);
		const cy = idx.div(cw);
		If(cy.lessThan(ch), () => {
			const sum = vec4(0).toVar();
			const inputCx = cx.toFloat().add(float(.5)).mul(inputWidthU).div(float(computeWidth)).sub(float(.5));
			const inputCy = cy.toFloat().add(float(.5)).mul(inputHeightU).div(float(computeHeight)).sub(float(.5));
			const inputCxI = inputCx.round().toInt();
			const inputCyI = inputCy.round().toInt().clamp(int(0), inputHm1);
			for (let i = -HALF_KERNEL; i <= HALF_KERNEL; i++) {
				const texel = textureLoad(inputTexture, uvec2(inputCxI.add(int(i)).clamp(int(0), inputWm1).toUint(), inputCyI.toUint()), int(0));
				sum.assign(sum.add(texel.mul(weightsH.element(int(i + HALF_KERNEL)))));
			}
			textureStore(intermediateWriteNode, uvec2(cx, cy), sum).toWriteOnly();
		});
	})().compute(computeWidth * computeHeight, [256]);
	const verticalPass = Fn(() => {
		const idx = instanceIndex;
		const cx = idx.mod(cw);
		const cy = idx.div(cw);
		If(cy.lessThan(ch), () => {
			const sum = vec4(0).toVar();
			for (let i = -HALF_KERNEL; i <= HALF_KERNEL; i++) {
				const texel = textureLoad(intermediateReadNode, uvec2(cx, cy.toInt().add(int(i)).clamp(int(0), computeHm1).toUint()), int(0));
				sum.assign(sum.add(texel.mul(weightsV.element(int(i + HALF_KERNEL)))));
			}
			textureStore(outputNode, uvec2(cx, cy), sum).toWriteOnly();
		});
	})().compute(computeWidth * computeHeight, [256]);
	function updateWeights() {
		const pixelRadius = lastRadius < 0 ? 4 : lastRadius;
		const sigmaH = Math.max(pixelRadius * .5, .001);
		const sigmaV = Math.max(sigmaH / scaleY, .001);
		const newH = new Array(KERNEL_SIZE);
		const newV = new Array(KERNEL_SIZE);
		let totalH = 0, totalV = 0;
		for (let i = -HALF_KERNEL; i <= HALF_KERNEL; i++) {
			const h = Math.exp(-(i * i) / (2 * sigmaH * sigmaH));
			const v = Math.exp(-(i * i) / (2 * sigmaV * sigmaV));
			newH[i + HALF_KERNEL] = h;
			newV[i + HALF_KERNEL] = v;
			totalH += h;
			totalV += v;
		}
		for (let i = 0; i < KERNEL_SIZE; i++) {
			newH[i] /= totalH;
			newV[i] /= totalV;
		}
		weightsH.array = newH;
		weightsV.array = newV;
	}
	function updateRadius(pixelRadius) {
		if (Math.abs(pixelRadius - lastRadius) < .01) return;
		lastRadius = pixelRadius;
		updateWeights();
	}
	updateRadius(4);
	return {
		computeNodes: [horizontalPass, verticalPass],
		outputTexture: outputTex,
		updateRadius,
		computeWidth,
		computeHeight,
		setInputDimensions: (newW, newH) => {
			inputWidthU.value = newW;
			inputHeightU.value = newH;
			const newScaleY = newH / computeHeight;
			if (Math.abs(newScaleY - scaleY) > .001) {
				scaleY = newScaleY;
				updateWeights();
			}
		}
	};
}
function createVariableGaussianBlurCompute(inputTexture, inputWidth, inputHeight, onCleanup, halfKernel = DEFAULT_HALF_KERNEL, computeWidth = DEFAULT_COMPUTE_WIDTH, computeHeight = 640) {
	const HALF_KERNEL = halfKernel;
	const FIXED_WEIGHTS = buildFixedWeights(HALF_KERNEL);
	const blurMapTex = new StorageTexture(computeWidth, computeHeight);
	blurMapTex.type = FloatType;
	const blurMapWriteNode = storageTexture(blurMapTex);
	const blurMapReadNode = storageTexture(blurMapTex).toReadOnly();
	const intermediateTex = new StorageTexture(computeWidth, computeHeight);
	intermediateTex.type = FloatType;
	const outputTex = new StorageTexture(computeWidth, computeHeight);
	outputTex.type = FloatType;
	onCleanup(() => {
		blurMapTex.dispose();
		intermediateTex.dispose();
		outputTex.dispose();
	});
	const intermediateWriteNode = storageTexture(intermediateTex);
	const intermediateReadNode = storageTexture(intermediateTex).toReadOnly();
	const outputNode = storageTexture(outputTex);
	const weights = uniformArray(FIXED_WEIGHTS, "float");
	const inputWidthU = uniform(inputWidth);
	const inputHeightU = uniform(inputHeight);
	const scaleYU = uniform(inputHeight / computeHeight);
	const cw = uint(computeWidth);
	const ch = uint(computeHeight);
	const computeHm1 = int(computeHeight - 1);
	const inputWm1 = int(inputWidthU.sub(1));
	const inputHm1 = int(inputHeightU.sub(1));
	const invHalf = float(1 / HALF_KERNEL);
	const passthroughThreshold = float(PASSTHROUGH_THRESHOLD);
	return {
		computeNodes: [Fn(() => {
			const idx = instanceIndex;
			const cx = idx.mod(cw);
			const cy = idx.div(cw);
			If(cy.lessThan(ch), () => {
				const blurRadius = textureLoad(blurMapReadNode, uvec2(cx, cy), int(0)).r;
				const inputCx = cx.toFloat().add(float(.5)).mul(inputWidthU).div(float(computeWidth)).sub(float(.5));
				const inputCy = cy.toFloat().add(float(.5)).mul(inputHeightU).div(float(computeHeight)).sub(float(.5));
				const inputCxF = inputCx;
				const inputCyI = inputCy.round().toInt().clamp(int(0), inputHm1);
				If(blurRadius.lessThan(passthroughThreshold), () => {
					const texel = textureLoad(inputTexture, uvec2(inputCxF.round().toInt().clamp(int(0), inputWm1).toUint(), inputCyI.toUint()), int(0));
					textureStore(intermediateWriteNode, uvec2(cx, cy), texel).toWriteOnly();
				}).Else(() => {
					const sum = vec4(0).toVar();
					for (let i = -HALF_KERNEL; i <= HALF_KERNEL; i++) {
						const offset = float(i).mul(invHalf).mul(blurRadius);
						const texel = textureLoad(inputTexture, uvec2(inputCxF.add(offset).round().toInt().clamp(int(0), inputWm1).toUint(), inputCyI.toUint()), int(0));
						sum.assign(sum.add(texel.mul(weights.element(int(i + HALF_KERNEL)))));
					}
					textureStore(intermediateWriteNode, uvec2(cx, cy), sum).toWriteOnly();
				});
			});
		})().compute(computeWidth * computeHeight, [256]), Fn(() => {
			const idx = instanceIndex;
			const cx = idx.mod(cw);
			const cy = idx.div(cw);
			If(cy.lessThan(ch), () => {
				const blurRadius = textureLoad(blurMapReadNode, uvec2(cx, cy), int(0)).r;
				If(blurRadius.lessThan(passthroughThreshold), () => {
					const texel = textureLoad(intermediateReadNode, uvec2(cx, cy), int(0));
					textureStore(outputNode, uvec2(cx, cy), texel).toWriteOnly();
				}).Else(() => {
					const sum = vec4(0).toVar();
					const yi = cy.toInt();
					const radiusComputeY = blurRadius.div(scaleYU);
					for (let i = -HALF_KERNEL; i <= HALF_KERNEL; i++) {
						const offset = float(i).mul(invHalf).mul(radiusComputeY);
						const texel = textureLoad(intermediateReadNode, uvec2(cx, yi.add(offset.round().toInt()).clamp(int(0), computeHm1).toUint()), int(0));
						sum.assign(sum.add(texel.mul(weights.element(int(i + HALF_KERNEL)))));
					}
					textureStore(outputNode, uvec2(cx, cy), sum).toWriteOnly();
				});
			});
		})().compute(computeWidth * computeHeight, [256])],
		outputTexture: outputTex,
		blurMapTexture: blurMapTex,
		blurMapWriteNode,
		computeWidth,
		computeHeight,
		setInputDimensions: (newW, newH) => {
			inputWidthU.value = newW;
			inputHeightU.value = newH;
			scaleYU.value = newH / computeHeight;
		}
	};
}
export { createVariableGaussianBlurCompute as i, DEFAULT_COMPUTE_WIDTH as n, createGaussianBlurCompute as r, DEFAULT_COMPUTE_HEIGHT as t };
