// shaders/dist/core/ConcentricSpin-CO3Kx1I1.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { Fn, convertToTexture, cos, float, floor, fract, mix, mod, screenUV, sin, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "ConcentricSpin",
	category: "Distortions",
	description: "Concentric rings that each rotate the underlying image by different amounts",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 20,
			description: "Maximum rotation angle per ring",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Effect"
			}
		},
		rings: {
			default: 8,
			description: "Number of concentric rings",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Rings",
				group: "Effect"
			}
		},
		smoothness: {
			default: .03,
			description: "Softness of transitions between rings",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Smoothness",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Randomization seed for per-ring rotation variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		speed: {
			default: .1,
			description: "Speed of continuous ring rotation",
			ui: {
				type: "range",
				min: -5,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		speedRandomness: {
			default: .5,
			description: "How much each ring varies in rotation speed and direction",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Speed Randomness",
				group: "Animation"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the concentric rings",
			ui: {
				type: "position",
				label: "Center Position",
				group: "Position"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms, childNode, onCleanup } = params;
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const center = vec2(uniforms.center.uniform.x, uniforms.center.uniform.y.oneMinus());
		const maxAngle = uniforms.intensity.uniform.mul(.01745);
		const rings = uniforms.rings.uniform;
		const edgeSoftness = uniforms.smoothness.uniform;
		const seed = uniforms.seed.uniform;
		const speedRandomness = uniforms.speedRandomness.uniform;
		const aspect = viewportSize.x.div(viewportSize.y);
		const animTime = createAnimatedTime(params, uniforms.speed);
		const hash = (x) => fract(sin(x.mul(12.9898)).mul(43758.5453)).sub(.5).mul(2);
		return Fn(() => {
			const uv$1 = screenUV;
			const dx = uv$1.x.sub(center.x).mul(aspect);
			const dy = uv$1.y.sub(center.y);
			const ringCoord = vec2(dx, dy).length().mul(rings);
			const ringIndex = floor(ringCoord);
			const ringFrac = fract(ringCoord);
			const seedOffset = seed.mul(.137);
			const staticA = hash(ringIndex.add(seedOffset)).mul(maxAngle);
			const staticB = hash(ringIndex.add(1).add(seedOffset)).mul(maxAngle);
			const ringSpeedA = mix(float(1), hash(ringIndex.add(42.7)), speedRandomness);
			const ringSpeedB = mix(float(1), hash(ringIndex.add(1).add(42.7)), speedRandomness);
			const animA = animTime.mul(.25).mul(ringSpeedA);
			const animB = animTime.mul(.25).mul(ringSpeedB);
			const totalA = staticA.add(animA);
			const totalB = staticB.add(animB);
			const TWO_PI = float(Math.PI * 2);
			const diff = mod(totalB.sub(totalA).add(Math.PI), TWO_PI).sub(Math.PI);
			const halfEdge = edgeSoftness.mul(.5).add(.001);
			const blend = smoothstep(float(.5).sub(halfEdge), float(.5).add(halfEdge), ringFrac);
			const angle = totalA.add(diff.mul(blend));
			const cosA = cos(angle);
			const sinA = sin(angle);
			const rotX = dx.mul(cosA).sub(dy.mul(sinA));
			const rotY = dx.mul(sinA).add(dy.mul(cosA));
			const distortedUV = vec2(rotX.div(aspect).add(center.x), rotY.add(center.y));
			return unpremultiplyAlpha(applyEdgeHandling(distortedUV, childTexture.sample(distortedUV), childTexture, uniforms.edges.uniform.value));
		})();
	}
};
var ConcentricSpin_default = componentDefinition;
export { componentDefinition as n, ConcentricSpin_default as t };
