// shaders/dist/core/ConicGradient-DRvVwiSj.js
import { a as transformColor, l as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { atan, float, fract, screenUV, vec2, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "ConicGradient",
	category: "Textures",
	description: "Colors sweep in a full circle around a center point, like a color wheel",
	props: {
		colorA: {
			default: "#FF0080",
			transform: transformColor,
			description: "Starting color of the sweep",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#00BFFF",
			transform: transformColor,
			description: "Ending color of the sweep (wraps back to Color A)",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the sweep",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation offset in degrees — shifts where Color A begins",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		repeat: {
			default: 1,
			description: "Number of times the gradient repeats around the circle. Values above 1 create a starburst pattern.",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 24,
				step: 1,
				label: "Repeat",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "oklch",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectUV = vec2(uv$1.x.mul(aspect), uv$1.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus());
		const dx = aspectUV.x.sub(centerPos.x);
		const normalized = atan(aspectUV.y.sub(centerPos.y), dx).negate().div(float(2 * Math.PI)).add(float(.5));
		const rotationFraction = uniforms.rotation.uniform.div(float(360));
		const t = fract(normalized.add(rotationFraction).mul(uniforms.repeat.uniform));
		return mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, t, uniforms.colorSpace.uniform.value);
	}
};
var ConicGradient_default = componentDefinition;
export { componentDefinition as n, ConicGradient_default as t };
