// shaders/dist/core/ContourLines-s4waN3uU.js
import { a as transformColor } from "./transformations-CJcUeZIC.js";
import { dot, float, fract, fwidth, mix, oneMinus, pow, smoothstep, vec3, vec4 } from "three/tsl";
var transformSource = (value) => {
	return value === "alpha" ? 1 : 0;
};
var transformColorMode = (value) => {
	return value === "custom" ? 1 : 0;
};
var transformBoolean = (value) => {
	return value ? 1 : 0;
};
const componentDefinition = {
	name: "ContourLines",
	category: "Stylize",
	description: "Draw topographical contour lines based on luminance or alpha",
	requiresChild: true,
	props: {
		levels: {
			default: 5,
			description: "Number of contour levels",
			ui: {
				type: ["range", "map"],
				min: 2,
				max: 30,
				step: 1,
				label: "Levels",
				group: "Effect"
			}
		},
		lineWidth: {
			default: 2,
			description: "Width of the contour lines in pixels",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 5,
				step: .1,
				label: "Line Width",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Edge softness of the lines (0 = sharp, 1 = soft)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .05,
				label: "Softness",
				group: "Effect"
			}
		},
		gamma: {
			default: .5,
			description: "Contour distribution. <1 clusters in bright, >1 clusters in dark",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 2,
				step: .05,
				label: "Gamma",
				group: "Effect"
			}
		},
		invert: {
			default: false,
			description: "Invert the source values",
			transform: transformBoolean,
			ui: {
				type: "checkbox",
				label: "Invert",
				group: "Effect"
			}
		},
		source: {
			default: "luminance",
			description: "Use luminance or alpha channel for contours",
			transform: transformSource,
			compileTime: true,
			ui: {
				type: "select",
				options: [{
					label: "Luminance",
					value: "luminance"
				}, {
					label: "Alpha",
					value: "alpha"
				}],
				label: "Source",
				group: "Effect"
			}
		},
		colorMode: {
			default: "source",
			description: "Use source image colors or custom colors",
			transform: transformColorMode,
			compileTime: true,
			ui: {
				type: "select",
				options: [{
					label: "Source",
					value: "source"
				}, {
					label: "Custom",
					value: "custom"
				}],
				label: "Color Mode",
				group: "Colors"
			}
		},
		lineColor: {
			default: "#000000",
			description: "Color of the contour lines (custom mode)",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Line Color",
				group: "Colors",
				condition: { colorMode: "custom" }
			}
		},
		backgroundColor: {
			default: "transparent",
			description: "Background color (custom mode)",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Background Color",
				group: "Colors",
				condition: { colorMode: "custom" }
			}
		}
	},
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const levels = uniforms.levels.uniform;
		const lineWidth = uniforms.lineWidth.uniform;
		const softness = uniforms.softness.uniform;
		const gamma = uniforms.gamma.uniform;
		const isInverted = uniforms.invert.uniform.equal(float(1));
		const lineColorVec = uniforms.lineColor.uniform;
		const customLineColor = vec3(lineColorVec.x, lineColorVec.y, lineColorVec.z);
		const customLineAlpha = lineColorVec.w;
		const bgColorVec = uniforms.backgroundColor.uniform;
		const customBgColor = vec3(bgColorVec.x, bgColorVec.y, bgColorVec.z);
		const customBgAlpha = bgColorVec.w;
		const originalColor = childNode;
		const rawSourceValue = uniforms.source.uniform.value === 1 ? originalColor.a : dot(originalColor.rgb, vec3(.2126, .7152, .0722));
		const scaledValue = pow(isInverted.select(oneMinus(rawSourceValue), rawSourceValue), gamma).mul(levels).add(.5);
		const fw = fwidth(scaledValue).max(float(1e-4));
		const normalizedDist = fract(scaledValue.add(.5)).sub(.5).abs().div(fw);
		const transitionFactor = softness.mul(.99).add(.01);
		const lineMask = smoothstep(lineWidth.mul(oneMinus(transitionFactor)), lineWidth, normalizedDist).oneMinus();
		const colorMode = uniforms.colorMode.uniform.value;
		const lineColor = colorMode === 1 ? customLineColor : originalColor.rgb;
		const lineAlpha = colorMode === 1 ? customLineAlpha : float(1);
		const bgColor = colorMode === 1 ? customBgColor : vec3(0, 0, 0);
		const bgAlpha = colorMode === 1 ? customBgAlpha : float(0);
		return vec4(mix(bgColor, lineColor, lineMask), mix(bgAlpha, lineAlpha, lineMask));
	}
};
var ContourLines_default = componentDefinition;
export { componentDefinition as n, ContourLines_default as t };
