// shaders/dist/core/CRTScreen-corekL_R.js
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { convertToTexture, float, fract, length, mix, screenUV, sin, smoothstep, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "CRTScreen",
	category: "Stylize",
	description: "Retro CRT monitor simulation with scanlines",
	requiresRTT: true,
	requiresChild: true,
	props: {
		pixelSize: {
			default: 128,
			description: "Size of individual TV pixels (lower = more pixels)",
			ui: {
				type: ["range", "map"],
				min: 8,
				max: 128,
				step: 1,
				label: "Pixel Size",
				group: "Effect"
			}
		},
		colorShift: {
			default: 1,
			description: "Chromatic aberration amount",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 10,
				step: .1,
				label: "Color Shift",
				group: "Effect"
			}
		},
		scanlineIntensity: {
			default: .3,
			description: "Strength of horizontal scanlines",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Scanline Intensity",
				group: "Effect"
			}
		},
		scanlineFrequency: {
			default: 200,
			description: "Number of scanlines across screen",
			ui: {
				type: ["range", "map"],
				min: 100,
				max: 800,
				step: 10,
				label: "Scanline Frequency",
				group: "Effect"
			}
		},
		brightness: {
			default: 1,
			description: "Screen brightness boost",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 2,
				step: .1,
				label: "Brightness",
				group: "Appearance"
			}
		},
		contrast: {
			default: 1,
			description: "Screen contrast enhancement",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 2,
				step: .1,
				label: "Contrast",
				group: "Appearance"
			}
		},
		vignetteIntensity: {
			default: 1,
			description: "Strength of corner darkening effect (0 = off)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Vignette Intensity",
				group: "Effect"
			}
		},
		vignetteRadius: {
			default: .5,
			description: "How far the vignette extends inward (0 = edges only, 1 = reaches center)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Vignette Radius",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const pixelSize = uniforms.pixelSize.uniform;
		const colorShift = uniforms.colorShift.uniform;
		const scanlineIntensity = uniforms.scanlineIntensity.uniform;
		const scanlineFrequency = uniforms.scanlineFrequency.uniform;
		const brightness = uniforms.brightness.uniform;
		const contrast = uniforms.contrast.uniform;
		const scaledColorShift = colorShift.mul(float(.002));
		const redUV = screenUV.add(vec2(scaledColorShift, float(0)));
		const greenUV = screenUV;
		const blueUV = screenUV.sub(vec2(scaledColorShift, float(0)));
		const redSample = childTexture.sample(redUV).r;
		const greenSample = childTexture.sample(greenUV).g;
		const blueSample = childTexture.sample(blueUV).b;
		const adjustedColor = vec3(redSample, greenSample, blueSample).sub(float(.5)).mul(contrast).add(float(.5)).mul(brightness);
		const scanlineWave = sin(screenUV.y.mul(scanlineFrequency).mul(float(3.14159 * 2)));
		const scanlineEffect = float(1).sub(scanlineIntensity.mul(scanlineWave.mul(float(.5)).add(float(.5))));
		const scanlinedColor = adjustedColor.mul(scanlineEffect);
		const phosphorScale = pixelSize.mul(float(.5));
		const pixelPhase = fract(screenUV.mul(phosphorScale));
		const redMask = sin(pixelPhase.x.mul(float(6.28318))).mul(float(.1)).add(float(.95));
		const greenMask = sin(pixelPhase.x.mul(float(6.28318)).add(float(2.09))).mul(float(.1)).add(float(.95));
		const blueMask = sin(pixelPhase.x.mul(float(6.28318)).add(float(4.18))).mul(float(.1)).add(float(.95));
		const phosphorColor = vec3(scanlinedColor.r.mul(redMask), scanlinedColor.g.mul(greenMask), scanlinedColor.b.mul(blueMask));
		const aspect = viewportSize.x.div(viewportSize.y);
		const centeredUV = screenUV.sub(vec2(.5, .5));
		const vignetteDistance = length(vec2(centeredUV.x.mul(aspect), centeredUV.y));
		const vignetteRadiusProp = uniforms.vignetteRadius.uniform;
		const vignetteIntensity = uniforms.vignetteIntensity.uniform;
		const innerEdge = float(1).sub(vignetteRadiusProp.mul(float(.9)));
		const vignetteValue = smoothstep(innerEdge, innerEdge.add(float(.5)), vignetteDistance).oneMinus();
		const vignette = mix(float(1), vignetteValue, vignetteIntensity);
		return unpremultiplyAlpha(vec4(phosphorColor.mul(vignette), float(1)));
	}
};
var CRTScreen_default = componentDefinition;
export { componentDefinition as n, CRTScreen_default as t };
