// shaders/dist/core/CursorTrail-CshoSM5Y.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { StorageTexture } from "three/webgpu";
import { DataTexture, FloatType, LinearFilter as LinearFilter$1, RGBAFormat } from "three";
import { Fn, If, attributeArray, float, instanceIndex, screenUV, step, storageTexture, texture, textureStore, uint, uniform, uvec2, vec2, vec4 } from "three/tsl";
var GRID_SIZE = 128;
const componentDefinition = {
	name: "CursorTrail",
	category: "Interactive",
	description: "Animated trail effect that tracks cursor movement",
	usesPointer: true,
	props: {
		colorA: {
			default: "#00aaff",
			transform: transformColor,
			description: "Color of fresh trails",
			ui: {
				type: "color",
				label: "Start Color",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ff00aa",
			transform: transformColor,
			description: "Color trails transition to as they fade",
			ui: {
				type: "color",
				label: "End Color",
				group: "Colors"
			}
		},
		radius: {
			default: .5,
			description: "Base radius of trail circles",
			ui: {
				type: "range",
				min: .5,
				max: 2,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		length: {
			default: .5,
			description: "How long trail circles persist (in seconds)",
			ui: {
				type: "range",
				min: .1,
				max: 2,
				step: .1,
				label: "Trail Length",
				group: "Animation"
			}
		},
		shrink: {
			default: 1,
			description: "How much circles shrink as they fade out (0 = no shrink, 1 = full shrink)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .1,
				label: "Shrink Amount",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	computeNode: ({ uniforms, onCleanup, renderer }) => {
		const trailBuffer = attributeArray(GRID_SIZE * GRID_SIZE, "vec4");
		const trailTex = new StorageTexture(GRID_SIZE, GRID_SIZE);
		trailTex.type = FloatType;
		const trailTexNode = storageTexture(trailTex);
		onCleanup(() => {
			trailTex.dispose();
		});
		const fadeRateU = uniform(.97);
		const ageRateU = uniform(.03);
		const shrinkU = uniform(1);
		const cursorXU = uniform(.5);
		const cursorYU = uniform(.5);
		const radiusU = uniform(.05);
		const amountU = uniform(0);
		const aspectU = uniform(1);
		const mouseActiveU = uniform(0);
		const gs = uint(GRID_SIZE);
		const decayPass = Fn(() => {
			const idx = instanceIndex;
			const cell = trailBuffer.element(idx);
			const intensity = cell.x.mul(fadeRateU);
			const size = float(1).sub(float(1).sub(intensity).mul(shrinkU));
			const age = cell.z.add(ageRateU).min(float(1));
			trailBuffer.element(idx).assign(vec4(intensity, size, age, float(0)));
		})().compute(GRID_SIZE * GRID_SIZE, [256]);
		const stampPass = Fn(() => {
			const idx = instanceIndex;
			const j = idx.mod(gs);
			const i = idx.div(gs);
			If(mouseActiveU.greaterThan(float(.5)), () => {
				const cellX = j.toFloat().add(float(.5)).div(float(GRID_SIZE));
				const cellY = i.toFloat().add(float(.5)).div(float(GRID_SIZE));
				const ddx = aspectU.greaterThanEqual(float(1)).select(cellX.sub(cursorXU).mul(aspectU), cellX.sub(cursorXU));
				const ddy = aspectU.greaterThanEqual(float(1)).select(cellY.sub(cursorYU), cellY.sub(cursorYU).div(aspectU));
				const distSq = ddx.mul(ddx).add(ddy.mul(ddy));
				const influenceRadius = radiusU.mul(float(3));
				If(distSq.lessThan(influenceRadius.mul(influenceRadius)), () => {
					const influence = distSq.negate().div(radiusU.mul(radiusU)).exp();
					const newIntensity = trailBuffer.element(idx).x.add(influence.mul(amountU)).min(float(1));
					trailBuffer.element(idx).assign(vec4(newIntensity, float(1), float(0), float(0)));
				});
			});
		})().compute(GRID_SIZE * GRID_SIZE, [256]);
		const outputPass = Fn(() => {
			const idx = instanceIndex;
			const j = idx.mod(gs);
			const i = idx.div(gs);
			const cell = trailBuffer.element(idx);
			textureStore(trailTexNode, uvec2(j, i), cell).toWriteOnly();
		})().compute(GRID_SIZE * GRID_SIZE, [256]);
		let prevX = .5;
		let prevY = .5;
		let lastTime = Date.now();
		let lastActiveTime = Date.now();
		return {
			outputs: { trailTexture: trailTex },
			getComputeNodes: (params) => {
				const currentTime = Date.now();
				const dt = Math.min((currentTime - lastTime) / 1e3, .016);
				lastTime = currentTime;
				const aspect = Math.max(1, params.dimensions.width) / Math.max(1, params.dimensions.height);
				const radius = uniforms.radius.uniform.value * .1;
				const length$1 = uniforms.length.uniform.value;
				const shrink = uniforms.shrink.uniform.value;
				const dx = params.pointer.x - prevX;
				const dy = params.pointer.y - prevY;
				const speed = Math.sqrt(dx * dx + dy * dy);
				const oldX = prevX;
				const oldY = prevY;
				prevX = params.pointer.x;
				prevY = params.pointer.y;
				if (speed > .001) lastActiveTime = currentTime;
				const settleMs = Math.min(3e4, Math.max(3e3, length$1 * 5e3));
				if (currentTime - lastActiveTime > settleMs) return null;
				fadeRateU.value = 1 - dt / Math.max(.1, length$1);
				ageRateU.value = dt / Math.max(.1, length$1);
				shrinkU.value = shrink;
				radiusU.value = radius;
				aspectU.value = aspect;
				renderer.compute(decayPass);
				if (speed > .001) {
					const stepSize = Math.max(.005, radius * .5);
					const numSteps = Math.min(20, Math.max(1, Math.ceil(speed / stepSize)));
					const amountPerStep = speed / numSteps * 50 * dt;
					mouseActiveU.value = 1;
					amountU.value = amountPerStep;
					for (let s = 0; s < numSteps; s++) {
						const t = (s + .5) / numSteps;
						cursorXU.value = oldX + dx * t;
						cursorYU.value = oldY + dy * t;
						renderer.compute(stampPass);
					}
				}
				return [outputPass];
			}
		};
	},
	fragmentNode: ({ uniforms, onBeforeRender, onCleanup, computeOutputs }) => {
		let trailField;
		if (computeOutputs?.trailTexture) trailField = texture(computeOutputs.trailTexture);
		else {
			const trailData = new Float32Array(GRID_SIZE * GRID_SIZE * 4);
			const trailTexture = new DataTexture(trailData, GRID_SIZE, GRID_SIZE, RGBAFormat, FloatType);
			trailTexture.magFilter = LinearFilter$1;
			trailTexture.minFilter = LinearFilter$1;
			trailTexture.needsUpdate = true;
			trailField = texture(trailTexture);
			let prevX = .5;
			let prevY = .5;
			let lastTime = Date.now();
			onBeforeRender(({ pointer, dimensions }) => {
				const currentTime = Date.now();
				const dt = Math.min((currentTime - lastTime) / 1e3, .016);
				lastTime = currentTime;
				const aspect = Math.max(1, dimensions.width) / Math.max(1, dimensions.height);
				const radius = uniforms.radius.uniform.value * .1;
				const length$1 = uniforms.length.uniform.value;
				const shrink = uniforms.shrink.uniform.value;
				const dx = pointer.x - prevX;
				const dy = pointer.y - prevY;
				const speed = Math.sqrt(dx * dx + dy * dy);
				const fadeRate = 1 - dt / Math.max(.1, length$1);
				const ageRate = dt / Math.max(.1, length$1);
				for (let i = 0; i < GRID_SIZE * GRID_SIZE * 4; i += 4) {
					trailData[i] *= fadeRate;
					trailData[i + 1] = 1 - (1 - trailData[i]) * shrink;
					trailData[i + 2] = Math.min(1, trailData[i + 2] + ageRate);
				}
				if (speed > .001) {
					const influenceRadius = radius * 3;
					const stepSize = Math.max(.005, radius * .5);
					const numSteps = Math.min(20, Math.max(1, Math.ceil(speed / stepSize)));
					const amountPerStep = speed / numSteps * 50 * dt;
					for (let s = 0; s < numSteps; s++) {
						const t = (s + .5) / numSteps;
						const px = prevX + dx * t;
						const py = prevY + dy * t;
						const minJ = Math.max(0, Math.floor((px - influenceRadius) * GRID_SIZE));
						const maxJ = Math.min(GRID_SIZE - 1, Math.ceil((px + influenceRadius) * GRID_SIZE));
						const minI = Math.max(0, Math.floor((py - influenceRadius) * GRID_SIZE));
						const maxI = Math.min(GRID_SIZE - 1, Math.ceil((py + influenceRadius) * GRID_SIZE));
						for (let i = minI; i <= maxI; i++) for (let j = minJ; j <= maxJ; j++) {
							const idx = (i * GRID_SIZE + j) * 4;
							const cellX = (j + .5) / GRID_SIZE;
							const cellY = (i + .5) / GRID_SIZE;
							const cellDx = aspect >= 1 ? (cellX - px) * aspect : cellX - px;
							const cellDy = aspect >= 1 ? cellY - py : (cellY - py) / aspect;
							const dist = Math.sqrt(cellDx * cellDx + cellDy * cellDy);
							if (dist < influenceRadius) {
								const influence = Math.exp(-dist * dist / (radius * radius));
								trailData[idx] = Math.min(1, trailData[idx] + influence * amountPerStep);
								trailData[idx + 1] = 1;
								trailData[idx + 2] = 0;
							}
						}
					}
				}
				trailTexture.needsUpdate = true;
				prevX = pointer.x;
				prevY = pointer.y;
			});
			onCleanup(() => {
				trailTexture.dispose();
			});
		}
		const pixelSize = float(1 / GRID_SIZE);
		const centerSample = trailField.sample(screenUV);
		const sample1 = trailField.sample(screenUV.add(vec2(pixelSize, float(0))));
		const sample2 = trailField.sample(screenUV.add(vec2(float(0), pixelSize)));
		const sample3 = trailField.sample(screenUV.add(vec2(pixelSize.negate(), float(0))));
		const sample4 = trailField.sample(screenUV.add(vec2(float(0), pixelSize.negate())));
		const trailIntensity = centerSample.x.mul(float(.5)).add(sample1.x.mul(float(.125))).add(sample2.x.mul(float(.125))).add(sample3.x.mul(float(.125))).add(sample4.x.mul(float(.125)));
		const trailAge = centerSample.z.mul(float(.5)).add(sample1.z.mul(float(.125))).add(sample2.z.mul(float(.125))).add(sample3.z.mul(float(.125))).add(sample4.z.mul(float(.125)));
		const circleMask = step(float(.01), trailIntensity);
		const colorA = uniforms.colorA.uniform;
		const colorB = uniforms.colorB.uniform;
		const trailColor = mixColors(colorA, colorB, trailAge, uniforms.colorSpace.uniform.value);
		return vec4(trailColor.rgb, trailColor.a.mul(circleMask));
	}
};
var CursorTrail_default = componentDefinition;
export { componentDefinition as n, CursorTrail_default as t };
