// shaders/dist/core/DiamondGradient-D6eyjZOn.js
import { a as transformColor, l as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { abs, cos, float, fract, max, mix, screenUV, sin, vec2, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "DiamondGradient",
	category: "Textures",
	description: "Diamond-shaped gradient radiating from a center point using Manhattan distance",
	props: {
		colorA: {
			default: "#4ffb4a",
			transform: transformColor,
			description: "Color at the center of the diamond",
			ui: {
				type: "color",
				label: "Color A (Center)",
				group: "Colors"
			}
		},
		colorB: {
			default: "#4f1238",
			transform: transformColor,
			description: "Color at the outer edges of the diamond",
			ui: {
				type: "color",
				label: "Color B (Edge)",
				group: "Colors"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the diamond",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		size: {
			default: .7,
			description: "Extent of the gradient — controls how far Color A reaches before transitioning to Color B",
			ui: {
				type: ["range", "map"],
				min: .01,
				max: 2,
				step: .01,
				label: "Size",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation in degrees — tilts the diamond into a rhombus",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		repeat: {
			default: 1,
			description: "Number of times the gradient repeats outward. Values above 1 create concentric diamond or square bands.",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 16,
				step: .5,
				label: "Repeat",
				group: "Effect"
			}
		},
		roundness: {
			default: 0,
			description: "Morphs from a sharp diamond (0) to a square (1)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Roundness",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "oklch",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectUV = vec2(uv$1.x.mul(aspect), uv$1.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus());
		const dx = aspectUV.x.sub(centerPos.x);
		const dy = aspectUV.y.sub(centerPos.y);
		const rotRad = uniforms.rotation.uniform.mul(float(Math.PI / 180));
		const cosR = cos(rotRad);
		const sinR = sin(rotRad);
		const rdx = dx.mul(cosR).add(dy.mul(sinR));
		const rdy = dy.mul(cosR).sub(dx.mul(sinR));
		const absDx = abs(rdx);
		const absDy = abs(rdy);
		const t = mix(absDx.add(absDy), max(absDx, absDy), uniforms.roundness.uniform).div(uniforms.size.uniform).mul(uniforms.repeat.uniform);
		const finalT = uniforms.repeat.uniform.greaterThan(float(1.0001)).select(fract(t), t.clamp(float(0), float(1)));
		return mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, finalT, uniforms.colorSpace.uniform.value);
	}
};
var DiamondGradient_default = componentDefinition;
export { componentDefinition as n, DiamondGradient_default as t };
