// shaders/dist/core/DiffuseBlur-BjvXHmJx.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { convertToTexture, dot, fract, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "DiffuseBlur",
	category: "Blurs",
	description: "Grain-like pixel displacement at random",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 30,
			description: "Intensity of the diffuse blur effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const uv$1 = screenUV;
		const sinVal = sin(dot(uv$1.mul(1e3), vec2(25.9796, 156.466)));
		const offset = fract(vec2(sinVal.mul(43758.5453), sinVal.mul(28461.7231))).mul(2).sub(1).mul(uniforms.intensity.uniform).div(viewportSize);
		const displacedUV = uv$1.add(offset);
		return unpremultiplyAlpha(applyEdgeHandling(displacedUV, childTexture.sample(displacedUV), childTexture, uniforms.edges.uniform.value));
	}
};
var DiffuseBlur_default = componentDefinition;
export { componentDefinition as n, DiffuseBlur_default as t };
