// shaders/dist/core/Dither-z6s-TFg7.js
import { a as transformColor } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { DataTexture, FloatType, NearestFilter, RGBAFormat, RepeatWrapping } from "three";
import { convertToTexture, dot, float, floor, fract, mix, screenUV, sin, step, texture, vec2, vec3, vec4, viewportSize } from "three/tsl";
var transformPattern = (value) => {
	return {
		"bayer2": 0,
		"bayer4": 1,
		"bayer8": 2,
		"clusteredDot": 3,
		"blueNoise": 4,
		"whiteNoise": 5
	}[value] ?? 1;
};
var transformColorMode = (value) => {
	return value === "source" ? 1 : 0;
};
const componentDefinition = {
	name: "Dither",
	category: "Stylize",
	description: "Dithering effect with multiple pattern options",
	requiresRTT: true,
	requiresChild: true,
	props: {
		pattern: {
			default: "bayer4",
			transform: transformPattern,
			compileTime: true,
			description: "Dithering pattern algorithm",
			ui: {
				type: "select",
				options: [
					{
						label: "Bayer 2x2",
						value: "bayer2"
					},
					{
						label: "Bayer 4x4",
						value: "bayer4"
					},
					{
						label: "Bayer 8x8",
						value: "bayer8"
					},
					{
						label: "Clustered Dot",
						value: "clusteredDot"
					},
					{
						label: "Blue Noise",
						value: "blueNoise"
					},
					{
						label: "White Noise",
						value: "whiteNoise"
					}
				],
				label: "Pattern",
				group: "Effect"
			}
		},
		pixelSize: {
			default: 4,
			description: "Size of dithering pixels",
			ui: {
				type: "range",
				min: 1,
				max: 20,
				step: 1,
				label: "Pixel Size",
				group: "Effect"
			}
		},
		threshold: {
			default: .5,
			description: "Luminance threshold for dithering",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Threshold",
				group: "Effect"
			}
		},
		spread: {
			default: 1,
			description: "How much of the luminance range participates in dithering (lower = more solid areas)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Spread",
				group: "Effect"
			}
		},
		colorMode: {
			default: "custom",
			transform: transformColorMode,
			compileTime: true,
			description: "How colors are determined",
			ui: {
				type: "select",
				options: [{
					label: "Custom Colors",
					value: "custom"
				}, {
					label: "Source Colors",
					value: "source"
				}],
				label: "Color Mode",
				group: "Colors"
			}
		},
		colorA: {
			default: "transparent",
			transform: transformColor,
			description: "Dark color for dithering",
			ui: {
				type: "color",
				label: "Dark Color",
				condition: { colorMode: "custom" },
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Light color for dithering",
			ui: {
				type: "color",
				label: "Light Color",
				condition: { colorMode: "custom" },
				group: "Colors"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const DITHER_SIZE = 8;
		const ditherData = new Float32Array(DITHER_SIZE * DITHER_SIZE * 4);
		const bQ = (x, y) => y * 3 + x * 2 - x * y * 4;
		const cM = [
			[
				12,
				5,
				6,
				13
			],
			[
				4,
				0,
				1,
				7
			],
			[
				11,
				3,
				2,
				8
			],
			[
				15,
				10,
				9,
				14
			]
		];
		for (let y = 0; y < DITHER_SIZE; y++) for (let x = 0; x < DITHER_SIZE; x++) {
			const i = (y * DITHER_SIZE + x) * 4;
			ditherData[i] = bQ(x % 2, y % 2) / 4;
			ditherData[i + 1] = (bQ(x % 2, y % 2) * 4 + bQ(Math.floor(x % 4 / 2), Math.floor(y % 4 / 2))) / 16;
			ditherData[i + 2] = (bQ(x % 2, y % 2) * 16 + bQ(Math.floor(x % 4 / 2), Math.floor(y % 4 / 2)) * 4 + bQ(Math.floor(x / 4), Math.floor(y / 4))) / 64;
			ditherData[i + 3] = cM[y % 4][x % 4] / 16;
		}
		const ditherTex = new DataTexture(ditherData, DITHER_SIZE, DITHER_SIZE, RGBAFormat, FloatType);
		ditherTex.magFilter = NearestFilter;
		ditherTex.minFilter = NearestFilter;
		ditherTex.wrapS = RepeatWrapping;
		ditherTex.wrapT = RepeatWrapping;
		ditherTex.needsUpdate = true;
		onCleanup(() => {
			ditherTex.dispose();
		});
		const pixelSize = uniforms.pixelSize.uniform;
		const threshold = uniforms.threshold.uniform;
		const spread = uniforms.spread.uniform;
		const pixelCoord = floor(screenUV.mul(viewportSize).div(pixelSize));
		const pixelatedUV = pixelCoord.add(.5).mul(pixelSize).div(viewportSize);
		const sourceColor = unpremultiplyAlpha(childTexture.sample(pixelatedUV));
		const luminance = dot(sourceColor.rgb, vec3(.299, .587, .114)).mul(sourceColor.a);
		let ditherValue;
		switch (uniforms.pattern.uniform.value) {
			case 0:
				ditherValue = texture(ditherTex).sample(pixelCoord.div(float(DITHER_SIZE))).r;
				break;
			case 1:
				ditherValue = texture(ditherTex).sample(pixelCoord.div(float(DITHER_SIZE))).g;
				break;
			case 2:
				ditherValue = texture(ditherTex).sample(pixelCoord.div(float(DITHER_SIZE))).b;
				break;
			case 3:
				ditherValue = texture(ditherTex).sample(pixelCoord.div(float(DITHER_SIZE))).a;
				break;
			case 4: {
				const blueNoiseCoord = pixelCoord.add(.5);
				ditherValue = fract(float(52.9829189).mul(fract(blueNoiseCoord.x.mul(.06711056).add(blueNoiseCoord.y.mul(.00583715)))));
				break;
			}
			default: ditherValue = fract(sin(dot(pixelCoord, vec2(12.9898, 78.233))).mul(43758.5453));
		}
		const ditherResult = step(float(.5).add(ditherValue.sub(.5).mul(spread)), luminance.add(threshold.sub(.5)));
		let finalRGB;
		let finalAlpha;
		if (uniforms.colorMode.uniform.value === 0) {
			finalRGB = mix(uniforms.colorA.uniform.rgb, uniforms.colorB.uniform.rgb, ditherResult);
			finalAlpha = mix(uniforms.colorA.uniform.a, uniforms.colorB.uniform.a, ditherResult);
		} else {
			finalRGB = mix(sourceColor.rgb.mul(.3), sourceColor.rgb.mul(1.3).min(vec3(1)), ditherResult);
			finalAlpha = sourceColor.a;
		}
		return vec4(finalRGB, finalAlpha);
	}
};
var Dither_default = componentDefinition;
export { componentDefinition as n, Dither_default as t };
