// shaders/dist/core/DOMTexture-Ec0k6AnK.js
import { CanvasTexture, LinearFilter, SRGBColorSpace } from "three/webgpu";
import { screenUV, texture } from "three/tsl";
const componentDefinition = {
	name: "DOMTexture",
	category: "Textures",
	description: "Render live HTML/DOM content as a WebGPU texture layer via the html-in-canvas API. Requires Chrome Canary with chrome://flags/#canvas-draw-element enabled.",
	capturesDOM: true,
	experimental: {
		message: "Requires Chrome Canary with chrome://flags/#canvas-draw-element enabled. Do not use for production use.",
		docs: "This component is powered by the WICG html-in-canvas proposal, which is currently only available in Chrome Canary behind a feature flag. It is not suitable for production use and may change as the specification evolves.",
		docsLink: {
			url: "https://github.com/WICG/html-in-canvas",
			label: "View the spec"
		}
	},
	props: {},
	fragmentNode: ({ onCleanup, onBeforeRender, onResize, renderer, domCanvas }) => {
		const dpr = Math.min(window.devicePixelRatio, 2);
		const backingCanvas = document.createElement("canvas");
		backingCanvas.width = Math.round(window.innerWidth * dpr);
		backingCanvas.height = Math.round(window.innerHeight * dpr);
		const ctx = backingCanvas.getContext("2d");
		ctx.fillStyle = "rgba(0,0,0,0)";
		ctx.fillRect(0, 0, backingCanvas.width, backingCanvas.height);
		const threeTexture = new CanvasTexture(backingCanvas);
		threeTexture.colorSpace = SRGBColorSpace;
		threeTexture.generateMipmaps = false;
		threeTexture.minFilter = LinearFilter;
		threeTexture.needsUpdate = true;
		const textureNode = texture(threeTexture);
		let gpuTexture = null;
		let gpuDevice = null;
		let contentElement = null;
		let hasLoggedError = false;
		let paintSetup = false;
		let hasFreshSnapshot = false;
		onBeforeRender(() => {
			if (!domCanvas) return;
			if (!contentElement) {
				contentElement = domCanvas.firstElementChild;
				if (contentElement) {
					const rect = contentElement.getBoundingClientRect();
					const physW = Math.round(rect.width * window.devicePixelRatio);
					const physH = Math.round(rect.height * window.devicePixelRatio);
					if (physW > 0 && physH > 0 && (physW !== backingCanvas.width || physH !== backingCanvas.height)) {
						backingCanvas.width = physW;
						backingCanvas.height = physH;
						threeTexture.needsUpdate = true;
					}
				}
			}
			if (!contentElement) return;
			if (!gpuDevice) try {
				gpuDevice = renderer.backend?.device ?? null;
			} catch {}
			if (!gpuTexture && gpuDevice) try {
				gpuTexture = renderer.backend.get(threeTexture)?.texture ?? null;
			} catch {}
			if (!paintSetup) {
				domCanvas.getContext("2d");
				domCanvas.onpaint = () => {
					hasFreshSnapshot = true;
				};
				domCanvas.requestPaint?.();
				paintSetup = true;
				return;
			}
			if (hasFreshSnapshot && gpuDevice && gpuTexture && contentElement) try {
				gpuDevice.queue.copyElementImageToTexture(contentElement, { texture: gpuTexture });
				hasFreshSnapshot = false;
			} catch (e) {
				if (!hasLoggedError) {
					console.warn("[DOMTexture] copyElementImageToTexture failed.\nRequires Chrome Canary with chrome://flags/#canvas-draw-element enabled.\nThe element must be a direct child of a <canvas layoutsubtree>.", e);
					hasLoggedError = true;
				}
			}
		});
		onResize(() => {
			const newDpr = Math.min(window.devicePixelRatio, 2);
			backingCanvas.width = Math.round(window.innerWidth * newDpr);
			backingCanvas.height = Math.round(window.innerHeight * newDpr);
			threeTexture.dispose();
			threeTexture.image = backingCanvas;
			threeTexture.needsUpdate = true;
			gpuTexture = null;
			hasFreshSnapshot = false;
			if (domCanvas) domCanvas.requestPaint?.();
		});
		onCleanup(() => {
			if (domCanvas) domCanvas.onpaint = null;
			threeTexture.dispose();
		});
		return textureNode.sample(screenUV);
	}
};
var DOMTexture_default = componentDefinition;
export { componentDefinition as n, DOMTexture_default as t };
