// shaders/dist/core/DropShadow-CR3cecIQ.js
import { a as transformColor, i as transformBoolean } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { Fn, clamp, convertToTexture, cos, float, max, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
var weights = [
	.056,
	.135,
	.278,
	.487,
	.726,
	.923,
	1,
	.923,
	.726,
	.487,
	.278,
	.135,
	.056
];
var weightSum = 6.214;
const componentDefinition = {
	name: "DropShadow",
	category: "Stylize",
	description: "Adds a soft shadow behind the child content based on its alpha silhouette",
	requiresRTT: true,
	requiresChild: true,
	props: {
		color: {
			default: "#000000",
			transform: transformColor,
			description: "Shadow color",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		distance: {
			default: .1,
			description: "How far the shadow is offset from the content",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .005,
				label: "Distance",
				group: "Effect"
			}
		},
		angle: {
			default: 135,
			description: "Direction the shadow is cast (compass degrees: 0=up, 90=right, 135=lower-right, 180=down)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		blur: {
			default: 5,
			description: "Shadow softness (blur radius in pixels)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .5,
				label: "Blur",
				group: "Effect"
			}
		},
		intensity: {
			default: .5,
			description: "Shadow intensity — how strong/visible the shadow is",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Intensity",
				group: "Effect"
			}
		},
		cutout: {
			default: false,
			transform: transformBoolean,
			compileTime: true,
			description: "Hide the original layer and show only the shadow",
			ui: {
				type: "checkbox",
				label: "Cutout",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const originalColor = childTexture.sample(screenUV);
		const angleRad = uniforms.angle.uniform.mul(float(Math.PI / 180));
		const dist = uniforms.distance.uniform.mul(float(.2));
		const aspect = viewportSize.x.div(viewportSize.y);
		const offsetUV = vec2(sin(angleRad).mul(dist).div(aspect), cos(angleRad).negate().mul(dist));
		const shadowUV = screenUV.sub(offsetUV);
		const horizontalTexture = convertToTexture(Fn(() => {
			const blurRadius = uniforms.blur.uniform;
			const pixelSize = vec2(float(1)).div(viewportSize);
			const total = float(0).toVar();
			for (let x = -6; x <= 6; x++) {
				const weight = float(weights[x + 6] / weightSum);
				const offset = vec2(float(x), float(0)).mul(blurRadius).mul(pixelSize);
				total.assign(total.add(childTexture.sample(shadowUV.add(offset)).a.mul(weight)));
			}
			return vec4(total, float(0), float(0), float(1));
		})());
		onCleanup(() => {
			if (horizontalTexture?.renderTarget?.dispose) horizontalTexture.renderTarget.dispose();
		});
		const shadowAlpha = Fn(() => {
			const blurRadius = uniforms.blur.uniform;
			const pixelSize = vec2(float(1)).div(viewportSize);
			const total = float(0).toVar();
			for (let y = -6; y <= 6; y++) {
				const weight = float(weights[y + 6] / weightSum);
				const offset = vec2(float(0), float(y)).mul(blurRadius).mul(pixelSize);
				total.assign(total.add(horizontalTexture.sample(screenUV.add(offset)).r.mul(weight)));
			}
			return total;
		})().mul(uniforms.intensity.uniform);
		const shadowColor = uniforms.color.uniform;
		const shadowA = shadowColor.a.mul(shadowAlpha);
		const cutout = uniforms.cutout.uniform.value === 1;
		let resultRGB;
		let resultA;
		if (cutout) {
			resultA = clamp(shadowA.sub(originalColor.a), float(0), float(1));
			resultRGB = shadowColor.rgb.mul(shadowColor.a).mul(resultA);
		} else {
			const shadowPremulRGB = shadowColor.rgb.mul(shadowColor.a).mul(shadowAlpha);
			const oneMinusOrigA = float(1).sub(originalColor.a);
			resultRGB = originalColor.rgb.add(shadowPremulRGB.mul(oneMinusOrigA));
			resultA = max(originalColor.a.add(shadowA.mul(oneMinusOrigA)), originalColor.a);
		}
		return unpremultiplyAlpha(vec4(resultRGB, resultA));
	}
};
var DropShadow_default = componentDefinition;
export { componentDefinition as n, DropShadow_default as t };
