// shaders/dist/core/Emboss-DGaubb9x.js
import { l as transformPosition } from "./transformations-CJcUeZIC.js";
import { n as createAnalyticSdfSampler, r as createSvgSdfSampler } from "./sdf-D_EAGXhN.js";
import { Fn, If, abs, clamp, convertToTexture, cos, float, max, screenUV, sin, smoothstep, sqrt, vec2, vec3, vec4, viewportSize } from "three/tsl";
var _INV_LIGHT_LEN = 1 / Math.sqrt(1.49);
var _LZN = .7 * _INV_LIGHT_LEN;
function applyEmbossEffect(childTexture, sdfSampler, uniforms) {
	return Fn(() => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const centerPos = vec2(uniforms.center.uniform.x, uniforms.center.uniform.y.oneMinus());
		const scale = uniforms.scale.uniform;
		const sdfUV = vec2(uv$1.x.sub(centerPos.x).mul(aspect).div(scale).add(float(.5)), uv$1.y.sub(centerPos.y).div(scale).add(float(.5)));
		const pxH = float(1).div(viewportSize.y);
		const eps = float(.003);
		const tw = float(.003);
		const sdfC = sdfSampler(sdfUV).r;
		const absDepth = abs(uniforms.depth.uniform).max(float(1e-4));
		const spread = absDepth.mul(float(.06));
		const pxH_sdf = pxH.div(scale);
		const output = vec4(0).toVar();
		If(sdfC.greaterThan(spread.mul(float(2)).add(max(pxH_sdf, tw))), () => {
			output.assign(childTexture.sample(uv$1));
		}).Else(() => {
			const sdfR = sdfSampler(sdfUV.add(vec2(eps, float(0)))).r;
			const sdfL = sdfSampler(sdfUV.sub(vec2(eps, float(0)))).r;
			const sdfU = sdfSampler(sdfUV.add(vec2(float(0), eps))).r;
			const sdfD = sdfSampler(sdfUV.sub(vec2(float(0), eps))).r;
			const depth = uniforms.depth.uniform;
			const hC = depth.mul(smoothstep(tw, tw.negate(), sdfC));
			const hR = depth.mul(smoothstep(tw, tw.negate(), sdfR));
			const hL = depth.mul(smoothstep(tw, tw.negate(), sdfL));
			const hU = depth.mul(smoothstep(tw, tw.negate(), sdfU));
			const hD = depth.mul(smoothstep(tw, tw.negate(), sdfD));
			const dhx = hR.sub(hL).div(eps.mul(float(2)));
			const dhy = hU.sub(hD).div(eps.mul(float(2)));
			const nLen = sqrt(dhx.mul(dhx).add(dhy.mul(dhy)).add(float(1)));
			const nx = dhx.negate().div(nLen);
			const ny = dhy.negate().div(nLen);
			const nz = float(1).div(nLen);
			const lightRad = uniforms.lightAngle.uniform.mul(float(Math.PI / 180));
			const lx = cos(lightRad);
			const ly = sin(lightRad);
			const lxn = lx.mul(float(_INV_LIGHT_LEN));
			const lyn = ly.mul(float(_INV_LIGHT_LEN));
			const lzn = float(_LZN);
			const edgeLighting = nx.mul(lxn).add(ny.mul(lyn)).add(nz.mul(lzn)).sub(lzn).mul(uniforms.lightIntensity.uniform);
			const highlight = max(edgeLighting, float(0));
			const edgeShadow = max(edgeLighting.negate(), float(0));
			const shadowTw = spread;
			const s1UV = sdfUV.add(vec2(lx.mul(spread.mul(float(.5))), ly.mul(spread.mul(float(.5)))));
			const s2UV = sdfUV.add(vec2(lx.mul(spread), ly.mul(spread)));
			const h1 = depth.mul(smoothstep(shadowTw, shadowTw.negate(), sdfSampler(s1UV).r));
			const h2 = depth.mul(smoothstep(shadowTw, shadowTw.negate(), sdfSampler(s2UV).r));
			const diff1 = h1.sub(hC).div(absDepth);
			const diff2 = h2.sub(hC).div(absDepth);
			const castShadow = smoothstep(float(0), float(.5), diff1).mul(float(.55)).add(smoothstep(float(0), float(.5), diff2).mul(float(.45)));
			const inside = smoothstep(tw, tw.negate(), sdfC);
			const dispScale = float(1).sub(depth.mul(float(.05)).mul(inside));
			const warpedUV = vec2(centerPos.x.add(uv$1.x.sub(centerPos.x).mul(dispScale)), centerPos.y.add(uv$1.y.sub(centerPos.y).mul(dispScale)));
			const totalShadow = clamp(edgeShadow.add(castShadow), float(0), float(1)).mul(uniforms.shadowIntensity.uniform);
			const childColor = childTexture.sample(warpedUV);
			const finalRgb = clamp(vec3(childColor.r, childColor.g, childColor.b).mul(float(1).sub(totalShadow)).add(vec3(highlight, highlight, highlight)), float(0), float(1));
			output.assign(vec4(finalRgb.x, finalRgb.y, finalRgb.z, childColor.a));
		});
		return output;
	})();
}
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
const componentDefinition = {
	name: "Emboss",
	category: "Shape Effects",
	description: "Embossed / debossed relief shading on top of child content, driven by a custom shape",
	requiresRTT: true,
	requiresChild: true,
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the embossed shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		scale: {
			default: 1,
			description: "Scale of the embossed shape (1 = default size)",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .05,
				label: "Scale",
				group: "Position"
			}
		},
		depth: {
			default: -.5,
			description: "Relief depth — negative = inset (debossed), positive = raised (embossed)",
			ui: {
				type: "range",
				min: -1,
				max: 1,
				step: .01,
				label: "Depth",
				group: "Relief"
			}
		},
		lightAngle: {
			default: 260,
			description: "Directional light angle in degrees — controls highlight and shadow direction",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Light Angle",
				group: "Lighting"
			}
		},
		lightIntensity: {
			default: .6,
			description: "Strength of the directional edge highlights and shadows",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .05,
				label: "Light Intensity",
				group: "Lighting"
			}
		},
		shadowIntensity: {
			default: .3,
			description: "Darkness of the relief shadow",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Shadow Intensity",
				group: "Relief"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file — when non-empty, activates SVG mode and triggers a shader recompile"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type — triggers recompile when shape is switched. When empty, derived from shape JSON at mount time."
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup, onBeforeRender }) => {
		if (!childNode) return vec4(0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const shapeSdfUrl = uniforms.shapeSdfUrl.uniform.value;
		const initialConfig = (() => {
			const val = uniforms.shape.uniform.value;
			if (typeof val === "object" && val !== null) return val;
			try {
				return JSON.parse(val);
			} catch {
				return {
					type: "circleSDF",
					radius: .35,
					centerX: .5,
					centerY: .5
				};
			}
		})();
		if (shapeSdfUrl) return applyEmbossEffect(childTexture, createSvgSdfSampler(shapeSdfUrl, onBeforeRender, onCleanup), uniforms);
		else return applyEmbossEffect(childTexture, createAnalyticSdfSampler(uniforms.shapeType.uniform.value || initialConfig.type || "circleSDF", initialConfig, uniforms.shape, onBeforeRender), uniforms);
	}
};
var Emboss_default = componentDefinition;
export { componentDefinition as n, Emboss_default as t };
