// shaders/dist/core/FallingLines-DqIZ8wTH.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { Fn, abs, cos, float, floor, fract, fwidth, screenUV, sin, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
var colHash = Fn(([idx]) => {
	return vec2(fract(sin(idx.mul(127.1)).mul(43758.5453123)), fract(sin(idx.mul(311.7)).mul(38291.3421)));
});
const componentDefinition = {
	name: "FallingLines",
	category: "Textures",
	description: "Directional falling lines with a leading-to-trailing color fade",
	requiresRTT: false,
	props: {
		colorA: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color at the leading edge of each line",
			ui: {
				type: "color",
				label: "Lead Color",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff00",
			transform: transformColor,
			description: "Color at the trailing edge (transparent by default)",
			ui: {
				type: "color",
				label: "Trail Color",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for interpolation between lead and trail colors",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		angle: {
			default: 90,
			description: "Direction of movement in degrees (90=down, 270=up, 0=right, 180=left)",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		speed: {
			default: .5,
			description: "Movement speed",
			ui: {
				type: "range",
				min: 0,
				max: 3,
				step: .05,
				label: "Speed",
				group: "Animation"
			}
		},
		speedVariance: {
			default: .3,
			description: "Per-line speed variance (0=uniform, 1=high variance)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .05,
				label: "Speed Variance",
				group: "Animation"
			}
		},
		density: {
			default: 15,
			description: "Number of line columns across the canvas",
			ui: {
				type: "range",
				min: 1,
				max: 60,
				step: 1,
				label: "Density",
				group: "Effect"
			}
		},
		trailLength: {
			default: .35,
			description: "Streak length relative to spacing (0=point, 1=continuous)",
			ui: {
				type: ["range", "map"],
				min: .01,
				max: 1,
				step: .01,
				label: "Trail Length",
				group: "Effect"
			}
		},
		balance: {
			default: .5,
			description: "Color mix midpoint (0.5=linear, 0=all trailing/colorB, 1=all leading/colorA)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Balance",
				group: "Colors"
			}
		},
		strokeWidth: {
			default: .15,
			description: "Line thickness as fraction of column width (0=hairline, 1=full width)",
			ui: {
				type: ["range", "map"],
				min: .02,
				max: 1,
				step: .02,
				label: "Stroke Width",
				group: "Effect"
			}
		},
		rounding: {
			default: 1,
			description: "Rounds the leading edge (0=flat/square, 1=fully rounded cap)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Rounding",
				group: "Effect"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const aspect = viewportSize.x.div(viewportSize.y);
		const corrUV = vec2(screenUV.x.mul(aspect), screenUV.y);
		const canvasCenter = vec2(aspect.mul(.5), float(.5));
		const centeredUV = corrUV.sub(canvasCenter);
		const angleRad = uniforms.angle.uniform.mul(Math.PI / 180);
		const cosA = cos(angleRad);
		const sinA = sin(angleRad);
		const u = centeredUV.x.mul(sinA).negate().add(centeredUV.y.mul(cosA));
		const v = centeredUV.x.mul(cosA).add(centeredUV.y.mul(sinA));
		const density = uniforms.density.uniform;
		const colF = u.mul(density);
		const colIdx = floor(colF);
		const colLocalU = fract(colF);
		const h = colHash(colIdx);
		const speedMod = float(1).add(h.x.sub(.5).mul(uniforms.speedVariance.uniform).mul(2));
		const phase = h.y;
		const animOffset = fract(createAnimatedTime(params, uniforms.speed).mul(speedMod));
		const vPeriodic = fract(v.sub(animOffset).sub(phase));
		const trailLen = uniforms.trailLength.uniform;
		const withinStreak = vPeriodic.div(trailLen).clamp(0, 1);
		const fwV = fwidth(v).clamp(.001, .1);
		const fwU = fwidth(u).mul(density).clamp(.001, .1);
		const streakMask = smoothstep(trailLen.sub(fwV), trailLen.add(fwV), vPeriodic).oneMinus();
		const hw = uniforms.strokeWidth.uniform.mul(.5);
		const distFromCenter = abs(colLocalU.sub(.5));
		const widthMask = smoothstep(hw.sub(fwU), hw.add(fwU), distFromCenter).oneMinus();
		const rectMask = streakMask.mul(widthMask);
		const maxDepth = hw.mul(uniforms.rounding.uniform);
		const t = vPeriodic.sub(trailLen).max(float(0)).mul(density).div(maxDepth.max(float(1e-4))).clamp(float(0), float(1));
		const capHW = hw.mul(float(1).sub(t.mul(t)).max(float(0)).sqrt());
		const capAlpha = smoothstep(capHW.sub(fwU).max(float(0)), capHW.add(fwU).max(float(.001)), distFromCenter).oneMinus().mul(smoothstep(float(0), fwU, capHW));
		const finalMask = rectMask.max(capAlpha);
		const balanceNum = float(2).mul(uniforms.balance.uniform).sub(float(1));
		const balanceDen = float(2).mul(float(1).sub(uniforms.balance.uniform)).max(.001);
		const balancedT = withinStreak.add(balanceNum).div(balanceDen).clamp(0, 1);
		const mixedColor = mixColors(uniforms.colorB.uniform, uniforms.colorA.uniform, balancedT, uniforms.colorSpace.uniform.value);
		const finalAlpha = finalMask.mul(mixedColor.a);
		return vec4(mixedColor.rgb, finalAlpha);
	}
};
var FallingLines_default = componentDefinition;
export { componentDefinition as n, FallingLines_default as t };
