// shaders/dist/core/FilmGrain-Dbw02Jz9.js
import { clamp, dot, float, fract, pow, sin, uniform, vec2, vec3, vec4, viewportCoordinate } from "three/tsl";
const componentDefinition = {
	name: "FilmGrain",
	category: "Stylize",
	description: "Analog film grain texture overlay, weighted toward darker areas",
	requiresChild: true,
	props: {
		strength: {
			default: .5,
			description: "Intensity of the film grain noise",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .025,
				label: "Strength",
				group: "Effect"
			}
		},
		bias: {
			default: 2,
			description: "Concentrates grain in darker areas. Higher values focus grain more heavily on shadows; 0 applies grain uniformly.",
			ui: {
				type: "range",
				min: 0,
				max: 10,
				step: .1,
				label: "Bias",
				group: "Effect"
			}
		},
		animated: {
			default: false,
			description: "When enabled, the grain pattern changes each frame for a dynamic film effect",
			ui: {
				type: "checkbox",
				label: "Animated",
				group: "Effect"
			},
			transform: (value) => value ? 1 : 0
		}
	},
	fragmentNode: ({ uniforms, childNode, onBeforeRender }) => {
		const strength = uniforms.strength.uniform;
		const bias = uniforms.bias.uniform;
		const animated = uniforms.animated.uniform;
		const sourceColor = childNode || vec4(0, 0, 0, 0);
		const animTime = uniform(0);
		onBeforeRender(({ deltaTime }) => {
			animTime.value += deltaTime * animated.value * 10;
		});
		const grain = fract(sin(dot(viewportCoordinate, vec2(12.9898, 78.233)).add(animTime)).mul(float(43758.5453))).mul(float(2)).sub(float(1));
		const brightness = clamp(dot(sourceColor.rgb, vec3(.2126, .7152, .0722)), 0, 1);
		const darkFactor = pow(float(1).sub(brightness).add(float(1e-6)), bias);
		const grainIntensity = grain.mul(darkFactor).mul(strength.mul(float(.1)));
		return vec4(sourceColor.rgb.add(grainIntensity), sourceColor.a);
	}
};
var FilmGrain_default = componentDefinition;
export { componentDefinition as n, FilmGrain_default as t };
