// shaders/dist/core/FloatingParticles-LMsFTJKp.js
import { a as transformColor } from "./transformations-CJcUeZIC.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { StorageTexture } from "three/webgpu";
import { FloatType } from "three";
import { Fn, cos, dot, float, floor, fract, instanceIndex, length, min, mix, screenUV, sin, smoothstep, storageTexture, texture, textureStore, time, uint, uniform, uvec2, vec2, vec3, vec4, viewportSize } from "three/tsl";
var PARTICLE_TEX = 1024;
var hash1_2 = Fn(([p]) => {
	const p3 = fract(vec3(p.x, p.y, p.x).mul(.1031));
	const d = dot(p3, vec3(p3.y, p3.z, p3.x).add(33.33));
	const p3b = p3.add(d);
	return fract(p3b.x.add(p3b.y).mul(p3b.z));
});
var hash2_2 = Fn(([p]) => {
	const p3 = fract(vec3(p.x, p.y, p.x).mul(vec3(.1031, .103, .0973)));
	const d = dot(p3, vec3(p3.y, p3.z, p3.x).add(33.33));
	const p3b = p3.add(d);
	const s = p3b.x.add(p3b.y);
	return fract(vec2(s.mul(p3b.z), s.mul(p3b.y)));
});
var voronoiPointFromRoot = Fn(([root, deg]) => {
	const point = hash2_2(root).sub(.5);
	const s = sin(deg);
	const c = cos(deg);
	return vec2(c.mul(point.x).sub(s.mul(point.y)).mul(.66), s.mul(point.x).add(c.mul(point.y)).mul(.66)).add(root).add(.5);
});
var degFromRootUV = Fn(([uv$1, randomness]) => {
	return time.mul(randomness).mul(hash1_2(uv$1).sub(.5)).mul(2);
});
var getParticleDistance = Fn(([uv$1, cellUV, randomness]) => {
	const pointUV = voronoiPointFromRoot(cellUV, degFromRootUV(cellUV, randomness));
	return length(uv$1.sub(pointUV));
});
var degFromRootUV_compute = Fn(([uv$1, randomness, frameTimeU]) => {
	return frameTimeU.mul(randomness).mul(hash1_2(uv$1).sub(.5)).mul(2);
});
var getParticleDistance_compute = Fn(([uv$1, cellUV, randomness, frameTimeU]) => {
	const pointUV = voronoiPointFromRoot(cellUV, degFromRootUV_compute(cellUV, randomness, frameTimeU));
	return length(uv$1.sub(pointUV));
});
const componentDefinition = {
	name: "FloatingParticles",
	category: "Textures",
	description: "Animated floating particles with twinkle effects",
	requiresRTT: false,
	props: {
		randomness: {
			default: .25,
			description: "Randomness of particle animation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Randomness",
				group: "Effect"
			}
		},
		speed: {
			default: .25,
			description: "Speed of particle movement",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		angle: {
			default: 90,
			description: "Movement angle in degrees (0=right, 90=down, 180=left, 270=up)",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Animation"
			}
		},
		particleSize: {
			default: 2,
			description: "Size of particles",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 20,
				step: .1,
				label: "Particle Size",
				group: "Effect"
			}
		},
		particleSoftness: {
			default: 0,
			description: "Edge softness of particles (0 = sharp, 1 = very soft)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .1,
				label: "Particle Softness",
				group: "Effect"
			}
		},
		twinkle: {
			default: .5,
			description: "Intensity of the twinkle effect (0 = off, 1 = full twinkle)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Twinkle",
				group: "Animation"
			}
		},
		count: {
			default: 5,
			description: "Number of particle layers",
			ui: {
				type: "range",
				min: 1,
				max: 5,
				step: 1,
				label: "Count",
				group: "Effect"
			}
		},
		particleColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color of the particles",
			ui: {
				type: "color",
				label: "Particle Color",
				group: "Colors"
			}
		},
		speedVariance: {
			default: .3,
			description: "Per-layer speed variance (0 = all layers same speed, 1 = high variance)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .1,
				label: "Speed Variance",
				group: "Animation"
			}
		},
		angleVariance: {
			default: 30,
			description: "Per-layer angle variance in degrees (0 = all layers same angle, 180 = full variance)",
			ui: {
				type: "range",
				min: 0,
				max: 180,
				step: 1,
				label: "Angle Variance",
				group: "Animation"
			}
		},
		particleDensity: {
			default: 3,
			description: "Particle density (lower = more spread out, higher = more dense)",
			ui: {
				type: "range",
				min: .5,
				max: 3,
				step: .1,
				label: "Particle Density",
				group: "Effect"
			}
		}
	},
	computeNode: (params) => {
		const { uniforms, onCleanup } = params;
		const texSize = PARTICLE_TEX;
		const count = texSize * texSize;
		const gs = uint(texSize);
		const outTex = new StorageTexture(texSize, texSize);
		outTex.type = FloatType;
		const outTexNode = storageTexture(outTex);
		onCleanup(() => outTex.dispose());
		const animTimeU = uniform(0);
		const frameTimeU = uniform(0);
		const aspectU = uniform(16 / 9);
		const computeRenderParticle = Fn(([uv$1]) => {
			const rootUV = floor(uv$1);
			const randomness = uniforms.randomness.uniform;
			const dCenter = getParticleDistance_compute(uv$1, rootUV, randomness, frameTimeU);
			const dLeft = getParticleDistance_compute(uv$1, rootUV.add(vec2(-1, 0)), randomness, frameTimeU);
			const dRight = getParticleDistance_compute(uv$1, rootUV.add(vec2(1, 0)), randomness, frameTimeU);
			const dTop = getParticleDistance_compute(uv$1, rootUV.add(vec2(0, -1)), randomness, frameTimeU);
			const dBottom = getParticleDistance_compute(uv$1, rootUV.add(vec2(0, 1)), randomness, frameTimeU);
			const minDist = min(min(min(min(dCenter, dLeft), dRight), dTop), dBottom);
			const softness = uniforms.particleSoftness.uniform;
			const scaledSize = uniforms.particleSize.uniform.mul(.01);
			const innerRadius = scaledSize.mul(.6);
			const baseIntensity = smoothstep(innerRadius, innerRadius.add(softness.mul(scaledSize.mul(2.4))), minDist).oneMinus();
			const twinkleIntensity = uniforms.twinkle.uniform;
			const particlePhase = hash1_2(rootUV);
			const twinkleValue = sin(frameTimeU.mul(2).add(particlePhase.mul(6.28318))).mul(.5).add(.5);
			const twinkleModifier = mix(float(1), twinkleValue, twinkleIntensity);
			return baseIntensity.mul(twinkleModifier).mul(uniforms.particleColor.uniform.rgb);
		});
		const computeLayeredParticles = Fn(([uv$1]) => {
			const sizeMod = float(1.05);
			const alphaMod = float(.9);
			const layers = uniforms.count.uniform;
			const particles = vec3(0).toVar();
			const layerScale = float(1).toVar();
			const alpha = float(1).toVar();
			const offset = vec2(0).toVar();
			const angleRad = uniforms.angle.uniform.mul(3.14159).div(180);
			const baseMovementDir = vec2(cos(angleRad), sin(angleRad));
			const baseMovement = animTimeU.mul(baseMovementDir);
			const speedVar = uniforms.speedVariance.uniform;
			const angleVar = uniforms.angleVariance.uniform.mul(3.14159).div(180);
			for (let i = 0; i < 5; i++) {
				const layerActive = float(i).lessThan(layers).and(alpha.greaterThan(.02));
				const layerHash = hash2_2(vec2(alpha, float(i).mul(7.919)));
				const speedMod = float(1).add(layerHash.x.sub(.5).mul(speedVar));
				const angleDelta = layerHash.y.sub(.5).mul(angleVar).mul(2);
				const variedAngle = angleRad.add(angleDelta).toVar();
				const variedDir = vec2(cos(variedAngle), sin(variedAngle));
				const movement = baseMovement.mul(speedMod).add(variedDir.sub(baseMovementDir).mul(animTimeU));
				const layerParticles = computeRenderParticle(uv$1.mul(layerScale).add(movement).add(offset)).mul(alpha);
				particles.assign(particles.add(layerActive.select(layerParticles, vec3(0))));
				offset.assign(offset.add(layerHash.mul(10)));
				alpha.assign(alpha.mul(alphaMod));
				layerScale.assign(layerScale.mul(sizeMod));
			}
			return particles;
		});
		const particlePass = Fn(() => {
			const idx = instanceIndex;
			const j = idx.mod(gs);
			const i = idx.div(gs);
			const u = j.toFloat().add(.5).div(float(texSize));
			const v = i.toFloat().add(.5).div(float(texSize));
			const particles = computeLayeredParticles(vec2(u.mul(2).sub(1).mul(aspectU), v.mul(2).sub(1)).mul(uniforms.particleDensity.uniform));
			const pa = length(particles).clamp(0, 1);
			textureStore(outTexNode, uvec2(j, i), vec4(particles, pa)).toWriteOnly();
		})().compute(count, [256]);
		return {
			outputs: { particleTexture: outTex },
			getComputeNodes: (frameParams) => {
				animTimeU.value += frameParams.deltaTime * uniforms.speed.uniform.value;
				frameTimeU.value += frameParams.deltaTime;
				if (frameParams.dimensions.height > 0) aspectU.value = frameParams.dimensions.width / frameParams.dimensions.height;
				return [particlePass];
			}
		};
	},
	fragmentNode: (params) => {
		const { uniforms, childNode, computeOutputs } = params;
		if (computeOutputs?.particleTexture) {
			const particleField = texture(computeOutputs.particleTexture);
			const px = float(1 / PARTICLE_TEX);
			const s0 = particleField.sample(screenUV);
			const s1 = particleField.sample(screenUV.add(vec2(px, float(0))));
			const s2 = particleField.sample(screenUV.add(vec2(float(0), px)));
			const s3 = particleField.sample(screenUV.add(vec2(px.negate(), float(0))));
			const s4 = particleField.sample(screenUV.add(vec2(float(0), px.negate())));
			const w0 = float(.5), w1 = float(.125);
			const particles$1 = s0.xyz.mul(w0).add(s1.xyz.mul(w1)).add(s2.xyz.mul(w1)).add(s3.xyz.mul(w1)).add(s4.xyz.mul(w1));
			const particleAlpha$1 = s0.w.mul(w0).add(s1.w.mul(w1)).add(s2.w.mul(w1)).add(s3.w.mul(w1)).add(s4.w.mul(w1));
			if (childNode) {
				const mixedColor = mix(childNode.xyz, particles$1, particleAlpha$1.greaterThan(.01).select(particleAlpha$1, 0));
				const pa = particleAlpha$1.clamp(0, 1);
				return vec4(mixedColor, pa.add(childNode.w.mul(float(1).sub(pa))));
			}
			return vec4(particles$1, particleAlpha$1);
		}
		const baseUV = screenUV;
		const animTime = createAnimatedTime(params, uniforms.speed);
		const renderParticle = Fn(([uv$2]) => {
			const rootUV = floor(uv$2);
			const randomness = uniforms.randomness.uniform;
			const dCenter = getParticleDistance(uv$2, rootUV, randomness);
			const dLeft = getParticleDistance(uv$2, rootUV.add(vec2(-1, 0)), randomness);
			const dRight = getParticleDistance(uv$2, rootUV.add(vec2(1, 0)), randomness);
			const dTop = getParticleDistance(uv$2, rootUV.add(vec2(0, -1)), randomness);
			const dBottom = getParticleDistance(uv$2, rootUV.add(vec2(0, 1)), randomness);
			const minDist = min(min(min(min(dCenter, dLeft), dRight), dTop), dBottom);
			const softness = uniforms.particleSoftness.uniform;
			const scaledSize = uniforms.particleSize.uniform.mul(.01);
			const innerRadius = scaledSize.mul(.6);
			const baseIntensity = smoothstep(innerRadius, innerRadius.add(softness.mul(scaledSize.mul(2.4))), minDist).oneMinus();
			const twinkleIntensity = uniforms.twinkle.uniform;
			const particlePhase = hash1_2(rootUV);
			const twinkleValue = sin(time.mul(2).add(particlePhase.mul(6.28318))).mul(.5).add(.5);
			const twinkleModifier = mix(float(1), twinkleValue, twinkleIntensity);
			return baseIntensity.mul(twinkleModifier).mul(uniforms.particleColor.uniform.rgb);
		});
		const layeredParticles = Fn(([uv$2]) => {
			const sizeMod = float(1.05);
			const alphaMod = float(.9);
			const layers = uniforms.count.uniform;
			const particles$1 = vec3(0).toVar();
			const size = float(1).toVar();
			const alpha = float(1).toVar();
			const offset = vec2(0).toVar();
			const angleRad = uniforms.angle.uniform.mul(3.14159).div(180);
			const baseMovementDir = vec2(cos(angleRad), sin(angleRad));
			const baseMovement = animTime.mul(baseMovementDir);
			const speedVar = uniforms.speedVariance.uniform;
			const angleVar = uniforms.angleVariance.uniform.mul(3.14159).div(180);
			for (let i = 0; i < 5; i++) {
				const layerActive = float(i).lessThan(layers).and(alpha.greaterThan(.02));
				const layerHash = hash2_2(vec2(alpha, float(i).mul(7.919)));
				const speedMod = float(1).add(layerHash.x.sub(.5).mul(speedVar));
				const angleDelta = layerHash.y.sub(.5).mul(angleVar).mul(2);
				const variedAngle = angleRad.add(angleDelta).toVar();
				const variedDir = vec2(cos(variedAngle), sin(variedAngle));
				const movement = baseMovement.mul(speedMod).add(variedDir.sub(baseMovementDir).mul(animTime));
				const layerParticles = renderParticle(uv$2.mul(size).add(movement).add(offset)).mul(alpha);
				particles$1.assign(particles$1.add(layerActive.select(layerParticles, vec3(0))));
				offset.assign(offset.add(layerHash.mul(10)));
				alpha.assign(alpha.mul(alphaMod));
				size.assign(size.mul(sizeMod));
			}
			return particles$1;
		});
		const uv$1 = baseUV.mul(2).sub(1);
		const aspectRatio = viewportSize.x.div(viewportSize.y);
		const particles = layeredParticles(vec2(uv$1.x.mul(aspectRatio), uv$1.y).mul(uniforms.particleDensity.uniform));
		const particleAlpha = length(particles);
		if (childNode) {
			const mixedColor = mix(childNode.xyz, particles, particleAlpha.greaterThan(.01).select(particleAlpha, 0));
			const pa = particleAlpha.clamp(0, 1);
			return vec4(mixedColor, pa.add(childNode.w.mul(float(1).sub(pa))));
		}
		return vec4(particles, particleAlpha);
	}
};
var FloatingParticles_default = componentDefinition;
export { componentDefinition as n, FloatingParticles_default as t };
