// shaders/dist/core/FlowField-D3CAHBBG.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { convertToTexture, cos, float, length, max, mx_noise_float, screenUV, sin, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "FlowField",
	category: "Distortions",
	description: "Fluid-like distortion with constant smooth motion",
	requiresRTT: true,
	requiresChild: true,
	props: {
		strength: {
			default: .15,
			description: "Intensity of the flow distortion",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .5,
				step: .05,
				label: "Strength",
				group: "Effect"
			}
		},
		detail: {
			default: 2,
			description: "Scale of the flow patterns",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 5,
				step: .1,
				label: "Detail",
				group: "Effect"
			}
		},
		speed: {
			default: 0,
			description: "Speed of the flow",
			ui: {
				type: "range",
				min: 0,
				max: 20,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		evolutionSpeed: {
			default: 0,
			description: "How fast the flow field pattern reshapes over time",
			ui: {
				type: "range",
				min: 0,
				max: 20,
				step: .1,
				label: "Evolution Speed",
				group: "Animation"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms, childNode, onCleanup } = params;
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const time$1 = createAnimatedTime(params, uniforms.speed).mul(.1);
		const evolutionTime = createAnimatedTime(params, uniforms.evolutionSpeed).mul(.1);
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(screenUV.x.mul(aspect), screenUV.y);
		const scale = uniforms.detail.uniform;
		const strength = uniforms.strength.uniform;
		const drift1 = time$1.mul(.7);
		const evolve1 = evolutionTime.mul(.2);
		const angle1 = mx_noise_float(vec3(aspectCorrectedUV.x.mul(scale).add(drift1.mul(.3)), aspectCorrectedUV.y.mul(scale).add(drift1.mul(.5)), evolve1)).mul(Math.PI * 2);
		const flow1 = vec2(cos(angle1), sin(angle1));
		const drift2 = time$1.mul(.5);
		const evolve2 = evolutionTime.mul(.15);
		const angle2 = mx_noise_float(vec3(aspectCorrectedUV.x.mul(scale.mul(2.3)).add(drift2.mul(-.4)).add(173.5), aspectCorrectedUV.y.mul(scale.mul(2.3)).add(drift2.mul(.3)).add(291.7), evolve2.add(50))).mul(Math.PI * 2);
		const flow2 = vec2(cos(angle2), sin(angle2));
		const drift3 = time$1.mul(.3);
		const evolve3 = evolutionTime.mul(.1);
		const angle3 = mx_noise_float(vec3(aspectCorrectedUV.x.mul(scale.mul(4.7)).add(drift3.mul(.2)).add(527.3), aspectCorrectedUV.y.mul(scale.mul(4.7)).add(drift3.mul(-.6)).add(839.1), evolve3.add(100))).mul(Math.PI * 2);
		const flow3 = vec2(cos(angle3), sin(angle3));
		const combinedFlow = flow1.add(flow2.mul(.5)).add(flow3.mul(.25));
		const flowLength = max(length(combinedFlow), float(.001));
		const normalizedFlow = combinedFlow.div(flowLength);
		const finalFlow = vec2(normalizedFlow.x.mul(strength).div(aspect), normalizedFlow.y.mul(strength));
		const distortedUV = vec2(screenUV.x.add(finalFlow.x), screenUV.y.add(finalFlow.y));
		return unpremultiplyAlpha(applyEdgeHandling(distortedUV, childTexture.sample(distortedUV), childTexture, uniforms.edges.uniform.value));
	}
};
var FlowField_default = componentDefinition;
export { componentDefinition as n, FlowField_default as t };
