// shaders/dist/core/FlowingGradient-BRQ_K-k3.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { Fn, abs, clamp, cos, float, mix, mx_noise_float, pow, screenUV, sin, smoothstep, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "FlowingGradient",
	category: "Textures",
	description: "Liquid silk gradient with organic flowing color bands",
	props: {
		colorA: {
			default: "#0a0015",
			transform: transformColor,
			description: "Deep background color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#6b17e6",
			transform: transformColor,
			description: "Primary accent color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		colorC: {
			default: "#ff4d6a",
			transform: transformColor,
			description: "Secondary accent color",
			ui: {
				type: "color",
				label: "Color C",
				group: "Colors"
			}
		},
		colorD: {
			default: "#ff6b35",
			transform: transformColor,
			description: "Tertiary accent color",
			ui: {
				type: "color",
				label: "Color D",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "oklch",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 10,
				step: .1,
				label: "Speed",
				group: "Effect"
			}
		},
		distortion: {
			default: .5,
			description: "Organic distortion intensity",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "Distortion",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for variation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const colorA = uniforms.colorA.uniform;
		const colorB = uniforms.colorB.uniform;
		const colorC = uniforms.colorC.uniform;
		const colorD = uniforms.colorD.uniform;
		const csMode = uniforms.colorSpace.uniform.value;
		const distortion = uniforms.distortion.uniform;
		const seed = uniforms.seed.uniform;
		const t = createAnimatedTime(params, uniforms.speed).mul(.075);
		const aspect = viewportSize.x.div(viewportSize.y);
		return Fn(() => {
			const uv$1 = screenUV.sub(.5).toVar();
			const seedOff = seed.mul(.17);
			const angle = mx_noise_float(vec3(uv$1.x.mul(.8).add(t.mul(.4)), uv$1.y.mul(.8).sub(t.mul(.25)), seedOff)).mul(4 * Math.PI).mul(distortion);
			const uvCorr = vec2(uv$1.x, uv$1.y.div(aspect));
			const ca = cos(angle);
			const sa = sin(angle);
			uv$1.assign(vec2(uvCorr.x.mul(ca).sub(uvCorr.y.mul(sa)), uvCorr.x.mul(sa).add(uvCorr.y.mul(ca)).mul(aspect)));
			const ws = t.mul(2);
			uv$1.x.assign(uv$1.x.add(sin(uv$1.y.mul(5).add(ws)).div(50).mul(distortion)));
			uv$1.y.assign(uv$1.y.add(sin(uv$1.x.mul(7.5).add(ws)).div(25).mul(distortion)));
			const wp1 = vec3(uv$1.x.mul(2), uv$1.y.mul(2), t.mul(.5).add(seedOff));
			const wx1 = mx_noise_float(wp1.add(vec3(5.2, 1.3, 0)));
			const wy1 = mx_noise_float(wp1.add(vec3(1.7, 9.2, 0)));
			uv$1.assign(uv$1.add(vec2(wx1, wy1).mul(.25).mul(distortion)));
			const wp2 = vec3(uv$1.x.mul(1.5), uv$1.y.mul(1.5), t.mul(.3).add(seedOff));
			const wx2 = mx_noise_float(wp2.add(vec3(8.3, 2.8, 0)));
			const wy2 = mx_noise_float(wp2.add(vec3(3.1, 7.4, 0)));
			uv$1.assign(uv$1.add(vec2(wx2, wy2).mul(.2).mul(distortion)));
			const zoneNoise = mx_noise_float(vec3(uv$1.x.mul(1.8), uv$1.y.mul(1.8), t.mul(.3).add(seedOff)));
			const edgeHardness = pow(wx1.mul(.5).add(.5), float(1.5));
			const edgeWidth = mix(float(.5), float(.008), edgeHardness);
			const t1 = smoothstep(edgeWidth.negate(), edgeWidth, zoneNoise);
			const t2 = smoothstep(edgeWidth.negate(), edgeWidth, wy1);
			const baseColor = mixColors(mixColors(colorA, colorB, t1, csMode), mixColors(colorC, colorD, t1, csMode), t2, csMode);
			const foldMask1 = float(1).sub(abs(t1.mul(2).sub(1)));
			const foldMask2 = float(1).sub(abs(t2.mul(2).sub(1)));
			const foldHighlight = foldMask1.max(foldMask2).mul(edgeHardness);
			const surfaceLight = wx2.mul(.5).add(.5).mul(.12).add(.94);
			const baseAlpha = baseColor.w;
			return vec4(clamp(baseColor.rgb.mul(surfaceLight).mul(float(1).add(foldHighlight.mul(.35))), 0, 1), baseAlpha);
		})();
	}
};
var FlowingGradient_default = componentDefinition;
export { componentDefinition as n, FlowingGradient_default as t };
