// shaders/dist/core/FractalNoise-Ccufnbw6.js
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { clamp, cos, float, mx_noise_float, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
var GOLDEN = 2.39996;
const componentDefinition = {
	name: "FractalNoise",
	category: "Textures",
	description: "Multi-octave fractal Brownian motion noise texture with true noise evolution",
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "First color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Second color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		octaves: {
			default: 4,
			compileTime: true,
			description: "Number of noise octaves (more = more detail)",
			ui: {
				type: "range",
				min: 1,
				max: 8,
				step: 1,
				label: "Octaves",
				group: "Effect"
			}
		},
		detail: {
			default: 2,
			description: "How much finer each successive octave becomes",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 4,
				step: .1,
				label: "Detail",
				group: "Effect"
			}
		},
		contrast: {
			default: .5,
			description: "How strongly finer octaves contribute — higher values create more texture contrast",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 1,
				step: .05,
				label: "Contrast",
				group: "Effect"
			}
		},
		speed: {
			default: .15,
			description: "Speed at which the noise pattern evolves in place",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Speed",
				group: "Animation"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Rotation angle in degrees",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for pattern variation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const angleRad = uniforms.angle.uniform.mul(.0174533);
		const cosA = cos(angleRad);
		const sinA = sin(angleRad);
		const au = uv$1.x.mul(aspect).sub(aspect.mul(.5));
		const av = uv$1.y.sub(.5);
		const ru = au.mul(cosA).add(av.mul(sinA));
		const rv = au.mul(sinA).negate().add(av.mul(cosA));
		const animTime = createAnimatedTime(params, uniforms.speed);
		const scale = float(2);
		const seedVal = uniforms.seed.uniform;
		const l = uniforms.detail.uniform;
		const p = uniforms.contrast.uniform;
		const d = (i) => ({
			x: Math.cos(i * GOLDEN),
			y: Math.sin(i * GOLDEN)
		});
		const coord = (octaveScale, i) => vec2(ru.mul(octaveScale).add(animTime.mul(float(d(i).x))).add(seedVal), rv.mul(octaveScale).add(animTime.mul(float(d(i).y))).add(seedVal.mul(.7)));
		const n1 = mx_noise_float(coord(scale, 0));
		const n2 = mx_noise_float(coord(scale.mul(l), 1)).mul(p);
		const n3 = mx_noise_float(coord(scale.mul(l).mul(l), 2)).mul(p).mul(p);
		const n4 = mx_noise_float(coord(scale.mul(l).mul(l).mul(l), 3)).mul(p).mul(p).mul(p);
		const n5 = mx_noise_float(coord(scale.mul(l).mul(l).mul(l).mul(l), 4)).mul(p).mul(p).mul(p).mul(p);
		const n6 = mx_noise_float(coord(scale.mul(l).mul(l).mul(l).mul(l).mul(l), 5)).mul(p).mul(p).mul(p).mul(p).mul(p);
		const n7 = mx_noise_float(coord(scale.mul(l).mul(l).mul(l).mul(l).mul(l).mul(l), 6)).mul(p).mul(p).mul(p).mul(p).mul(p).mul(p);
		const n8 = mx_noise_float(coord(scale.mul(l).mul(l).mul(l).mul(l).mul(l).mul(l).mul(l), 7)).mul(p).mul(p).mul(p).mul(p).mul(p).mul(p).mul(p);
		const octaveCount = uniforms.octaves.uniform.value;
		let accumulated = n1;
		let totalWeight = float(1);
		if (octaveCount >= 2) {
			accumulated = accumulated.add(n2);
			totalWeight = totalWeight.add(p);
		}
		if (octaveCount >= 3) {
			accumulated = accumulated.add(n3);
			totalWeight = totalWeight.add(p.mul(p));
		}
		if (octaveCount >= 4) {
			accumulated = accumulated.add(n4);
			totalWeight = totalWeight.add(p.mul(p).mul(p));
		}
		if (octaveCount >= 5) {
			accumulated = accumulated.add(n5);
			totalWeight = totalWeight.add(p.mul(p).mul(p).mul(p));
		}
		if (octaveCount >= 6) {
			accumulated = accumulated.add(n6);
			totalWeight = totalWeight.add(p.mul(p).mul(p).mul(p).mul(p));
		}
		if (octaveCount >= 7) {
			accumulated = accumulated.add(n7);
			totalWeight = totalWeight.add(p.mul(p).mul(p).mul(p).mul(p).mul(p));
		}
		if (octaveCount >= 8) {
			accumulated = accumulated.add(n8);
			totalWeight = totalWeight.add(p.mul(p).mul(p).mul(p).mul(p).mul(p).mul(p));
		}
		const normalized = clamp(accumulated.div(totalWeight).mul(.5).add(.5), float(0), float(1));
		const result = mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, normalized, uniforms.colorSpace.uniform.value);
		return vec4(result.rgb, result.a);
	}
};
var FractalNoise_default = componentDefinition;
export { componentDefinition as n, FractalNoise_default as t };
