// shaders/dist/core/Glass-CXT1jdLw.js
import { a as transformColor, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { r as createGaussianBlurCompute } from "./computeBlur-CprKe0xa.js";
import { n as createAnalyticSdfSampler, r as createSvgSdfSampler } from "./sdf-D_EAGXhN.js";
import { Fn, If, clamp, convertToTexture, dot, exp2, float, max, mix, pow, screenUV, smoothstep, sqrt, texture, uniform, vec2, vec3, vec4, viewportSize } from "three/tsl";
var GOLDEN_ANGLE = 2.3999632297286535;
var BLUR_SAMPLE_COUNT = 9;
const BLUR_DISK = [];
for (let i = 0; i < BLUR_SAMPLE_COUNT; i++) {
	const angle = i * GOLDEN_ANGLE;
	const r = Math.sqrt(i / BLUR_SAMPLE_COUNT);
	BLUR_DISK.push([Math.cos(angle) * r, Math.sin(angle) * r]);
}
var INV_HLEN_GLASS = 1 / Math.sqrt(5);
function applyGlassEffect(childTexture, sdfSampler, uniforms, onBeforeRender, blurredTexture) {
	const lx = uniform(0);
	const ly = uniform(0);
	const updateLightAngle = () => {
		const rad = uniforms.lightAngle.uniform.value * Math.PI / 180;
		lx.value = Math.cos(rad);
		ly.value = Math.sin(rad);
	};
	updateLightAngle();
	onBeforeRender?.(updateLightAngle);
	return Fn(() => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const sharp = uniforms.edgeSoftness.uniform.mul(float(.5)).max(float(.001));
		const pixelSize = vec2(1).div(viewportSize);
		const pxH = float(1).div(viewportSize.y);
		const centerPos = vec2(uniforms.center.uniform.x, uniforms.center.uniform.y.oneMinus());
		const scale = uniforms.scale.uniform;
		const sdfUV = vec2(uv$1.x.sub(centerPos.x).mul(aspect).div(scale).add(float(.5)), uv$1.y.sub(centerPos.y).div(scale).add(float(.5)));
		const sdfRaw = sdfSampler(sdfUV).r;
		const sdf = sdfRaw.div(scale);
		const output = vec4(0).toVar();
		If(sdf.greaterThan(pxH.mul(float(2))), () => {
			const baseSample = childTexture.sample(uv$1);
			const earlyAlpha = uniforms.cutout.uniform.equal(float(1)).select(float(0), baseSample.a);
			output.assign(vec4(baseSample.r, baseSample.g, baseSample.b, earlyAlpha));
		}).Else(() => {
			const EPS = float(.01);
			const gradX = sdfSampler(sdfUV.add(vec2(EPS, float(0)))).r.sub(sdfRaw).div(EPS);
			const gradY = sdfSampler(sdfUV.add(vec2(float(0), EPS))).r.sub(sdfRaw).div(EPS);
			const rb1 = clamp(sdf.negate().div(sharp).mul(float(32)), float(0), float(1));
			const rb2base = clamp(sdf.sub(pxH).negate().div(sharp).mul(float(16)), float(0), float(1)).sub(clamp(sdf.negate().div(sharp).mul(float(16)), float(0), float(1)));
			const lightFacing = gradX.mul(lx).add(gradY.mul(ly)).mul(float(.5)).add(float(.5));
			const rb2 = rb2base.mul(lightFacing).mul(uniforms.highlight.uniform);
			const thicknessRange = uniforms.thickness.uniform.mul(float(.3)).max(float(.005));
			const depthNorm = clamp(sdf.negate().div(thicknessRange), float(0), float(1));
			const refrStrength = float(1).sub(depthNorm).mul(float(1).sub(depthNorm));
			const refrScale = uniforms.refraction.uniform.mul(float(.15));
			const offsetX = gradX.negate().mul(refrScale).mul(refrStrength).div(aspect);
			const offsetY = gradY.negate().mul(refrScale).mul(refrStrength);
			const lensUV = centerPos.add(uv$1.sub(centerPos).div(uniforms.innerZoom.uniform)).add(vec2(offsetX, offsetY));
			const blurEnabled = typeof uniforms.blur.uniform.value === "number" ? uniforms.blur.uniform.value > 0 : true;
			const aberrationEnabled = typeof uniforms.aberration.uniform.value === "number" ? uniforms.aberration.uniform.value > 0 : true;
			let blurred;
			if (blurredTexture && blurEnabled) {
				const blurTex = texture(blurredTexture);
				if (!aberrationEnabled) blurred = blurTex.sample(lensUV);
				else {
					const chrScale = uniforms.aberration.uniform.mul(float(.06));
					const chrOffX = offsetX.mul(chrScale);
					const chrOffY = offsetY.mul(chrScale);
					const rUV = lensUV.add(vec2(chrOffX, chrOffY));
					const bUV = lensUV.sub(vec2(chrOffX, chrOffY));
					const rSample = blurTex.sample(rUV);
					const gSample = blurTex.sample(lensUV);
					const bSample = blurTex.sample(bUV);
					blurred = vec4(rSample.r, gSample.g, bSample.b, gSample.a);
				}
			} else if (!blurEnabled && !aberrationEnabled) blurred = childTexture.sample(lensUV);
			else if (!blurEnabled && aberrationEnabled) {
				const chrScale = uniforms.aberration.uniform.mul(float(.06));
				const chrOffX = offsetX.mul(chrScale);
				const chrOffY = offsetY.mul(chrScale);
				const rUV = lensUV.add(vec2(chrOffX, chrOffY));
				const bUV = lensUV.sub(vec2(chrOffX, chrOffY));
				const rSample = childTexture.sample(rUV);
				const gSample = childTexture.sample(lensUV);
				const bSample = childTexture.sample(bUV);
				blurred = vec4(rSample.r, gSample.g, bSample.b, gSample.a);
			} else if (blurEnabled && !aberrationEnabled) {
				const acc = vec4(0).toVar();
				for (const [ox, oy] of BLUR_DISK) {
					const blurOff = pixelSize.mul(vec2(float(ox), float(oy))).mul(uniforms.blur.uniform.mul(2));
					acc.assign(acc.add(childTexture.sample(lensUV.add(blurOff))));
				}
				blurred = acc.mul(float(1 / BLUR_SAMPLE_COUNT));
			} else {
				const chrScale = uniforms.aberration.uniform.mul(float(.06));
				const chrOffX = offsetX.mul(chrScale);
				const chrOffY = offsetY.mul(chrScale);
				const rUV = lensUV.add(vec2(chrOffX, chrOffY));
				const bUV = lensUV.sub(vec2(chrOffX, chrOffY));
				const accR = vec4(0).toVar();
				const accG = vec4(0).toVar();
				const accB = vec4(0).toVar();
				for (const [ox, oy] of BLUR_DISK) {
					const blurOff = pixelSize.mul(vec2(float(ox), float(oy))).mul(uniforms.blur.uniform.mul(2));
					accR.assign(accR.add(childTexture.sample(rUV.add(blurOff))));
					accG.assign(accG.add(childTexture.sample(lensUV.add(blurOff))));
					accB.assign(accB.add(childTexture.sample(bUV.add(blurOff))));
				}
				const samplesInv = float(1 / BLUR_SAMPLE_COUNT);
				blurred = vec4(accR.mul(samplesInv).r, accG.mul(samplesInv).g, accB.mul(samplesInv).b, accG.mul(samplesInv).a);
			}
			const tintCol = vec3(uniforms.tintColor.uniform.x, uniforms.tintColor.uniform.y, uniforms.tintColor.uniform.z);
			const tintAmt = uniforms.tintIntensity.uniform;
			const blurredRgb = vec3(blurred.r, blurred.g, blurred.b);
			const lumWeights = vec3(.299, .587, .114);
			const origLum = dot(blurredRgb, lumWeights);
			const tinted = mix(blurredRgb, tintCol, tintAmt);
			const tintedLum = dot(tinted, lumWeights);
			const lumPreserved = tinted.mul(origLum.div(max(tintedLum, float(1e-4))));
			const tintedGlass = vec4(uniforms.tintPreserveLuminosity.uniform.equal(float(1)).select(lumPreserved, tinted), blurred.a);
			const nz = float(2);
			const nLen = sqrt(gradX.mul(gradX).add(gradY.mul(gradY)).add(nz.mul(nz)));
			const nx = gradX.div(nLen);
			const ny = gradY.div(nLen);
			const nnz = nz.div(nLen);
			const nDotH = nx.mul(lx.mul(float(INV_HLEN_GLASS))).add(ny.mul(ly.mul(float(INV_HLEN_GLASS)))).add(nnz.mul(float(2 * INV_HLEN_GLASS)));
			const shininess = exp2(float(8).sub(uniforms.highlightSoftness.uniform.mul(float(7))));
			const specGlint = pow(clamp(nDotH, float(0), float(1)), shininess).mul(uniforms.highlight.uniform).mul(refrStrength);
			const fresnelWidth = uniforms.fresnelSoftness.uniform.mul(float(.06)).max(float(.001));
			const fresnelDepth = clamp(sdf.negate().div(fresnelWidth), float(0), float(1));
			const fresnelRim = float(1).sub(fresnelDepth).mul(float(1).sub(fresnelDepth)).mul(uniforms.fresnel.uniform).mul(rb1);
			const hlCol = uniforms.highlightColor.uniform;
			const frCol = uniforms.fresnelColor.uniform;
			const lighting = tintedGlass.add(vec4(hlCol.x.mul(rb2), hlCol.y.mul(rb2), hlCol.z.mul(rb2), rb2)).add(vec4(hlCol.x.mul(specGlint), hlCol.y.mul(specGlint), hlCol.z.mul(specGlint), float(0))).add(vec4(frCol.x.mul(fresnelRim), frCol.y.mul(fresnelRim), frCol.z.mul(fresnelRim), float(0)));
			const transition = smoothstep(float(0), float(1), rb1);
			const isCutout = uniforms.cutout.uniform.equal(float(1));
			const composited = mix(childTexture.sample(uv$1), lighting, transition);
			const cutoutAlpha = isCutout.select(transition, float(1));
			output.assign(vec4(composited.r, composited.g, composited.b, composited.a.mul(cutoutAlpha)));
		});
		return output;
	})();
}
var transformBoolean = (value) => value ? 1 : 0;
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
const componentDefinition = {
	name: "Glass",
	category: "Shape Effects",
	description: "Optically realistic glass lens driven in a custom shape",
	requiresRTT: true,
	requiresChild: true,
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the glass shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		scale: {
			default: 1,
			description: "Scale of the glass shape (1 = default size)",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 3,
				step: .05,
				label: "Scale",
				group: "Position"
			}
		},
		cutout: {
			default: false,
			description: "Cut out the alpha outside the glass shape",
			transform: transformBoolean,
			ui: {
				type: "checkbox",
				label: "Cutout",
				group: "Glass"
			}
		},
		refraction: {
			default: 1,
			description: "Lens refraction — how aggressively the edges warp content beneath (0 = none, 1 = max)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .01,
				label: "Refraction",
				group: "Glass"
			}
		},
		edgeSoftness: {
			default: .1,
			description: "Edge softness — higher values give a wider, softer fade at the glass boundary",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Edge Softness",
				group: "Glass"
			}
		},
		blur: {
			default: 0,
			compileTimeWhen: (prev, next) => prev > 0 !== next > 0,
			description: "Frosted blur amount — 0 = clear glass, higher = frosted/diffuse",
			ui: {
				type: "range",
				min: 0,
				max: 20,
				step: .1,
				label: "Blur",
				group: "Glass"
			}
		},
		thickness: {
			default: .2,
			transform: (v) => v * .5,
			description: "Glass depth — how far inward from the edge the refraction extends",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Thickness",
				group: "Glass"
			}
		},
		aberration: {
			default: .5,
			compileTimeWhen: (prev, next) => prev > 0 !== next > 0,
			description: "Chromatic aberration — splits RGB channels along the refraction vector",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Aberration",
				group: "Glass"
			}
		},
		innerZoom: {
			default: 1,
			description: "Inner zoom level — magnifies content seen through the glass",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 3,
				step: .05,
				label: "Inner Zoom",
				group: "Glass"
			}
		},
		lightAngle: {
			default: 300,
			description: "Light angle in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Light Angle",
				group: "Highlight"
			}
		},
		highlight: {
			default: .05,
			description: "Directional edge highlight — bright rim on the light-facing boundary",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .05,
				label: "Highlight",
				group: "Highlight"
			}
		},
		highlightColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color of the directional edge highlight and specular glint",
			ui: {
				type: "color",
				label: "Highlight Color",
				group: "Highlight"
			}
		},
		highlightSoftness: {
			default: .5,
			description: "Specular highlight softness",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Highlight Softness",
				group: "Highlight"
			}
		},
		fresnel: {
			default: .1,
			description: "Fresnel rim glow — a soft luminous halo around the glass boundary",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Fresnel",
				group: "Fresnel"
			}
		},
		fresnelSoftness: {
			default: .1,
			description: "Fresnel rim width — higher values spread the glow further inward",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Fresnel Softness",
				group: "Fresnel"
			}
		},
		fresnelColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color of the fresnel rim glow",
			ui: {
				type: "color",
				label: "Fresnel Color",
				group: "Fresnel"
			}
		},
		tintColor: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color tint applied to the internal directional gradient",
			ui: {
				type: "color",
				label: "Tint Color",
				group: "Tint"
			}
		},
		tintIntensity: {
			default: 0,
			description: "Intensity of the color tint applied to the glass interior",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Tint Intensity",
				group: "Tint"
			}
		},
		tintPreserveLuminosity: {
			default: true,
			transform: transformBoolean,
			description: "Preserve original brightness when tinting",
			ui: {
				type: "checkbox",
				label: "Preserve Luminosity",
				group: "Tint"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file — when non-empty, activates SVG mode and triggers a shader recompile"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type — triggers recompile when shape is switched. When empty, derived from shape JSON at mount time."
		}
	},
	computeNode: ({ uniforms, childNode, onCleanup, dimensions, renderer, onResize }) => {
		if (!childNode) return null;
		const blurValue = uniforms.blur.uniform.value;
		if (typeof blurValue === "number" && blurValue <= 0) return null;
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const dpr = renderer?.getPixelRatio?.() ?? 1;
		const blur = createGaussianBlurCompute(childTexture, Math.round(dimensions.width * dpr), Math.round(dimensions.height * dpr), onCleanup);
		const { computeNodes, outputTexture, updateRadius } = blur;
		onResize(({ width, height }) => {
			blur.setInputDimensions(Math.round(width * dpr), Math.round(height * dpr));
		});
		return {
			outputs: {
				childTexture,
				blurredTexture: outputTexture
			},
			getComputeNodes: () => {
				const currentBlur = uniforms.blur.uniform.value;
				updateRadius(currentBlur * 2);
				return computeNodes;
			}
		};
	},
	fragmentNode: ({ uniforms, childNode, onCleanup, onBeforeRender, computeOutputs }) => {
		if (!childNode) return vec4(0);
		const childTexture = computeOutputs?.childTexture ?? convertToTexture(childNode);
		if (!computeOutputs) onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const blurredTexture = computeOutputs?.blurredTexture;
		const shapeSdfUrl = uniforms.shapeSdfUrl.uniform.value;
		const initialConfig = (() => {
			const val = uniforms.shape.uniform.value;
			if (typeof val === "object" && val !== null) return val;
			try {
				return JSON.parse(val);
			} catch {
				return {
					type: "circleSDF",
					radius: .35,
					centerX: .5,
					centerY: .5
				};
			}
		})();
		if (shapeSdfUrl) return applyGlassEffect(childTexture, createSvgSdfSampler(shapeSdfUrl, onBeforeRender, onCleanup), uniforms, onBeforeRender, blurredTexture);
		else return applyGlassEffect(childTexture, createAnalyticSdfSampler(uniforms.shapeType.uniform.value || initialConfig.type || "circleSDF", initialConfig, uniforms.shape, onBeforeRender), uniforms, onBeforeRender, blurredTexture);
	}
};
var Glass_default = componentDefinition;
export { componentDefinition as n, Glass_default as t };
