// shaders/dist/core/GlassTiles-CXJHjEBe.js
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { add, convertToTexture, cos, div, float, floor, mul, screenUV, sin, sub, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "GlassTiles",
	category: "Distortions",
	description: "Refraction-like distortion in a tile grid pattern",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 2,
			description: "The intensity of the glass tiles effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 10,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		tileCount: {
			default: 20,
			description: "Number of tiles across the longest dimension",
			ui: {
				type: ["range", "map"],
				min: 5,
				max: 50,
				step: 1,
				label: "Tile Count",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation angle of the tile grid in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		},
		roundness: {
			default: 0,
			description: "Makes tiles more circular instead of square",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Roundness",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Glass Tiles shader.");
			return vec4(0);
		}
		const texture$1 = convertToTexture(childNode);
		onCleanup(() => {
			if (texture$1?.renderTarget?.dispose) texture$1.renderTarget.dispose();
		});
		const aspectRatioUniform = viewportSize.x.div(viewportSize.y);
		const baseUV = screenUV;
		const intensity = uniforms.intensity.uniform;
		const baseTileCount = uniforms.tileCount.uniform;
		const rotationDegrees = uniforms.rotation.uniform;
		const roundnessAmount = uniforms.roundness.uniform;
		const isWide = aspectRatioUniform.greaterThan(float(1));
		const tileCount = vec2(isWide.select(baseTileCount, mul(baseTileCount, aspectRatioUniform)), isWide.select(div(baseTileCount, aspectRatioUniform), baseTileCount));
		const aspectCorrectedUV = vec2(mul(baseUV.x, aspectRatioUniform), baseUV.y);
		const rotationRadians = mul(rotationDegrees, Math.PI / 180);
		const cosAngle = cos(rotationRadians);
		const sinAngle = sin(rotationRadians);
		const centered = sub(aspectCorrectedUV, vec2(mul(.5, aspectRatioUniform), .5));
		const rotatedUV = add(vec2(sub(mul(centered.x, cosAngle), mul(centered.y, sinAngle)), add(mul(centered.x, sinAngle), mul(centered.y, cosAngle))), vec2(mul(.5, aspectRatioUniform), .5));
		const gridUV = vec2(div(rotatedUV.x, aspectRatioUniform), rotatedUV.y);
		const fromCenter = sub(div(sub(gridUV, div(floor(mul(gridUV, tileCount)), tileCount)), div(vec2(1), tileCount)), vec2(.5, .5));
		const clampedRoundness = sub(1, mul(mul(fromCenter.x, fromCenter.x).add(mul(fromCenter.y, fromCenter.y)), mul(roundnessAmount, 4))).max(0);
		const baseDistortion = mul(fromCenter, mul(mul(intensity, .025), clampedRoundness));
		const distortedUV = add(baseUV, vec2(div(baseDistortion.x, aspectRatioUniform), baseDistortion.y));
		return unpremultiplyAlpha(vec4(texture$1.sample(distortedUV)));
	}
};
var GlassTiles_default = componentDefinition;
export { componentDefinition as n, GlassTiles_default as t };
