// shaders/dist/core/Glitch-Bg9W-KNm.js
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { Fn, abs, clamp, convertToTexture, float, floor, fract, mix, screenUV, sin, step, time, vec2, vec3, vec4, viewportSize } from "three/tsl";
var hash11 = Fn(([p]) => {
	return fract(sin(p.mul(78.233)).mul(43758.5453));
});
const componentDefinition = {
	name: "Glitch",
	category: "Stylize",
	description: "Digital glitch that melts pixels and distorts colors",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: .5,
			description: "Overall glitch strength and frequency of glitch bursts",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Intensity",
				group: "Effect"
			}
		},
		speed: {
			default: 1,
			description: "How fast the glitch pattern evolves",
			ui: {
				type: "range",
				min: .1,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Effect"
			}
		},
		rgbShift: {
			default: 5,
			description: "Amount of chromatic aberration (RGB channel splitting)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .5,
				label: "RGB Shift",
				group: "Effect"
			}
		},
		blockDensity: {
			default: 10,
			description: "Base number of horizontal glitch bands",
			ui: {
				type: ["range", "map"],
				min: 2,
				max: 50,
				step: 1,
				label: "Block Density",
				group: "Effect"
			}
		},
		colorBarIntensity: {
			default: .2,
			description: "Intensity of vivid neon color bar overlay in glitch regions",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Color Bars",
				group: "Effect"
			}
		},
		mirrorAmount: {
			default: .3,
			description: "Chance of glitch blocks showing mirrored/flipped content",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Mirror",
				group: "Effect"
			}
		},
		scanlineIntensity: {
			default: .2,
			description: "Visibility of CRT-style horizontal scanlines in distorted areas",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Scanlines",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const intensity = uniforms.intensity.uniform;
		const speed = uniforms.speed.uniform;
		const rgbShiftAmount = uniforms.rgbShift.uniform;
		const blockDensity = uniforms.blockDensity.uniform;
		const colorBarMix = uniforms.colorBarIntensity.uniform;
		const mirrorChance = uniforms.mirrorAmount.uniform;
		const scanlines = uniforms.scanlineIntensity.uniform;
		const t = time.mul(speed);
		const slowFrame = floor(t.mul(2.5));
		const slowHash = hash11(slowFrame);
		const slowPulse = slowHash.mul(slowHash).mul(slowHash);
		const fastFrame = floor(t.mul(10));
		const fastHash = hash11(fastFrame.add(73));
		const fastPulse = fastHash.mul(fastHash);
		const threshold = float(1).sub(intensity);
		const majorBurst = step(threshold, slowPulse);
		const minorFlicker = step(threshold.add(.15), fastPulse);
		const glitchStrength = clamp(majorBurst.mul(.8).add(minorFlicker.mul(.4)), 0, 1).mul(intensity);
		const bandY1 = floor(screenUV.y.mul(blockDensity));
		const bandSeed1 = bandY1.mul(.37).add(slowFrame.mul(7.13));
		const bandHash1 = hash11(bandSeed1);
		const bandActive1 = step(float(1).sub(glitchStrength.mul(.5)), bandHash1);
		const displace1 = hash11(bandSeed1.add(42)).sub(.5).mul(.25).mul(bandActive1);
		const bandSeed2 = floor(screenUV.y.mul(blockDensity.mul(2))).mul(.53).add(fastFrame.mul(3.37));
		const bandHash2 = hash11(bandSeed2);
		const bandActive2 = step(float(1).sub(glitchStrength.mul(.7)), bandHash2);
		const displace2 = hash11(bandSeed2.add(99)).sub(.5).mul(.12).mul(bandActive2);
		const bandSeed3 = floor(screenUV.y.mul(blockDensity.mul(4))).mul(.71).add(fastFrame.mul(11.71));
		const bandHash3 = hash11(bandSeed3);
		const bandActive3 = step(float(1).sub(glitchStrength.mul(.9)), bandHash3);
		const displace3 = hash11(bandSeed3.add(173)).sub(.5).mul(.05).mul(bandActive3);
		const colX = floor(screenUV.x.mul(4));
		const blockSeed = colX.mul(.41).add(bandY1.mul(.67)).add(slowFrame.mul(2.31));
		const blockHash = hash11(blockSeed);
		const blockActive = step(float(1).sub(glitchStrength.mul(.35)), blockHash);
		const blockDisplace = hash11(blockSeed.add(77)).sub(.5).mul(.18).mul(blockActive);
		const totalDisplaceX = displace1.add(displace2).add(displace3).add(blockDisplace);
		const vertJitter = bandActive1.mul(hash11(bandSeed1.add(200)).sub(.5)).mul(.012);
		const displacedUV = vec2(clamp(screenUV.x.add(totalDisplaceX.mul(glitchStrength)), 0, 1), clamp(screenUV.y.add(vertJitter.mul(glitchStrength)), 0, 1));
		const mirrorSeed = colX.mul(.83).add(bandY1.mul(1.17)).add(slowFrame.mul(4.57));
		const hMirrorHash = hash11(mirrorSeed);
		const vMirrorHash = hash11(mirrorSeed.add(500));
		const hMirrorActive = step(float(1).sub(mirrorChance.mul(glitchStrength)), hMirrorHash).mul(bandActive1);
		const vMirrorActive = step(float(1).sub(mirrorChance.mul(glitchStrength).mul(.4)), vMirrorHash).mul(bandActive1);
		const mirroredUV = vec2(mix(displacedUV.x, float(1).sub(displacedUV.x), hMirrorActive), mix(displacedUV.y, float(1).sub(displacedUV.y), vMirrorActive));
		const rgbOffset = rgbShiftAmount.mul(.003).mul(glitchStrength.add(abs(totalDisplaceX).mul(2)));
		const redUV = vec2(clamp(mirroredUV.x.add(rgbOffset), 0, 1), mirroredUV.y);
		const blueUV = vec2(clamp(mirroredUV.x.sub(rgbOffset), 0, 1), mirroredUV.y);
		const rSample = childTexture.sample(redUV).r;
		const gSample = childTexture.sample(mirroredUV).g;
		const bSample = childTexture.sample(blueUV).b;
		const glitchedColor = vec3(rSample, gSample, bSample);
		const fillStyleHash = hash11(blockSeed.add(555));
		const barOffset = hash11(bandY1.add(slowFrame.mul(3.33)).add(333));
		const barIndex = floor(fract(screenUV.x.mul(3).add(barOffset.mul(7))).mul(7));
		const smpteColor = vec3(float(1).sub(step(float(2), barIndex)).add(step(float(4), barIndex).mul(float(1).sub(step(float(6), barIndex)))), float(1).sub(step(float(4), barIndex)), float(1).sub(step(float(.5), fract(barIndex.mul(.5)))));
		const sh6 = fract(hash11(bandY1.add(13)).mul(.8).add(slowFrame.mul(.07))).mul(6);
		const withBars = mix(glitchedColor, mix(vec3(clamp(abs(sh6.sub(3)).sub(1), 0, 1), clamp(float(2).sub(abs(sh6.sub(2))), 0, 1), clamp(float(2).sub(abs(sh6.sub(4))), 0, 1)).mul(1.3), smpteColor, step(float(.5), fillStyleHash)), bandActive1.mul(blockActive).mul(colorBarMix).mul(glitchStrength));
		const anyDistortion = clamp(bandActive1.add(bandActive2).add(bandActive3), 0, 1);
		const scanlineWave = sin(screenUV.y.mul(viewportSize.y).mul(float(3.14159265))).mul(.5).add(.5);
		const scanlineEffect = float(1).sub(scanlines.mul(scanlineWave).mul(.3).mul(anyDistortion));
		return unpremultiplyAlpha(vec4(withBars.mul(scanlineEffect), float(1)));
	}
};
var Glitch_default = componentDefinition;
export { componentDefinition as n, Glitch_default as t };
