// shaders/dist/core/Glow-5PP4-rM9.js
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { i as createVariableGaussianBlurCompute, n as DEFAULT_COMPUTE_WIDTH, t as DEFAULT_COMPUTE_HEIGHT } from "./computeBlur-CprKe0xa.js";
import { StorageTexture } from "three/webgpu";
import { FloatType } from "three";
import { Fn, If, clamp, convertToTexture, dot, float, instanceIndex, int, max, mix, pow, screenUV, storageTexture, texture, textureLoad, textureStore, uint, uniform, uvec2, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Glow",
	category: "Stylize",
	description: "Soft glow effect with adjustable intensity",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 1,
			description: "Glow intensity (brightness of the glow effect)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 50,
				step: .5,
				label: "Intensity",
				group: "Effect"
			}
		},
		threshold: {
			default: .5,
			description: "Brightness threshold for glow extraction (lower = more glow)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Threshold",
				group: "Effect"
			}
		},
		size: {
			default: 25,
			description: "Glow spread in pixels (clean up to ~72px, mild banding above)",
			compileTimeWhen: (prev, next) => prev === 0 !== (next === 0),
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Glow Size",
				group: "Effect"
			}
		}
	},
	computeNode: ({ childNode, onCleanup, dimensions, renderer, uniforms, onResize }) => {
		if (!childNode) return null;
		if (uniforms.size.uniform.value === 0) return null;
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const dpr = renderer?.getPixelRatio?.() ?? 1;
		const texWidth = Math.round(dimensions.width * dpr);
		const texHeight = Math.round(dimensions.height * dpr);
		const computeWidth = DEFAULT_COMPUTE_WIDTH;
		const computeHeight = 640;
		const brightTex = new StorageTexture(computeWidth, computeHeight);
		brightTex.type = FloatType;
		const brightWriteNode = storageTexture(brightTex);
		const brightReadNode = storageTexture(brightTex).toReadOnly();
		onCleanup(() => {
			brightTex.dispose();
		});
		const { computeNodes: blurNodes, outputTexture: blurredTex, blurMapWriteNode } = createVariableGaussianBlurCompute(brightReadNode, computeWidth, computeHeight, onCleanup);
		const inputWidthU = uniform(texWidth);
		const inputHeightU = uniform(texHeight);
		const cw = uint(computeWidth);
		const ch = uint(computeHeight);
		const sizeMapInfo = uniforms.size.mapComputeInfo;
		const extractAndFill = Fn(() => {
			const idx = instanceIndex;
			const cx = idx.mod(cw);
			const cy = idx.div(cw);
			If(cy.lessThan(ch), () => {
				const u = cx.toFloat().add(float(.5)).div(float(computeWidth));
				const v = cy.toFloat().add(float(.5)).div(float(computeHeight));
				const srcX = u.mul(inputWidthU).round().toInt().clamp(int(0), int(inputWidthU.sub(1))).toUint();
				const srcY = v.mul(inputHeightU).round().toInt().clamp(int(0), int(inputHeightU.sub(1))).toUint();
				const sample = textureLoad(childTexture, uvec2(srcX, srcY), int(0));
				const lum = dot(sample.rgb, vec3(.2126, .7152, .0722));
				const mask = max(float(0), lum.sub(uniforms.threshold.uniform));
				textureStore(brightWriteNode, uvec2(cx, cy), vec4(sample.rgb.mul(mask), sample.a)).toWriteOnly();
				let radius;
				if (sizeMapInfo) {
					const { sourceTexture, channel, inputMin, inputMax, outputMin, outputMax, curve } = sizeMapInfo;
					const mapSample = textureLoad(sourceTexture, uvec2(srcX, srcY), int(0));
					let raw;
					if (channel === "alpha") raw = mapSample.a;
					else if (channel === "alphaInverted") raw = mapSample.a.oneMinus();
					else if (channel === "luminanceInverted") raw = dot(mapSample.rgb, vec3(.2126, .7152, .0722)).oneMinus();
					else raw = dot(mapSample.rgb, vec3(.2126, .7152, .0722));
					const inputRange = inputMax.sub(inputMin).max(float(1e-4));
					radius = mix(outputMin, outputMax, pow(clamp(raw.sub(inputMin).div(inputRange), float(0), float(1)), pow(float(2), curve.negate().mul(float(2)))));
				} else radius = uniforms.size.uniform;
				textureStore(blurMapWriteNode, uvec2(cx, cy), vec4(radius, float(0), float(0), float(1))).toWriteOnly();
			});
		})().compute(computeWidth * computeHeight, [256]);
		onResize(({ width, height }) => {
			const newW = Math.round(width * dpr);
			const newH = Math.round(height * dpr);
			inputWidthU.value = newW;
			inputHeightU.value = newH;
		});
		return {
			outputs: {
				childTexture,
				blurredTexture: blurredTex
			},
			getComputeNodes: () => [extractAndFill, ...blurNodes]
		};
	},
	fragmentNode: ({ uniforms, childNode, onCleanup, computeOutputs }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		if (uniforms.size.uniform.value === 0) return childNode;
		if (computeOutputs?.blurredTexture) {
			const original = computeOutputs.childTexture.sample(screenUV);
			const bloom = texture(computeOutputs.blurredTexture).sample(screenUV);
			const sizePin = uniforms.size.uniform.mul(float(0));
			const intensity$1 = uniforms.intensity.uniform;
			return unpremultiplyAlpha(vec4(original.rgb.add(bloom.rgb.mul(intensity$1)).add(sizePin), original.a));
		}
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const originalColor = childTexture.sample(screenUV);
		const weights = [
			.056,
			.135,
			.278,
			.487,
			.726,
			.923,
			1,
			.923,
			.726,
			.487,
			.278,
			.135,
			.056
		];
		const weightSum = 6.214;
		const horizontalTexture = convertToTexture(Fn(() => {
			const blurRadius = uniforms.size.uniform;
			const threshold = uniforms.threshold.uniform;
			const pixelSize = vec2(1).div(viewportSize);
			const total = vec4(0).toVar();
			for (let x = -6; x <= 6; x++) {
				const weight = float(weights[x + 6] / weightSum);
				const offset = vec2(float(x), float(0)).mul(blurRadius).mul(pixelSize);
				const sampleCoord = screenUV.add(offset);
				const sampleColor = childTexture.sample(sampleCoord);
				const luminance = dot(sampleColor.rgb, vec3(.299, .587, .114));
				const brightnessMask = max(float(0), luminance.sub(threshold));
				const brightColor = sampleColor.mul(brightnessMask);
				total.assign(total.add(brightColor.mul(weight)));
			}
			return total;
		})());
		onCleanup(() => {
			if (horizontalTexture?.renderTarget?.dispose) horizontalTexture.renderTarget.dispose();
		});
		const glowEffect = Fn(() => {
			const blurRadius = uniforms.size.uniform;
			const pixelSize = vec2(1).div(viewportSize);
			const total = vec4(0).toVar();
			for (let y = -6; y <= 6; y++) {
				const weight = float(weights[y + 6] / weightSum);
				const offset = vec2(float(0), float(y)).mul(blurRadius).mul(pixelSize);
				const sampleCoord = screenUV.add(offset);
				total.assign(total.add(horizontalTexture.sample(sampleCoord).mul(weight)));
			}
			return total;
		})();
		const intensity = uniforms.intensity.uniform;
		const intensifiedGlow = glowEffect.mul(intensity);
		return unpremultiplyAlpha(vec4(originalColor.rgb.add(intensifiedGlow.rgb), max(originalColor.a, intensifiedGlow.a)));
	}
};
var Glow_default = componentDefinition;
export { componentDefinition as n, Glow_default as t };
