// shaders/dist/core/Grid-DwyhjFAT.js
import { a as transformColor } from "./transformations-CJcUeZIC.js";
import { cos, float, floor, fract, fwidth, max, screenUV, sin, smoothstep, step, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Grid",
	category: "Textures",
	description: "Simple grid lines pattern with adjustable thickness and rotation",
	props: {
		color: {
			default: "#ffffff",
			transform: transformColor,
			description: "The color of the grid lines",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		cells: {
			default: 10,
			description: "Number of cells along the shortest canvas edge (creates square cells)",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 50,
				step: 1,
				label: "Cells",
				group: "Effect"
			}
		},
		thickness: {
			default: 1,
			description: "Thickness of grid lines (normalized, 0.0-1.0)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .1,
				label: "Thickness",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the grid in degrees. At 45° this produces a crosshatch/diamond pattern.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		}
	},
	mapSampleUVs: (uniforms) => {
		const aspect = viewportSize.x.div(viewportSize.y);
		const correctedUV = vec2(screenUV.x.mul(aspect), screenUV.y.oneMinus());
		const rotRad = uniforms.rotation.uniform.mul(float(Math.PI / 180));
		const cosR = cos(rotRad);
		const sinR = sin(rotRad);
		const center = vec2(aspect.mul(float(.5)), float(.5));
		const centered = correctedUV.sub(center);
		const cellCentered = floor(vec2(centered.x.mul(cosR).sub(centered.y.mul(sinR)), centered.x.mul(sinR).add(centered.y.mul(cosR))).add(center).mul(uniforms.cells.uniform)).add(.5).div(uniforms.cells.uniform).sub(center);
		const unrotated = vec2(cellCentered.x.mul(cosR).add(cellCentered.y.mul(sinR)), cellCentered.x.mul(sinR).negate().add(cellCentered.y.mul(cosR))).add(center);
		const cellCenterScreen = vec2(unrotated.x.div(aspect), unrotated.y.oneMinus());
		return {
			thickness: cellCenterScreen,
			cells: cellCenterScreen,
			rotation: cellCenterScreen
		};
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const correctedUV = vec2(uv$1.x.mul(aspect), uv$1.y.oneMinus());
		const rotRad = uniforms.rotation.uniform.mul(float(Math.PI / 180));
		const cosR = cos(rotRad);
		const sinR = sin(rotRad);
		const center = vec2(aspect.mul(float(.5)), float(.5));
		const centered = correctedUV.sub(center);
		const rotated = vec2(centered.x.mul(cosR).sub(centered.y.mul(sinR)), centered.x.mul(sinR).add(centered.y.mul(cosR))).add(center);
		const cells = uniforms.cells.uniform;
		const gridUV = rotated.mul(cells);
		const gridFract = fract(gridUV);
		const lineWidth = uniforms.thickness.uniform.mul(.01);
		const distToLineX = gridFract.x.min(gridFract.x.oneMinus());
		const distToLineY = gridFract.y.min(gridFract.y.oneMinus());
		const pixelSize = fwidth(gridUV);
		const gridMask = max(smoothstep(lineWidth.add(pixelSize.x), lineWidth, distToLineX), smoothstep(lineWidth.add(pixelSize.y), lineWidth, distToLineY)).mul(step(1e-4, lineWidth));
		return vec4(uniforms.color.uniform.rgb, uniforms.color.uniform.a.mul(gridMask));
	}
};
var Grid_default = componentDefinition;
export { componentDefinition as n, Grid_default as t };
