// shaders/dist/core/HexGrid-BUsgyF8P.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { abs, dot, float, fwidth, max, mod, screenUV, smoothstep, step, vec2, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "HexGrid",
	category: "Textures",
	description: "Honeycomb hexagonal grid pattern",
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "Cell fill color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Grid line color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		cells: {
			default: 8,
			description: "Number of hexagons across the shortest canvas edge",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 40,
				step: 1,
				label: "Cells",
				group: "Effect"
			}
		},
		thickness: {
			default: 1,
			description: "Thickness of the hex grid lines",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 10,
				step: .1,
				label: "Thickness",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const scaledUV = vec2(uv$1.x.mul(aspect), uv$1.y).mul(uniforms.cells.uniform);
		const s = vec2(float(Math.sqrt(3)), float(1));
		const sh = s.mul(float(.5));
		const a = mod(scaledUV, s).sub(sh);
		const b = mod(scaledUV.sub(sh), s).sub(sh);
		const gv = dot(a, a).lessThan(dot(b, b)).select(a, b);
		const pAbs = vec2(abs(gv.x), abs(gv.y));
		const hexDist = max(dot(pAbs, vec2(float(Math.sqrt(3) / 2), float(.5))), pAbs.y);
		const edgeDist = float(.5).sub(hexDist);
		const lineWidth = uniforms.thickness.uniform.mul(float(.01));
		const pixel = fwidth(scaledUV.x).mul(float(.5));
		const lineMask = smoothstep(lineWidth.add(pixel), lineWidth, edgeDist).mul(step(float(1e-4), lineWidth));
		return mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, lineMask, uniforms.colorSpace.uniform.value);
	}
};
var HexGrid_default = componentDefinition;
export { componentDefinition as n, HexGrid_default as t };
