// shaders/dist/core/HueShift-D3_5D7jx.js
import { cos, sin, vec3, vec4 } from "three/tsl";
const componentDefinition = {
	name: "HueShift",
	category: "Adjustments",
	description: "Rotate hue around the color wheel",
	requiresChild: true,
	props: { shift: {
		default: 0,
		description: "The amount to shift the hue by",
		compileTimeWhen: (prev, next) => prev === 0 !== (next === 0),
		ui: {
			type: ["range", "map"],
			min: -180,
			max: 180,
			step: 1,
			label: "Shift",
			group: "Appearance"
		},
		transform: (value) => value * Math.PI / 180
	} },
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Hue Shift shader.");
			return vec4(0);
		}
		if (uniforms.shift.uniform.value === 0) return childNode;
		const angle = uniforms.shift.uniform;
		const cosA = cos(angle);
		const sinA = sin(angle);
		const k = 1 / 3;
		const sqrt3 = 1.7320508075688772;
		const kOneMinusCos = cosA.oneMinus().mul(k);
		const sinOverSqrt3 = sinA.div(sqrt3);
		const m00 = cosA.add(kOneMinusCos);
		const m01 = kOneMinusCos.sub(sinOverSqrt3);
		const m02 = kOneMinusCos.add(sinOverSqrt3);
		const m10 = kOneMinusCos.add(sinOverSqrt3);
		const m11 = cosA.add(kOneMinusCos);
		const m12 = kOneMinusCos.sub(sinOverSqrt3);
		const m20 = kOneMinusCos.sub(sinOverSqrt3);
		const m21 = kOneMinusCos.add(sinOverSqrt3);
		const m22 = cosA.add(kOneMinusCos);
		const r = childNode.r;
		const g = childNode.g;
		const b = childNode.b;
		return vec4(vec3(r.mul(m00).add(g.mul(m01)).add(b.mul(m02)), r.mul(m10).add(g.mul(m11)).add(b.mul(m12)), r.mul(m20).add(g.mul(m21)).add(b.mul(m22))), childNode.a);
	}
};
var HueShift_default = componentDefinition;
export { componentDefinition as n, HueShift_default as t };
