// shaders/dist/core/Kaleidoscope-D4jRcZCt.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { atan, convertToTexture, cos, float, length, mod, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Kaleidoscope",
	category: "Distortions",
	description: "Create a kaleidoscope effect with radial mirrored segments",
	requiresRTT: true,
	requiresChild: true,
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the kaleidoscope effect",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		segments: {
			default: 6,
			description: "Number of radial segments in the kaleidoscope",
			ui: {
				type: ["range", "map"],
				min: 2,
				max: 24,
				step: 1,
				label: "Segments",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			description: "Rotation offset for the entire kaleidoscope pattern",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(screenUV.x.mul(aspect), screenUV.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus());
		const delta = aspectCorrectedUV.sub(centerPos);
		const radius = length(delta);
		const currentAngle = atan(delta.y, delta.x);
		const rotationRad = uniforms.angle.uniform.mul(Math.PI / 180);
		const rotatedAngle = currentAngle.add(rotationRad);
		const segmentAngle = float(Math.PI * 2).div(uniforms.segments.uniform);
		const wrappedAngle = mod(rotatedAngle, segmentAngle.mul(2));
		const finalAngle = wrappedAngle.greaterThan(segmentAngle).select(segmentAngle.mul(2).sub(wrappedAngle), wrappedAngle);
		const kaleidoscopeDelta = vec2(cos(finalAngle).mul(radius), sin(finalAngle).mul(radius));
		const kaleidoscopePos = centerPos.add(kaleidoscopeDelta);
		const finalUV = vec2(kaleidoscopePos.x.div(aspect), kaleidoscopePos.y);
		return unpremultiplyAlpha(applyEdgeHandling(finalUV, childTexture.sample(finalUV), childTexture, uniforms.edges.uniform.value));
	}
};
var Kaleidoscope_default = componentDefinition;
export { componentDefinition as n, Kaleidoscope_default as t };
