// shaders/dist/core/LensFlare-D7vypE2Q.js
import { l as transformPosition } from "./transformations-CJcUeZIC.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { Fn, abs, atan, clamp, cos, dot, float, max, mix, pow, screenUV, sin, smoothstep, vec2, vec3, vec4, viewportSize } from "three/tsl";
var spectralColor = Fn(([t]) => {
	const TAU = float(6.2831853);
	return vec3(cos(TAU.mul(t)).mul(.5).add(.5), cos(TAU.mul(t.sub(.333))).mul(.5).add(.5), cos(TAU.mul(t.sub(.666))).mul(.5).add(.5));
});
const componentDefinition = {
	name: "LensFlare",
	category: "Stylize",
	description: "Realistic camera lens flare with artifacts.",
	props: {
		lightPosition: {
			default: {
				x: .3,
				y: .3
			},
			transform: transformPosition,
			description: "Position of the light source",
			ui: {
				type: "position",
				label: "Light Position",
				group: "Light"
			}
		},
		intensity: {
			default: .5,
			description: "Master brightness of the entire lens flare effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .01,
				label: "Intensity",
				group: "Light"
			}
		},
		ghostIntensity: {
			default: .4,
			description: "Brightness of internal reflection ghost discs along the flare axis",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Ghosts",
				group: "Reflections"
			}
		},
		ghostSpread: {
			default: .7,
			description: "Spacing between ghost reflections along the flare axis",
			ui: {
				type: "range",
				min: .1,
				max: 2,
				step: .01,
				label: "Ghost Spread",
				group: "Reflections"
			}
		},
		ghostChroma: {
			default: .3,
			description: "Rainbow chromatic fringing around ghost element edges",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Ghost Chroma",
				group: "Reflections"
			}
		},
		haloIntensity: {
			default: .4,
			description: "Brightness of the circular halo ring from internal reflection",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Halo",
				group: "Reflections"
			}
		},
		haloRadius: {
			default: .6,
			description: "Radius of the halo ring",
			ui: {
				type: "range",
				min: .1,
				max: 1,
				step: .01,
				label: "Halo Size",
				group: "Reflections"
			}
		},
		haloChroma: {
			default: .6,
			description: "Spectral dispersion on the halo creating rainbow color separation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Halo Chroma",
				group: "Reflections"
			}
		},
		haloSoftness: {
			default: .8,
			description: "Thickness and softness of the halo ring",
			ui: {
				type: ["range", "map"],
				min: .01,
				max: 3,
				step: .01,
				label: "Halo Softness",
				group: "Reflections"
			}
		},
		starburstIntensity: {
			default: .3,
			description: "Brightness of diffraction spikes radiating from the light source",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Starburst",
				group: "Effects"
			}
		},
		starburstPoints: {
			default: 6,
			description: "Number of starburst spikes (simulates aperture blade count)",
			ui: {
				type: ["range", "map"],
				min: 4,
				max: 16,
				step: 1,
				label: "Blades",
				group: "Effects"
			}
		},
		streakIntensity: {
			default: .15,
			description: "Brightness of horizontal anamorphic light streak",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Streak",
				group: "Effects"
			}
		},
		streakLength: {
			default: .5,
			description: "Horizontal extent of the anamorphic streak",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 1,
				step: .01,
				label: "Streak Length",
				group: "Effects"
			}
		},
		glareIntensity: {
			default: .2,
			description: "Soft veiling glare that washes out contrast around the light",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Glare",
				group: "Effects"
			}
		},
		glareSize: {
			default: .5,
			description: "Size of the soft glare glow",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 1,
				step: .01,
				label: "Glare Size",
				group: "Effects"
			}
		},
		edgeFade: {
			default: .2,
			description: "How much the flare fades when the light source is near the screen edge (0 = no fade, 1 = heavy fade)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Fade",
				group: "Light"
			}
		},
		speed: {
			default: .5,
			description: "Speed of subtle flare shimmer and starburst rotation",
			ui: {
				type: "range",
				min: 0,
				max: 3,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const uIntensity = uniforms.intensity.uniform;
		const uGhostInt = uniforms.ghostIntensity.uniform;
		const uGhostSpread = uniforms.ghostSpread.uniform;
		const uGhostChroma = uniforms.ghostChroma.uniform;
		const uHaloInt = uniforms.haloIntensity.uniform;
		const uHaloRad = uniforms.haloRadius.uniform;
		const uHaloChroma = uniforms.haloChroma.uniform;
		const uHaloSoft = uniforms.haloSoftness.uniform;
		const uStarburstInt = uniforms.starburstIntensity.uniform;
		const uStarburstPts = uniforms.starburstPoints.uniform;
		const uStreakInt = uniforms.streakIntensity.uniform;
		const uStreakLen = uniforms.streakLength.uniform;
		const uGlareInt = uniforms.glareIntensity.uniform;
		const uGlareSz = uniforms.glareSize.uniform;
		const aspect = viewportSize.x.div(viewportSize.y);
		const uv$1 = screenUV;
		const lightPos = vec2(uniforms.lightPosition.uniform.x, uniforms.lightPosition.uniform.y.oneMinus());
		const flareAxis = vec2(.5, .5).sub(lightPos);
		const t = createAnimatedTime(params, uniforms.speed);
		const shimmer = sin(t.mul(3.7)).mul(.04).add(sin(t.mul(7.3)).mul(.03)).add(sin(t.mul(13.1)).mul(.02)).add(.97);
		const uEdgeFade = uniforms.edgeFade.uniform;
		const edgeDist = max(abs(lightPos.x.sub(.5)).mul(2), abs(lightPos.y.sub(.5)).mul(2));
		const fadeStart = float(1).sub(uEdgeFade.mul(.5));
		const edgeFade = mix(float(1), smoothstep(fadeStart, float(1), edgeDist).oneMinus(), uEdgeFade);
		const masterFade = uIntensity.mul(edgeFade).mul(shimmer);
		const toLightAR = vec2(uv$1.x.sub(lightPos.x).mul(aspect), uv$1.y.sub(lightPos.y));
		const lightDist = toLightAR.length();
		const ghostConfigs = [
			{
				offset: .15,
				size: .2,
				bright: .3,
				hollow: .1,
				r: 1,
				g: .95,
				b: .9
			},
			{
				offset: .3,
				size: .12,
				bright: .25,
				hollow: .65,
				r: 1,
				g: .85,
				b: .55
			},
			{
				offset: .5,
				size: .08,
				bright: .35,
				hollow: .15,
				r: .62,
				g: .82,
				b: 1
			},
			{
				offset: .72,
				size: .14,
				bright: .18,
				hollow: .75,
				r: .75,
				g: 1,
				b: .7
			},
			{
				offset: .95,
				size: .06,
				bright: .3,
				hollow: 0,
				r: 1,
				g: .6,
				b: .85
			},
			{
				offset: 1.25,
				size: .1,
				bright: .15,
				hollow: .8,
				r: .55,
				g: .88,
				b: 1
			},
			{
				offset: 1.55,
				size: .07,
				bright: .2,
				hollow: .25,
				r: 1,
				g: .92,
				b: .6
			}
		];
		let ghostAccum = vec3(0, 0, 0);
		for (let i = 0; i < ghostConfigs.length; i++) {
			const cfg = ghostConfigs[i];
			const ghostPos = lightPos.add(flareAxis.mul(float(cfg.offset).mul(uGhostSpread).mul(2)));
			const dist = vec2(uv$1.x.sub(ghostPos.x).mul(aspect), uv$1.y.sub(ghostPos.y)).length();
			const sz = float(cfg.size);
			const disc = smoothstep(sz.mul(.12), sz, dist).oneMinus();
			const innerFade = smoothstep(sz.mul(.05), sz.mul(.45), dist).oneMinus();
			const shape = disc.mul(float(1).sub(innerFade.mul(float(cfg.hollow))));
			const gPhase = float(i * 2.17);
			const gShimmer = sin(t.mul(4.3).add(gPhase)).mul(.08).add(sin(t.mul(9.1).add(gPhase.mul(1.7))).mul(.05)).add(.95);
			const ringCenter = sz.mul(.78);
			const chromaSpread = sz.mul(uGhostChroma).mul(.14);
			const ringWidth = sz.mul(.1);
			const rRing = smoothstep(float(0), ringWidth, abs(dist.sub(ringCenter.add(chromaSpread)))).oneMinus();
			const gRing = smoothstep(float(0), ringWidth, abs(dist.sub(ringCenter))).oneMinus();
			const bRing = smoothstep(float(0), ringWidth, abs(dist.sub(ringCenter.sub(chromaSpread)))).oneMinus();
			const brightness = float(cfg.bright).mul(uGhostInt).mul(gShimmer);
			const contribution = vec3(shape.mul(float(cfg.r)).add(rRing.mul(uGhostChroma)).mul(brightness), shape.mul(float(cfg.g)).add(gRing.mul(uGhostChroma)).mul(brightness), shape.mul(float(cfg.b)).add(bRing.mul(uGhostChroma)).mul(brightness));
			ghostAccum = ghostAccum.add(contribution);
		}
		const antiLight = vec2(float(1).sub(lightPos.x), float(1).sub(lightPos.y));
		const toAntiLight = vec2(uv$1.x.sub(antiLight.x).mul(aspect), uv$1.y.sub(antiLight.y));
		const antiLightDist = toAntiLight.length();
		const haloChromaSpread = uHaloRad.mul(uHaloChroma).mul(.06);
		const haloWidth = uHaloSoft.mul(.12).add(.005);
		const rHalo = smoothstep(float(0), haloWidth, abs(antiLightDist.sub(uHaloRad.add(haloChromaSpread)))).oneMinus();
		const gHalo = smoothstep(float(0), haloWidth, abs(antiLightDist.sub(uHaloRad))).oneMinus();
		const bHalo = smoothstep(float(0), haloWidth, abs(antiLightDist.sub(uHaloRad.sub(haloChromaSpread)))).oneMinus();
		const haloAngVar = sin(atan(toAntiLight.y, toAntiLight.x).mul(3).add(t.mul(.5))).mul(.12).add(.88);
		const haloColor = vec3(rHalo, gHalo, bHalo).mul(uHaloInt).mul(haloAngVar);
		const lightAngle = atan(toLightAR.y, toLightAR.x);
		const rotation = atan(flareAxis.y, flareAxis.x.mul(aspect)).mul(.5).add(t.mul(.06));
		const spikeAngle = lightAngle.add(rotation);
		const primarySpikes = pow(abs(cos(spikeAngle.mul(uStarburstPts.div(2)))), float(3.5));
		const secondarySpikes = pow(abs(cos(spikeAngle.add(.25).mul(uStarburstPts.div(2)))), float(7)).mul(.4);
		const radialFall = float(1).div(float(1).add(lightDist.mul(lightDist).mul(60)));
		const starSpectral = spectralColor(spikeAngle.div(float(6.2831853)).add(.5));
		const starTint = mix(vec3(1, .97, .92), starSpectral, float(.12));
		const starburstColor = primarySpikes.add(secondarySpikes).mul(radialFall).mul(uStarburstInt).mul(starTint);
		const dy = abs(uv$1.y.sub(lightPos.y));
		const dx = abs(uv$1.x.sub(lightPos.x));
		const vertFall = float(-800).mul(dy).mul(dy).exp();
		const lenSq = uStreakLen.mul(uStreakLen).mul(.25).add(.001);
		const rStreak = float(-1).mul(dx).mul(dx).div(lenSq.mul(.7)).exp();
		const gStreak = float(-1).mul(dx).mul(dx).div(lenSq).exp();
		const bStreak = float(-1).mul(dx).mul(dx).div(lenSq.mul(1.4)).exp();
		const streakColor = vec3(vertFall.mul(rStreak).mul(.5), vertFall.mul(gStreak).mul(.65), vertFall.mul(bStreak).mul(.9)).mul(uStreakInt);
		const glareSharpness = float(1).div(uGlareSz.mul(uGlareSz).add(.01));
		const glareColor = float(-1).mul(lightDist).mul(lightDist).mul(glareSharpness).exp().mul(uGlareInt).mul(vec3(1, .98, .95));
		const coreSoft = float(-500).mul(lightDist).mul(lightDist).exp();
		const coreSharp = smoothstep(float(0), float(.025), lightDist).oneMinus();
		const coreColor = coreSoft.add(coreSharp.mul(1.5)).mul(vec3(1, .98, .96)).mul(uIntensity);
		const finalColor = ghostAccum.add(haloColor).add(starburstColor).add(streakColor).add(glareColor).add(coreColor).mul(masterFade);
		return vec4(finalColor, clamp(dot(finalColor, vec3(.299, .587, .114)).mul(2), 0, 1));
	}
};
var LensFlare_default = componentDefinition;
export { componentDefinition as n, LensFlare_default as t };
