// shaders/dist/core/LinearBlur-2hspd2YP.js
import { r as transformAngle } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { Fn, convertToTexture, cos, float, radians, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
var rawWeights = Array.from({ length: 32 }, (_, i) => {
	const t = (i / 31 - .5) * 2;
	return Math.exp(-.5 * t * t / .64);
});
var wSum = rawWeights.reduce((a, b) => a + b);
var blurWeights = rawWeights.map((w) => w / wSum);
const componentDefinition = {
	name: "LinearBlur",
	category: "Blurs",
	description: "Directional motion blur in a specific angle",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 30,
			description: "Intensity of the linear blur effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Direction of the linear blur (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const angleRad = radians(uniforms.angle.uniform);
		const aspect = viewportSize.x.div(viewportSize.y);
		const blurDirection = vec2(cos(angleRad).div(aspect), sin(angleRad));
		return Fn(() => {
			const uv$1 = screenUV;
			const blurVector = blurDirection.mul(uniforms.intensity.uniform).div(viewportSize).mul(2);
			const total = vec4(0).toVar();
			for (let i = 0; i < 32; i++) {
				const coord = uv$1.add(blurVector.mul(i / 31 - .5));
				total.assign(total.add(childTexture.sample(coord).mul(float(blurWeights[i]))));
			}
			return unpremultiplyAlpha(total);
		})();
	}
};
var LinearBlur_default = componentDefinition;
export { componentDefinition as n, LinearBlur_default as t };
