// shaders/dist/core/LinearGradient-glCc9IaB.js
import { a as transformColor, c as transformEdges, l as transformPosition, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { abs, cos, dot, float, fract, length, mod, normalize, radians, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "LinearGradient",
	category: "Textures",
	description: "Create smooth linear color gradients",
	props: {
		colorA: {
			default: "#1aff00",
			transform: transformColor,
			description: "The starting color of the gradient",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#0000ff",
			transform: transformColor,
			description: "The ending color of the gradient",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		start: {
			default: {
				x: 0,
				y: .5
			},
			transform: transformPosition,
			description: "The starting point of the gradient",
			ui: {
				type: "position",
				label: "Start",
				group: "Position"
			}
		},
		end: {
			default: {
				x: 1,
				y: .5
			},
			transform: transformPosition,
			description: "The ending point of the gradient",
			ui: {
				type: "position",
				label: "End",
				group: "Position"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Additional rotation angle of the gradient (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			transform: transformEdges,
			compileTime: true,
			description: "How to handle areas beyond the gradient endpoints",
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const startPos = vec2(uniforms.start.uniform[0], uniforms.start.uniform[1].oneMinus());
		const endPos = vec2(uniforms.end.uniform[0], uniforms.end.uniform[1].oneMinus());
		const gradientVector = endPos.sub(startPos);
		const gradientLength = length(gradientVector);
		const gradientDir = normalize(gradientVector);
		const safeViewportY = viewportSize.y.max(1e-6);
		const aspect = viewportSize.x.div(safeViewportY);
		const angleRad = radians(uniforms.angle.uniform).negate();
		const cosAngle = cos(angleRad);
		const sinAngle = sin(angleRad);
		const midpoint = startPos.add(endPos).mul(.5);
		const centeredUV = uv$1.sub(midpoint);
		const aspectCorrectedCentered = vec2(centeredUV.x.mul(aspect), centeredUV.y);
		const rotatedCorrected = vec2(aspectCorrectedCentered.x.mul(cosAngle).sub(aspectCorrectedCentered.y.mul(sinAngle)), aspectCorrectedCentered.x.mul(sinAngle).add(aspectCorrectedCentered.y.mul(cosAngle)));
		const t = dot(vec2(rotatedCorrected.x.div(aspect), rotatedCorrected.y).add(midpoint).sub(startPos), gradientDir).div(gradientLength.max(1e-6));
		const edgeModeValue = uniforms.edges.uniform.value;
		let finalT;
		switch (edgeModeValue) {
			case 1:
				finalT = t;
				break;
			case 2: {
				const mirrorT = mod(abs(t), float(2));
				finalT = mirrorT.greaterThan(float(1)).select(float(2).sub(mirrorT), mirrorT);
				break;
			}
			case 3:
				finalT = fract(t);
				break;
			default: finalT = t.clamp(float(0), float(1));
		}
		const mixedColor = mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, finalT.clamp(float(0), float(1)), uniforms.colorSpace.uniform.value);
		let finalAlpha;
		if (edgeModeValue === 1) {
			const inBoundsX = t.greaterThanEqual(float(0)).select(float(1), float(0));
			const inBoundsY = t.lessThanEqual(float(1)).select(float(1), float(0));
			const inBounds = inBoundsX.mul(inBoundsY);
			finalAlpha = mixedColor.w.mul(inBounds);
		} else finalAlpha = mixedColor.w;
		return vec4(mixedColor.x, mixedColor.y, mixedColor.z, finalAlpha);
	}
};
var LinearGradient_default = componentDefinition;
export { componentDefinition as n, LinearGradient_default as t };
