// shaders/dist/core/Marble-DkHLcqFD.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { float, mx_noise_float, screenUV, sin, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Marble",
	category: "Textures",
	description: "Classic marble swirl and vein texture using noise-warped sine waves",
	props: {
		colorA: {
			default: "#ffffff",
			transform: transformColor,
			description: "Base background color of the marble",
			ui: {
				type: "color",
				label: "Base",
				group: "Colors"
			}
		},
		colorB: {
			default: "#3a2d54",
			transform: transformColor,
			description: "Secondary marble tone",
			ui: {
				type: "color",
				label: "Tone",
				group: "Colors"
			}
		},
		colorC: {
			default: "#0f0f0f",
			transform: transformColor,
			description: "Deepest marble color",
			ui: {
				type: "color",
				label: "Depth",
				group: "Colors"
			}
		},
		scale: {
			default: 2,
			description: "Scale and density of the marble vein pattern",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 10,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		turbulence: {
			default: 10,
			description: "Amount of noise-driven distortion applied to the veins",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 50,
				step: .5,
				label: "Turbulence",
				group: "Effect"
			}
		},
		speed: {
			default: .05,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: .25,
				step: .005,
				label: "Speed",
				group: "Animation"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for pattern variation",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const au = uv$1.x.mul(aspect);
		const av = uv$1.y;
		const animTime = createAnimatedTime(params, uniforms.speed);
		const scale = uniforms.scale.uniform;
		const seedVal = uniforms.seed.uniform;
		const n1 = mx_noise_float(vec2(au.mul(scale).add(seedVal), av.mul(scale).add(animTime)));
		const n2 = mx_noise_float(vec2(au.mul(scale).mul(2).add(seedVal.mul(1.7)), av.mul(scale).mul(2).add(animTime.mul(.7)))).mul(.5);
		const n3 = mx_noise_float(vec2(au.mul(scale).mul(4).add(seedVal.mul(3.1)), av.mul(scale).mul(4).add(animTime.mul(1.3)))).mul(.25);
		const totalNoise = n1.add(n2).add(n3);
		const turbulence = uniforms.turbulence.uniform;
		const marbleT = sin(au.mul(scale).mul(float(Math.PI)).add(totalNoise.mul(turbulence))).mul(.5).add(.5);
		const n4 = mx_noise_float(vec2(av.mul(scale).mul(1.5).add(seedVal.mul(2)), au.mul(scale).mul(1.2))).mul(.3).add(.5);
		const t1 = smoothstep(float(.3), float(.6), marbleT);
		const t2 = smoothstep(float(.6), float(.9), marbleT);
		const colABC = mixColors(mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, t1, uniforms.colorSpace.uniform.value), uniforms.colorC.uniform, t2.mul(n4), uniforms.colorSpace.uniform.value);
		return vec4(colABC.rgb, colABC.a);
	}
};
var Marble_default = componentDefinition;
export { componentDefinition as n, Marble_default as t };
