// shaders/dist/core/Mirror-Dv5fOEp6.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { convertToTexture, cos, mix, screenUV, sin, step, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Mirror",
	category: "Distortions",
	description: "Mirror content across a line defined by center point and angle",
	requiresRTT: true,
	requiresChild: true,
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The point the mirror line passes through",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		angle: {
			default: 0,
			description: "The angle of the mirror line in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(screenUV.x.mul(aspect), screenUV.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus());
		const delta = aspectCorrectedUV.sub(centerPos);
		const angleRad = uniforms.angle.uniform.mul(Math.PI / 180);
		const lineDirection = vec2(cos(angleRad), sin(angleRad));
		const lineNormal = vec2(lineDirection.y.negate(), lineDirection.x);
		const signedDistance = delta.dot(lineNormal);
		const reflectedDelta = delta.sub(lineNormal.mul(signedDistance.mul(2)));
		const mirroredUV = centerPos.add(reflectedDelta);
		const finalMirroredUV = vec2(mirroredUV.x.div(aspect), mirroredUV.y);
		const shouldMirror = step(0, signedDistance);
		return unpremultiplyAlpha(mix(childTexture.sample(screenUV), applyEdgeHandling(finalMirroredUV, childTexture.sample(finalMirroredUV), childTexture, uniforms.edges.uniform.value), shouldMirror));
	}
};
var Mirror_default = componentDefinition;
export { componentDefinition as n, Mirror_default as t };
