// shaders/dist/core/MultiPointGradient-Dvt1GUiv.js
import { a as transformColor, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { Fn, float, pow, screenUV, sqrt, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "MultiPointGradient",
	category: "Textures",
	description: "Five individually placed color points blended together by proximity — drag each point to shape the gradient",
	props: {
		colorA: {
			default: "#4776E6",
			transform: transformColor,
			description: "Color of control point A",
			ui: {
				type: "color",
				label: "Color",
				group: "Point A"
			}
		},
		positionA: {
			default: {
				x: .2,
				y: .2
			},
			transform: transformPosition,
			description: "Position of control point A",
			ui: {
				type: "position",
				label: "Position",
				group: "Point A"
			}
		},
		colorB: {
			default: "#C44DFF",
			transform: transformColor,
			description: "Color of control point B",
			ui: {
				type: "color",
				label: "Color",
				group: "Point B"
			}
		},
		positionB: {
			default: {
				x: .8,
				y: .2
			},
			transform: transformPosition,
			description: "Position of control point B",
			ui: {
				type: "position",
				label: "Position",
				group: "Point B"
			}
		},
		colorC: {
			default: "#1ABC9C",
			transform: transformColor,
			description: "Color of control point C",
			ui: {
				type: "color",
				label: "Color",
				group: "Point C"
			}
		},
		positionC: {
			default: {
				x: .2,
				y: .8
			},
			transform: transformPosition,
			description: "Position of control point C",
			ui: {
				type: "position",
				label: "Position",
				group: "Point C"
			}
		},
		colorD: {
			default: "#F8BBD9",
			transform: transformColor,
			description: "Color of control point D",
			ui: {
				type: "color",
				label: "Color",
				group: "Point D"
			}
		},
		positionD: {
			default: {
				x: .8,
				y: .8
			},
			transform: transformPosition,
			description: "Position of control point D",
			ui: {
				type: "position",
				label: "Position",
				group: "Point D"
			}
		},
		colorE: {
			default: "#FF8C42",
			transform: transformColor,
			description: "Color of control point E",
			ui: {
				type: "color",
				label: "Color",
				group: "Point E"
			}
		},
		positionE: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Position of control point E",
			ui: {
				type: "position",
				label: "Position",
				group: "Point E"
			}
		},
		smoothness: {
			default: 2,
			description: "Controls how smoothly colors blend.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .05,
				label: "Smoothness",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const smoothness = uniforms.smoothness.uniform;
		const aspect = viewportSize.x.div(viewportSize.y);
		const uvAspect = vec2(screenUV.x.mul(aspect), screenUV.y);
		const points = [
			{
				color: uniforms.colorA.uniform,
				pos: vec2(uniforms.positionA.uniform.x.mul(aspect), uniforms.positionA.uniform.y.oneMinus())
			},
			{
				color: uniforms.colorB.uniform,
				pos: vec2(uniforms.positionB.uniform.x.mul(aspect), uniforms.positionB.uniform.y.oneMinus())
			},
			{
				color: uniforms.colorC.uniform,
				pos: vec2(uniforms.positionC.uniform.x.mul(aspect), uniforms.positionC.uniform.y.oneMinus())
			},
			{
				color: uniforms.colorD.uniform,
				pos: vec2(uniforms.positionD.uniform.x.mul(aspect), uniforms.positionD.uniform.y.oneMinus())
			},
			{
				color: uniforms.colorE.uniform,
				pos: vec2(uniforms.positionE.uniform.x.mul(aspect), uniforms.positionE.uniform.y.oneMinus())
			}
		];
		return Fn(() => {
			const wRGB = vec3(0).toVar();
			const wA = float(0).toVar();
			const wSum = float(0).toVar();
			for (const { color, pos } of points) {
				const dx = uvAspect.x.sub(pos.x);
				const dy = uvAspect.y.sub(pos.y);
				const dist = sqrt(dx.mul(dx).add(dy.mul(dy))).add(float(.001));
				const effectivePower = float(8).div(smoothness.add(float(.5))).min(float(8));
				const weight = float(1).div(pow(dist, effectivePower));
				wRGB.assign(wRGB.add(color.rgb.mul(weight)));
				wA.assign(wA.add(color.a.mul(weight)));
				wSum.assign(wSum.add(weight));
			}
			return vec4(wRGB.div(wSum), wA.div(wSum));
		})();
	}
};
var MultiPointGradient_default = componentDefinition;
export { componentDefinition as n, MultiPointGradient_default as t };
