// shaders/dist/core/Neon-DQDB7hAE.js
import { a as transformColor, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { n as createAnalyticSdfSampler, r as createSvgSdfSampler } from "./sdf-D_EAGXhN.js";
import { Fn, If, atan, clamp, cos, exp, exp2, float, max, mix, pow, screenUV, sin, smoothstep, sqrt, uniform, vec2, vec3, vec4, viewportSize } from "three/tsl";
var _LZ = .7;
var _INV_LIGHT_LEN = 1 / Math.sqrt(1 + _LZ * _LZ);
var _LZN = _LZ * _INV_LIGHT_LEN;
var _HZ = _LZN + 1;
var _INV_H_LEN = 1 / Math.sqrt(_INV_LIGHT_LEN * _INV_LIGHT_LEN + _HZ * _HZ);
function applyNeonEffect(sdfSampler, uniforms, timeNode) {
	return Fn(() => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const centerPos = vec2(uniforms.center.uniform.x, uniforms.center.uniform.y.oneMinus());
		const scale = uniforms.scale.uniform;
		const sdfUV = vec2(uv$1.x.sub(centerPos.x).mul(aspect).div(scale).add(float(.5)), uv$1.y.sub(centerPos.y).div(scale).add(float(.5)));
		const smoothAmt = uniforms.cornerSmoothing.uniform;
		const smoothEPS = float(.008).add(smoothAmt.mul(float(.03)));
		const sdfCenter = sdfSampler(sdfUV).r;
		const sdfRight = sdfSampler(sdfUV.add(vec2(smoothEPS, float(0)))).r;
		const sdfLeft = sdfSampler(sdfUV.sub(vec2(smoothEPS, float(0)))).r;
		const sdfUp = sdfSampler(sdfUV.add(vec2(float(0), smoothEPS))).r;
		const sdfDown = sdfSampler(sdfUV.sub(vec2(float(0), smoothEPS))).r;
		const gradX = sdfRight.sub(sdfLeft).div(smoothEPS.mul(float(2)));
		const gradY = sdfUp.sub(sdfDown).div(smoothEPS.mul(float(2)));
		const sdf = mix(sdfCenter, sdfCenter.add(sdfRight).add(sdfLeft).add(sdfUp).add(sdfDown).div(float(5)), smoothAmt);
		const pxH = float(1).div(viewportSize.y);
		const glowR = uniforms.glowRadius.uniform.max(float(.001));
		const softPx = max(pxH.div(scale), float(.004));
		const output = vec4(0).toVar();
		If(sdf.greaterThan(glowR.mul(float(1.5)).add(softPx)), () => {
			output.assign(vec4(0, 0, 0, 0));
		}).Else(() => {
			const gradLen = sqrt(gradX.mul(gradX).add(gradY.mul(gradY))).max(float(1e-4));
			const gnx = gradX.div(gradLen);
			const gny = gradY.div(gradLen);
			const tubeR = uniforms.tubeThickness.uniform;
			const depthNorm = clamp(sdf.negate().div(tubeR.max(float(.001))), float(0), float(1));
			const outerMask = smoothstep(softPx, softPx.negate(), sdf);
			const innerMask = smoothstep(tubeR.add(softPx), tubeR.sub(softPx), sdf.negate());
			const tubeMask = outerMask.mul(innerMask);
			const t = depthNorm;
			const cylinderZ = sqrt(t.mul(float(2).sub(t)).max(float(0)));
			const cylinderR = float(1).sub(t);
			const nx = gnx.mul(cylinderR);
			const ny = gny.mul(cylinderR);
			const nz = cylinderZ;
			const lightRad = uniforms.lightAngle.uniform.mul(float(Math.PI / 180));
			const lx = cos(lightRad);
			const ly = sin(lightRad);
			const lxn = lx.mul(float(_INV_LIGHT_LEN));
			const lyn = ly.mul(float(_INV_LIGHT_LEN));
			const lzn = float(_LZN);
			const diffuse = clamp(nx.mul(lxn).add(ny.mul(lyn)).add(nz.mul(lzn)), float(0), float(1));
			const NdotH = nx.mul(lxn.mul(float(_INV_H_LEN))).add(ny.mul(lyn.mul(float(_INV_H_LEN)))).add(nz.mul(float(_HZ * _INV_H_LEN)));
			const shininess = exp2(float(8).sub(uniforms.specularSize.uniform.mul(float(7))));
			const specular = pow(clamp(NdotH, float(0), float(1)), shininess).mul(uniforms.specularIntensity.uniform);
			const flowDx = sdfUV.x.sub(float(.5));
			const flowWave = sin(atan(sdfUV.y.sub(float(.5)), flowDx).mul(float(2)).add(timeNode.mul(uniforms.flowSpeed.uniform))).mul(float(.5)).add(float(.5));
			const flowMod = mix(float(1), flowWave, uniforms.flowAmount.uniform);
			const primaryCol = vec3(uniforms.color.uniform.x, uniforms.color.uniform.y, uniforms.color.uniform.z);
			const secondCol = vec3(uniforms.secondaryColor.uniform.x, uniforms.secondaryColor.uniform.y, uniforms.secondaryColor.uniform.z);
			const blendAmt = uniforms.secondaryBlend.uniform;
			const baseColor = mix(primaryCol, mix(secondCol, primaryCol, diffuse), blendAmt);
			const ambient = float(.12);
			const selfEmission = cylinderZ.mul(float(.35));
			const shadedRgb = baseColor.mul(ambient.add(diffuse.mul(float(.55))).add(selfEmission));
			const specCol = vec3(specular, specular, specular);
			const tubeRgb = shadedRgb.add(specCol);
			const hotCoreAmt = smoothstep(float(.55), float(1), depthNorm).mul(uniforms.hotCoreIntensity.uniform).mul(flowMod);
			const tubeFlowed = mix(tubeRgb, mix(primaryCol, vec3(1, 1, 1), float(.75)), hotCoreAmt).mul(mix(float(1), flowMod, float(.4)));
			const glowI = uniforms.glowIntensity.uniform;
			const sdfOutside = max(sdf, float(0));
			const innerGlowFalloff = float(12).div(glowR);
			const innerGlow = exp(sdfOutside.negate().mul(innerGlowFalloff)).mul(float(.7));
			const outerGlowFalloff = float(4).div(glowR);
			const outerGlow = exp(sdfOutside.negate().mul(outerGlowFalloff)).mul(float(.35));
			const totalGlow = innerGlow.add(outerGlow).mul(glowI);
			const outsideFactor = smoothstep(float(-.003), float(.003), sdf);
			const glowMasked = totalGlow.mul(outsideFactor);
			const glowCol = vec3(uniforms.glowColor.uniform.x, uniforms.glowColor.uniform.y, uniforms.glowColor.uniform.z);
			const flickerTime = timeNode.mul(uniforms.flickerSpeed.uniform);
			const f1 = sin(flickerTime.mul(float(17.13)));
			const f2 = sin(flickerTime.mul(float(31.71)));
			const f3 = sin(flickerTime.mul(float(47.29)));
			const combined = f1.add(f2.mul(float(.5))).add(f3.mul(float(.3)));
			const threshold = mix(float(-2), float(.8), uniforms.flickerAmount.uniform);
			const flickerValue = smoothstep(threshold.sub(float(.1)), threshold.add(float(.1)), combined);
			const intensityMul = uniforms.intensity.uniform.mul(flickerValue);
			const finalTubeRgb = tubeFlowed.mul(intensityMul);
			const finalGlowRgb = glowCol.mul(glowMasked).mul(intensityMul);
			const finalR = mix(finalGlowRgb.x, finalTubeRgb.x, tubeMask);
			const finalG = mix(finalGlowRgb.y, finalTubeRgb.y, tubeMask);
			const finalB = mix(finalGlowRgb.z, finalTubeRgb.z, tubeMask);
			const finalA = max(tubeMask, clamp(glowMasked.mul(float(.6)), float(0), float(1)));
			output.assign(vec4(finalR, finalG, finalB, finalA));
		});
		return output;
	})();
}
var DEFAULT_SHAPE_CONFIG = JSON.stringify({
	type: "circleSDF",
	radius: .35
});
const componentDefinition = {
	name: "Neon",
	category: "Shape Effects",
	description: "Photorealistic neon tube / 3D pipe effect driven by a custom shape",
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the neon shape",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		scale: {
			default: 1,
			description: "Scale of the neon shape (1 = default size)",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .05,
				label: "Scale",
				group: "Position"
			}
		},
		color: {
			default: "#00ddff",
			transform: transformColor,
			description: "Primary neon tube color",
			ui: {
				type: "color",
				label: "Tube Color",
				group: "Colors"
			}
		},
		secondaryColor: {
			default: "#ff00aa",
			transform: transformColor,
			description: "Shadow-side color for a two-tone / dual-lit pipe look",
			ui: {
				type: "color",
				label: "Secondary Color",
				group: "Colors"
			}
		},
		secondaryBlend: {
			default: .5,
			description: "Blend between mono (0) and two-tone (1) tube coloring",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Blend",
				group: "Colors"
			}
		},
		glowColor: {
			default: "#00ddff",
			transform: transformColor,
			description: "Color of the outer glow / bloom",
			ui: {
				type: "color",
				label: "Glow Color",
				group: "Colors"
			}
		},
		tubeThickness: {
			default: .2,
			transform: (v) => v * .05,
			description: "How far inward from the boundary the tube extends. Low = thin neon outline, high = thick 3D pipe",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Tube Thickness",
				group: "Tube"
			}
		},
		intensity: {
			default: 1.5,
			description: "Overall brightness multiplier",
			ui: {
				type: "range",
				min: .5,
				max: 4,
				step: .05,
				label: "Intensity",
				group: "Tube"
			}
		},
		hotCoreIntensity: {
			default: .6,
			description: "Bright white-hot center line — the gas discharge glow inside the tube",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Hot Core",
				group: "Tube"
			}
		},
		glowIntensity: {
			default: .6,
			description: "Outer glow / bloom strength",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .05,
				label: "Glow Intensity",
				group: "Glow"
			}
		},
		glowRadius: {
			default: .25,
			description: "How far the glow extends beyond the tube",
			ui: {
				type: "range",
				min: .01,
				max: 1,
				step: .01,
				label: "Glow Radius",
				group: "Glow"
			}
		},
		lightAngle: {
			default: 300,
			description: "Directional light angle in degrees — controls 3D shading on the tube",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Light Angle",
				group: "Lighting"
			}
		},
		specularIntensity: {
			default: .5,
			description: "Specular highlight brightness on the tube surface",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .05,
				label: "Specular",
				group: "Lighting"
			}
		},
		specularSize: {
			default: .5,
			description: "Specular highlight size — 0 = tight pinpoint, 1 = broad sheen",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Specular Size",
				group: "Lighting"
			}
		},
		cornerSmoothing: {
			default: .15,
			description: "Rounds sharp corners to mimic how real glass tubes curve at bends",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Corner Smoothing",
				group: "Tube"
			}
		},
		flickerSpeed: {
			default: 0,
			description: "Flicker animation speed — 0 = off, higher = faster sporadic on/off",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Flicker"
			}
		},
		flickerAmount: {
			default: .2,
			description: "How often the neon flickers off — 0 = always on, 1 = frequent outages",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Amount",
				group: "Flicker"
			}
		},
		flowSpeed: {
			default: 0,
			description: "Flow animation speed — 0 = off, light rotates through the tube",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Flow"
			}
		},
		flowAmount: {
			default: .3,
			description: "Strength of the flowing brightness variation — 0 = uniform, 1 = dramatic",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Amount",
				group: "Flow"
			}
		},
		shape: {
			default: DEFAULT_SHAPE_CONFIG,
			description: "Serialized shape configuration (JSON)",
			ui: {
				type: "shape",
				label: "Shape",
				group: "Shape"
			}
		},
		shapeSdfUrl: {
			default: "",
			compileTime: true,
			description: "URL to a pre-generated SDF .bin file — when non-empty, activates SVG mode and triggers a shader recompile"
		},
		shapeType: {
			default: "",
			compileTime: true,
			description: "Active SDF shape type — triggers recompile when shape is switched. When empty, derived from shape JSON at mount time."
		}
	},
	fragmentNode: ({ uniforms, onCleanup, onBeforeRender }) => {
		const timeU = uniform(0);
		onBeforeRender(({ deltaTime }) => {
			timeU.value += deltaTime;
		});
		const shapeSdfUrl = uniforms.shapeSdfUrl.uniform.value;
		const initialConfig = (() => {
			const val = uniforms.shape.uniform.value;
			if (typeof val === "object" && val !== null) return val;
			try {
				return JSON.parse(val);
			} catch {
				return {
					type: "circleSDF",
					radius: .35,
					centerX: .5,
					centerY: .5
				};
			}
		})();
		if (shapeSdfUrl) return applyNeonEffect(createSvgSdfSampler(shapeSdfUrl, onBeforeRender, onCleanup), uniforms, timeU);
		else return applyNeonEffect(createAnalyticSdfSampler(uniforms.shapeType.uniform.value || initialConfig.type || "circleSDF", initialConfig, uniforms.shape, onBeforeRender), uniforms, timeU);
	}
};
var Neon_default = componentDefinition;
export { componentDefinition as n, Neon_default as t };
