// shaders/dist/core/Paper-DNP2-hPM.js
import { abs, clamp, convertToTexture, float, mx_noise_float, screenUV, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Paper",
	category: "Stylize",
	description: "Applies realistic paper grain and surface roughness to child content",
	requiresRTT: true,
	requiresChild: true,
	props: {
		roughness: {
			default: .3,
			description: "Surface roughness — higher values create more pronounced brightness variation",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Roughness",
				group: "Texture"
			}
		},
		grainScale: {
			default: 1,
			description: "Scale of the paper grain — lower = coarser, higher = finer",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .05,
				label: "Grain Scale",
				group: "Texture"
			}
		},
		displacement: {
			default: .15,
			description: "Surface micro-roughness — shifts pixels at grain scale like real paper fiber bumps",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Displacement",
				group: "Texture"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for pattern variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Texture"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectUV = vec2(uv$1.x.mul(aspect), uv$1.y);
		const seedOffset = uniforms.seed.uniform;
		const grainFreq = uniforms.grainScale.uniform.mul(float(120)).add(float(120));
		const warpFreq = grainFreq.mul(float(.25));
		const warpPos = aspectUV.mul(warpFreq).add(seedOffset.add(float(50)));
		const warpX = mx_noise_float(vec3(warpPos.x, warpPos.y, float(22.3)));
		const warpY = mx_noise_float(vec3(warpPos.x, warpPos.y, float(29.7)));
		const dispAmt = uniforms.displacement.uniform.mul(float(.002));
		const displacedUV = uv$1.add(vec2(warpX.mul(dispAmt), warpY.mul(dispAmt)));
		const childColor = childTexture.sample(displacedUV);
		const grainUV = vec2(displacedUV.x.mul(aspect), displacedUV.y).add(vec2(warpX.mul(float(.005)), warpY.mul(float(.005))));
		const p1 = grainUV.mul(grainFreq.mul(float(.1))).add(seedOffset);
		const n1 = mx_noise_float(vec3(p1.x, p1.y, float(1.3)));
		const p2 = grainUV.mul(grainFreq.mul(float(.4))).add(seedOffset.add(float(33.7)));
		const n2 = abs(mx_noise_float(vec3(p2.x, p2.y, float(4.1))));
		const p3 = grainUV.mul(grainFreq).add(seedOffset.add(float(71.2)));
		const n3 = mx_noise_float(vec3(p3.x, p3.y, float(9.8)));
		const p4 = grainUV.mul(grainFreq.mul(float(2))).add(seedOffset.add(float(107.3)));
		const n4 = mx_noise_float(vec3(p4.x, p4.y, float(15.6)));
		const grain = n1.mul(float(.3)).add(n2.mul(float(.3))).add(n3.mul(float(.25))).add(n4.mul(float(.15)));
		const rough = uniforms.roughness.uniform;
		const brightness = clamp(float(1).add(grain.mul(rough.mul(float(.4)))), float(0), float(1.5));
		const finalRgb = vec3(childColor.r, childColor.g, childColor.b).mul(brightness);
		return vec4(finalRgb.x, finalRgb.y, finalRgb.z, childColor.a);
	}
};
var Paper_default = componentDefinition;
export { componentDefinition as n, Paper_default as t };
