// shaders/dist/core/Pixelate-BbT3Rrf2.js
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { convertToTexture, float, floor, fract, length, screenUV, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Pixelate",
	category: "Stylize",
	description: "Pixelation effect with adjustable cell size",
	requiresRTT: true,
	requiresChild: true,
	props: {
		scale: {
			default: 50,
			description: "Number of pixels along the longest edge (higher = smaller pixels)",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 200,
				step: 1,
				label: "Scale",
				group: "Effect"
			}
		},
		gap: {
			default: 0,
			description: "Space between pixels as a fraction of cell size (0 = no gap, 1 = fully invisible)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .95,
				step: .01,
				label: "Gap",
				group: "Effect"
			}
		},
		roundness: {
			default: 0,
			description: "Roundness of each pixel's corners (0 = square, 1 = circle)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Roundness",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const aspect = viewportSize.x.div(viewportSize.y);
		const pixelSize = uniforms.scale.uniform;
		const isWide = aspect.greaterThan(float(1));
		const pixelCountX = isWide.select(pixelSize, pixelSize.mul(aspect)).max(float(1));
		const pixelCountY = isWide.select(pixelSize.div(aspect), pixelSize).max(float(1));
		const pixelatedUV = vec2(floor(screenUV.x.mul(pixelCountX)).div(pixelCountX), floor(screenUV.y.mul(pixelCountY)).div(pixelCountY));
		const sampledColor = unpremultiplyAlpha(childTexture.sample(pixelatedUV));
		const localUV = vec2(fract(screenUV.x.mul(pixelCountX)).sub(.5), fract(screenUV.y.mul(pixelCountY)).sub(.5));
		const halfSize = float(.5).sub(uniforms.gap.uniform.mul(.5));
		const cornerRadius = uniforms.roundness.uniform.mul(halfSize);
		const mask = length(vec2(localUV.x.abs().sub(halfSize).add(cornerRadius).max(float(0)), localUV.y.abs().sub(halfSize).add(cornerRadius).max(float(0)))).sub(cornerRadius).lessThanEqual(float(0)).select(float(1), float(0));
		return vec4(sampledColor.rgb, sampledColor.a.mul(mask));
	}
};
var Pixelate_default = componentDefinition;
export { componentDefinition as n, Pixelate_default as t };
