// shaders/dist/core/Plasma-D9rhkH2y.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { clamp, exp, float, mx_noise_float, pow, screenUV, vec3, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Plasma",
	category: "Textures",
	description: "Animated effect of glowing plasma",
	props: {
		density: {
			default: 2,
			description: "Density of the plasma pattern",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 4,
				step: .1,
				label: "Density",
				group: "Effect"
			}
		},
		speed: {
			default: 2,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Effect"
			}
		},
		intensity: {
			default: 1.5,
			description: "Brightness and spread of the plasma glow",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 3,
				step: .1,
				label: "Intensity",
				group: "Colors"
			}
		},
		warp: {
			default: .4,
			description: "How much the flow distorts and swirls",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Warp",
				group: "Effect"
			}
		},
		contrast: {
			default: 1,
			description: "Push darks darker and lights lighter",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 3,
				step: .1,
				label: "Contrast",
				group: "Colors"
			}
		},
		balance: {
			default: 50,
			description: "Skew color balance toward A (higher) or B (lower)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Balance",
				group: "Colors"
			}
		},
		colorA: {
			default: "#7018be",
			description: "Primary color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#000000",
			description: "Secondary color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const density = uniforms.density.uniform;
		const intensity = uniforms.intensity.uniform;
		const warp = uniforms.warp.uniform;
		const t = createAnimatedTime(params, uniforms.speed).mul(.125);
		const aspect = viewportSize.x.div(viewportSize.y);
		const pos = vec3(screenUV.x.mul(aspect), screenUV.y, float(0)).mul(exp(density.sub(1))).toVar();
		const warpAmount = warp.mul(4);
		const qx = mx_noise_float(pos.add(vec3(t.mul(.3), t.mul(-.2), 0)));
		const qy = mx_noise_float(pos.add(vec3(5.2, 1.3, t.mul(.25))));
		const warped1 = pos.add(vec3(qx, qy, float(0)).mul(warpAmount));
		const rx = mx_noise_float(warped1.add(vec3(1.7, 9.2, t.mul(.15))));
		const ry = mx_noise_float(warped1.add(vec3(8.3, 2.8, t.mul(.13))));
		const contrasted = clamp(pow(mx_noise_float(pos.add(vec3(rx, ry, float(0)).mul(warpAmount))).mul(.5).add(.5), float(5).div(intensity)).sub(.5).mul(uniforms.contrast.uniform).add(.5), 0, 1);
		const balanceShift = uniforms.balance.uniform.div(100).sub(.5);
		const balanced = clamp(contrasted.add(balanceShift), 0, 1);
		return mixColors(uniforms.colorB.uniform, uniforms.colorA.uniform, balanced, uniforms.colorSpace.uniform.value);
	}
};
var Plasma_default = componentDefinition;
export { componentDefinition as n, Plasma_default as t };
