// shaders/dist/core/PolarCoordinates-2RBCk-fo.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { PI, atan, convertToTexture, length, mix, screenUV, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "PolarCoordinates",
	category: "Distortions",
	description: "Convert rectangular coordinates to polar space",
	requiresRTT: true,
	requiresChild: true,
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point for polar coordinate conversion",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		wrap: {
			default: 1,
			description: "Controls how much of the angular range to use (1 = full 360°, 0.5 = 180°)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "Wrap",
				group: "Effect"
			}
		},
		radius: {
			default: 1,
			description: "Controls how much of the radius range to use (affects the radial mapping)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		intensity: {
			default: 1,
			description: "Blends between original UVs (0) and polar coordinates (1)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		edges: {
			default: "transparent",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(screenUV.x.mul(aspect), screenUV.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus());
		const delta = aspectCorrectedUV.sub(centerPos);
		const normalizedAngle = atan(delta.y, delta.x).add(PI).div(PI.mul(2));
		const r = length(delta);
		const blendedUV = mix(screenUV, vec2(normalizedAngle.mul(uniforms.wrap.uniform), r.mul(uniforms.radius.uniform)), uniforms.intensity.uniform);
		return unpremultiplyAlpha(applyEdgeHandling(blendedUV, childTexture.sample(blendedUV), childTexture, uniforms.edges.uniform.value));
	}
};
var PolarCoordinates_default = componentDefinition;
export { componentDefinition as n, PolarCoordinates_default as t };
