// shaders/dist/core/RadialGradient-CQIctl__.js
import { a as transformColor, l as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { cos, distance, float, fract, screenUV, sin, vec2, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "RadialGradient",
	category: "Textures",
	description: "Radial gradient radiating from a center point",
	props: {
		colorA: {
			default: "#ff0000",
			transform: transformColor,
			description: "The starting color at the center of the gradient",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#0000ff",
			transform: transformColor,
			description: "The ending color at the edge of the gradient",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the radial gradient",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		radius: {
			default: 1,
			description: "The radius of the gradient (normalized, 0.0-1.0)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 2,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		repeat: {
			default: 1,
			description: "Number of times the gradient repeats. Values above 1 create concentric rings.",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 20,
				step: .5,
				label: "Repeat",
				group: "Effect"
			}
		},
		aspect: {
			default: 1,
			description: "Stretches the gradient into an ellipse. Values below 1 compress vertically, above 1 compress horizontally.",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 4,
				step: .05,
				label: "Aspect",
				group: "Effect"
			}
		},
		skewAngle: {
			default: 0,
			description: "Rotates the ellipse axis in degrees. Only visible when Aspect is not 1.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Skew Angle",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspectRatio = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(uv$1.x.mul(aspectRatio), uv$1.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspectRatio), uniforms.center.uniform.y.oneMinus());
		const dx = aspectCorrectedUV.x.sub(centerPos.x);
		const dy = aspectCorrectedUV.y.sub(centerPos.y);
		const skewRad = uniforms.skewAngle.uniform.mul(float(Math.PI / 180));
		const cosS = cos(skewRad);
		const sinS = sin(skewRad);
		const rdx = dx.mul(cosS).add(dy.mul(sinS));
		const rdy = dx.mul(sinS).negate().add(dy.mul(cosS));
		const dist = distance(vec2(rdx.mul(uniforms.aspect.uniform), rdy), vec2(float(0), float(0)));
		const safeRadius = uniforms.radius.uniform.max(float(1e-6));
		const t = dist.div(safeRadius).mul(uniforms.repeat.uniform);
		const finalT = uniforms.repeat.uniform.greaterThan(float(1.0001)).select(fract(t), t.clamp(float(0), float(1)));
		return mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, finalT, uniforms.colorSpace.uniform.value);
	}
};
var RadialGradient_default = componentDefinition;
export { componentDefinition as n, RadialGradient_default as t };
