// shaders/dist/core/ReflectivePlane-UIT4xlmn.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { i as createVariableGaussianBlurCompute } from "./computeBlur-CprKe0xa.js";
import { Fn, If, convertToTexture, float, instanceIndex, mix, screenUV, smoothstep, step, texture, textureStore, uint, uniform, uvec2, vec2, vec4, viewportSize } from "three/tsl";
var HALF_KERNEL = 30;
var BLUR_TO_RADIUS_PX = 12;
var POISSON_DISK = [
	[0, 0],
	[-.94, -.34],
	[.95, .31],
	[-.09, -.93],
	[.34, .94],
	[-.91, .41],
	[.88, -.46],
	[-.4, .71],
	[.51, .58],
	[-.49, -.59],
	[.4, -.68],
	[-.71, .05],
	[.07, .21]
];
var POISSON_WEIGHT = 1 / POISSON_DISK.length;
const componentDefinition = {
	name: "ReflectivePlane",
	category: "Stylize",
	description: "Reflective floor that mirrors the content above it",
	requiresRTT: true,
	requiresChild: true,
	props: {
		height: {
			default: .7,
			description: "Vertical position of the reflective surface",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Floor Height",
				group: "Plane"
			}
		},
		distance: {
			default: .5,
			description: "How far below the floor the reflection remains visible before fully fading to transparent.",
			ui: {
				type: ["range", "map"],
				min: .01,
				max: 1,
				step: .01,
				label: "Reflection Distance",
				group: "Plane"
			}
		},
		falloff: {
			default: .5,
			description: "Width of the fade zone, as a fraction of reflection distance.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 3,
				step: .01,
				label: "Reflection Falloff",
				group: "Plane"
			}
		},
		blur: {
			default: 3,
			description: "Maximum blur applied to the reflection far from the surface.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .05,
				label: "Blur",
				group: "Surface"
			}
		},
		blurDistance: {
			default: .3,
			description: "How far below the surface the blur takes to ramp from sharp to maximum.",
			ui: {
				type: "range",
				min: .01,
				max: 1,
				step: .01,
				label: "Blur Distance",
				group: "Surface"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle reflected samples that fall outside the source content.",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Surface"
			}
		}
	},
	computeNode: ({ childNode, onCleanup, dimensions, renderer, uniforms, onResize }) => {
		if (!childNode) return null;
		if (uniforms.height.mapComputeInfo || uniforms.blur.mapComputeInfo || uniforms.blurDistance.mapComputeInfo) return null;
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const dpr = renderer?.getPixelRatio?.() ?? 1;
		const blur = createVariableGaussianBlurCompute(childTexture, Math.round(dimensions.width * dpr), Math.round(dimensions.height * dpr), onCleanup, HALF_KERNEL);
		const { computeNodes: blurNodes, outputTexture, blurMapWriteNode, computeWidth, computeHeight } = blur;
		const heightU = uniform(.5);
		const blurAmountU = uniform(1);
		const blurDistanceU = uniform(.5);
		const cw = uint(computeWidth);
		const ch = uint(computeHeight);
		function readScalar(entry) {
			return entry.getCpuValue?.() ?? (typeof entry.uniform.value === "number" ? entry.uniform.value : null);
		}
		function updateUniforms() {
			const h = readScalar(uniforms.height);
			if (h !== null) heightU.value = h;
			const b = readScalar(uniforms.blur);
			if (b !== null) blurAmountU.value = b;
			const bd = readScalar(uniforms.blurDistance);
			if (bd !== null) blurDistanceU.value = Math.max(.01, bd);
		}
		const fillBlurMap = Fn(() => {
			const idx = instanceIndex;
			const cx = idx.mod(cw);
			const cy = idx.div(cw);
			If(cy.lessThan(ch), () => {
				const v = cy.toFloat().add(float(.5)).div(float(computeHeight));
				const distBelowAtOutput = heightU.sub(v).max(float(0));
				const ramp = smoothstep(float(0), blurDistanceU, distBelowAtOutput);
				const radius = blurAmountU.mul(ramp).mul(float(BLUR_TO_RADIUS_PX));
				textureStore(blurMapWriteNode, uvec2(cx, cy), vec4(radius, float(0), float(0), float(1))).toWriteOnly();
			});
		})().compute(computeWidth * computeHeight, [256]);
		onResize(({ width, height }) => {
			blur.setInputDimensions(Math.round(width * dpr), Math.round(height * dpr));
		});
		return {
			outputs: {
				childTexture,
				blurredTexture: outputTexture
			},
			getComputeNodes: () => {
				updateUniforms();
				return [fillBlurMap, ...blurNodes];
			}
		};
	},
	fragmentNode: ({ uniforms, childNode, onCleanup, computeOutputs }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		if (computeOutputs?.blurredTexture && computeOutputs?.childTexture) {
			const sharp = computeOutputs.childTexture;
			const blurred = texture(computeOutputs.blurredTexture);
			return Fn(() => {
				const floorY = uniforms.height.uniform;
				const uv$1 = screenUV;
				const distBelow = uv$1.y.sub(floorY).max(float(0));
				const originalSample = sharp.sample(uv$1);
				const reflectedUV = vec2(uv$1.x, floorY.mul(2).sub(uv$1.y));
				const reflSample = applyEdgeHandling(reflectedUV, blurred.sample(reflectedUV), blurred, uniforms.edges.uniform.value);
				const distanceMax = uniforms.distance.uniform.max(float(.001));
				const fadeWidth = distanceMax.mul(uniforms.falloff.uniform).max(float(1e-4));
				const fadeStart = distanceMax.sub(fadeWidth);
				const visibility = float(1).sub(smoothstep(fadeStart, distanceMax, distBelow));
				const effReflA = reflSample.a.mul(visibility);
				const reflPremultRgb = reflSample.rgb.mul(visibility);
				const propsPin = uniforms.height.uniform.mul(float(0)).add(uniforms.blur.uniform.mul(float(0))).add(uniforms.blurDistance.uniform.mul(float(0)));
				const oneMinusReflA = effReflA.oneMinus();
				return unpremultiplyAlpha(mix(originalSample, vec4(reflPremultRgb.add(originalSample.rgb.mul(oneMinusReflA)).add(propsPin), effReflA.add(originalSample.a.mul(oneMinusReflA))), step(floorY, uv$1.y)));
			})();
		}
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		return Fn(() => {
			const floorY = uniforms.height.uniform;
			const uv$1 = screenUV;
			const aspect = viewportSize.x.div(viewportSize.y);
			const distBelow = uv$1.y.sub(floorY).max(float(0));
			const originalSample = childTexture.sample(uv$1);
			const reflectedUV = vec2(uv$1.x, floorY.mul(2).sub(uv$1.y));
			const clampedBlurDistance = uniforms.blurDistance.uniform.max(float(.01));
			const blurRamp = smoothstep(float(0), clampedBlurDistance, distBelow);
			const blurRadius = uniforms.blur.uniform.mul(blurRamp).mul(.02);
			const blurredColor = vec4(0).toVar();
			for (let i = 0; i < POISSON_DISK.length; i++) {
				const [ox, oy] = POISSON_DISK[i];
				const offset = vec2(float(ox).mul(blurRadius).div(aspect), float(oy).mul(blurRadius));
				const sampleUV = reflectedUV.add(offset);
				const edgeHandled = applyEdgeHandling(sampleUV, childTexture.sample(sampleUV), childTexture, uniforms.edges.uniform.value);
				blurredColor.assign(blurredColor.add(edgeHandled.mul(float(POISSON_WEIGHT))));
			}
			const distanceMax = uniforms.distance.uniform.max(float(.001));
			const fadeWidth = distanceMax.mul(uniforms.falloff.uniform).max(float(1e-4));
			const fadeStart = distanceMax.sub(fadeWidth);
			const visibility = float(1).sub(smoothstep(fadeStart, distanceMax, distBelow));
			const effReflA = blurredColor.a.mul(visibility);
			const reflPremultRgb = blurredColor.rgb.mul(visibility);
			const oneMinusReflA = effReflA.oneMinus();
			return unpremultiplyAlpha(mix(originalSample, vec4(reflPremultRgb.add(originalSample.rgb.mul(oneMinusReflA)), effReflA.add(originalSample.a.mul(oneMinusReflA))), step(floorY, uv$1.y)));
		})();
	}
};
var ReflectivePlane_default = componentDefinition;
export { componentDefinition as n, ReflectivePlane_default as t };
