// shaders/dist/core/Ring-UT__kmzY.js
import { a as transformColor, l as transformPosition, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { l as ringSdf } from "./sdf-D_EAGXhN.js";
import { float, screenUV, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
var transformStrokePosition = (value) => {
	return {
		"outside": 0,
		"center": 1,
		"inside": 2
	}[value?.toLowerCase()] ?? 1;
};
const componentDefinition = {
	name: "Ring",
	category: "Shapes",
	description: "Annular ring (donut) with adjustable radius and band thickness",
	props: {
		color: {
			default: "#ffffff",
			transform: transformColor,
			description: "Fill color of the ring",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center position of the ring",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		radius: {
			default: .3,
			description: "Distance from center to the ring's midline in UV space",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Radius",
				group: "Shape"
			}
		},
		thickness: {
			default: .07,
			description: "Half-width of the ring band — total ring width is twice this value",
			ui: {
				type: "range",
				min: .005,
				max: .3,
				step: .005,
				label: "Thickness",
				group: "Shape"
			}
		},
		softness: {
			default: 0,
			description: "Edge softness for antialiasing (applied to both inner and outer ring edges)",
			ui: {
				type: "range",
				min: 0,
				max: .1,
				step: .001,
				label: "Softness",
				group: "Effect"
			}
		},
		strokeThickness: {
			default: 0,
			description: "Stroke thickness. Zero means no stroke.",
			ui: {
				type: "range",
				min: 0,
				max: .1,
				step: .005,
				label: "Stroke Thickness",
				group: "Stroke"
			}
		},
		strokeColor: {
			default: "#000000",
			transform: transformColor,
			description: "Color of the stroke outline",
			ui: {
				type: "color",
				label: "Stroke Color",
				group: "Colors"
			}
		},
		strokePosition: {
			default: "center",
			transform: transformStrokePosition,
			description: "Position of the stroke relative to the ring edge",
			ui: {
				type: "select",
				options: [
					{
						label: "Outside",
						value: "outside"
					},
					{
						label: "Center",
						value: "center"
					},
					{
						label: "Inside",
						value: "inside"
					}
				],
				label: "Stroke Position",
				group: "Stroke"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for blending fill and stroke colors",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Blending",
				group: "Colors"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectUV = vec2(uv$1.x.mul(aspect), uv$1.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus());
		const sdf = ringSdf(aspectUV.x.sub(centerPos.x), aspectUV.y.sub(centerPos.y), uniforms.radius.uniform, uniforms.thickness.uniform);
		const edgeSoftness = uniforms.softness.uniform;
		const strokeThickness = uniforms.strokeThickness.uniform;
		const strokePositionMode = uniforms.strokePosition.uniform;
		const hasStroke = strokeThickness.greaterThan(float(0));
		const isCenter = strokePositionMode.greaterThanEqual(float(.5)).and(strokePositionMode.lessThan(float(1.5)));
		const isInside = strokePositionMode.greaterThanEqual(float(1.5));
		const halfThickness = strokeThickness.mul(.5);
		const strokeInner = isInside.select(strokeThickness.negate(), isCenter.select(halfThickness.negate(), float(0)));
		const strokeOuter = isInside.select(float(0), isCenter.select(halfThickness, strokeThickness));
		const overallMask = smoothstep(strokeOuter.sub(edgeSoftness), strokeOuter, sdf).oneMinus();
		const strokeBlend = hasStroke.select(smoothstep(strokeInner.sub(edgeSoftness), strokeInner, sdf), float(0));
		const blendedColor = mixColors(vec4(uniforms.color.uniform.rgb, uniforms.color.uniform.a), vec4(uniforms.strokeColor.uniform.rgb, uniforms.strokeColor.uniform.a), strokeBlend, uniforms.colorSpace.uniform.value);
		return vec4(blendedColor.xyz, blendedColor.w.mul(overallMask));
	}
};
var Ring_default = componentDefinition;
export { componentDefinition as n, Ring_default as t };
