// shaders/dist/core/Saturation-DFOp2yD9.js
import { dot, mix, vec3, vec4 } from "three/tsl";
const componentDefinition = {
	name: "Saturation",
	category: "Adjustments",
	description: "Adjust color saturation intensity",
	requiresChild: true,
	props: { intensity: {
		default: 1,
		description: "The intensity of the saturation effect (1 being no change)",
		compileTimeWhen: (prev, next) => prev === 1 !== (next === 1),
		ui: {
			type: ["range", "map"],
			min: 0,
			max: 3,
			step: .1,
			label: "Intensity",
			group: "Appearance"
		}
	} },
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Saturation shader.");
			return vec4(0);
		}
		if (uniforms.intensity.uniform.value === 1) return childNode;
		const luminance = dot(childNode.rgb, vec3(.2126, .7152, .0722));
		return vec4(mix(vec3(luminance, luminance, luminance), childNode.rgb, uniforms.intensity.uniform), childNode.a);
	}
};
var Saturation_default = componentDefinition;
export { componentDefinition as n, Saturation_default as t };
