// shaders/dist/core/sdf-D_EAGXhN.js
import { DataTexture, FloatType, LinearFilter, RedFormat } from "three";
import { abs, atan, cos, float, floor, max, min, mix, sign, sin, sqrt, texture, uniform, vec4 } from "three/tsl";
function circleSdf(dx, dy, radius) {
	return sqrt(dx.mul(dx).add(dy.mul(dy))).sub(radius);
}
function polygonSdf(dx, dy, radius, sides) {
	const len = sqrt(dx.mul(dx).add(dy.mul(dy)));
	const angle = atan(dy, dx);
	const sectorAngle = float(2 * Math.PI).div(sides);
	const sectorIdx = floor(angle.div(sectorAngle).add(float(.5)));
	const canonicalAngle = angle.sub(sectorIdx.mul(sectorAngle));
	return len.mul(cos(canonicalAngle)).sub(radius);
}
function flowerSdf(dx, dy, outerRadius, sides, innerRatio) {
	const angle = atan(dy, dx);
	const len = sqrt(dx.mul(dx).add(dy.mul(dy)));
	const innerRadius = outerRadius.mul(innerRatio);
	const tAngle = angle.mul(sides).div(float(2 * Math.PI));
	const t = abs(tAngle.sub(floor(tAngle)).mul(float(2)).sub(float(1)));
	const boundaryR = innerRadius.add(outerRadius.sub(innerRadius).mul(t));
	return len.sub(boundaryR);
}
function starSdf(dx, dy, outerRadius, sides, innerRatio) {
	const innerRadius = outerRadius.mul(innerRatio);
	const len = sqrt(dx.mul(dx).add(dy.mul(dy)));
	const angle = atan(dy, dx);
	const sectorAngle = float(2 * Math.PI).div(sides);
	const sectorIdx = floor(angle.div(sectorAngle).add(float(.5)));
	const bn = angle.sub(sectorIdx.mul(sectorAngle));
	const fpx = abs(len.mul(sin(bn)));
	const fpy = len.mul(cos(bn));
	const an = float(Math.PI).div(sides);
	const sinAn = sin(an);
	const cosAn = cos(an);
	const ex = innerRadius.mul(sinAn);
	const ey = innerRadius.mul(cosAn).sub(outerRadius);
	const qx = fpx;
	const qy = fpy.sub(outerRadius);
	const dotQE = qx.mul(ex).add(qy.mul(ey));
	const dotEE = ex.mul(ex).add(ey.mul(ey));
	const t = dotQE.div(dotEE).clamp(float(0), float(1));
	const nx = fpx.sub(ex.mul(t));
	const ny = fpy.sub(outerRadius.add(ey.mul(t)));
	const dist = sqrt(nx.mul(nx).add(ny.mul(ny)));
	const cross = ex.mul(qy).sub(ey.mul(qx));
	return dist.mul(sign(cross));
}
function ringSdf(dx, dy, radius, thickness) {
	return abs(sqrt(dx.mul(dx).add(dy.mul(dy))).sub(radius)).sub(thickness);
}
function crossSdf(dx, dy, size, thickness, rounding) {
	const px = abs(dx);
	const py = abs(dy);
	return min(max(px.sub(size), py.sub(thickness)), max(py.sub(size), px.sub(thickness))).sub(rounding);
}
function roundedRectSdf(dx, dy, width, height, rounding) {
	const qx = abs(dx).sub(width).add(rounding);
	const qy = abs(dy).sub(height).add(rounding);
	const outerDist = sqrt(max(qx, float(0)).mul(max(qx, float(0))).add(max(qy, float(0)).mul(max(qy, float(0)))));
	const innerDist = min(max(qx, qy), float(0));
	return outerDist.add(innerDist).sub(rounding);
}
function ellipseSdf(dx, dy, radiusX, radiusY) {
	const nx = dx.div(radiusX);
	const ny = dy.div(radiusY);
	return sqrt(nx.mul(nx).add(ny.mul(ny))).sub(float(1)).mul(min(radiusX, radiusY));
}
function vesicaSdf(dx, dy, radius, spread) {
	const px = abs(dx);
	const py = abs(dy);
	const d = radius.mul(spread);
	const b = sqrt(max(radius.mul(radius).sub(d.mul(d)), float(0)));
	const cond = py.sub(b).mul(d).greaterThan(px.mul(b));
	const dist1 = sqrt(px.mul(px).add(py.sub(b).mul(py.sub(b))));
	const dist2 = sqrt(px.add(d).mul(px.add(d)).add(py.mul(py))).sub(radius);
	return cond.select(dist1, dist2);
}
function crescentSdf(dx, dy, outerRadius, innerRatio, offset) {
	const py = abs(dy);
	const ra = outerRadius;
	const rb = outerRadius.mul(innerRatio);
	const d = offset;
	const a = ra.mul(ra).sub(rb.mul(rb)).add(d.mul(d)).div(d.mul(float(2)).max(float(.001)));
	const b = sqrt(max(ra.mul(ra).sub(a.mul(a)), float(0)));
	const lhs = d.mul(dx.mul(b).sub(py.mul(a)));
	const rhs = d.mul(d).mul(max(b.sub(py), float(0)));
	const d1x = dx.sub(a);
	const d1y = py.sub(b);
	const dist1 = sqrt(d1x.mul(d1x).add(d1y.mul(d1y)));
	const lenP = sqrt(dx.mul(dx).add(py.mul(py)));
	const innerX = dx.sub(d);
	const lenInner = sqrt(innerX.mul(innerX).add(py.mul(py)));
	const dist2 = max(lenP.sub(ra), lenInner.sub(rb).negate());
	return lhs.greaterThan(rhs).select(dist1, dist2);
}
function trapezoidSdf(dx, dy, r1, r2, he) {
	const px = abs(dx);
	const k1x = r2;
	const k1y = he;
	const k2x = r2.sub(r1);
	const k2y = he.mul(float(2));
	const clampWidth = dy.lessThan(float(0)).select(r1, r2);
	const cax = px.sub(min(px, clampWidth));
	const cay = abs(dy).sub(he);
	const diffX = k1x.sub(px);
	const diffY = k1y.sub(dy);
	const dotNum = diffX.mul(k2x).add(diffY.mul(k2y));
	const dotDen = k2x.mul(k2x).add(k2y.mul(k2y)).max(float(1e-4));
	const t = dotNum.div(dotDen).clamp(float(0), float(1));
	const cbx = px.sub(k1x).add(k2x.mul(t));
	const cby = dy.sub(k1y).add(k2y.mul(t));
	const s = cbx.lessThan(float(0)).and(cay.lessThan(float(0))).select(float(-1), float(1));
	const dca = cax.mul(cax).add(cay.mul(cay));
	const dcb = cbx.mul(cbx).add(cby.mul(cby));
	return s.mul(sqrt(min(dca, dcb)));
}
async function loadSdfFromUrl(url, sdfData) {
	const response = await fetch(url, { mode: "cors" });
	if (!response.ok) throw new Error(`SDF fetch failed: HTTP ${response.status}`);
	const buffer = await response.arrayBuffer();
	if (buffer.byteLength === 512 * 512 * 2) {
		const uint16 = new Uint16Array(buffer);
		const n = Math.min(uint16.length, sdfData.length);
		for (let i = 0; i < n; i++) sdfData[i] = uint16[i] / 32767.5 - 1;
	} else {
		const floats = new Float32Array(buffer);
		const n = Math.min(floats.length, sdfData.length);
		for (let i = 0; i < n; i++) sdfData[i] = floats[i];
	}
}
function createSvgSdfSampler(shapeSdfUrl, onBeforeRender, onCleanup) {
	const sdfData = new Float32Array(512 * 512).fill(.5);
	const sdfTexture = new DataTexture(sdfData, 512, 512, RedFormat, FloatType);
	sdfTexture.magFilter = LinearFilter;
	sdfTexture.minFilter = LinearFilter;
	sdfTexture.needsUpdate = true;
	onCleanup(() => sdfTexture.dispose());
	let sdfNeedsUpdate = false;
	if (shapeSdfUrl) loadSdfFromUrl(shapeSdfUrl, sdfData).then(() => {
		sdfNeedsUpdate = true;
	}).catch((err) => console.warn("[SDF] Failed to load SDF texture:", err));
	onBeforeRender(() => {
		if (sdfNeedsUpdate) {
			sdfNeedsUpdate = false;
			sdfTexture.needsUpdate = true;
		}
	});
	const sdfNode = texture(sdfTexture);
	return (uv$1) => sdfNode.sample(uv$1);
}
function createAnalyticSdfSampler(shapeType, initialConfig, shapeUniform, onBeforeRender) {
	const radiusU = uniform(initialConfig.radius ?? .35);
	const sidesU = uniform(initialConfig.sides ?? 6);
	const roundingU = uniform(initialConfig.rounding ?? 0);
	const innerRatioU = uniform(initialConfig.innerRatio ?? initialConfig.thickness ?? .4);
	const rotationU = uniform(initialConfig.rotation ?? 0);
	const heightU = uniform(initialConfig.height ?? .25);
	const offsetU = uniform(initialConfig.offset ?? .2);
	let lastShapeJson = "";
	let lastCfg = initialConfig;
	onBeforeRender(() => {
		const raw = shapeUniform.uniform.value;
		if (typeof raw === "object" && raw !== null) lastCfg = raw;
		else {
			const str = raw;
			if (str !== lastShapeJson) {
				lastShapeJson = str;
				try {
					lastCfg = JSON.parse(str);
				} catch {
					return;
				}
			}
		}
		radiusU.value = lastCfg.radius ?? lastCfg.width ?? lastCfg.bottomWidth ?? .35;
		sidesU.value = lastCfg.sides ?? 6;
		roundingU.value = lastCfg.rounding ?? 0;
		innerRatioU.value = lastCfg.innerRatio ?? lastCfg.thickness ?? lastCfg.spread ?? lastCfg.topWidth ?? lastCfg.topRatio ?? .4;
		rotationU.value = lastCfg.rotation ?? 0;
		heightU.value = lastCfg.height ?? .25;
		offsetU.value = lastCfg.offset ?? .2;
	});
	const needsRotation = shapeType !== "circleSDF" && shapeType !== "ringSDF";
	return (uv$1) => {
		const dx = uv$1.x.sub(float(.5));
		const dy = uv$1.y.sub(float(.5));
		let rdx = dx, rdy = dy;
		if (needsRotation) {
			const rotRad = rotationU.mul(float(Math.PI / 180));
			const cosR = cos(rotRad);
			const sinR = sin(rotRad);
			rdx = dx.mul(cosR).add(dy.mul(sinR));
			rdy = dy.mul(cosR).sub(dx.mul(sinR));
		}
		let dist;
		switch (shapeType) {
			case "polygonSDF":
				dist = mix(polygonSdf(rdx, rdy, radiusU, sidesU), circleSdf(dx, dy, radiusU), roundingU);
				break;
			case "flowerSDF":
				dist = flowerSdf(rdx, rdy, radiusU, sidesU, innerRatioU);
				break;
			case "ringSDF":
				dist = ringSdf(dx, dy, radiusU, innerRatioU);
				break;
			case "crossSDF":
				dist = crossSdf(rdx, rdy, radiusU, innerRatioU, roundingU);
				break;
			case "starSDF":
				dist = starSdf(rdx, rdy, radiusU, sidesU, innerRatioU);
				break;
			case "roundedRectSDF":
				dist = roundedRectSdf(rdx, rdy, radiusU, heightU, roundingU);
				break;
			case "ellipseSDF":
				dist = ellipseSdf(rdx, rdy, radiusU, heightU);
				break;
			case "vesicaSDF":
				dist = vesicaSdf(rdx, rdy, radiusU, innerRatioU);
				break;
			case "crescentSDF":
				dist = crescentSdf(rdx, rdy, radiusU, innerRatioU, offsetU);
				break;
			case "trapezoidSDF":
				dist = trapezoidSdf(rdx, rdy, innerRatioU, radiusU, heightU);
				break;
			case "circleSDF":
			default:
				dist = circleSdf(dx, dy, radiusU);
				break;
		}
		return vec4(dist, float(0), float(0), float(1));
	};
}
export { crossSdf as a, polygonSdf as c, starSdf as d, trapezoidSdf as f, crescentSdf as i, ringSdf as l, createAnalyticSdfSampler as n, ellipseSdf as o, vesicaSdf as p, createSvgSdfSampler as r, flowerSdf as s, circleSdf as t, roundedRectSdf as u };
