// shaders/dist/core/shaderRegistry-DZDkRCt8.js
import { n as componentDefinition } from "./AngularBlur-C1oRJ8Y9.js";
import { n as componentDefinition$1 } from "./Ascii-BwSMmD6Z.js";
import { n as componentDefinition$2 } from "./Aurora-D5Pjk4ve.js";
import { n as componentDefinition$3 } from "./BarShift-2XBhva1A.js";
import { n as componentDefinition$4 } from "./Beam-B3BX08lR.js";
import { n as componentDefinition$5 } from "./Blob-B7IFHhVB.js";
import { n as componentDefinition$6 } from "./Blur-5_t3tR_g.js";
import { n as componentDefinition$7 } from "./BrickPattern-BfTsjp4I.js";
import { n as componentDefinition$8 } from "./BrightnessContrast-GppIqtkO.js";
import { n as componentDefinition$9 } from "./Bulge-DMOATSEs.js";
import { n as componentDefinition$10 } from "./ChannelBlur-DWgS3nhn.js";
import { n as componentDefinition$11 } from "./Checkerboard-BVBGTvd-.js";
import { n as componentDefinition$12 } from "./Chevron-BKZjAXR_.js";
import { n as componentDefinition$13 } from "./ChromaFlow-C0guw_LP.js";
import { n as componentDefinition$14 } from "./ChromaticAberration-vygkvgz-.js";
import { n as componentDefinition$15 } from "./Circle-BycqVCMt.js";
import { n as componentDefinition$16 } from "./ColorWheel-BQUKQPM-.js";
import { n as componentDefinition$17 } from "./ConcentricSpin-CO3Kx1I1.js";
import { n as componentDefinition$18 } from "./ConicGradient-DRvVwiSj.js";
import { n as componentDefinition$19 } from "./ContourLines-s4waN3uU.js";
import { n as componentDefinition$20 } from "./Crescent-ppz7W7eM.js";
import { n as componentDefinition$21 } from "./Cross-BfNOtQ6s.js";
import { n as componentDefinition$22 } from "./CRTScreen-corekL_R.js";
import { n as componentDefinition$23 } from "./Crystal-QsQoy8fZ.js";
import { n as componentDefinition$24 } from "./CursorRipples-CLP7Sinv.js";
import { n as componentDefinition$25 } from "./CursorTrail-CshoSM5Y.js";
import { n as componentDefinition$26 } from "./DiamondGradient-D6eyjZOn.js";
import { n as componentDefinition$27 } from "./DiffuseBlur-BjvXHmJx.js";
import { n as componentDefinition$28 } from "./Dither-z6s-TFg7.js";
import { n as componentDefinition$29 } from "./DOMTexture-Ec0k6AnK.js";
import { n as componentDefinition$30 } from "./DotGrid-BxAAGO4F.js";
import { n as componentDefinition$31 } from "./DropShadow-CR3cecIQ.js";
import { n as componentDefinition$32 } from "./Duotone-PwTfBBtM.js";
import { n as componentDefinition$33 } from "./Ellipse-sWhNvW9-.js";
import { n as componentDefinition$34 } from "./Emboss-DGaubb9x.js";
import { n as componentDefinition$35 } from "./FallingLines-DqIZ8wTH.js";
import { n as componentDefinition$36 } from "./FilmGrain-Dbw02Jz9.js";
import { n as componentDefinition$37 } from "./FloatingParticles-LMsFTJKp.js";
import { n as componentDefinition$38 } from "./Flower-BbRNrXIa.js";
import { n as componentDefinition$39 } from "./FlowField-D3CAHBBG.js";
import { n as componentDefinition$40 } from "./FlowingGradient-BRQ_K-k3.js";
import { n as componentDefinition$41 } from "./FlutedGlass-CyFWN_SM.js";
import { n as componentDefinition$42 } from "./Fog-CLGVVVeo.js";
import { n as componentDefinition$43 } from "./Form3D-8tPyUgh4.js";
import { n as componentDefinition$44 } from "./FractalNoise-Ccufnbw6.js";
import { n as componentDefinition$45 } from "./Glass-CXT1jdLw.js";
import { n as componentDefinition$46 } from "./GlassTiles-CXJHjEBe.js";
import { n as componentDefinition$47 } from "./Glitch-Bg9W-KNm.js";
import { n as componentDefinition$48 } from "./Glow-5PP4-rM9.js";
import { n as componentDefinition$49 } from "./Godrays-DHXKsw5B.js";
import { n as componentDefinition$50 } from "./Grayscale-DgXoUcmg.js";
import { n as componentDefinition$51 } from "./Grid-DwyhjFAT.js";
import { n as componentDefinition$52 } from "./GridDistortion-DJ8S-96Y.js";
import { n as componentDefinition$53 } from "./Group-DmgNNlDd.js";
import { n as componentDefinition$54 } from "./Halftone-x9boWkFW.js";
import { n as componentDefinition$55 } from "./HexGrid-BUsgyF8P.js";
import { n as componentDefinition$56 } from "./HueShift-D3_5D7jx.js";
import { n as componentDefinition$57 } from "./ImageTexture-6ZzBt79u.js";
import { n as componentDefinition$58 } from "./Invert-kXxY97QJ.js";
import { n as componentDefinition$59 } from "./Kaleidoscope-D4jRcZCt.js";
import { n as componentDefinition$60 } from "./LensFlare-D7vypE2Q.js";
import { n as componentDefinition$61 } from "./LinearBlur-2hspd2YP.js";
import { n as componentDefinition$62 } from "./LinearGradient-glCc9IaB.js";
import { n as componentDefinition$63 } from "./Liquify-uZjs0wcu.js";
import { n as componentDefinition$64 } from "./Marble-DkHLcqFD.js";
import { n as componentDefinition$65 } from "./Mirror-Dv5fOEp6.js";
import { n as componentDefinition$66 } from "./MultiPointGradient-Dvt1GUiv.js";
import { n as componentDefinition$67 } from "./Neon-DQDB7hAE.js";
import { n as componentDefinition$68 } from "./Paper-DNP2-hPM.js";
import { n as componentDefinition$69 } from "./Perspective-CAIchX9p.js";
import { n as componentDefinition$70 } from "./Pixelate-BbT3Rrf2.js";
import { n as componentDefinition$71 } from "./Plasma-D9rhkH2y.js";
import { n as componentDefinition$72 } from "./PolarCoordinates-2RBCk-fo.js";
import { n as componentDefinition$73 } from "./Polygon-C6_siOs0.js";
import { n as componentDefinition$74 } from "./Posterize-DNxTRgvF.js";
import { n as componentDefinition$75 } from "./ProgressiveBlur-BJwJ5mE0.js";
import { n as componentDefinition$76 } from "./RadialGradient-CQIctl__.js";
import { n as componentDefinition$77 } from "./RectangularCoordinates-F2VzXmal.js";
import { n as componentDefinition$78 } from "./ReflectivePlane-UIT4xlmn.js";
import { n as componentDefinition$79 } from "./Ring-UT__kmzY.js";
import { n as componentDefinition$80 } from "./Ripples-CWVLSUP0.js";
import { n as componentDefinition$81 } from "./RoundedRect-BPWHTT1j.js";
import { n as componentDefinition$82 } from "./Saturation-DFOp2yD9.js";
import { n as componentDefinition$83 } from "./Sharpness-C-IEswPL.js";
import { n as componentDefinition$84 } from "./Shatter-CUYQp-qy.js";
import { n as componentDefinition$85 } from "./SimplexNoise-LAxWw8fJ.js";
import { n as componentDefinition$86 } from "./SineWave-CEQWxC9Q.js";
import { n as componentDefinition$87 } from "./Smoke-DO4yylWk.js";
import { n as componentDefinition$88 } from "./SmokeFill-Dm0sZS5F.js";
import { n as componentDefinition$89 } from "./Solarize-mE7LU1li.js";
import { n as componentDefinition$90 } from "./SolidColor-BljkSdvg.js";
import { n as componentDefinition$91 } from "./Spherize-DARagREe.js";
import { n as componentDefinition$92 } from "./Spiral-CumsIpkp.js";
import { n as componentDefinition$93 } from "./Star-8fAlzR9L.js";
import { n as componentDefinition$94 } from "./Strands-BB4RIsFz.js";
import { n as componentDefinition$95 } from "./Stretch-3S2atBkJ.js";
import { n as componentDefinition$96 } from "./Stripes-BGFVSZIt.js";
import { n as componentDefinition$97 } from "./StudioBackground-BtJ6b1Ki.js";
import { n as componentDefinition$98 } from "./SunBurst-BJRFFYhs.js";
import { n as componentDefinition$99 } from "./Swirl-KA9cEnLI.js";
import { n as componentDefinition$100 } from "./TiltShift-DtQ3dRL-.js";
import { n as componentDefinition$101 } from "./Tint-W_EvzN1-.js";
import { n as componentDefinition$102 } from "./Trapezoid-cQzS6-bh.js";
import { n as componentDefinition$103 } from "./Tritone-B3hI2nAi.js";
import { n as componentDefinition$104 } from "./Truchet-s9PmowCP.js";
import { n as componentDefinition$105 } from "./Twirl-UDDXGkFl.js";
import { n as componentDefinition$106 } from "./Vesica-ChQBYWuw.js";
import { n as componentDefinition$107 } from "./VHS-DW1H7Wuy.js";
import { n as componentDefinition$108 } from "./Vibrance-DUj7hwzE.js";
import { n as componentDefinition$109 } from "./VideoTexture-ClutlOqj.js";
import { n as componentDefinition$110 } from "./Vignette-F9yxi-UM.js";
import { n as componentDefinition$111 } from "./Voronoi-D8HHP_WR.js";
import { n as componentDefinition$112 } from "./WaveDistortion-CDIc5Uyk.js";
import { n as componentDefinition$113 } from "./Weave-ou5shgl3.js";
import { n as componentDefinition$114 } from "./WebcamTexture-D4e06ajM.js";
import { n as componentDefinition$115 } from "./WorleyNoise-Uf6IPm7A.js";
import { n as componentDefinition$116 } from "./ZoomBlur-DJ-RNKHM.js";
var createShaderRegistry = () => {
	const registry = {};
	const shaderDefinitions = {
		AngularBlur: componentDefinition,
		Ascii: componentDefinition$1,
		Aurora: componentDefinition$2,
		BarShift: componentDefinition$3,
		Beam: componentDefinition$4,
		Blob: componentDefinition$5,
		Blur: componentDefinition$6,
		BrickPattern: componentDefinition$7,
		BrightnessContrast: componentDefinition$8,
		Bulge: componentDefinition$9,
		ChannelBlur: componentDefinition$10,
		Checkerboard: componentDefinition$11,
		Chevron: componentDefinition$12,
		ChromaFlow: componentDefinition$13,
		ChromaticAberration: componentDefinition$14,
		Circle: componentDefinition$15,
		ColorWheel: componentDefinition$16,
		ConcentricSpin: componentDefinition$17,
		ConicGradient: componentDefinition$18,
		ContourLines: componentDefinition$19,
		Crescent: componentDefinition$20,
		Cross: componentDefinition$21,
		CRTScreen: componentDefinition$22,
		Crystal: componentDefinition$23,
		CursorRipples: componentDefinition$24,
		CursorTrail: componentDefinition$25,
		DiamondGradient: componentDefinition$26,
		DiffuseBlur: componentDefinition$27,
		Dither: componentDefinition$28,
		DOMTexture: componentDefinition$29,
		DotGrid: componentDefinition$30,
		DropShadow: componentDefinition$31,
		Duotone: componentDefinition$32,
		Ellipse: componentDefinition$33,
		Emboss: componentDefinition$34,
		FallingLines: componentDefinition$35,
		FilmGrain: componentDefinition$36,
		FloatingParticles: componentDefinition$37,
		Flower: componentDefinition$38,
		FlowField: componentDefinition$39,
		FlowingGradient: componentDefinition$40,
		FlutedGlass: componentDefinition$41,
		Fog: componentDefinition$42,
		Form3D: componentDefinition$43,
		FractalNoise: componentDefinition$44,
		Glass: componentDefinition$45,
		GlassTiles: componentDefinition$46,
		Glitch: componentDefinition$47,
		Glow: componentDefinition$48,
		Godrays: componentDefinition$49,
		Grayscale: componentDefinition$50,
		Grid: componentDefinition$51,
		GridDistortion: componentDefinition$52,
		Group: componentDefinition$53,
		Halftone: componentDefinition$54,
		HexGrid: componentDefinition$55,
		HueShift: componentDefinition$56,
		ImageTexture: componentDefinition$57,
		Invert: componentDefinition$58,
		Kaleidoscope: componentDefinition$59,
		LensFlare: componentDefinition$60,
		LinearBlur: componentDefinition$61,
		LinearGradient: componentDefinition$62,
		Liquify: componentDefinition$63,
		Marble: componentDefinition$64,
		Mirror: componentDefinition$65,
		MultiPointGradient: componentDefinition$66,
		Neon: componentDefinition$67,
		Paper: componentDefinition$68,
		Perspective: componentDefinition$69,
		Pixelate: componentDefinition$70,
		Plasma: componentDefinition$71,
		PolarCoordinates: componentDefinition$72,
		Polygon: componentDefinition$73,
		Posterize: componentDefinition$74,
		ProgressiveBlur: componentDefinition$75,
		RadialGradient: componentDefinition$76,
		RectangularCoordinates: componentDefinition$77,
		ReflectivePlane: componentDefinition$78,
		Ring: componentDefinition$79,
		Ripples: componentDefinition$80,
		RoundedRect: componentDefinition$81,
		Saturation: componentDefinition$82,
		Sharpness: componentDefinition$83,
		Shatter: componentDefinition$84,
		SimplexNoise: componentDefinition$85,
		SineWave: componentDefinition$86,
		Smoke: componentDefinition$87,
		SmokeFill: componentDefinition$88,
		Solarize: componentDefinition$89,
		SolidColor: componentDefinition$90,
		Spherize: componentDefinition$91,
		Spiral: componentDefinition$92,
		Star: componentDefinition$93,
		Strands: componentDefinition$94,
		Stretch: componentDefinition$95,
		Stripes: componentDefinition$96,
		StudioBackground: componentDefinition$97,
		SunBurst: componentDefinition$98,
		Swirl: componentDefinition$99,
		TiltShift: componentDefinition$100,
		Tint: componentDefinition$101,
		Trapezoid: componentDefinition$102,
		Tritone: componentDefinition$103,
		Truchet: componentDefinition$104,
		Twirl: componentDefinition$105,
		Vesica: componentDefinition$106,
		VHS: componentDefinition$107,
		Vibrance: componentDefinition$108,
		VideoTexture: componentDefinition$109,
		Vignette: componentDefinition$110,
		Voronoi: componentDefinition$111,
		WaveDistortion: componentDefinition$112,
		Weave: componentDefinition$113,
		WebcamTexture: componentDefinition$114,
		WorleyNoise: componentDefinition$115,
		ZoomBlur: componentDefinition$116
	};
	Object.entries(shaderDefinitions).forEach(([fileName, def]) => {
		const propsMetadata = {};
		if (def.props) Object.entries(def.props).forEach(([propName, propConfig]) => {
			if (propConfig.ui) propsMetadata[propName] = {
				ui: propConfig.ui,
				default: propConfig.default,
				description: propConfig.description
			};
		});
		registry[def.name] = {
			name: def.name,
			fileName,
			category: def.category || "Uncategorized",
			definition: def,
			propsMetadata
		};
	});
	return registry;
};
const shaderRegistry = createShaderRegistry();
function getAllShaders() {
	return Object.values(shaderRegistry);
}
function getShaderByName(name) {
	return shaderRegistry[name];
}
function getShadersByCategory(category) {
	return Object.values(shaderRegistry).filter((shader) => shader.category === category);
}
function getShaderCategories() {
	const categories = new Set(Object.values(shaderRegistry).map((shader) => shader.category));
	return Array.from(categories).sort();
}
export { shaderRegistry as a, getShadersByCategory as i, getShaderByName as n, getShaderCategories as r, getAllShaders as t };
