// shaders/dist/core/Shatter-CUYQp-qy.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { StorageTexture } from "three/webgpu";
import { DataTexture, FloatType, NearestFilter, RGBAFormat } from "three";
import { Fn, If, convertToTexture, float, instanceIndex, int, screenUV, smoothstep, storageTexture, texture, textureStore, uint, uniformArray, uvec2, vec2, vec4 } from "three/tsl";
var seededRandom = (seed) => {
	const x = Math.sin(seed) * 1e4;
	return x - Math.floor(x);
};
var SHARD_COUNT = 16;
var VORONOI_SIZE = 1024;
const componentDefinition = {
	name: "Shatter",
	category: "Interactive",
	description: "Broken glass effect with tectonic plate displacement",
	requiresRTT: true,
	requiresChild: true,
	usesPointer: true,
	props: {
		crackWidth: {
			default: 1,
			description: "Thickness of crack lines",
			ui: {
				type: ["range", "map"],
				min: .5,
				max: 5,
				step: .1,
				label: "Crack Width",
				group: "Effect"
			}
		},
		intensity: {
			default: 4,
			description: "How much shards shift",
			ui: {
				type: "range",
				min: 0,
				max: 20,
				step: 1,
				label: "Intensity",
				group: "Effect"
			}
		},
		radius: {
			default: .4,
			description: "Cursor influence radius",
			ui: {
				type: "range",
				min: .1,
				max: 1,
				step: .1,
				label: "Radius",
				group: "Effect"
			}
		},
		decay: {
			default: 1,
			description: "How fast shards return to rest",
			ui: {
				type: "range",
				min: .1,
				max: 10,
				step: .1,
				label: "Decay",
				group: "Effect"
			}
		},
		seed: {
			default: 2,
			description: "Random seed for pattern",
			ui: {
				type: "range",
				min: 0,
				max: 50,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		chromaticSplit: {
			default: 1,
			description: "RGB separation for prismatic glass effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .1,
				label: "Chromatic Split",
				group: "Effect"
			}
		},
		refractionStrength: {
			default: 5,
			description: "How much cracks bend/distort the underlying image",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 10,
				step: .1,
				label: "Refraction",
				group: "Effect"
			}
		},
		shardLighting: {
			default: .1,
			description: "Subtle lighting on tilted shards for 3D depth",
			ui: {
				type: "range",
				min: 0,
				max: .5,
				step: .1,
				label: "Shard Lighting",
				group: "Effect"
			}
		},
		edges: {
			default: "mirror",
			description: "How to handle edges when displacement pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	computeNode: ({ uniforms, onCleanup }) => {
		const cellPositions = uniformArray(new Array(SHARD_COUNT * 2).fill(0), "float");
		const voronoiTex = new StorageTexture(VORONOI_SIZE, VORONOI_SIZE);
		voronoiTex.type = FloatType;
		voronoiTex.magFilter = NearestFilter;
		voronoiTex.minFilter = NearestFilter;
		voronoiTex.generateMipmaps = false;
		const voronoiTexNode = storageTexture(voronoiTex);
		onCleanup(() => {
			voronoiTex.dispose();
		});
		const vs = uint(VORONOI_SIZE);
		const invVS = float(1 / VORONOI_SIZE);
		const voronoiKernel = Fn(() => {
			const idx = instanceIndex;
			const x = idx.mod(vs);
			const y = idx.div(vs);
			If(y.lessThan(vs), () => {
				const u = x.toFloat().add(float(.5)).mul(invVS);
				const v = y.toFloat().add(float(.5)).mul(invVS);
				const d1 = float(99999).toVar();
				const d2 = float(99999).toVar();
				const idx1 = float(0).toVar();
				const idx2 = float(0).toVar();
				for (let i = 0; i < SHARD_COUNT; i++) {
					const cellX = cellPositions.element(int(i * 2));
					const cellY = cellPositions.element(int(i * 2 + 1));
					const ddx = u.sub(cellX);
					const ddy = v.sub(cellY);
					const distSq = ddx.mul(ddx).add(ddy.mul(ddy));
					If(distSq.lessThan(d1), () => {
						d2.assign(d1);
						idx2.assign(idx1);
						d1.assign(distSq);
						idx1.assign(float(i));
					}).Else(() => {
						If(distSq.lessThan(d2), () => {
							d2.assign(distSq);
							idx2.assign(float(i));
						});
					});
				}
				textureStore(voronoiTexNode, uvec2(x, y), vec4(idx1, idx2, float(0), float(0))).toWriteOnly();
			});
		})().compute(VORONOI_SIZE * VORONOI_SIZE, [256]);
		let currentSeed = uniforms.seed.uniform.value;
		let needsDispatch = true;
		const generateCells = (seed) => {
			const positions = new Array(SHARD_COUNT * 2);
			for (let i = 0; i < SHARD_COUNT; i++) {
				positions[i * 2] = seededRandom(seed + i * 2);
				positions[i * 2 + 1] = seededRandom(seed + i * 2 + 1);
			}
			cellPositions.array = positions;
		};
		generateCells(currentSeed);
		return {
			outputs: { voronoiTexture: voronoiTex },
			getComputeNodes: () => {
				const newSeed = uniforms.seed.uniform.value;
				if (newSeed !== currentSeed) {
					currentSeed = newSeed;
					generateCells(currentSeed);
					needsDispatch = true;
				}
				if (needsDispatch) {
					needsDispatch = false;
					return [voronoiKernel];
				}
				return null;
			}
		};
	},
	fragmentNode: ({ uniforms, onBeforeRender, childNode, onCleanup, dimensions, computeOutputs }) => {
		const FRICTION = 1;
		const cellData = new Float32Array(SHARD_COUNT * 4);
		const cellPosArray = new Array(SHARD_COUNT * 2).fill(0);
		const cellPosUniform = uniformArray(cellPosArray, "float");
		const displacementData = new Float32Array(SHARD_COUNT * 4);
		const dispArray = new Array(SHARD_COUNT * 2).fill(0);
		const dispUniform = uniformArray(dispArray, "float");
		const syncCellPosUniform = () => {
			for (let i = 0; i < SHARD_COUNT; i++) {
				cellPosArray[i * 2] = cellData[i * 4];
				cellPosArray[i * 2 + 1] = cellData[i * 4 + 1];
			}
			cellPosUniform.array = cellPosArray;
		};
		const syncDispUniform = () => {
			for (let i = 0; i < SHARD_COUNT; i++) {
				dispArray[i * 2] = displacementData[i * 4];
				dispArray[i * 2 + 1] = displacementData[i * 4 + 1];
			}
			dispUniform.array = dispArray;
		};
		let voronoiField;
		if (computeOutputs?.voronoiTexture) voronoiField = texture(computeOutputs.voronoiTexture);
		else {
			const voronoiData = new Float32Array(VORONOI_SIZE * VORONOI_SIZE * 4);
			const voronoiTexture = new DataTexture(voronoiData, VORONOI_SIZE, VORONOI_SIZE, RGBAFormat, FloatType);
			voronoiTexture.magFilter = NearestFilter;
			voronoiTexture.minFilter = NearestFilter;
			voronoiField = texture(voronoiTexture);
			const generateVoronoi = () => {
				for (let y = 0; y < VORONOI_SIZE; y++) for (let x = 0; x < VORONOI_SIZE; x++) {
					const u = (x + .5) / VORONOI_SIZE;
					const v = (y + .5) / VORONOI_SIZE;
					let d1 = Infinity, d2 = Infinity, i1 = 0, i2 = 0;
					for (let i = 0; i < SHARD_COUNT; i++) {
						const dx = u - cellData[i * 4];
						const dy = v - cellData[i * 4 + 1];
						const distSq = dx * dx + dy * dy;
						if (distSq < d1) {
							d2 = d1;
							i2 = i1;
							d1 = distSq;
							i1 = i;
						} else if (distSq < d2) {
							d2 = distSq;
							i2 = i;
						}
					}
					const offset = (y * VORONOI_SIZE + x) * 4;
					voronoiData[offset] = i1;
					voronoiData[offset + 1] = i2;
				}
				voronoiTexture.needsUpdate = true;
			};
			let currentSeed = uniforms.seed.uniform.value;
			const generateCells = (seed) => {
				for (let i = 0; i < SHARD_COUNT; i++) {
					cellData[i * 4] = seededRandom(seed + i * 2);
					cellData[i * 4 + 1] = seededRandom(seed + i * 2 + 1);
					cellData[i * 4 + 2] = seededRandom(seed + i * 3);
					cellData[i * 4 + 3] = seededRandom(seed + i * 3 + 1);
				}
				syncCellPosUniform();
			};
			generateCells(currentSeed);
			generateVoronoi();
			onBeforeRender(() => {
				const newSeed = uniforms.seed.uniform.value;
				if (newSeed !== currentSeed) {
					currentSeed = newSeed;
					generateCells(currentSeed);
					generateVoronoi();
					displacementData.fill(0);
				}
			});
			onCleanup(() => {
				voronoiTexture.dispose();
			});
		}
		const generateCellsFull = (seed) => {
			for (let i = 0; i < SHARD_COUNT; i++) {
				cellData[i * 4] = seededRandom(seed + i * 2);
				cellData[i * 4 + 1] = seededRandom(seed + i * 2 + 1);
				cellData[i * 4 + 2] = seededRandom(seed + i * 3);
				cellData[i * 4 + 3] = seededRandom(seed + i * 3 + 1);
			}
			syncCellPosUniform();
		};
		let currentSeedForDisp = uniforms.seed.uniform.value;
		generateCellsFull(currentSeedForDisp);
		displacementData.fill(0);
		syncDispUniform();
		let prevX = .5, prevY = .5, lastTime = Date.now();
		onBeforeRender(({ pointer }) => {
			const currentTime = Date.now();
			const dt = Math.min((currentTime - lastTime) / 1e3, .016);
			lastTime = currentTime;
			const aspect = dimensions.width / dimensions.height;
			const newSeed = uniforms.seed.uniform.value;
			if (newSeed !== currentSeedForDisp) {
				currentSeedForDisp = newSeed;
				generateCellsFull(currentSeedForDisp);
				displacementData.fill(0);
			}
			const intensity = uniforms.intensity.uniform.value;
			const radius = uniforms.radius.uniform.value;
			const decay = uniforms.decay.uniform.value;
			let velX = dt > 0 ? (pointer.x - prevX) / dt : 0;
			let velY = dt > 0 ? (pointer.y - prevY) / dt : 0;
			let speed = Math.sqrt(velX * velX + velY * velY);
			const maxVelocity = 5 + intensity * 2;
			if (speed > maxVelocity) {
				const scale = maxVelocity / speed;
				velX *= scale;
				velY *= scale;
				speed = maxVelocity;
			}
			for (let i = 0; i < SHARD_COUNT; i++) {
				const cellX = cellData[i * 4];
				const cellY = cellData[i * 4 + 1];
				const randomDirX = cellData[i * 4 + 2] - .5;
				const randomDirY = cellData[i * 4 + 3] - .5;
				const dx = aspect >= 1 ? (cellX - pointer.x) * aspect : cellX - pointer.x;
				const dy = aspect >= 1 ? cellY - pointer.y : (cellY - pointer.y) / aspect;
				const dist = Math.sqrt(dx * dx + dy * dy);
				let currentDx = displacementData[i * 4];
				let currentDy = displacementData[i * 4 + 1];
				const decayTime = Math.max(.01, decay);
				const decayFactor = Math.exp(-dt / decayTime);
				currentDx *= decayFactor;
				currentDy *= decayFactor;
				let velocityDx = 0, velocityDy = 0;
				if (dist < radius && speed > .01) {
					const influence = Math.max(0, 1 - dist / radius);
					const influenceCurve = influence * influence;
					const pushForce = influenceCurve * speed * intensity * dt * .5;
					velocityDx = velX * pushForce;
					velocityDy = velY * pushForce;
					const jitterForce = influenceCurve * speed * intensity * dt * .1;
					velocityDx += randomDirX * jitterForce;
					velocityDy += randomDirY * jitterForce;
				}
				const lerpFactor = Math.min(1, FRICTION * dt);
				displacementData[i * 4] = currentDx + velocityDx * lerpFactor;
				displacementData[i * 4 + 1] = currentDy + velocityDy * lerpFactor;
				displacementData[i * 4 + 2] = 0;
				displacementData[i * 4 + 3] = 0;
			}
			syncDispUniform();
			prevX = pointer.x;
			prevY = pointer.y;
		});
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const voronoiSample = voronoiField.sample(screenUV);
		const nearestI = voronoiSample.r.toInt();
		const secondI = voronoiSample.g.toInt();
		const nearestCellPos = vec2(cellPosUniform.element(nearestI.mul(2)), cellPosUniform.element(nearestI.mul(2).add(1)));
		const secondCellPos = vec2(cellPosUniform.element(secondI.mul(2)), cellPosUniform.element(secondI.mul(2).add(1)));
		const minDist1 = screenUV.sub(nearestCellPos).length();
		const minDist2 = screenUV.sub(secondCellPos).length();
		const displacement = vec2(dispUniform.element(nearestI.mul(2)), dispUniform.element(nearestI.mul(2).add(1)));
		const displacementMag = displacement.length();
		const edgeDiff = minDist2.sub(minDist1);
		const crackThreshold = uniforms.crackWidth.uniform.mul(.005);
		const baseCrackIntensity = float(1).sub(smoothstep(float(0), crackThreshold, edgeDiff));
		const displacementFactor = smoothstep(float(0), float(.01), displacementMag);
		const crackIntensity = baseCrackIntensity.mul(displacementFactor);
		const toNearest = screenUV.sub(nearestCellPos).normalize();
		const edgeNormal = vec2(toNearest.y.negate(), toNearest.x);
		const displacedUV = screenUV.sub(displacement);
		const refractionOffset = edgeNormal.mul(uniforms.refractionStrength.uniform).mul(.01);
		const chromaticOffset = uniforms.chromaticSplit.uniform.mul(.005);
		const rOffset = refractionOffset.add(edgeNormal.mul(chromaticOffset));
		const gOffset = refractionOffset;
		const bOffset = refractionOffset.sub(edgeNormal.mul(chromaticOffset));
		const rUV = displacedUV.add(rOffset.mul(crackIntensity));
		const gUV = displacedUV.add(gOffset.mul(crackIntensity));
		const bUV = displacedUV.add(bOffset.mul(crackIntensity));
		const rFinal = applyEdgeHandling(rUV, childTexture.sample(rUV), childTexture, uniforms.edges.uniform.value);
		const gFinal = applyEdgeHandling(gUV, childTexture.sample(gUV), childTexture, uniforms.edges.uniform.value);
		const bFinal = applyEdgeHandling(bUV, childTexture.sample(bUV), childTexture, uniforms.edges.uniform.value);
		const refractedColor = vec4(rFinal.r, gFinal.g, bFinal.b, gFinal.a);
		const normalColor = applyEdgeHandling(displacedUV, childTexture.sample(displacedUV), childTexture, uniforms.edges.uniform.value);
		let shadedRGB = normalColor.rgb.mul(float(1).sub(crackIntensity)).add(refractedColor.rgb.mul(crackIntensity));
		const displacementLength = displacement.length();
		const tiltX = displacement.x.div(displacementLength.add(.001));
		const tiltY = displacement.y.div(displacementLength.add(.001));
		const lightDir3D = vec2(.3, .6).normalize();
		const normalDot = tiltX.mul(lightDir3D.x).add(tiltY.mul(lightDir3D.y));
		const lightingFactor = float(1).add(normalDot.mul(uniforms.shardLighting.uniform));
		const lightingIntensity = smoothstep(float(0), float(.02), displacementLength);
		const finalLighting = float(1).add(lightingFactor.sub(1).mul(lightingIntensity));
		return unpremultiplyAlpha(vec4(shadedRGB.mul(finalLighting), normalColor.a));
	}
};
var Shatter_default = componentDefinition;
export { componentDefinition as n, Shatter_default as t };
