// shaders/dist/core/SimplexNoise-LAxWw8fJ.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { clamp, exp, mx_noise_float, screenUV, vec2, vec3, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "SimplexNoise",
	category: "Textures",
	description: "Organic noise with animated movement",
	props: {
		colorA: {
			default: "#ffffff",
			description: "First color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#000000",
			description: "Second color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		scale: {
			default: 2,
			description: "Pattern scale (higher = larger patterns)",
			ui: {
				type: "range",
				min: -2,
				max: 5,
				step: .1,
				label: "Scale",
				group: "Effect"
			}
		},
		balance: {
			default: 0,
			description: "Balance between colors (negative = more colorB, positive = more colorA)",
			ui: {
				type: ["range", "map"],
				min: -1,
				max: 1,
				step: .1,
				label: "Balance",
				group: "Effect"
			}
		},
		contrast: {
			default: 0,
			description: "Pattern contrast (higher = sharper transitions)",
			ui: {
				type: ["range", "map"],
				min: -2,
				max: 5,
				step: .1,
				label: "Contrast",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for pattern variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(uv$1.x.mul(aspect), uv$1.y);
		const animTime = createAnimatedTime(params, uniforms.speed);
		const pos = aspectCorrectedUV.mul(exp(uniforms.scale.uniform)).add(uniforms.seed.uniform);
		const noise = mx_noise_float(vec3(pos.x, pos.y, animTime.mul(.5)));
		const contrastMultiplier = uniforms.contrast.uniform.add(1);
		const k = clamp(noise.mul(contrastMultiplier).add(uniforms.balance.uniform).mul(.5).add(.5), 0, 1);
		return mixColors(uniforms.colorB.uniform, uniforms.colorA.uniform, k, uniforms.colorSpace.uniform.value);
	}
};
var SimplexNoise_default = componentDefinition;
export { componentDefinition as n, SimplexNoise_default as t };
