// shaders/dist/core/Solarize-mE7LU1li.js
import { dot, mix, vec3, vec4 } from "three/tsl";
const componentDefinition = {
	name: "Solarize",
	category: "Adjustments",
	description: "Inverts tones above a luminance threshold — a classic darkroom and photo effect",
	requiresChild: true,
	props: {
		threshold: {
			default: .5,
			description: "Luminance level above which colors are inverted. Pixels brighter than this threshold get flipped.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Threshold",
				group: "Effect"
			}
		},
		strength: {
			default: 1,
			description: "Blend between original (0) and fully solarized (1)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Strength",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Solarize shader.");
			return vec4(0);
		}
		const threshold = uniforms.threshold.uniform;
		const strength = uniforms.strength.uniform;
		const luminance = dot(childNode.rgb, vec3(.299, .587, .114));
		const inverted = vec3(1).sub(childNode.rgb);
		const solarized = luminance.greaterThan(threshold).select(inverted, childNode.rgb);
		return vec4(mix(childNode.rgb, solarized, strength), childNode.a);
	}
};
var Solarize_default = componentDefinition;
export { componentDefinition as n, Solarize_default as t };
