// shaders/dist/core/Stretch-3S2atBkJ.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { clamp, convertToTexture, cos, float, mix, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Stretch",
	category: "Distortions",
	description: "Stretch content towards a direction from a center point",
	requiresRTT: true,
	requiresChild: true,
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the stretch effect",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		strength: {
			default: 1,
			description: "The intensity of the stretch effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Strength",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			description: "The direction of the stretch in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		falloff: {
			default: 0,
			description: "Controls the sharpness of the transition (0 = sharp edge, 1 = gradual transition)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Falloff",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(screenUV.x.mul(aspect), screenUV.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus());
		const delta = aspectCorrectedUV.sub(centerPos);
		const angleRad = uniforms.angle.uniform.mul(Math.PI / 180);
		const directionVector = vec2(cos(angleRad), sin(angleRad));
		const projection = delta.dot(directionVector);
		const parallel = directionVector.mul(projection);
		const perpendicular = delta.sub(parallel);
		const falloffAmount = uniforms.falloff.uniform;
		const transitionWidth = mix(float(.001), float(75), falloffAmount);
		const effectMask = clamp(projection.div(transitionWidth), float(0), float(1));
		const scaledStrength = uniforms.strength.uniform.mul(float(100));
		const stretchFactor = float(1).add(scaledStrength.mul(effectMask));
		const finalProjection = projection.div(stretchFactor);
		const stretchedDelta = directionVector.mul(finalProjection).add(perpendicular);
		const stretchedUV = centerPos.add(stretchedDelta);
		const finalUV = vec2(stretchedUV.x.div(aspect), stretchedUV.y);
		return unpremultiplyAlpha(applyEdgeHandling(finalUV, childTexture.sample(finalUV), childTexture, uniforms.edges.uniform.value));
	}
};
var Stretch_default = componentDefinition;
export { componentDefinition as n, Stretch_default as t };
