// shaders/dist/core/Stripes-BGFVSZIt.js
import { a as transformColor, o as transformColorSpace, r as transformAngle, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { cos, fract, fwidth, screenUV, sin, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Stripes",
	category: "Textures",
	description: "Alternating colored stripes with animation",
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "First stripe color",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Second stripe color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		angle: {
			default: 45,
			transform: transformAngle,
			description: "Angle of stripes in degrees",
			ui: {
				type: ["range", "map"],
				min: -180,
				max: 180,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		density: {
			default: 5,
			description: "Number of stripe pairs visible",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: 1,
				label: "Density",
				group: "Effect"
			}
		},
		balance: {
			default: .5,
			description: "Ratio of the two colors",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Balance",
				group: "Effect"
			}
		},
		softness: {
			default: 0,
			description: "Edge softness",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Softness",
				group: "Effect"
			}
		},
		speed: {
			default: .2,
			description: "Animation speed",
			ui: {
				type: "range",
				min: -1,
				max: 1,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		offset: {
			default: 0,
			description: "Phase offset for pattern positioning",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Offset",
				group: "Animation"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(uv$1.x.mul(aspect), uv$1.y);
		const angleRad = uniforms.angle.uniform.mul(.0174533);
		const cosA = cos(angleRad);
		const sinA = sin(angleRad);
		const rotatedCoord = aspectCorrectedUV.x.mul(cosA).add(aspectCorrectedUV.y.mul(sinA));
		const t = createAnimatedTime(params, uniforms.speed);
		const pattern = rotatedCoord.mul(uniforms.density.uniform).add(t).add(uniforms.offset.uniform);
		const stripe = fract(pattern);
		const fw = fwidth(pattern);
		const balance = uniforms.balance.uniform;
		const soft = uniforms.softness.uniform;
		const mask = smoothstep(balance.sub(soft).sub(fw), balance.add(soft).add(fw), stripe);
		const mixedColor = mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, mask, uniforms.colorSpace.uniform.value);
		return vec4(mixedColor.rgb, mixedColor.a);
	}
};
var Stripes_default = componentDefinition;
export { componentDefinition as n, Stripes_default as t };
