// shaders/dist/core/StudioBackground-BtJ6b1Ki.js
import { a as transformColor, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { Fn, clamp, cos, exp, float, fract, mix, screenUV, sin, smoothstep, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "StudioBackground",
	category: "Textures",
	description: "Multi-light studio background with ambient motion.",
	props: {
		color: {
			default: "#d8dbec",
			transform: transformColor,
			description: "Base studio surface color",
			ui: {
				type: "color",
				label: "Surface Color",
				group: "Colors"
			}
		},
		keyColor: {
			default: "#d5e4ea",
			transform: transformColor,
			description: "Color of the overhead key light",
			ui: {
				type: "color",
				label: "Key Color",
				group: "Key Light"
			}
		},
		keyIntensity: {
			default: 40,
			description: "Intensity of the key light",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Key Light"
			}
		},
		keySoftness: {
			default: 50,
			description: "How diffuse the key light is",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Softness",
				group: "Key Light"
			}
		},
		fillColor: {
			default: "#d5e4ea",
			transform: transformColor,
			description: "Color of the side fill lights",
			ui: {
				type: "color",
				label: "Fill Color",
				group: "Fill Lights"
			}
		},
		fillIntensity: {
			default: 10,
			description: "Intensity of the fill lights",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Fill Lights"
			}
		},
		fillSoftness: {
			default: 70,
			description: "How diffuse the fill lights are",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Softness",
				group: "Fill Lights"
			}
		},
		fillAngle: {
			default: 70,
			description: "How far apart the fill lights are from center",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Spread",
				group: "Fill Lights"
			}
		},
		backColor: {
			default: "#c8d4e8",
			transform: transformColor,
			description: "Color of the upward back wash",
			ui: {
				type: "color",
				label: "Back Color",
				group: "Back Light"
			}
		},
		backIntensity: {
			default: 20,
			description: "Intensity of the back wash",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Back Light"
			}
		},
		backSoftness: {
			default: 80,
			description: "How diffuse the back wash is",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Softness",
				group: "Back Light"
			}
		},
		brightness: {
			default: 20,
			description: "Overall ambient light level",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Ambient",
				group: "Lighting"
			}
		},
		vignette: {
			default: 0,
			description: "Edge darkening",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Vignette",
				group: "Lighting"
			}
		},
		center: {
			default: {
				x: .5,
				y: .8
			},
			transform: transformPosition,
			description: "Where the spotlight meets the floor",
			ui: {
				type: "position",
				label: "Stage Center",
				group: "Scene"
			}
		},
		lightTarget: {
			default: 100,
			description: "How far toward the floor vs wall the spotlights aim",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Light Depth",
				group: "Scene"
			}
		},
		wallCurvature: {
			default: 10,
			description: "How rounded the cove is",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Wall Curvature",
				group: "Scene"
			}
		},
		ambientIntensity: {
			default: 50,
			description: "Intensity of drifting ambient lights",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Ambient Detail",
				group: "Environment"
			}
		},
		ambientSpeed: {
			default: 2,
			description: "Drift speed",
			ui: {
				type: "range",
				min: -5,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Environment"
			}
		},
		seed: {
			default: 0,
			description: "Seed for ambient pattern",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Environment"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const surfaceColor = uniforms.color.uniform;
		const brightness = uniforms.brightness.uniform.mul(.02);
		const keyColor = uniforms.keyColor.uniform;
		const keyInt = uniforms.keyIntensity.uniform.mul(.015);
		const fillColor = uniforms.fillColor.uniform;
		const fillInt = uniforms.fillIntensity.uniform.mul(.015);
		const fillOffset = uniforms.fillAngle.uniform.mul(.005);
		const backColor = uniforms.backColor.uniform;
		const backInt = uniforms.backIntensity.uniform.mul(.015);
		const centerX = uniforms.center.uniform.x;
		const centerY = uniforms.center.uniform.y.oneMinus();
		const lightTarget = uniforms.lightTarget.uniform.mul(.01);
		const curvature = uniforms.wallCurvature.uniform.mul(.01);
		const vignetteStrength = uniforms.vignette.uniform.mul(.025);
		const ambIntensity = uniforms.ambientIntensity.uniform.mul(.012);
		const seed = uniforms.seed.uniform;
		const aspect = viewportSize.x.div(viewportSize.y);
		const animTime = createAnimatedTime(params, uniforms.ambientSpeed);
		const keyDecay = mix(float(5), float(.8), uniforms.keySoftness.uniform.mul(.01));
		const fillDecay = mix(float(5), float(.8), uniforms.fillSoftness.uniform.mul(.01));
		const backDecay = mix(float(5), float(.8), uniforms.backSoftness.uniform.mul(.01));
		const hash01 = (x) => fract(sin(x.mul(12.9898)).mul(43758.5453));
		const ambLights = Array.from({ length: 4 }, (_, i) => {
			const s = seed.add(float(i * 73.1));
			const sizeVal = hash01(s.add(311)).mul(.12).add(.1);
			return {
				radiusX: hash01(s).mul(.5).add(.2),
				radiusY: hash01(s.add(37)).mul(.25).add(.1),
				freqX: hash01(s.add(91)).mul(.3).add(.1),
				freqY: hash01(s.add(143)).mul(.25).add(.08),
				phaseX: hash01(s.add(200)).mul(Math.PI * 2),
				phaseY: hash01(s.add(257)).mul(Math.PI * 2),
				sizeSq: sizeVal.mul(sizeVal)
			};
		});
		return Fn(() => {
			const uv$1 = screenUV;
			const transitionWidth = mix(float(.015), float(.15), curvature);
			const isFloor = smoothstep(centerY.sub(transitionWidth), centerY.add(transitionWidth), uv$1.y);
			const notFloor = float(1).sub(isFloor);
			const wallNormY = clamp(uv$1.y.div(centerY), 0, 1);
			const wallNormY2 = wallNormY.mul(wallNormY);
			const floorNormY = clamp(uv$1.y.sub(centerY).div(float(1).sub(centerY)), 0, 1);
			const uvXCentered = uv$1.x.sub(centerX).mul(aspect);
			const coneWidth = mix(float(.05), float(.45), wallNormY2);
			const wallFallY = float(1).sub(wallNormY).mul(1.2);
			const wallFallY2 = wallFallY.mul(wallFallY);
			const floorPoolY = centerY.add(float(1).sub(centerY).mul(.3));
			const floorDy = uv$1.y.sub(floorPoolY).mul(4);
			const floorDy2 = floorDy.mul(floorDy);
			const ambient = mix(mix(float(.2), float(.4), wallNormY), mix(float(.38), float(.18), floorNormY), isFloor).mul(brightness);
			const curveDist = uv$1.y.sub(centerY).div(transitionWidth.add(.001));
			const curveGlow = exp(curveDist.mul(curveDist).mul(-.5)).mul(curvature).mul(brightness).mul(.1);
			const neutralBase = ambient.add(curveGlow);
			const lightAccum = vec3(neutralBase, neutralBase, neutralBase).toVar();
			const wallBias = float(1).sub(lightTarget).mul(1.4).add(.3);
			const floorBias = lightTarget.mul(1.4).add(.3);
			const downSpot = (sourceX, decay) => {
				const beamX = mix(sourceX, centerX, wallNormY2);
				const wDx = uv$1.x.sub(beamX).mul(aspect).div(coneWidth);
				const wall = exp(wDx.mul(wDx).add(wallFallY2).mul(-1).mul(decay)).mul(notFloor).mul(wallBias);
				const poolX = mix(centerX, sourceX, float(.25));
				const fDx = uv$1.x.sub(poolX).mul(aspect).mul(1.4);
				const floor$1 = exp(fDx.mul(fDx).add(floorDy2).mul(-1).mul(decay.mul(1.5))).mul(isFloor).mul(floorBias);
				return wall.add(floor$1);
			};
			const keyPat = downSpot(centerX, keyDecay).mul(keyInt);
			lightAccum.assign(lightAccum.add(keyColor.rgb.mul(keyPat)));
			const leftPat = downSpot(centerX.sub(fillOffset), fillDecay).mul(fillInt);
			const rightPat = downSpot(centerX.add(fillOffset), fillDecay).mul(fillInt);
			lightAccum.assign(lightAccum.add(fillColor.rgb.mul(leftPat.add(rightPat))));
			const backBottomY = mix(float(1), centerY, float(.15));
			const bfDx = uvXCentered.mul(.5);
			const bfDy = uv$1.y.sub(backBottomY).mul(4);
			const backFloor = exp(bfDx.mul(bfDx).add(bfDy.mul(bfDy)).mul(-1).mul(backDecay)).mul(isFloor);
			const bwDx = uvXCentered.mul(.6);
			const bwDy = centerY.sub(uv$1.y).mul(3);
			const backWall = exp(bwDx.mul(bwDx).add(bwDy.mul(bwDy)).mul(-1).mul(backDecay.mul(.8))).mul(notFloor).mul(wallNormY2);
			const backPat = backFloor.add(backWall).mul(backInt);
			lightAccum.assign(lightAccum.add(backColor.rgb.mul(backPat)));
			const bnDx = uvXCentered.mul(.7);
			const bnDy = uv$1.y.sub(centerY).add(.02).mul(10);
			const bounce = exp(bnDx.mul(bnDx).add(bnDy.mul(bnDy)).mul(-2.5)).mul(keyInt.mul(.12)).mul(notFloor);
			lightAccum.assign(lightAccum.add(vec3(bounce, bounce, bounce)));
			const ambGlowScale = ambIntensity.mul(.25);
			for (let i = 0; i < 4; i++) {
				const l = ambLights[i];
				const lx = centerX.add(sin(animTime.mul(l.freqX).add(l.phaseX)).mul(l.radiusX));
				const ly = centerY.sub(.2).add(cos(animTime.mul(l.freqY).add(l.phaseY)).mul(l.radiusY));
				const dx = uv$1.x.sub(lx).mul(aspect);
				const dy = uv$1.y.sub(ly);
				const glow = exp(dx.mul(dx).add(dy.mul(dy)).div(l.sizeSq).mul(-.5)).mul(ambGlowScale);
				lightAccum.assign(lightAccum.add(vec3(glow, glow, glow)));
			}
			const vigX = uv$1.x.sub(.5).mul(aspect);
			const vigY = uv$1.y.sub(.5);
			const vigD2 = vigX.mul(vigX).add(vigY.mul(vigY));
			const vig = clamp(float(1).sub(vigD2.mul(vignetteStrength)), 0, 1);
			lightAccum.assign(clamp(lightAccum, 0, 3));
			return vec4(surfaceColor.rgb.mul(lightAccum).mul(vig).clamp(0, 1), float(1));
		})();
	}
};
var StudioBackground_default = componentDefinition;
export { componentDefinition as n, StudioBackground_default as t };
