// shaders/dist/core/SunBurst-BJRFFYhs.js
import { a as transformColor, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { atan, float, mix, pow, screenUV, sin, smoothstep, sqrt, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "SunBurst",
	category: "Textures",
	description: "Radial sunburst rays emanating from a center point",
	props: {
		color: {
			default: "#ffdd88",
			transform: transformColor,
			description: "Ray color",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		background: {
			default: "#000000",
			transform: transformColor,
			description: "Background color",
			ui: {
				type: "color",
				label: "Background",
				group: "Colors"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the sunburst",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		rayCount: {
			default: 12,
			description: "Number of rays",
			ui: {
				type: "range",
				min: 3,
				max: 64,
				step: 1,
				label: "Ray Count",
				group: "Effect"
			}
		},
		softness: {
			default: .3,
			description: "Softness of ray edges",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Softness",
				group: "Effect"
			}
		},
		radius: {
			default: .8,
			description: "How far the rays extend from the center",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1.2,
				step: .01,
				label: "Radius",
				group: "Effect"
			}
		},
		feather: {
			default: .5,
			description: "How gradually the rays fade at their outer edge",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .01,
				label: "Feather",
				group: "Effect"
			}
		},
		speed: {
			default: .2,
			description: "Rotation speed — positive values rotate clockwise",
			ui: {
				type: "range",
				min: -2,
				max: 2,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const cx = uniforms.center.uniform.x;
		const cy = uniforms.center.uniform.y.oneMinus();
		const dx = uv$1.x.mul(aspect).sub(cx.mul(aspect));
		const dy = uv$1.y.sub(cy);
		const dist = sqrt(dx.mul(dx).add(dy.mul(dy)));
		const angle = atan(dy, dx);
		const animTime = createAnimatedTime(params, uniforms.speed);
		const rayCount = uniforms.rayCount.uniform;
		const rayMask = pow(sin(angle.mul(rayCount).sub(animTime)).mul(.5).add(.5), float(1).div(uniforms.softness.uniform.add(float(.05))).mul(float(.3)));
		const radius = uniforms.radius.uniform;
		const feather = uniforms.feather.uniform;
		const falloffStart = radius.mul(float(1).sub(feather));
		const outerMask = float(1).sub(smoothstep(falloffStart, radius, dist));
		const alpha = rayMask.mul(outerMask);
		const col = uniforms.color.uniform;
		const bg = uniforms.background.uniform;
		return vec4(mix(bg.rgb, col.rgb, alpha), mix(bg.a, col.a, alpha));
	}
};
var SunBurst_default = componentDefinition;
export { componentDefinition as n, SunBurst_default as t };
