// shaders/dist/core/Swirl-KA9cEnLI.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { Fn, cos, screenUV, sin, smoothstep, vec2, vec4 } from "three/tsl";
const componentDefinition = {
	name: "Swirl",
	category: "Textures",
	description: "Flowing swirl pattern with multi-layered noise",
	props: {
		colorA: {
			default: "#1275d8",
			description: "Primary gradient color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#e19136",
			description: "Secondary gradient color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		speed: {
			default: 1,
			description: "Flow animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		detail: {
			default: 1,
			description: "Level of detail and intricacy in the swirl patterns",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .1,
				label: "Detail",
				group: "Effect"
			}
		},
		blend: {
			default: 50,
			description: "Skew color balance toward A (lower values) or B (higher values)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Blend",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const t = createAnimatedTime(params, uniforms.speed);
		const detail = uniforms.detail.uniform;
		const csMode = uniforms.colorSpace.uniform.value;
		return Fn(() => {
			const freq1 = detail;
			const d1 = vec2(screenUV.x.add(sin(screenUV.y.mul(freq1.mul(1.7)).add(t.mul(.8))).mul(.12).add(cos(screenUV.x.mul(freq1.mul(.9)).sub(t.mul(.5))).mul(.05))), screenUV.y.add(cos(screenUV.x.mul(freq1.mul(1.3)).sub(t.mul(.6))).mul(.12).add(sin(screenUV.y.mul(freq1.mul(1.1)).add(t.mul(.7))).mul(.05)))).toVar();
			const pattern1 = sin(d1.x.mul(freq1.mul(2.1)).add(d1.y.mul(freq1.mul(1.8))).add(t.mul(.4)));
			const freq2 = detail.mul(2.1);
			const d2 = vec2(d1.x.add(cos(d1.y.mul(freq2.mul(2.7)).sub(t.mul(.45))).mul(.07).add(sin(d1.x.mul(freq2.mul(1.9)).add(t.mul(.6))).mul(.04))), d1.y.add(sin(d1.x.mul(freq2.mul(2.3)).add(t.mul(.65))).mul(.07).add(cos(d1.y.mul(freq2.mul(1.6)).sub(t.mul(.4))).mul(.04)))).toVar();
			const pattern2 = cos(d2.x.mul(freq2.mul(1.4)).sub(d2.y.mul(freq2.mul(1.9))).add(t.mul(.35)));
			const freq3 = detail.mul(3.7);
			const d3 = vec2(d2.x.add(sin(d2.y.mul(freq3.mul(1.8)).add(t.mul(.85))).mul(.04).add(cos(d2.x.mul(freq3.mul(1.3)).sub(t.mul(.55))).mul(.025)).add(sin(d2.x.add(d2.y).mul(freq3.mul(.7)).add(t.mul(.9))).mul(.02))), d2.y.add(cos(d2.x.mul(freq3.mul(1.6)).sub(t.mul(.75))).mul(.04).add(sin(d2.y.mul(freq3.mul(1.1)).add(t.mul(.5))).mul(.025)).add(cos(d2.x.add(d2.y).mul(freq3.mul(.8)).sub(t.mul(.95))).mul(.02)))).toVar();
			const pattern3 = sin(d3.x.mul(freq3.mul(1.1)).add(d3.y.mul(freq3.mul(1.5))).sub(t.mul(.55)));
			const combinedPattern = pattern1.mul(.45).add(pattern2.mul(.35)).add(pattern3.mul(.2));
			const blendBias = uniforms.blend.uniform.sub(50).mul(.006);
			const blendFactor = smoothstep(.3, .7, combinedPattern.mul(.5).add(.5).add(blendBias));
			const colorMix = mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, blendFactor, csMode);
			const shimmer = sin(t.mul(2.5).add(combinedPattern.mul(8))).mul(.015).add(1);
			const finalColor = colorMix.mul(shimmer);
			return vec4(finalColor.rgb, finalColor.a);
		})();
	}
};
var Swirl_default = componentDefinition;
export { componentDefinition as n, Swirl_default as t };
