// shaders/dist/core/TiltShift-DtQ3dRL-.js
import { l as transformPosition, r as transformAngle } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { i as createVariableGaussianBlurCompute } from "./computeBlur-CprKe0xa.js";
import { Fn, If, abs, clamp, convertToTexture, cos, dot, float, instanceIndex, int, mix, pow, radians, screenUV, sin, smoothstep, texture, textureLoad, textureStore, uint, uniform, uvec2, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "TiltShift",
	category: "Blurs",
	description: "Selective focus blur mimicking tilt-shift photography",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 50,
			description: "Maximum blur intensity at edges",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Effect"
			}
		},
		width: {
			default: .3,
			description: "Width of the sharp focus area",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Width",
				group: "Effect"
			}
		},
		falloff: {
			default: .3,
			description: "Distance over which blur transitions to full strength",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Falloff",
				group: "Effect"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Rotation angle of the focus line (in degrees)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the focus line",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		}
	},
	computeNode: ({ childNode, onCleanup, dimensions, renderer, uniforms, onResize }) => {
		if (!childNode) return null;
		if (uniforms.angle.mapComputeInfo || uniforms.width.mapComputeInfo || uniforms.falloff.mapComputeInfo) return null;
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const dpr = renderer?.getPixelRatio?.() ?? 1;
		const texWidth = Math.round(dimensions.width * dpr);
		const texHeight = Math.round(dimensions.height * dpr);
		const blur = createVariableGaussianBlurCompute(childTexture, texWidth, texHeight, onCleanup, 14);
		const { computeNodes: blurNodes, outputTexture, blurMapWriteNode, computeWidth, computeHeight } = blur;
		const angleU = uniform(0);
		const centerXU = uniform(.5);
		const centerYU = uniform(.5);
		const focusWidthU = uniform(.15);
		const falloffU = uniform(.3);
		const aspectU = uniform(1);
		const inputWidthU = uniform(texWidth);
		const inputHeightU = uniform(texHeight);
		const cw = uint(computeWidth);
		const ch = uint(computeHeight);
		function readScalar(entry) {
			return entry.getCpuValue?.() ?? (typeof entry.uniform.value === "number" ? entry.uniform.value : null);
		}
		function updateGeometryUniforms() {
			const angleVal = readScalar(uniforms.angle);
			if (angleVal !== null) angleU.value = angleVal;
			const centerXVal = uniforms.center.uniform.value?.x ?? uniforms.center.uniform.x?.value;
			if (typeof centerXVal === "number") centerXU.value = centerXVal;
			const centerYVal = uniforms.center.uniform.value?.y ?? uniforms.center.uniform.y?.value;
			if (typeof centerYVal === "number") centerYU.value = centerYVal;
			const widthVal = readScalar(uniforms.width);
			if (widthVal !== null) focusWidthU.value = widthVal * .5;
			const falloffVal = readScalar(uniforms.falloff);
			if (falloffVal !== null) falloffU.value = falloffVal;
			aspectU.value = dimensions.width / dimensions.height;
		}
		const intensityMapInfo = uniforms.intensity.mapComputeInfo;
		if (intensityMapInfo) {
			const { sourceTexture, channel, inputMin, inputMax, outputMin, outputMax, curve } = intensityMapInfo;
			const fillBlurMap$1 = Fn(() => {
				const idx = instanceIndex;
				const cx = idx.mod(cw);
				const cy = idx.div(cw);
				If(cy.lessThan(ch), () => {
					const u = cx.toFloat().add(float(.5)).div(float(computeWidth));
					const v = cy.toFloat().add(float(.5)).div(float(computeHeight));
					const angleRad = angleU.mul(float(Math.PI / 180));
					const perpVector = vec2(angleRad.sin().negate(), angleRad.cos());
					const centerPos = vec2(centerXU, float(1).sub(centerYU));
					const centeredUV = vec2(u, v).sub(centerPos);
					const distFromLine = abs(dot(vec2(centeredUV.x.mul(aspectU), centeredUV.y), perpVector));
					const blurAmount = smoothstep(focusWidthU, focusWidthU.add(falloffU), distFromLine);
					const mapSample = textureLoad(sourceTexture, uvec2(u.mul(inputWidthU).round().toInt().clamp(int(0), int(inputWidthU.sub(1))).toUint(), v.mul(inputHeightU).round().toInt().clamp(int(0), int(inputHeightU.sub(1))).toUint()), int(0));
					let raw;
					if (channel === "alpha") raw = mapSample.a;
					else if (channel === "alphaInverted") raw = mapSample.a.oneMinus();
					else if (channel === "luminanceInverted") raw = dot(mapSample.rgb, vec3(.2126, .7152, .0722)).oneMinus();
					else raw = dot(mapSample.rgb, vec3(.2126, .7152, .0722));
					const inputRange = inputMax.sub(inputMin).max(float(1e-4));
					const maxRadius = mix(outputMin, outputMax, pow(clamp(raw.sub(inputMin).div(inputRange), float(0), float(1)), pow(float(2), curve.negate().mul(float(2))))).mul(float(.36));
					const radius = blurAmount.mul(maxRadius);
					textureStore(blurMapWriteNode, uvec2(cx, cy), vec4(radius, float(0), float(0), float(1))).toWriteOnly();
				});
			})().compute(computeWidth * computeHeight, [256]);
			onResize(({ width, height }) => {
				const newW = Math.round(width * dpr);
				const newH = Math.round(height * dpr);
				inputWidthU.value = newW;
				inputHeightU.value = newH;
				blur.setInputDimensions(newW, newH);
			});
			return {
				outputs: {
					childTexture,
					blurredTexture: outputTexture
				},
				getComputeNodes: () => {
					updateGeometryUniforms();
					return [fillBlurMap$1, ...blurNodes];
				}
			};
		}
		const maxRadiusU = uniform(18);
		const fillBlurMap = Fn(() => {
			const idx = instanceIndex;
			const cx = idx.mod(cw);
			const cy = idx.div(cw);
			If(cy.lessThan(ch), () => {
				const u = cx.toFloat().add(float(.5)).div(float(computeWidth));
				const v = cy.toFloat().add(float(.5)).div(float(computeHeight));
				const angleRad = angleU.mul(float(Math.PI / 180));
				const perpVector = vec2(angleRad.sin().negate(), angleRad.cos());
				const centerPos = vec2(centerXU, float(1).sub(centerYU));
				const centeredUV = vec2(u, v).sub(centerPos);
				const distFromLine = abs(dot(vec2(centeredUV.x.mul(aspectU), centeredUV.y), perpVector));
				const radius = smoothstep(focusWidthU, focusWidthU.add(falloffU), distFromLine).mul(maxRadiusU);
				textureStore(blurMapWriteNode, uvec2(cx, cy), vec4(radius, float(0), float(0), float(1))).toWriteOnly();
			});
		})().compute(computeWidth * computeHeight, [256]);
		onResize(({ width, height }) => {
			const newW = Math.round(width * dpr);
			const newH = Math.round(height * dpr);
			inputWidthU.value = newW;
			inputHeightU.value = newH;
			blur.setInputDimensions(newW, newH);
		});
		return {
			outputs: {
				childTexture,
				blurredTexture: outputTexture
			},
			getComputeNodes: () => {
				updateGeometryUniforms();
				const intensityVal = readScalar(uniforms.intensity);
				if (intensityVal !== null) maxRadiusU.value = intensityVal * .36;
				return [fillBlurMap, ...blurNodes];
			}
		};
	},
	fragmentNode: ({ uniforms, childNode, onCleanup, computeOutputs }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		if (computeOutputs?.blurredTexture) {
			const sharp = computeOutputs.childTexture.sample(screenUV);
			const blurred = texture(computeOutputs.blurredTexture).sample(screenUV);
			const aspect$1 = viewportSize.x.div(viewportSize.y);
			const angleRad$1 = radians(uniforms.angle.uniform);
			const perpVector$1 = vec2(angleRad$1.sin().negate(), angleRad$1.cos());
			const centerPosition$1 = vec2(uniforms.center.uniform[0], uniforms.center.uniform[1].oneMinus());
			const centeredUV$1 = screenUV.sub(centerPosition$1);
			const distFromLine$1 = abs(dot(vec2(centeredUV$1.x.mul(aspect$1), centeredUV$1.y), perpVector$1));
			const focusWidth$1 = uniforms.width.uniform.mul(.5);
			const blurAmount$1 = smoothstep(focusWidth$1, focusWidth$1.add(uniforms.falloff.uniform), distFromLine$1);
			const propsPin = uniforms.intensity.uniform.mul(float(0)).add(uniforms.width.uniform.mul(float(0))).add(uniforms.falloff.uniform.mul(float(0))).add(uniforms.angle.uniform.mul(float(0)));
			const mixed = mix(sharp, blurred, blurAmount$1);
			return unpremultiplyAlpha(vec4(mixed.rgb.add(propsPin), mixed.a));
		}
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const aspect = viewportSize.x.div(viewportSize.y);
		const angleRad = radians(uniforms.angle.uniform);
		const cosAngle = cos(angleRad);
		const perpVector = vec2(sin(angleRad).negate(), cosAngle);
		const centerPosition = vec2(uniforms.center.uniform[0], uniforms.center.uniform[1].oneMinus());
		const centeredUV = screenUV.sub(centerPosition);
		const distFromLine = abs(dot(vec2(centeredUV.x.mul(aspect), centeredUV.y), perpVector));
		const focusWidth = uniforms.width.uniform.mul(.5);
		const falloff = uniforms.falloff.uniform;
		const blurAmount = smoothstep(focusWidth, focusWidth.add(falloff), distFromLine);
		const scaledIntensity = uniforms.intensity.uniform.mul(.06);
		const blurRadius = blurAmount.mul(scaledIntensity);
		const weights = [
			.056,
			.135,
			.278,
			.487,
			.726,
			.923,
			1,
			.923,
			.726,
			.487,
			.278,
			.135,
			.056
		];
		const weightSum = 6.21;
		const horizontalTexture = convertToTexture(Fn(() => {
			const uv$1 = screenUV;
			const pixelSize = vec2(1).div(viewportSize);
			const total = vec4(0).toVar();
			for (let x = -6; x <= 6; x++) {
				const weight = float(weights[x + 6] / weightSum);
				const offset = vec2(float(x), float(0)).mul(blurRadius).mul(pixelSize);
				const sampleCoord = uv$1.add(offset);
				const sample = childTexture.sample(sampleCoord).mul(weight);
				total.assign(total.add(sample));
			}
			return total;
		})());
		onCleanup(() => {
			if (horizontalTexture?.renderTarget?.dispose) horizontalTexture.renderTarget.dispose();
		});
		return Fn(() => {
			const uv$1 = screenUV;
			const pixelSize = vec2(1).div(viewportSize);
			const total = vec4(0).toVar();
			for (let y = -6; y <= 6; y++) {
				const weight = float(weights[y + 6] / weightSum);
				const offset = vec2(float(0), float(y)).mul(blurRadius).mul(pixelSize);
				const sampleCoord = uv$1.add(offset);
				const sample = horizontalTexture.sample(sampleCoord).mul(weight);
				total.assign(total.add(sample));
			}
			const blurred = total;
			return unpremultiplyAlpha(mix(childTexture.sample(uv$1), blurred, blurAmount));
		})();
	}
};
var TiltShift_default = componentDefinition;
export { componentDefinition as n, TiltShift_default as t };
