// shaders/dist/core/Tint-W_EvzN1-.js
import { a as transformColor } from "./transformations-CJcUeZIC.js";
import { dot, max, mix, vec3, vec4 } from "three/tsl";
const componentDefinition = {
	name: "Tint",
	category: "Adjustments",
	description: "Apply a color tint to the image",
	requiresChild: true,
	props: {
		color: {
			default: "#ff8800",
			transform: transformColor,
			description: "Tint color",
			ui: {
				type: "color",
				label: "Tint Color",
				group: "Colors"
			}
		},
		amount: {
			default: .5,
			description: "Tint amount (0 = no tint, 1 = full tint)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Amount",
				group: "Effect"
			}
		},
		preserveLuminosity: {
			default: true,
			compileTime: true,
			description: "Preserve original brightness",
			transform: (v) => v ? 1 : 0,
			ui: {
				type: "checkbox",
				label: "Preserve Luminosity",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Tint shader.");
			return vec4(0);
		}
		const tintColor = uniforms.color.uniform;
		const amount = uniforms.amount.uniform;
		const rgb = vec3(childNode.r, childNode.g, childNode.b);
		const tinted = mix(rgb, tintColor.rgb, amount);
		const preserveLum = uniforms.preserveLuminosity.uniform.value === 1;
		let result;
		if (preserveLum) {
			const luminanceWeights = vec3(.299, .587, .114);
			const originalLum = dot(rgb, luminanceWeights);
			const tintedLum = dot(tinted, luminanceWeights);
			result = tinted.mul(originalLum.div(max(tintedLum, 1e-4)));
		} else result = tinted;
		return vec4(result, childNode.a);
	}
};
var Tint_default = componentDefinition;
export { componentDefinition as n, Tint_default as t };
