// shaders/dist/core/Tritone-B3hI2nAi.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { dot, smoothstep, vec3, vec4 } from "three/tsl";
const componentDefinition = {
	name: "Tritone",
	category: "Adjustments",
	description: "Map colors to three tones: shadows, midtones, highlights",
	requiresChild: true,
	props: {
		colorA: {
			default: "#ce1bea",
			transform: transformColor,
			description: "First color (used for shadows/darkest areas)",
			ui: {
				type: "color",
				label: "Color A (Shadows)",
				group: "Colors"
			}
		},
		colorB: {
			default: "#2fff00",
			transform: transformColor,
			description: "Second color (used for midtones)",
			ui: {
				type: "color",
				label: "Color B (Midtones)",
				group: "Colors"
			}
		},
		colorC: {
			default: "#ffff00",
			transform: transformColor,
			description: "Third color (used for highlights/brightest areas)",
			ui: {
				type: "color",
				label: "Color C (Highlights)",
				group: "Colors"
			}
		},
		blendMid: {
			default: .5,
			description: "Midpoint position between the three colors",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .1,
				label: "Midpoint",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const luminance = dot(vec3(childNode.r, childNode.g, childNode.b), vec3(.299, .587, .114));
		const shadowToMid = smoothstep(uniforms.blendMid.uniform.sub(.25), uniforms.blendMid.uniform, luminance);
		const csMode = uniforms.colorSpace.uniform.value;
		const lowerBlend = mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, shadowToMid, csMode);
		const midToHighlight = smoothstep(uniforms.blendMid.uniform, uniforms.blendMid.uniform.add(.25), luminance);
		return vec4(mixColors(lowerBlend, mixColors(uniforms.colorB.uniform, uniforms.colorC.uniform, midToHighlight, csMode), smoothstep(uniforms.blendMid.uniform.sub(.1), uniforms.blendMid.uniform.add(.1), luminance), csMode).rgb, childNode.a);
	}
};
var Tritone_default = componentDefinition;
export { componentDefinition as n, Tritone_default as t };
