// shaders/dist/core/Truchet-s9PmowCP.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { Fn, abs, dot, float, floor, fract, fwidth, length, min, screenUV, sin, smoothstep, step, vec2, viewportSize } from "three/tsl";
var truchetHash = Fn(([p]) => {
	return fract(sin(dot(p, vec2(127.1, 311.7))).mul(float(43758.5453)));
});
const componentDefinition = {
	name: "Truchet",
	category: "Textures",
	description: "Quarter-circle arc tiles that connect to form organic, maze-like flowing curves",
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "Background color between the arcs",
			ui: {
				type: "color",
				label: "Color A",
				group: "Colors"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Arc line color",
			ui: {
				type: "color",
				label: "Color B",
				group: "Colors"
			}
		},
		cells: {
			default: 10,
			description: "Number of tiles across the shortest canvas edge",
			ui: {
				type: ["range", "map"],
				min: 2,
				max: 40,
				step: 1,
				label: "Cells",
				group: "Effect"
			}
		},
		thickness: {
			default: 2,
			description: "Thickness of the arc lines",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 20,
				step: .1,
				label: "Thickness",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Random seed — changes which tiles flip, producing a different maze pattern",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const scaledUV = vec2(uv$1.x.mul(aspect), uv$1.y).mul(uniforms.cells.uniform);
		const cell = floor(scaledUV);
		const localUV = fract(scaledUV);
		const flip = truchetHash(cell.add(uniforms.seed.uniform)).greaterThan(float(.5));
		const c1x = flip.select(float(1), float(0));
		const c1y = flip.select(float(0), float(0));
		const c2x = flip.select(float(0), float(1));
		const c2y = flip.select(float(1), float(1));
		const arcDist = min(abs(length(localUV.sub(vec2(c1x, c1y))).sub(float(.5))), abs(length(localUV.sub(vec2(c2x, c2y))).sub(float(.5))));
		const lineWidth = uniforms.thickness.uniform.mul(float(.005));
		const pixel = fwidth(scaledUV.x).mul(float(.5));
		const lineMask = smoothstep(lineWidth.add(pixel), lineWidth, arcDist).mul(step(float(1e-4), lineWidth));
		return mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, lineMask, uniforms.colorSpace.uniform.value);
	}
};
var Truchet_default = componentDefinition;
export { componentDefinition as n, Truchet_default as t };
