// shaders/dist/core/Twirl-UDDXGkFl.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { convertToTexture, cos, length, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Twirl",
	category: "Distortions",
	description: "Rotate and twist content around a center point",
	requiresRTT: true,
	requiresChild: true,
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the twirl effect",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		intensity: {
			default: 1,
			description: "The strength of the twirl effect",
			ui: {
				type: ["range", "map"],
				min: -5,
				max: 5,
				step: .1,
				label: "Intensity",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Twirl shader.");
			return vec4(0);
		}
		const texture$1 = convertToTexture(childNode);
		onCleanup(() => {
			if (texture$1?.renderTarget?.dispose) texture$1.renderTarget.dispose();
		});
		const uvNode = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const centerPos = vec2(uniforms.center.uniform.x, uniforms.center.uniform.y.oneMinus());
		const delta = uvNode.sub(centerPos);
		const aspectCorrectedDelta = vec2(delta.x.mul(aspect), delta.y);
		const angle = uniforms.intensity.uniform.mul(length(aspectCorrectedDelta));
		const cosAngle = cos(angle);
		const sinAngle = sin(angle);
		const rotatedX = cosAngle.mul(aspectCorrectedDelta.x).sub(sinAngle.mul(aspectCorrectedDelta.y));
		const rotatedY = sinAngle.mul(aspectCorrectedDelta.x).add(cosAngle.mul(aspectCorrectedDelta.y));
		const twistedUV = vec2(rotatedX.div(aspect).add(centerPos.x), rotatedY.add(centerPos.y));
		return unpremultiplyAlpha(applyEdgeHandling(twistedUV, texture$1.sample(twistedUV), texture$1, uniforms.edges.uniform.value));
	}
};
var Twirl_default = componentDefinition;
export { componentDefinition as n, Twirl_default as t };
