// shaders/dist/core/VHS-DW1H7Wuy.js
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { clamp, convertToTexture, cos, dot, float, floor, fract, mix, mod, screenUV, sin, smoothstep, time, vec2, vec3, vec4 } from "three/tsl";
var SMEAR_SAMPLES = 6;
var FIELD_LINES = 487;
const componentDefinition = {
	name: "VHS",
	category: "Stylize",
	description: "Analog VHS tape with intermittent tape damage, chroma bleed, and per-scanline noise",
	requiresRTT: true,
	requiresChild: true,
	props: {
		wobble: {
			default: 1,
			description: "Overall amount of tape damage — waves, creases, and head-switching noise. Bursts on and off organically over time.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 5,
				step: .01,
				label: "Wobble",
				group: "Effect"
			}
		},
		scanlineNoise: {
			default: .6,
			description: "Per-scanline fine chroma/luma jitter. Adds the classic horizontal-streak detail.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Scanline Noise",
				group: "Effect"
			}
		},
		smear: {
			default: .2,
			description: "Horizontal chroma smear (color bleed) amount. Positive trails colour to the right (classic VHS), negative trails it to the left.",
			ui: {
				type: ["range", "map"],
				min: -2,
				max: 2,
				step: .01,
				label: "Smear",
				group: "Effect"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed of the tape effects.",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .1,
				label: "Speed",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const wobble = uniforms.wobble.uniform;
		const scanlineNoiseAmt = uniforms.scanlineNoise.uniform;
		const smear = uniforms.smear.uniform;
		const speed = uniforms.speed.uniform;
		const t = time.mul(speed);
		const hash2D = (p) => fract(sin(dot(p, vec2(12.9898, 78.233))).mul(43758.5453));
		const smoothNoise2D = (p) => {
			const i = floor(p);
			const f = fract(p);
			const u = f.mul(f).mul(f.mul(-2).add(3));
			const a = hash2D(i);
			const b = hash2D(i.add(vec2(1, 0)));
			const c = hash2D(i.add(vec2(0, 1)));
			const d = hash2D(i.add(vec2(1, 1)));
			return mix(mix(a, b, u.x), mix(c, d, u.x), u.y);
		};
		const burst = smoothstep(float(.45), float(.8), smoothNoise2D(vec2(t.mul(.4), float(0))));
		const rowInt = floor(screenUV.y.mul(FIELD_LINES));
		const scanlineHash = (seedOffset) => fract(sin(rowInt.mul(12.9898).add(t.add(seedOffset).mul(78.233))).mul(43758.5453)).sub(.5).mul(2);
		const fineGate = float(.25).add(burst.mul(.75));
		const chromaRowOffset = scanlineHash(0).mul(scanlineNoiseAmt).mul(.008).mul(fineGate);
		const lumaRowOffset = scanlineHash(69.42).mul(scanlineNoiseAmt).mul(.004).mul(fineGate);
		const wave1 = smoothNoise2D(vec2(screenUV.y.mul(3), t.mul(.8))).sub(.5);
		const wave2 = smoothNoise2D(vec2(screenUV.y.mul(30), t.mul(6))).sub(.5);
		const tapeWave = wave1.mul(.008).add(wave2.mul(.002)).mul(wobble);
		const creasePhase = smoothstep(float(.92), float(.99), sin(screenUV.y.mul(8).sub(t.mul(3.77))));
		const creaseNoise = smoothstep(float(.3), float(1), smoothNoise2D(vec2(screenUV.y.mul(4.77), t)));
		const creaseShift = creasePhase.mul(creaseNoise).mul(wobble).mul(-.018);
		const switchPhase = smoothstep(float(.06), float(0), screenUV.y);
		const switchNoise = smoothNoise2D(vec2(screenUV.y.mul(60), t.mul(14))).sub(.5);
		const switchX = switchPhase.mul(switchNoise).mul(wobble).mul(.09);
		const switchY = switchPhase.mul(wobble).mul(burst).mul(.02);
		const globalX = tapeWave.add(creaseShift).add(switchX);
		const lumaUV = vec2(screenUV.x.add(globalX).add(lumaRowOffset), screenUV.y.add(switchY));
		const chromaUV = vec2(screenUV.x.add(globalX).add(chromaRowOffset), screenUV.y.add(switchY));
		const lumaSample = childTexture.sample(lumaUV);
		const sharpY = lumaSample.rgb.dot(vec3(.299, .587, .114));
		const smearScale = smear.mul(.0075);
		let accumI = float(0);
		let accumQ = float(0);
		for (let i = 0; i < SMEAR_SAMPLES; i++) {
			const offsetX = float(-i).mul(smearScale);
			const sampled = childTexture.sample(vec2(chromaUV.x.add(offsetX), chromaUV.y));
			const iComp = sampled.rgb.dot(vec3(.596, -.274, -.322));
			const qComp = sampled.rgb.dot(vec3(.211, -.523, .312));
			const w = i / (SMEAR_SAMPLES - 1) * (2 / SMEAR_SAMPLES);
			accumI = accumI.add(iComp.mul(w));
			accumQ = accumQ.add(qComp.mul(w));
		}
		const finalYiq = vec3(sharpY, accumI, accumQ);
		const finalRgb = vec3(finalYiq.x.add(finalYiq.y.mul(.956)).add(finalYiq.z.mul(.621)), finalYiq.x.sub(finalYiq.y.mul(.272)).sub(finalYiq.z.mul(.647)), finalYiq.x.sub(finalYiq.y.mul(1.106)).add(finalYiq.z.mul(1.703)));
		const acBeat = float(1).add(cos(mod(t, 6.2831853).mul(2).add(screenUV.y.mul(.5))).mul(.015).mul(wobble));
		return unpremultiplyAlpha(vec4(clamp(finalRgb.mul(acBeat), vec3(0), vec3(1)), lumaSample.a));
	}
};
var VHS_default = componentDefinition;
export { componentDefinition as n, VHS_default as t };
