// shaders/dist/core/Vibrance-DUj7hwzE.js
import { add, max, mix, vec3, vec4 } from "three/tsl";
const componentDefinition = {
	name: "Vibrance",
	category: "Adjustments",
	description: "Selective saturation adjustment protecting skin tones",
	requiresChild: true,
	props: { intensity: {
		default: 0,
		description: "The intensity of the vibrance effect",
		ui: {
			type: ["range", "map"],
			min: -2,
			max: 2,
			step: .1,
			label: "Intensity",
			group: "Appearance"
		}
	} },
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Vibrance shader.");
			return vec4(0);
		}
		const color = childNode.rgb;
		const adjustment = uniforms.intensity.uniform;
		const average = add(color.r, add(color.g, color.b)).div(3);
		const mx = max(color.r, max(color.g, color.b));
		const amt = mx.sub(average).mul(adjustment).mul(-3);
		return vec4(mix(color, vec3(mx, mx, mx), amt), childNode.a);
	}
};
var Vibrance_default = componentDefinition;
export { componentDefinition as n, Vibrance_default as t };
