// shaders/dist/core/VideoTexture-ClutlOqj.js
import { t as needsVideoCanvasWorkaround } from "./browser-NUM-x2tw.js";
import { CanvasTexture, SRGBColorSpace, VideoTexture } from "three/webgpu";
import { float, max, min, or, screenUV, select, texture, uniform, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "VideoTexture",
	category: "Textures",
	description: "Display a video with customizable playback and object-fit modes",
	props: {
		url: {
			default: "https://shaders.com/sample.mp4",
			description: "Upload a video or provide a URL",
			ui: {
				type: "video-upload",
				label: "Video",
				group: "Media"
			}
		},
		objectFit: {
			default: "cover",
			description: "How the video should be sized within the viewport",
			transform: (value) => {
				return {
					"cover": 0,
					"contain": 1,
					"fill": 2,
					"scale-down": 3,
					"none": 4
				}[value] ?? 0;
			},
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Cover",
						value: "cover"
					},
					{
						label: "Contain",
						value: "contain"
					},
					{
						label: "Fill",
						value: "fill"
					},
					{
						label: "Scale Down",
						value: "scale-down"
					},
					{
						label: "None",
						value: "none"
					}
				],
				label: "Object Fit",
				group: "Appearance"
			}
		},
		loop: {
			default: true,
			description: "Loop the video playback",
			ui: {
				type: "checkbox",
				label: "Loop",
				group: "Playback"
			}
		}
	},
	fragmentNode: ({ uniforms, onCleanup, onBeforeRender }) => {
		const baseUV = screenUV;
		let videoElement = null;
		let currentUrl = "";
		let isLoading = false;
		let isDisposed = false;
		const videoAspectUniform = uniform(1);
		const videoWidthUniform = uniform(640);
		const videoHeightUniform = uniform(480);
		let canvas = null;
		let ctx = null;
		let activeTexture;
		if (needsVideoCanvasWorkaround()) {
			canvas = document.createElement("canvas");
			canvas.width = 640;
			canvas.height = 480;
			ctx = canvas.getContext("2d");
			activeTexture = new CanvasTexture(canvas);
		} else {
			const placeholderVideo = document.createElement("video");
			placeholderVideo.playsInline = true;
			placeholderVideo.muted = true;
			activeTexture = new VideoTexture(placeholderVideo);
		}
		activeTexture.colorSpace = SRGBColorSpace;
		const textureNode = texture(activeTexture);
		const loadVideo = async (url) => {
			if (!url || url.trim() === "" || isLoading || isDisposed) return;
			isLoading = true;
			try {
				const video = document.createElement("video");
				video.src = url;
				video.playsInline = true;
				video.crossOrigin = "anonymous";
				video.muted = true;
				video.loop = uniforms.loop?.uniform?.value ?? true;
				await new Promise((resolve, reject) => {
					video.onloadedmetadata = () => {
						resolve();
					};
					video.onerror = () => {
						reject(/* @__PURE__ */ new Error("Failed to load video"));
					};
					setTimeout(() => {
						reject(/* @__PURE__ */ new Error("Video load timeout"));
					}, 1e4);
				});
				if (isDisposed) {
					video.src = "";
					return;
				}
				const width = video.videoWidth || 640;
				const height = video.videoHeight || 480;
				videoAspectUniform.value = width / height;
				videoWidthUniform.value = width;
				videoHeightUniform.value = height;
				if (videoElement) {
					videoElement.pause();
					videoElement.src = "";
					videoElement = null;
				}
				videoElement = video;
				currentUrl = url;
				try {
					await video.play();
				} catch (error) {
					console.warn("[VideoTexture] Autoplay failed (browser policy):", error);
				}
				activeTexture.dispose();
				if (needsVideoCanvasWorkaround()) {
					canvas.width = width;
					canvas.height = height;
					activeTexture = new CanvasTexture(canvas);
				} else activeTexture = new VideoTexture(video);
				activeTexture.colorSpace = SRGBColorSpace;
				textureNode.value = activeTexture;
				isLoading = false;
			} catch (error) {
				console.error("[VideoTexture] Failed to load video:", error);
				isLoading = false;
			}
		};
		setTimeout(() => {
			const initialUrl = uniforms?.url?.uniform?.value ?? "";
			if (initialUrl && initialUrl.trim() !== "") loadVideo(initialUrl);
		}, 0);
		onBeforeRender(() => {
			const newUrl = uniforms?.url?.uniform?.value ?? "";
			if (newUrl !== currentUrl && !isLoading) loadVideo(newUrl);
			if (videoElement) {
				const loopValue = uniforms.loop?.uniform?.value ?? true;
				if (videoElement.loop !== loopValue) videoElement.loop = loopValue;
				if (needsVideoCanvasWorkaround() && !videoElement.paused && videoElement.readyState >= 2) {
					ctx.drawImage(videoElement, 0, 0, canvas.width, canvas.height);
					activeTexture.needsUpdate = true;
				}
			}
		});
		onCleanup(() => {
			isDisposed = true;
			if (videoElement) {
				videoElement.pause();
				videoElement.src = "";
				videoElement = null;
			}
			activeTexture.dispose();
			currentUrl = "";
		});
		const uv$1 = baseUV;
		const viewportAspect = viewportSize.x.div(viewportSize.y);
		const objectFitMode = uniforms.objectFit.uniform.value;
		let uvScale;
		switch (objectFitMode) {
			case 1: {
				const containScale = min(viewportAspect.div(videoAspectUniform), float(1));
				uvScale = vec2(videoAspectUniform.div(viewportAspect).mul(containScale), containScale);
				break;
			}
			case 2:
				uvScale = vec2(1, 1);
				break;
			case 3: {
				const scaleDownScale = min(min(viewportAspect.div(videoAspectUniform), float(1)), min(viewportSize.x.div(videoWidthUniform), viewportSize.y.div(videoHeightUniform)));
				uvScale = vec2(videoAspectUniform.div(viewportAspect).mul(scaleDownScale), scaleDownScale);
				break;
			}
			case 4: {
				const noneScale = min(viewportSize.x.div(videoWidthUniform), viewportSize.y.div(videoHeightUniform));
				uvScale = vec2(videoAspectUniform.div(viewportAspect).mul(noneScale), noneScale);
				break;
			}
			case 0:
			default: {
				const coverScale = max(viewportAspect.div(videoAspectUniform), float(1));
				uvScale = vec2(videoAspectUniform.div(viewportAspect).mul(coverScale), coverScale);
				break;
			}
		}
		const adjustedUV = uv$1.sub(vec2(.5)).div(uvScale).add(vec2(.5));
		const finalUV = vec2(adjustedUV.x, float(1).sub(adjustedUV.y));
		const sampledColor = textureNode.sample(finalUV);
		const isOutOfBounds = or(or(finalUV.x.lessThan(0), finalUV.x.greaterThan(1)), or(finalUV.y.lessThan(0), finalUV.y.greaterThan(1)));
		return vec4(sampledColor.rgb, select(isOutOfBounds, float(0), sampledColor.a));
	}
};
var VideoTexture_default = componentDefinition;
export { componentDefinition as n, VideoTexture_default as t };
