// shaders/dist/core/Vignette-F9yxi-UM.js
import { a as transformColor, l as transformPosition } from "./transformations-CJcUeZIC.js";
import { length, mix, screenUV, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Vignette",
	category: "Stylize",
	description: "Darkens or tints the edges of the frame, drawing attention toward the center",
	requiresChild: true,
	props: {
		color: {
			default: "#000000",
			transform: transformColor,
			description: "Color of the vignette at the edges",
			ui: {
				type: "color",
				label: "Color",
				group: "Colors"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center of the clear area where the vignette begins",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		},
		radius: {
			default: .5,
			description: "Distance from center where the vignette begins to fade in",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1.5,
				step: .01,
				label: "Radius",
				group: "Effect"
			}
		},
		falloff: {
			default: .5,
			description: "Width of the transition zone from clear to full vignette",
			ui: {
				type: ["range", "map"],
				min: .01,
				max: 1.5,
				step: .01,
				label: "Falloff",
				group: "Effect"
			}
		},
		intensity: {
			default: 1,
			description: "Strength of the vignette effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Intensity",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Vignette shader.");
			return vec4(0);
		}
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectUV = vec2(uv$1.x.mul(aspect), uv$1.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus());
		const dist = length(aspectUV.sub(centerPos));
		const vignetteMask = smoothstep(uniforms.radius.uniform, uniforms.radius.uniform.add(uniforms.falloff.uniform), dist);
		return vec4(mix(childNode.rgb, uniforms.color.uniform.rgb, vignetteMask.mul(uniforms.intensity.uniform)), childNode.a);
	}
};
var Vignette_default = componentDefinition;
export { componentDefinition as n, Vignette_default as t };
