// shaders/dist/core/Voronoi-D8HHP_WR.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { Fn, cos, dot, float, floor, fract, length, mix, pow, screenUV, sin, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
var voronoiHash = Fn(([p]) => {
	return fract(sin(vec2(dot(p, vec2(127.1, 311.7)), dot(p, vec2(269.5, 183.3)))).mul(float(43758.5453)));
});
const componentDefinition = {
	name: "Voronoi",
	category: "Textures",
	description: "Cellular pattern where each pixel is colored by its distance to the nearest of many scattered points",
	props: {
		colorA: {
			default: "#3186cf",
			transform: transformColor,
			description: "Color near each cell's center point",
			ui: {
				type: "color",
				label: "Color A (Center)",
				group: "Colors"
			}
		},
		colorB: {
			default: "#fc02dd",
			transform: transformColor,
			description: "Color at cell boundaries, far from any center point",
			ui: {
				type: "color",
				label: "Color B (Edge)",
				group: "Colors"
			}
		},
		colorBorder: {
			default: "#000000",
			transform: transformColor,
			description: "Color of the cell boundary lines",
			ui: {
				type: "color",
				label: "Border Color",
				group: "Colors"
			}
		},
		scale: {
			default: 6,
			description: "Number of cells across the canvas",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 20,
				step: .5,
				label: "Scale",
				group: "Effect"
			}
		},
		speed: {
			default: .5,
			description: "Animation speed — how fast the cell points drift",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		seed: {
			default: 0,
			description: "Random seed — shifts the cell pattern without changing the overall structure",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		edgeIntensity: {
			default: .5,
			description: "Controls how much of the cell interior is filled by the edge color. Low = center color dominates with a sharp boundary. High = edge color spreads further into the cell.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .01,
				label: "Edge Intensity",
				group: "Effect"
			}
		},
		edgeSoftness: {
			default: .05,
			description: "Width of the cell boundary lines.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .4,
				step: .005,
				label: "Edge Softness",
				group: "Effect"
			}
		},
		colorSpace: {
			default: "oklch",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const colorA = uniforms.colorA.uniform;
		const colorB = uniforms.colorB.uniform;
		const colorBorder = uniforms.colorBorder.uniform;
		const csMode = uniforms.colorSpace.uniform.value;
		const scale = uniforms.scale.uniform;
		const seed = uniforms.seed.uniform;
		const t = createAnimatedTime(params, uniforms.speed);
		const aspect = viewportSize.x.div(viewportSize.y);
		const scaledUV = vec2(screenUV.x.mul(aspect), screenUV.y).mul(scale);
		return Fn(([uv$1, animT, seedVal, edgeIntensity, edgeSoftness, borderColor]) => {
			const cell = floor(uv$1);
			const localUV = fract(uv$1);
			let d1 = float(10);
			let d2 = float(10);
			for (let ny = -1; ny <= 1; ny++) for (let nx = -1; nx <= 1; nx++) {
				const offset = vec2(float(nx), float(ny));
				const h = voronoiHash(cell.add(offset).add(seedVal));
				const pointX = h.x.add(sin(animT.add(h.x.mul(float(6.28)))).mul(float(.15))).clamp(float(.05), float(.95));
				const pointY = h.y.add(cos(animT.mul(float(.7)).add(h.y.mul(float(6.28)))).mul(float(.15))).clamp(float(.05), float(.95));
				const d = length(localUV.sub(offset.add(vec2(pointX, pointY))));
				const isCloser = d.lessThan(d1);
				d2 = isCloser.select(d1, d.lessThan(d2).select(d, d2));
				d1 = isCloser.select(d, d1);
			}
			const safeSum = d1.add(d2).max(float(1e-4));
			const cellColor = mixColors(colorA, colorB, pow(d1.mul(float(2)).div(safeSum).clamp(float(0), float(1)), float(4).sub(edgeIntensity.mul(float(3)))).clamp(float(0), float(1)), csMode);
			const scaledEdgeSoftness = edgeSoftness.mul(scale).div(float(6));
			const edgeMetric = d2.sub(d1).div(safeSum);
			const edgeMask = smoothstep(float(0), scaledEdgeSoftness.add(float(.001)), edgeMetric);
			return vec4(mix(borderColor.rgb, cellColor.rgb, edgeMask), cellColor.a);
		})(scaledUV, t, seed, uniforms.edgeIntensity.uniform, uniforms.edgeSoftness.uniform, colorBorder);
	}
};
var Voronoi_default = componentDefinition;
export { componentDefinition as n, Voronoi_default as t };
