// shaders/dist/core/WaveDistortion-CDIc5Uyk.js
import { t as applyEdgeHandling } from "./edges-CX_lJB9R.js";
import { c as transformEdges } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { PI, abs, convertToTexture, cos, float, fract, screenUV, sin, step, uniform, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "WaveDistortion",
	category: "Distortions",
	description: "Wave-based distortion with multiple waveform types",
	requiresRTT: true,
	requiresChild: true,
	props: {
		strength: {
			default: .3,
			description: "Distortion intensity",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Strength",
				group: "Effect"
			}
		},
		frequency: {
			default: 1,
			description: "Number of bends/waves",
			ui: {
				type: ["range", "map"],
				min: .1,
				max: 10,
				step: .1,
				label: "Frequency",
				group: "Effect"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		},
		angle: {
			default: 0,
			description: "Direction of wave distortion in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Angle",
				group: "Effect"
			}
		},
		waveType: {
			default: "sine",
			description: "Shape of the distortion wave",
			compileTime: true,
			transform: (value) => {
				return {
					"sine": 0,
					"triangle": 1,
					"square": 2,
					"sawtooth": 3,
					"bounce": 4
				}[value] ?? 0;
			},
			ui: {
				type: "select",
				options: [
					{
						label: "Sine",
						value: "sine"
					},
					{
						label: "Triangle",
						value: "triangle"
					},
					{
						label: "Square",
						value: "square"
					},
					{
						label: "Sawtooth",
						value: "sawtooth"
					},
					{
						label: "Bounce",
						value: "bounce"
					}
				],
				label: "Wave Type",
				group: "Effect"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges",
				group: "Effect"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms, childNode, onCleanup, onBeforeRender } = params;
		if (!childNode) return vec4(0, 0, 0, 0);
		const uvCoord = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const t = createAnimatedTime(params, uniforms.speed).mul(.5);
		let cosAngle, sinAngle;
		if (uniforms.angle.uniform.isUniformNode) {
			const precomputedCos = uniform(1);
			const precomputedSin = uniform(0);
			onBeforeRender(() => {
				const rad = uniforms.angle.uniform.value * Math.PI / 180;
				precomputedCos.value = Math.cos(rad);
				precomputedSin.value = Math.sin(rad);
			});
			cosAngle = precomputedCos;
			sinAngle = precomputedSin;
		} else {
			const angleRad = uniforms.angle.uniform.mul(PI).div(180);
			cosAngle = cos(angleRad);
			sinAngle = sin(angleRad);
		}
		const centeredUV = uvCoord.sub(vec2(.5, .5));
		const aspectCorrectedUV = vec2(centeredUV.x.mul(aspect), centeredUV.y);
		const phase = vec2(aspectCorrectedUV.x.mul(cosAngle).sub(aspectCorrectedUV.y.mul(sinAngle)), aspectCorrectedUV.x.mul(sinAngle).add(aspectCorrectedUV.y.mul(cosAngle))).y.add(.5).mul(uniforms.frequency.uniform).mul(PI.mul(2)).add(t);
		let finalWave;
		switch (uniforms.waveType.uniform.value) {
			case 1:
				finalWave = abs(fract(phase.div(PI.mul(2))).mul(2).sub(1)).mul(2).sub(1);
				break;
			case 2: {
				const normalizedPhase = fract(phase.div(PI.mul(2)));
				finalWave = step(float(.5), normalizedPhase).mul(2).sub(1);
				break;
			}
			case 3:
				finalWave = fract(phase.div(PI.mul(2))).mul(2).sub(1);
				break;
			case 4:
				finalWave = abs(sin(phase));
				break;
			default: finalWave = sin(phase);
		}
		const displacement = finalWave.mul(uniforms.strength.uniform).mul(.5);
		const displacementX = displacement.mul(cosAngle).div(aspect);
		const displacementY = displacement.mul(sinAngle);
		const distortedUV = vec2(uvCoord.x.add(displacementX), uvCoord.y.add(displacementY));
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		return unpremultiplyAlpha(applyEdgeHandling(distortedUV, childTexture.sample(distortedUV), childTexture, uniforms.edges.uniform.value));
	}
};
var WaveDistortion_default = componentDefinition;
export { componentDefinition as n, WaveDistortion_default as t };
