// shaders/dist/core/Weave-ou5shgl3.js
import { a as transformColor } from "./transformations-CJcUeZIC.js";
import { cos, float, floor, fract, fwidth, mod, screenUV, sin, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Weave",
	category: "Textures",
	description: "Interlaced textile weave pattern with two thread colors going over and under each other",
	props: {
		colorA: {
			default: "#c4c4c4",
			transform: transformColor,
			description: "Horizontal thread color",
			ui: {
				type: "color",
				label: "X Color",
				group: "Colors"
			}
		},
		colorB: {
			default: "#4d4d4d",
			transform: transformColor,
			description: "Vertical thread color",
			ui: {
				type: "color",
				label: "Y Color",
				group: "Colors"
			}
		},
		cells: {
			default: 10,
			description: "Number of threads across the shortest canvas edge",
			ui: {
				type: ["range", "map"],
				min: 2,
				max: 40,
				step: 1,
				label: "Cells",
				group: "Effect"
			}
		},
		gap: {
			default: .25,
			description: "Gap between threads (0 = no gap, 0.5 = maximum gap)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: .45,
				step: .01,
				label: "Gap",
				group: "Effect"
			}
		},
		rotation: {
			default: 0,
			description: "Rotation of the weave pattern in degrees",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 360,
				step: 1,
				label: "Rotation",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const correctedUV = vec2(uv$1.x.mul(aspect), uv$1.y);
		const rotRad = uniforms.rotation.uniform.mul(float(Math.PI / 180));
		const cosR = cos(rotRad);
		const sinR = sin(rotRad);
		const center = vec2(aspect.mul(float(.5)), float(.5));
		const centered = correctedUV.sub(center);
		const scaledUV = vec2(centered.x.mul(cosR).sub(centered.y.mul(sinR)), centered.x.mul(sinR).add(centered.y.mul(cosR))).add(center).mul(uniforms.cells.uniform);
		const cell = floor(scaledUV);
		const localUV = fract(scaledUV);
		const gap = uniforms.gap.uniform;
		const pixel = fwidth(scaledUV.x).mul(float(.5));
		const hBand = smoothstep(gap, gap.add(pixel), localUV.y).mul(smoothstep(gap, gap.add(pixel), float(1).sub(localUV.y)));
		const vBand = smoothstep(gap, gap.add(pixel), localUV.x).mul(smoothstep(gap, gap.add(pixel), float(1).sub(localUV.x)));
		const isHTopFloat = mod(cell.x.add(cell.y), float(2));
		const intersection = hBand.mul(vBand);
		const hVis = hBand.sub(intersection.mul(isHTopFloat)).clamp(float(0), float(1));
		const vVis = vBand.sub(intersection.mul(float(1).sub(isHTopFloat))).clamp(float(0), float(1));
		const colorA = uniforms.colorA.uniform;
		const colorB = uniforms.colorB.uniform;
		const weightA = colorA.a.mul(hVis);
		const weightB = colorB.a.mul(vVis);
		const totalAlpha = weightA.add(weightB).clamp(float(0), float(1));
		const safeWeight = weightA.add(weightB).max(float(.001));
		return vec4(colorA.x.mul(weightA).add(colorB.x.mul(weightB)).div(safeWeight), colorA.y.mul(weightA).add(colorB.y.mul(weightB)).div(safeWeight), colorA.z.mul(weightA).add(colorB.z.mul(weightB)).div(safeWeight), totalAlpha);
	}
};
var Weave_default = componentDefinition;
export { componentDefinition as n, Weave_default as t };
