// shaders/dist/core/WebcamTexture-D4e06ajM.js
import { t as needsVideoCanvasWorkaround } from "./browser-NUM-x2tw.js";
import { CanvasTexture, SRGBColorSpace, VideoTexture } from "three/webgpu";
import { float, max, min, or, screenUV, select, texture, uniform, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "WebcamTexture",
	category: "Textures",
	description: "Display a live webcam feed with customizable object-fit modes",
	props: {
		objectFit: {
			default: "cover",
			description: "How the webcam feed should be sized within the viewport",
			transform: (value) => {
				return {
					"cover": 0,
					"contain": 1,
					"fill": 2,
					"scale-down": 3,
					"none": 4
				}[value] ?? 0;
			},
			compileTime: true,
			ui: {
				type: "select",
				options: [
					{
						label: "Cover",
						value: "cover"
					},
					{
						label: "Contain",
						value: "contain"
					},
					{
						label: "Fill",
						value: "fill"
					},
					{
						label: "Scale Down",
						value: "scale-down"
					},
					{
						label: "None",
						value: "none"
					}
				],
				label: "Object Fit",
				group: "Appearance"
			}
		},
		mirror: {
			default: true,
			description: "Mirror the webcam feed horizontally (selfie mode)",
			ui: {
				type: "checkbox",
				label: "Mirror",
				group: "Effect"
			}
		}
	},
	fragmentNode: ({ uniforms, onCleanup, onBeforeRender }) => {
		const baseUV = screenUV;
		let mediaStream = null;
		let videoElement = null;
		let isInitialized = false;
		let isDisposed = false;
		const videoAspectUniform = uniform(1);
		const videoWidthUniform = uniform(640);
		const videoHeightUniform = uniform(480);
		let canvas = null;
		let ctx = null;
		let activeTexture;
		if (needsVideoCanvasWorkaround()) {
			canvas = document.createElement("canvas");
			canvas.width = 640;
			canvas.height = 480;
			ctx = canvas.getContext("2d");
			activeTexture = new CanvasTexture(canvas);
		} else {
			const placeholderVideo = document.createElement("video");
			placeholderVideo.playsInline = true;
			placeholderVideo.muted = true;
			activeTexture = new VideoTexture(placeholderVideo);
		}
		activeTexture.colorSpace = SRGBColorSpace;
		const textureNode = texture(activeTexture);
		const startWebcam = async () => {
			if (isInitialized || isDisposed) return;
			try {
				mediaStream = await navigator.mediaDevices.getUserMedia({
					video: {
						width: { ideal: 1280 },
						height: { ideal: 720 },
						facingMode: "user"
					},
					audio: false
				});
				if (isDisposed) {
					mediaStream.getTracks().forEach((track) => track.stop());
					mediaStream = null;
					return;
				}
				videoElement = document.createElement("video");
				videoElement.srcObject = mediaStream;
				videoElement.playsInline = true;
				videoElement.muted = true;
				await new Promise((resolve, reject) => {
					if (!videoElement) return reject(/* @__PURE__ */ new Error("Video element not created"));
					videoElement.onloadedmetadata = () => {
						resolve();
					};
					videoElement.onerror = () => {
						reject(/* @__PURE__ */ new Error("Failed to load video metadata"));
					};
				});
				if (isDisposed) return;
				await videoElement.play();
				if (isDisposed) return;
				const width = videoElement.videoWidth || 640;
				const height = videoElement.videoHeight || 480;
				videoAspectUniform.value = width / height;
				videoWidthUniform.value = width;
				videoHeightUniform.value = height;
				activeTexture.dispose();
				if (needsVideoCanvasWorkaround()) {
					canvas.width = width;
					canvas.height = height;
					activeTexture = new CanvasTexture(canvas);
				} else activeTexture = new VideoTexture(videoElement);
				activeTexture.colorSpace = SRGBColorSpace;
				textureNode.value = activeTexture;
				isInitialized = true;
				console.log(`[WebcamTexture] Webcam started: ${width}x${height}`);
			} catch (error) {
				if (error instanceof DOMException) if (error.name === "NotAllowedError") console.error("[WebcamTexture] Camera permission denied by user");
				else if (error.name === "NotFoundError") console.error("[WebcamTexture] No camera found on this device");
				else if (error.name === "NotReadableError") console.error("[WebcamTexture] Camera is already in use by another application");
				else console.error("[WebcamTexture] Camera error:", error.message);
				else console.error("[WebcamTexture] Failed to start webcam:", error);
			}
		};
		startWebcam();
		onBeforeRender(() => {
			if (needsVideoCanvasWorkaround() && videoElement && isInitialized && !videoElement.paused && videoElement.readyState >= 2) {
				ctx.drawImage(videoElement, 0, 0, canvas.width, canvas.height);
				activeTexture.needsUpdate = true;
			}
		});
		onCleanup(() => {
			isDisposed = true;
			if (mediaStream) {
				mediaStream.getTracks().forEach((track) => track.stop());
				mediaStream = null;
			}
			if (videoElement) {
				videoElement.srcObject = null;
				videoElement = null;
			}
			activeTexture.dispose();
			isInitialized = false;
		});
		const uv$1 = baseUV;
		const viewportAspect = viewportSize.x.div(viewportSize.y);
		const objectFitMode = uniforms.objectFit.uniform.value;
		let uvScale;
		switch (objectFitMode) {
			case 1: {
				const containScale = min(viewportAspect.div(videoAspectUniform), float(1));
				uvScale = vec2(videoAspectUniform.div(viewportAspect).mul(containScale), containScale);
				break;
			}
			case 2:
				uvScale = vec2(1, 1);
				break;
			case 3: {
				const scaleDownScale = min(min(viewportAspect.div(videoAspectUniform), float(1)), min(viewportSize.x.div(videoWidthUniform), viewportSize.y.div(videoHeightUniform)));
				uvScale = vec2(videoAspectUniform.div(viewportAspect).mul(scaleDownScale), scaleDownScale);
				break;
			}
			case 4: {
				const noneScale = min(viewportSize.x.div(videoWidthUniform), viewportSize.y.div(videoHeightUniform));
				uvScale = vec2(videoAspectUniform.div(viewportAspect).mul(noneScale), noneScale);
				break;
			}
			case 0:
			default: {
				const coverScale = max(viewportAspect.div(videoAspectUniform), float(1));
				uvScale = vec2(videoAspectUniform.div(viewportAspect).mul(coverScale), coverScale);
				break;
			}
		}
		const adjustedUV = uv$1.sub(vec2(.5)).div(uvScale).add(vec2(.5));
		const mirrorMode = uniforms.mirror.uniform;
		const finalUV = vec2(select(mirrorMode, float(1).sub(adjustedUV.x), adjustedUV.x), float(1).sub(adjustedUV.y));
		const sampledColor = textureNode.sample(finalUV);
		const isOutOfBounds = or(or(finalUV.x.lessThan(0), finalUV.x.greaterThan(1)), or(finalUV.y.lessThan(0), finalUV.y.greaterThan(1)));
		return vec4(sampledColor.rgb, select(isOutOfBounds, float(0), sampledColor.a));
	}
};
var WebcamTexture_default = componentDefinition;
export { componentDefinition as n, WebcamTexture_default as t };
