// shaders/dist/core/WorleyNoise-Uf6IPm7A.js
import { a as transformColor, o as transformColorSpace, t as colorSpaceOptions } from "./transformations-CJcUeZIC.js";
import { t as mixColors } from "./colorMixing-BPpDnR5I.js";
import { t as createAnimatedTime } from "./time-DUqSFWvT.js";
import { Fn, abs, clamp, cos, dot, float, floor, fract, lengthSq, max, min, mix, screenUV, sin, sqrt, vec2, viewportSize } from "three/tsl";
var TAU = float(Math.PI * 2);
var HASH_AMP = float(.15);
var CELL_CENTER = float(.5);
var CELL_MIN = float(.05);
var CELL_MAX = float(.95);
var ANIM_PHASE_SCALE = float(.7);
var SENTINEL = float(10);
var worleyHash = Fn(([p]) => {
	return fract(sin(vec2(dot(p, vec2(127.1, 311.7)), dot(p, vec2(269.5, 183.3)))).mul(float(43758.5453)));
});
var transformMode = (value) => {
	return {
		f1: 0,
		f2: 1,
		f2MinusF1: 2,
		f1PlusF2: 3,
		f1TimesF2: 4
	}[value] ?? 0;
};
var transformDistance = (value) => {
	return {
		euclidean: 0,
		manhattan: 1,
		chebyshev: 2
	}[value] ?? 0;
};
const componentDefinition = {
	name: "WorleyNoise",
	category: "Textures",
	description: "Cellular noise field — distance-based, with selectable feature combinations and fractal octaves",
	props: {
		colorA: {
			default: "#ffffff",
			transform: transformColor,
			description: "Color where the noise field is low (typically near cell centers)",
			ui: {
				type: "color",
				label: "Color A (Low)",
				group: "Colors"
			}
		},
		colorB: {
			default: "#000000",
			transform: transformColor,
			description: "Color where the noise field is high (typically near cell boundaries)",
			ui: {
				type: "color",
				label: "Color B (High)",
				group: "Colors"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			compileTime: true,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space",
				group: "Colors"
			}
		},
		scale: {
			default: 6,
			description: "Number of cells across the canvas at the base octave",
			ui: {
				type: ["range", "map"],
				min: 1,
				max: 30,
				step: .5,
				label: "Scale",
				group: "Effect"
			}
		},
		mode: {
			default: "f1",
			transform: transformMode,
			compileTime: true,
			description: "Field type. F1 = distance to nearest point. F2 = distance to second-nearest. F2 − F1 emphasises cell boundaries.",
			ui: {
				type: "select",
				options: [
					{
						label: "F1 (Cells)",
						value: "f1"
					},
					{
						label: "F2 (Wider Cells)",
						value: "f2"
					},
					{
						label: "F2 − F1 (Edges)",
						value: "f2MinusF1"
					},
					{
						label: "F1 + F2",
						value: "f1PlusF2"
					},
					{
						label: "F1 × F2",
						value: "f1TimesF2"
					}
				],
				label: "Mode",
				group: "Effect"
			}
		},
		distance: {
			default: "euclidean",
			transform: transformDistance,
			compileTime: true,
			description: "Distance metric. Euclidean = round cells. Manhattan = diamond. Chebyshev = square.",
			ui: {
				type: "select",
				options: [
					{
						label: "Euclidean",
						value: "euclidean"
					},
					{
						label: "Manhattan",
						value: "manhattan"
					},
					{
						label: "Chebyshev",
						value: "chebyshev"
					}
				],
				label: "Distance",
				group: "Effect"
			}
		},
		octaves: {
			default: 1,
			compileTime: true,
			description: "Number of fractal layers stacked at progressively finer scales",
			ui: {
				type: "range",
				min: 1,
				max: 4,
				step: 1,
				label: "Octaves",
				group: "Effect"
			}
		},
		lacunarity: {
			default: 2,
			description: "Scale multiplier between octaves (only active when Octaves > 1)",
			ui: {
				type: ["range", "map"],
				min: 1.5,
				max: 4,
				step: .1,
				label: "Lacunarity",
				group: "Effect",
				condition: { octaves: [
					2,
					3,
					4
				] }
			}
		},
		persistence: {
			default: .5,
			description: "Amplitude multiplier between octaves (only active when Octaves > 1)",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Persistence",
				group: "Effect",
				condition: { octaves: [
					2,
					3,
					4
				] }
			}
		},
		jitter: {
			default: 1,
			description: "How much each cell's point drifts inside its cell. 0 = rigid grid (banded look), 1 = fully random.",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 1,
				step: .05,
				label: "Jitter",
				group: "Effect"
			}
		},
		contrast: {
			default: 1,
			description: "Steepness of the gradient between low and high regions",
			ui: {
				type: ["range", "map"],
				min: .25,
				max: 4,
				step: .05,
				label: "Contrast",
				group: "Effect"
			}
		},
		balance: {
			default: 0,
			description: "Shifts the gradient midpoint. Negative pulls the field toward Color A, positive toward Color B.",
			ui: {
				type: ["range", "map"],
				min: -1,
				max: 1,
				step: .05,
				label: "Balance",
				group: "Effect"
			}
		},
		seed: {
			default: 0,
			description: "Random seed — shifts the cell pattern without changing its overall structure",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Seed",
				group: "Effect"
			}
		},
		speed: {
			default: .5,
			description: "Animation speed — how fast each cell's point drifts",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed",
				group: "Animation"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const colorA = uniforms.colorA.uniform;
		const colorB = uniforms.colorB.uniform;
		const csMode = uniforms.colorSpace.uniform.value;
		const modeNum = uniforms.mode.uniform.value;
		const distNum = uniforms.distance.uniform.value;
		const octavesNum = Math.max(1, Math.min(4, Math.round(uniforms.octaves.uniform.value)));
		const t = createAnimatedTime(params, uniforms.speed);
		const aspect = viewportSize.x.div(viewportSize.y);
		const baseUV = vec2(screenUV.x.mul(aspect), screenUV.y);
		const scale = uniforms.scale.uniform;
		const seed = uniforms.seed.uniform;
		const jitter = uniforms.jitter.uniform;
		const lacunarity = uniforms.lacunarity.uniform;
		const persistence = uniforms.persistence.uniform;
		let modeScale = 1.5;
		if (modeNum === 1) modeScale = 1;
		else if (modeNum === 2) modeScale = 2.5;
		else if (modeNum === 3) modeScale = .6;
		else if (modeNum === 4) modeScale = 2;
		if (distNum === 1) modeScale *= .7;
		else if (distNum === 2) modeScale *= 1.4;
		const isSquaredEuclidean = distNum === 0;
		const needsD2 = modeNum !== 0;
		const evalOctaveRaw = Fn(([oUV, animT, seedOff]) => {
			const animTSlow = animT.mul(ANIM_PHASE_SCALE);
			const cell = floor(oUV);
			const localUV = fract(oUV);
			let d1 = SENTINEL;
			let d2 = SENTINEL;
			for (let ny = -1; ny <= 1; ny++) for (let nx = -1; nx <= 1; nx++) {
				const offset = vec2(float(nx), float(ny));
				const h = worleyHash(cell.add(offset).add(seedOff));
				const px = h.x.add(sin(animT.add(h.x.mul(TAU))).mul(HASH_AMP));
				const py = h.y.add(cos(animTSlow.add(h.y.mul(TAU))).mul(HASH_AMP));
				const jx = mix(CELL_CENTER, px.clamp(CELL_MIN, CELL_MAX), jitter);
				const jy = mix(CELL_CENTER, py.clamp(CELL_MIN, CELL_MAX), jitter);
				const point = offset.add(vec2(jx, jy));
				const delta = localUV.sub(point);
				let d;
				if (distNum === 1) d = abs(delta.x).add(abs(delta.y));
				else if (distNum === 2) d = max(abs(delta.x), abs(delta.y));
				else d = lengthSq(delta);
				if (needsD2) {
					d2 = min(d2, max(d, d1));
					d1 = min(d1, d);
				} else d1 = min(d1, d);
			}
			let raw;
			if (isSquaredEuclidean) if (modeNum === 1) raw = sqrt(d2);
			else if (modeNum === 2) raw = sqrt(d2).sub(sqrt(d1));
			else if (modeNum === 3) raw = sqrt(d1).add(sqrt(d2));
			else if (modeNum === 4) raw = sqrt(d1.mul(d2));
			else raw = sqrt(d1);
			else if (modeNum === 1) raw = d2;
			else if (modeNum === 2) raw = d2.sub(d1);
			else if (modeNum === 3) raw = d1.add(d2);
			else if (modeNum === 4) raw = d1.mul(d2);
			else raw = d1;
			return raw;
		});
		let acc = float(0);
		let totalAmp = float(0);
		let amp = float(1);
		let freq = scale;
		for (let o = 0; o < octavesNum; o++) {
			const raw = evalOctaveRaw(baseUV.mul(freq), t.add(float(o * 17)), seed.add(float(o * 31)));
			acc = acc.add(raw.mul(amp));
			totalAmp = totalAmp.add(amp);
			amp = amp.mul(persistence);
			freq = freq.mul(lacunarity);
		}
		return mixColors(colorA, colorB, clamp(acc.div(totalAmp.max(float(1e-4))).mul(float(modeScale)).mul(float(2)).sub(float(1)).mul(uniforms.contrast.uniform).add(uniforms.balance.uniform).mul(float(.5)).add(float(.5)), float(0), float(1)), csMode);
	}
};
var WorleyNoise_default = componentDefinition;
export { componentDefinition as n, WorleyNoise_default as t };
