// shaders/dist/core/ZoomBlur-DJ-RNKHM.js
import { l as transformPosition } from "./transformations-CJcUeZIC.js";
import { t as unpremultiplyAlpha } from "./alpha-C4ptedXe.js";
import { Fn, convertToTexture, float, screenUV, vec2, vec4, viewportSize } from "three/tsl";
var rawWeights = Array.from({ length: 32 }, (_, i) => {
	const t = (i / 31 - .5) * 2;
	return Math.exp(-.5 * t * t / .64);
});
var wSum = rawWeights.reduce((a, b) => a + b);
var blurWeights = rawWeights.map((w) => w / wSum);
const componentDefinition = {
	name: "ZoomBlur",
	category: "Blurs",
	description: "Radial zoom blur expanding from a center point",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 30,
			description: "Intensity of the zoom blur effect",
			ui: {
				type: ["range", "map"],
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity",
				group: "Effect"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the zoom blur",
			ui: {
				type: "position",
				label: "Center",
				group: "Position"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const center = vec2(uniforms.center.uniform.x, uniforms.center.uniform.y.oneMinus());
		const radius = uniforms.intensity.uniform.mul(.01);
		const aspect = viewportSize.x.div(viewportSize.y);
		return Fn(() => {
			const uv$1 = screenUV;
			const total = vec4(0).toVar();
			const delta = uv$1.sub(center);
			const aspectCorrectedDelta = vec2(delta.x.mul(aspect), delta.y);
			for (let i = 0; i < 32; i++) {
				const scale = float(1).add(radius.mul(i / 31));
				const scaledDelta = aspectCorrectedDelta.div(scale);
				const scaledCoord = vec2(scaledDelta.x.div(aspect), scaledDelta.y).add(center);
				total.assign(total.add(childTexture.sample(scaledCoord).mul(float(blurWeights[i]))));
			}
			return unpremultiplyAlpha(total);
		})();
	}
};
var ZoomBlur_default = componentDefinition;
export { componentDefinition as n, ZoomBlur_default as t };
