// shaders/dist/partner/index.js
const DEFAULT_BASE_URL = "https://shaders.com";
const CACHE_TTL_MS = 5 * 60 * 1e3;
class PartnerClient {
  constructor(options) {
    this.cache = /* @__PURE__ */ new Map();
    if (!options.apiKey) throw new Error("PartnerClient requires an apiKey");
    this.apiKey = options.apiKey;
    this.baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/$/, "");
  }
  getCached(key) {
    const entry = this.cache.get(key);
    if (!entry) return null;
    if (Date.now() > entry.expiresAt) {
      this.cache.delete(key);
      return null;
    }
    return entry.data;
  }
  setCached(key, data) {
    this.cache.set(key, { data, expiresAt: Date.now() + CACHE_TTL_MS });
  }
  async listPresets(options = {}) {
    const page = Math.max(1, options.page ?? 1);
    const limit = Math.min(100, Math.max(1, options.limit ?? 20));
    const categories = options.categories && options.categories.length > 0 ? [...new Set(options.categories)].sort() : null;
    const cacheKey = `presets:p${page}:l${limit}:c${(categories == null ? void 0 : categories.join(",")) ?? "*"}`;
    const cached = this.getCached(cacheKey);
    if (cached) return cached;
    const url = new URL(`${this.baseUrl}/api/partner/v1/presets`);
    url.searchParams.set("page", String(page));
    url.searchParams.set("limit", String(limit));
    if (categories) {
      url.searchParams.set("categories", categories.join(","));
    }
    const response = await fetch(url.toString(), {
      headers: { Authorization: `Bearer ${this.apiKey}` }
    });
    if (!response.ok) {
      let message = response.statusText;
      try {
        const body = await response.json();
        if (body.message) message = body.message;
      } catch {
      }
      throw new Error(`Partner API error ${response.status}: ${message}`);
    }
    const data = await response.json();
    this.setCached(cacheKey, data);
    return data;
  }
  /**
   * Generate an SDF (signed distance field) `.bin` file from raw SVG content.
   * Resolves to the file's bytes as a `Uint8Array` — the caller can write it
   * to disk, upload it to their own CDN, or feed it straight into a shader
   * component that consumes SDF buffers.
   *
   * Backed by the public `/api/plugin/sdf/generate` endpoint, which is shared
   * with the Framer plugin. The endpoint is anonymous (no API key required);
   * Vercel-side rate limits apply. Maximum input size: 5 MB of SVG text.
   *
   * Not cached — the same SVG fed in twice will round-trip twice. Memoize
   * upstream if you need to.
   *
   * @example
   * ```ts
   * const fs = await import('node:fs/promises')
   * const svg = await fs.readFile('./logo.svg', 'utf8')
   * const sdf = await client.generateSdf(svg)
   * await fs.writeFile('./logo.sdf.bin', sdf)
   * ```
   */
  async generateSdf(svg, options = {}) {
    if (typeof svg !== "string" || svg.length === 0) {
      throw new Error("generateSdf requires a non-empty SVG string");
    }
    const timeoutMs = options.timeoutMs ?? 3e4;
    const controller = new AbortController();
    const timeoutHandle = setTimeout(() => controller.abort(), timeoutMs);
    try {
      const response = await fetch(`${this.baseUrl}/api/plugin/sdf/generate`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ svg }),
        signal: controller.signal
      });
      if (!response.ok) {
        let message = response.statusText;
        try {
          const body = await response.json();
          if (body.message) message = body.message;
          else if (body.error) message = body.error;
        } catch {
        }
        throw new Error(`SDF generation failed ${response.status}: ${message}`);
      }
      const buffer = await response.arrayBuffer();
      return new Uint8Array(buffer);
    } catch (err) {
      if (err instanceof Error && err.name === "AbortError") {
        throw new Error(`SDF generation timed out after ${timeoutMs}ms`);
      }
      throw err;
    } finally {
      clearTimeout(timeoutHandle);
    }
  }
}
export {
  PartnerClient
};
