// shaders/dist/js/createPreview.js
import { createShader } from "./createShader.js";
import { decodePreviewDefinition } from "./decode.js";
var DEFAULT_API_BASE = "https://shaders.com";
var WATERMARK_TEXT = "Unlock your Shaders Pro license";
var WATERMARK_LINK = "https://shaders.com/dashboard?pricing=true";
var RESERVED_IDENTIFIERS = new Set([
	"break",
	"case",
	"catch",
	"class",
	"const",
	"continue",
	"default",
	"delete",
	"do",
	"else",
	"export",
	"extends",
	"finally",
	"for",
	"function",
	"if",
	"import",
	"in",
	"instanceof",
	"new",
	"return",
	"super",
	"switch",
	"this",
	"throw",
	"try",
	"typeof",
	"var",
	"void",
	"while",
	"with",
	"yield"
]);
function slugifyIdentifier(label, fallback) {
	const cleaned = label.replace(/[^A-Za-z0-9]+/g, " ").trim();
	if (!cleaned) return fallback;
	const camel = cleaned.split(/\s+/).map((part, i) => i === 0 ? part.charAt(0).toLowerCase() + part.slice(1) : part.charAt(0).toUpperCase() + part.slice(1)).join("");
	const safe = /^[A-Za-z_$]/.test(camel) ? camel : `_${camel}`;
	return RESERVED_IDENTIFIERS.has(safe) ? `_${safe}` : safe;
}
function buildKeyPropIdentifierMap(keyProps) {
	const used = new Set(["style"]);
	const map = /* @__PURE__ */ new Map();
	if (!keyProps?.length) return map;
	keyProps.forEach((kp, i) => {
		if (!kp.targets?.length) return;
		const base = slugifyIdentifier(kp.label, `keyProp${i}`);
		let ident = base;
		let suffix = 1;
		while (used.has(ident)) {
			suffix++;
			ident = `${base}${suffix}`;
		}
		used.add(ident);
		map.set(ident, kp);
	});
	return map;
}
function findInTree(id, list) {
	for (const c of list) {
		if (c.id === id) return c;
		const children = c.children;
		if (children?.length) {
			const hit = findInTree(id, children);
			if (hit) return hit;
		}
	}
	return null;
}
function applyConfigurationInPlace(components, configuration, identMap) {
	if (!configuration) return;
	for (const [ident, value] of Object.entries(configuration)) {
		const kp = identMap.get(ident);
		if (!kp) continue;
		for (const t of kp.targets ?? []) {
			const comp = findInTree(t.component_id, components);
			if (comp?.props && t.prop in comp.props) comp.props[t.prop] = value;
		}
	}
}
async function createPreview(canvas, options) {
	if (!options.shader && !options.presetId || options.shader && options.presetId) throw new Error("Exactly one of shader (preview token) or presetId must be provided");
	const apiBase = options.apiBaseUrl || DEFAULT_API_BASE;
	const versionParam = options.presetId && options.version ? `?version=${encodeURIComponent(options.version)}` : "";
	const url = options.shader ? `${apiBase}/api/preview/shader/${encodeURIComponent(options.shader)}` : `${apiBase}/api/preview/preset/${encodeURIComponent(options.presetId)}${versionParam}`;
	const response = await fetch(url);
	if (!response.ok) throw new Error(`Failed to fetch preview: ${response.status} ${response.statusText}`);
	const payload = await response.json();
	const item = payload.preset ?? payload.shader;
	if (!item?.definition || typeof item.definition !== "string") throw new Error("Preview response missing encoded definition");
	const definition = decodePreviewDefinition(item.definition);
	const keyProps = Array.isArray(item.key_props) ? item.key_props : null;
	const identMap = buildKeyPropIdentifierMap(keyProps);
	const baselineMap = /* @__PURE__ */ new Map();
	for (const kp of identMap.values()) for (const t of kp.targets ?? []) {
		const comp = findInTree(t.component_id, definition.components);
		if (comp?.props && t.prop in comp.props) baselineMap.set(`${t.component_id}:${t.prop}`, comp.props[t.prop]);
	}
	applyConfigurationInPlace(definition.components, options.configuration, identMap);
	const instance = await createShader(canvas, definition, {
		isPreview: true,
		enablePerformanceTracking: true
	});
	const { overlay, savedPosition, parentEl } = createWatermarkOverlay(canvas);
	const originalDestroy = instance.destroy.bind(instance);
	const previewInstance = instance;
	previewInstance.destroy = () => {
		if (parentEl && savedPosition !== void 0) parentEl.style.position = savedPosition;
		overlay.remove();
		originalDestroy();
	};
	previewInstance.setConfiguration = (configuration) => {
		if (!keyProps?.length) return;
		for (const [ident, kp] of identMap) {
			const hasNewValue = Object.prototype.hasOwnProperty.call(configuration, ident) && configuration[ident] !== void 0;
			for (const t of kp.targets ?? []) {
				const value = hasNewValue ? configuration[ident] : baselineMap.get(`${t.component_id}:${t.prop}`);
				if (value === void 0) continue;
				instance.update(t.component_id, { [t.prop]: value });
			}
		}
	};
	return previewInstance;
}
function createWatermarkOverlay(canvas) {
	const parent = canvas.parentElement;
	let savedPosition;
	if (parent) {
		if (getComputedStyle(parent).position === "static") {
			savedPosition = parent.style.position;
			parent.style.position = "relative";
		}
	}
	const link = document.createElement("a");
	link.href = WATERMARK_LINK;
	link.target = "_blank";
	link.rel = "noopener noreferrer";
	link.textContent = WATERMARK_TEXT;
	Object.assign(link.style, {
		position: "absolute",
		bottom: "8px",
		right: "12px",
		fontSize: "11px",
		fontFamily: "system-ui, -apple-system, sans-serif",
		color: "rgba(255, 255, 255, 0.5)",
		textDecoration: "none",
		zIndex: "10",
		pointerEvents: "auto",
		transition: "color 0.2s ease"
	});
	link.addEventListener("mouseenter", () => {
		link.style.color = "rgba(255, 255, 255, 0.8)";
	});
	link.addEventListener("mouseleave", () => {
		link.style.color = "rgba(255, 255, 255, 0.5)";
	});
	if (parent) parent.appendChild(link);
	return {
		overlay: link,
		savedPosition,
		parentEl: parent
	};
}
export { createPreview };
