// shaders/dist/pixi/index.js
import { createRendererFromJSON } from "../core/index.js";
import { RendererType, CanvasSource, Texture, Sprite } from "pixi.js";
async function createShaderSprite(preset, pixiApp, options = {}) {
  const width = options.width || 512;
  const height = options.height || 512;
  const dpr = window.devicePixelRatio || 1;
  if (pixiApp.renderer.type !== RendererType.WEBGPU) {
    throw new Error('Shaders requires PixiJS to be initialized with WebGPU (preference: "webgpu")');
  }
  const pixiGpu = pixiApp.renderer.gpu;
  const device = pixiGpu.device;
  const adapter = pixiGpu.adapter;
  if (!device || !adapter) {
    throw new Error("Shaders could not access the WebGPU device from the PixiJS renderer");
  }
  const wrapper = document.createElement("div");
  wrapper.style.position = "absolute";
  wrapper.style.left = "-9999px";
  wrapper.style.top = "-9999px";
  wrapper.style.width = `${width}px`;
  wrapper.style.height = `${height}px`;
  wrapper.style.overflow = "hidden";
  wrapper.style.pointerEvents = "none";
  document.body.appendChild(wrapper);
  const canvas = document.createElement("canvas");
  canvas.width = width * dpr;
  canvas.height = height * dpr;
  canvas.style.width = `${width}px`;
  canvas.style.height = `${height}px`;
  wrapper.appendChild(canvas);
  const gpuContext = canvas.getContext("webgpu");
  if (!gpuContext) {
    throw new Error("Could not get WebGPU context from canvas");
  }
  gpuContext.configure({
    device,
    format: navigator.gpu.getPreferredCanvasFormat(),
    alphaMode: "premultiplied",
    usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_SRC | GPUTextureUsage.TEXTURE_BINDING
  });
  const shadersRenderer = createRendererFromJSON(preset, {
    gpu: { device, adapter },
    forceFullFrameRate: true
  });
  await shadersRenderer.initialize(canvas);
  let canvasSource = new CanvasSource({
    resource: canvas,
    resolution: dpr,
    antialias: false,
    autoDensity: false
  });
  let texture = new Texture({
    source: canvasSource
  });
  const sprite = new Sprite(texture);
  sprite.width = width;
  sprite.height = height;
  let currentWidth = width;
  let currentHeight = height;
  let pendingCanvasSource = null;
  let pendingTexture = null;
  let resizeSeq = 0;
  return {
    sprite,
    /**
     * Update shader rendering and sync to PixiJS
     * Call this in your render loop BEFORE rendering PixiJS
     */
    update: async () => {
      await shadersRenderer.renderFrame();
      if (!pendingCanvasSource) {
        canvasSource.update();
      }
    },
    /**
     * Update the shader preset dynamically
     */
    updatePreset: (newPreset) => {
      shadersRenderer.updatePreset(newPreset);
    },
    /**
     * Resize the shader to new dimensions
     * This updates the offscreen canvas, triggering the renderer's resize logic.
     * The texture swap is deferred until after Three.js's resize RAF fires, so the
     * sprite continues to show the last good frame instead of flashing blank.
     */
    resize: (newWidth, newHeight) => {
      if (newWidth === currentWidth && newHeight === currentHeight) return;
      currentWidth = newWidth;
      currentHeight = newHeight;
      if (pendingCanvasSource) {
        pendingCanvasSource.destroy();
        pendingTexture.destroy();
        pendingCanvasSource = null;
        pendingTexture = null;
      }
      const seq = ++resizeSeq;
      wrapper.style.width = `${newWidth}px`;
      wrapper.style.height = `${newHeight}px`;
      canvas.width = newWidth * dpr;
      canvas.height = newHeight * dpr;
      canvas.style.width = `${newWidth}px`;
      canvas.style.height = `${newHeight}px`;
      gpuContext.configure({
        device,
        format: navigator.gpu.getPreferredCanvasFormat(),
        alphaMode: "premultiplied",
        usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_SRC | GPUTextureUsage.TEXTURE_BINDING
      });
      pendingCanvasSource = new CanvasSource({
        resource: canvas,
        resolution: dpr,
        antialias: false,
        autoDensity: false
      });
      pendingTexture = new Texture({ source: pendingCanvasSource });
      requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          if (seq !== resizeSeq) return;
          if (pendingCanvasSource && pendingTexture) {
            pendingCanvasSource.update();
            const oldTexture = texture;
            const oldCanvasSource = canvasSource;
            canvasSource = pendingCanvasSource;
            texture = pendingTexture;
            sprite.texture = texture;
            pendingCanvasSource = null;
            pendingTexture = null;
            oldTexture.destroy();
            oldCanvasSource.destroy();
          }
        });
      });
    },
    /**
     * Clean up resources
     */
    dispose: () => {
      if (pendingCanvasSource) {
        pendingCanvasSource.destroy();
        pendingTexture == null ? void 0 : pendingTexture.destroy();
        pendingCanvasSource = null;
        pendingTexture = null;
      }
      shadersRenderer.dispose();
      texture.destroy(true);
      wrapper.remove();
    }
  };
}
export {
  createShaderSprite
};
