// shaders/dist/vue/Preview.vue_vue_type_script_setup_true_lang.js
import AngularBlur_default from "./AngularBlur.js";
import Ascii_default from "./Ascii.js";
import Aurora_default from "./Aurora.js";
import BarShift_default from "./BarShift.js";
import Beam_default from "./Beam.js";
import Blob_default from "./Blob.js";
import Blur_default from "./Blur.js";
import BrickPattern_default from "./BrickPattern.js";
import BrightnessContrast_default from "./BrightnessContrast.js";
import Bulge_default from "./Bulge.js";
import CRTScreen_default from "./CRTScreen.js";
import ChannelBlur_default from "./ChannelBlur.js";
import Checkerboard_default from "./Checkerboard.js";
import Chevron_default from "./Chevron.js";
import ChromaFlow_default from "./ChromaFlow.js";
import ChromaticAberration_default from "./ChromaticAberration.js";
import Circle_default from "./Circle.js";
import ColorWheel_default from "./ColorWheel.js";
import ConcentricSpin_default from "./ConcentricSpin.js";
import ConicGradient_default from "./ConicGradient.js";
import ContourLines_default from "./ContourLines.js";
import Crescent_default from "./Crescent.js";
import Cross_default from "./Cross.js";
import Crystal_default from "./Crystal.js";
import CursorRipples_default from "./CursorRipples.js";
import CursorTrail_default from "./CursorTrail.js";
import DOMTexture_default from "./DOMTexture.js";
import DiamondGradient_default from "./DiamondGradient.js";
import DiffuseBlur_default from "./DiffuseBlur.js";
import Dither_default from "./Dither.js";
import DotGrid_default from "./DotGrid.js";
import DropShadow_default from "./DropShadow.js";
import Duotone_default from "./Duotone.js";
import Ellipse_default from "./Ellipse.js";
import Emboss_default from "./Emboss.js";
import FallingLines_default from "./FallingLines.js";
import FilmGrain_default from "./FilmGrain.js";
import FloatingParticles_default from "./FloatingParticles.js";
import FlowField_default from "./FlowField.js";
import Flower_default from "./Flower.js";
import FlowingGradient_default from "./FlowingGradient.js";
import FlutedGlass_default from "./FlutedGlass.js";
import Fog_default from "./Fog.js";
import Form3D_default from "./Form3D.js";
import FractalNoise_default from "./FractalNoise.js";
import Glass_default from "./Glass.js";
import GlassTiles_default from "./GlassTiles.js";
import Glitch_default from "./Glitch.js";
import Glow_default from "./Glow.js";
import Godrays_default from "./Godrays.js";
import Grayscale_default from "./Grayscale.js";
import Grid_default from "./Grid.js";
import GridDistortion_default from "./GridDistortion.js";
import Group_default from "./Group.js";
import Halftone_default from "./Halftone.js";
import HexGrid_default from "./HexGrid.js";
import HueShift_default from "./HueShift.js";
import ImageTexture_default from "./ImageTexture.js";
import Invert_default from "./Invert.js";
import Kaleidoscope_default from "./Kaleidoscope.js";
import LensFlare_default from "./LensFlare.js";
import LinearBlur_default from "./LinearBlur.js";
import LinearGradient_default from "./LinearGradient.js";
import Liquify_default from "./Liquify.js";
import Marble_default from "./Marble.js";
import Mirror_default from "./Mirror.js";
import MultiPointGradient_default from "./MultiPointGradient.js";
import Neon_default from "./Neon.js";
import Paper_default from "./Paper.js";
import Perspective_default from "./Perspective.js";
import Pixelate_default from "./Pixelate.js";
import Plasma_default from "./Plasma.js";
import PolarCoordinates_default from "./PolarCoordinates.js";
import Polygon_default from "./Polygon.js";
import Posterize_default from "./Posterize.js";
import ProgressiveBlur_default from "./ProgressiveBlur.js";
import RadialGradient_default from "./RadialGradient.js";
import RectangularCoordinates_default from "./RectangularCoordinates.js";
import ReflectivePlane_default from "./ReflectivePlane.js";
import Ring_default from "./Ring.js";
import Ripples_default from "./Ripples.js";
import RoundedRect_default from "./RoundedRect.js";
import Saturation_default from "./Saturation.js";
import Sharpness_default from "./Sharpness.js";
import Shatter_default from "./Shatter.js";
import SimplexNoise_default from "./SimplexNoise.js";
import SineWave_default from "./SineWave.js";
import Smoke_default from "./Smoke.js";
import SmokeFill_default from "./SmokeFill.js";
import Solarize_default from "./Solarize.js";
import SolidColor_default from "./SolidColor.js";
import Spherize_default from "./Spherize.js";
import Spiral_default from "./Spiral.js";
import Star_default from "./Star.js";
import Strands_default from "./Strands.js";
import Stretch_default from "./Stretch.js";
import Stripes_default from "./Stripes.js";
import StudioBackground_default from "./StudioBackground.js";
import SunBurst_default from "./SunBurst.js";
import Swirl_default from "./Swirl.js";
import TiltShift_default from "./TiltShift.js";
import Tint_default from "./Tint.js";
import Trapezoid_default from "./Trapezoid.js";
import Tritone_default from "./Tritone.js";
import Truchet_default from "./Truchet.js";
import Twirl_default from "./Twirl.js";
import VHS_default from "./VHS.js";
import Vesica_default from "./Vesica.js";
import Vibrance_default from "./Vibrance.js";
import VideoTexture_default from "./VideoTexture.js";
import Vignette_default from "./Vignette.js";
import Voronoi_default from "./Voronoi.js";
import WaveDistortion_default from "./WaveDistortion.js";
import Weave_default from "./Weave.js";
import WebcamTexture_default from "./WebcamTexture.js";
import WorleyNoise_default from "./WorleyNoise.js";
import ZoomBlur_default from "./ZoomBlur.js";
import Shader_default from "./Shader.js";
import { Fragment, computed, createBlock, createCommentVNode, createElementBlock, createVNode, defineComponent, h, mergeProps, onMounted, openBlock, ref, renderList, toDisplayString, watch, withCtx } from "vue";
var _hoisted_1 = {
	key: 0,
	class: "shaders-preview",
	style: {
		"position": "relative",
		"width": "100%",
		"height": "100%"
	}
};
var _hoisted_2 = ["href"];
var RenderComponent = {
	name: "RenderComponent",
	props: [
		"config",
		"componentMap",
		"structureVersion"
	],
	setup(props) {
		return () => {
			const { config, componentMap } = props;
			const Component = componentMap[config.type];
			if (!Component) return null;
			const componentProps = config.props ? { ...config.props } : {};
			const childVNodes = config.children?.map((child, index) => h(RenderComponent, {
				key: `${child.id || "child"}_${index}_v${props.structureVersion || 0}`,
				config: {
					...child,
					renderOrder: child.renderOrder ?? index
				},
				componentMap,
				structureVersion: props.structureVersion
			}));
			const slots = childVNodes && childVNodes.length > 0 ? { default: () => childVNodes } : void 0;
			return h(Component, {
				id: config.id,
				...componentProps,
				renderOrder: config.renderOrder
			}, slots);
		};
	}
};
var Preview_vue_vue_type_script_setup_true_lang_default = /* @__PURE__ */ defineComponent({
	__name: "Preview",
	props: {
		shader: {},
		presetId: {},
		apiBaseUrl: { default: "https://shaders.com" },
		obfuscationKey: { default: "shaders-preview-key" },
		watermarkText: { default: "Unlock your Shaders Pro license" },
		watermarkLink: { default: "https://shaders.com/dashboard?pricing=true" },
		configuration: {},
		version: {}
	},
	setup(__props) {
		const componentMap = {
			AngularBlur: AngularBlur_default,
			Ascii: Ascii_default,
			Aurora: Aurora_default,
			BarShift: BarShift_default,
			Beam: Beam_default,
			Blob: Blob_default,
			Blur: Blur_default,
			BrickPattern: BrickPattern_default,
			BrightnessContrast: BrightnessContrast_default,
			Bulge: Bulge_default,
			CRTScreen: CRTScreen_default,
			ChannelBlur: ChannelBlur_default,
			Checkerboard: Checkerboard_default,
			Chevron: Chevron_default,
			ChromaFlow: ChromaFlow_default,
			ChromaticAberration: ChromaticAberration_default,
			Circle: Circle_default,
			ColorWheel: ColorWheel_default,
			ConcentricSpin: ConcentricSpin_default,
			ConicGradient: ConicGradient_default,
			ContourLines: ContourLines_default,
			Crescent: Crescent_default,
			Cross: Cross_default,
			Crystal: Crystal_default,
			CursorRipples: CursorRipples_default,
			CursorTrail: CursorTrail_default,
			DOMTexture: DOMTexture_default,
			DiamondGradient: DiamondGradient_default,
			DiffuseBlur: DiffuseBlur_default,
			Dither: Dither_default,
			DotGrid: DotGrid_default,
			DropShadow: DropShadow_default,
			Duotone: Duotone_default,
			Ellipse: Ellipse_default,
			Emboss: Emboss_default,
			FallingLines: FallingLines_default,
			FilmGrain: FilmGrain_default,
			FloatingParticles: FloatingParticles_default,
			FlowField: FlowField_default,
			Flower: Flower_default,
			FlowingGradient: FlowingGradient_default,
			FlutedGlass: FlutedGlass_default,
			Fog: Fog_default,
			Form3D: Form3D_default,
			FractalNoise: FractalNoise_default,
			Glass: Glass_default,
			GlassTiles: GlassTiles_default,
			Glitch: Glitch_default,
			Glow: Glow_default,
			Godrays: Godrays_default,
			Grayscale: Grayscale_default,
			Grid: Grid_default,
			GridDistortion: GridDistortion_default,
			Group: Group_default,
			Halftone: Halftone_default,
			HexGrid: HexGrid_default,
			HueShift: HueShift_default,
			ImageTexture: ImageTexture_default,
			Invert: Invert_default,
			Kaleidoscope: Kaleidoscope_default,
			LensFlare: LensFlare_default,
			LinearBlur: LinearBlur_default,
			LinearGradient: LinearGradient_default,
			Liquify: Liquify_default,
			Marble: Marble_default,
			Mirror: Mirror_default,
			MultiPointGradient: MultiPointGradient_default,
			Neon: Neon_default,
			Paper: Paper_default,
			Perspective: Perspective_default,
			Pixelate: Pixelate_default,
			Plasma: Plasma_default,
			PolarCoordinates: PolarCoordinates_default,
			Polygon: Polygon_default,
			Posterize: Posterize_default,
			ProgressiveBlur: ProgressiveBlur_default,
			RadialGradient: RadialGradient_default,
			RectangularCoordinates: RectangularCoordinates_default,
			ReflectivePlane: ReflectivePlane_default,
			Ring: Ring_default,
			Ripples: Ripples_default,
			RoundedRect: RoundedRect_default,
			Saturation: Saturation_default,
			Sharpness: Sharpness_default,
			Shatter: Shatter_default,
			SimplexNoise: SimplexNoise_default,
			SineWave: SineWave_default,
			Smoke: Smoke_default,
			SmokeFill: SmokeFill_default,
			Solarize: Solarize_default,
			SolidColor: SolidColor_default,
			Spherize: Spherize_default,
			Spiral: Spiral_default,
			Star: Star_default,
			Strands: Strands_default,
			Stretch: Stretch_default,
			Stripes: Stripes_default,
			StudioBackground: StudioBackground_default,
			SunBurst: SunBurst_default,
			Swirl: Swirl_default,
			TiltShift: TiltShift_default,
			Tint: Tint_default,
			Trapezoid: Trapezoid_default,
			Tritone: Tritone_default,
			Truchet: Truchet_default,
			Twirl: Twirl_default,
			VHS: VHS_default,
			Vesica: Vesica_default,
			Vibrance: Vibrance_default,
			VideoTexture: VideoTexture_default,
			Vignette: Vignette_default,
			Voronoi: Voronoi_default,
			WaveDistortion: WaveDistortion_default,
			Weave: Weave_default,
			WebcamTexture: WebcamTexture_default,
			WorleyNoise: WorleyNoise_default,
			ZoomBlur: ZoomBlur_default
		};
		function xorCrypt(data, key) {
			const keyBytes = new TextEncoder().encode(key);
			const result = new Uint8Array(data.length);
			for (let i = 0; i < data.length; i++) result[i] = data[i] ^ keyBytes[i % keyBytes.length];
			return result;
		}
		function decodePreviewDefinition(encoded, key) {
			const binaryString = atob(encoded);
			const encrypted = new Uint8Array(binaryString.length);
			for (let i = 0; i < binaryString.length; i++) encrypted[i] = binaryString.charCodeAt(i);
			const decrypted = xorCrypt(encrypted, key);
			const json = new TextDecoder().decode(decrypted);
			try {
				return JSON.parse(json);
			} catch {
				throw new Error("Failed to decode shader definition: invalid data or incorrect key");
			}
		}
		const RESERVED_IDENTIFIERS = new Set([
			"break",
			"case",
			"catch",
			"class",
			"const",
			"continue",
			"default",
			"delete",
			"do",
			"else",
			"export",
			"extends",
			"finally",
			"for",
			"function",
			"if",
			"import",
			"in",
			"instanceof",
			"new",
			"return",
			"super",
			"switch",
			"this",
			"throw",
			"try",
			"typeof",
			"var",
			"void",
			"while",
			"with",
			"yield"
		]);
		function slugifyIdentifier(label, fallback) {
			const cleaned = label.replace(/[^A-Za-z0-9]+/g, " ").trim();
			if (!cleaned) return fallback;
			const camel = cleaned.split(/\s+/).map((part, i) => i === 0 ? part.charAt(0).toLowerCase() + part.slice(1) : part.charAt(0).toUpperCase() + part.slice(1)).join("");
			const safe = /^[A-Za-z_$]/.test(camel) ? camel : `_${camel}`;
			return RESERVED_IDENTIFIERS.has(safe) ? `_${safe}` : safe;
		}
		function buildKeyPropIdentifierMap(keyProps) {
			const used = new Set(["style"]);
			const map = /* @__PURE__ */ new Map();
			if (!keyProps?.length) return map;
			keyProps.forEach((kp, i) => {
				if (!kp.targets?.length) return;
				const base = slugifyIdentifier(kp.label, `keyProp${i}`);
				let ident = base;
				let suffix = 1;
				while (used.has(ident)) {
					suffix++;
					ident = `${base}${suffix}`;
				}
				used.add(ident);
				map.set(ident, kp);
			});
			return map;
		}
		function findInTree(id, list) {
			for (const c of list) {
				if (c.id === id) return c;
				if (c.children?.length) {
					const hit = findInTree(id, c.children);
					if (hit) return hit;
				}
			}
			return null;
		}
		function applyConfiguration(definition, configuration, keyProps) {
			if (!configuration || !keyProps?.length) return definition;
			const entries = Object.entries(configuration);
			if (entries.length === 0) return definition;
			const identMap = buildKeyPropIdentifierMap(keyProps);
			const cloned = JSON.parse(JSON.stringify(definition.components));
			for (const [ident, value] of entries) {
				const kp = identMap.get(ident);
				if (!kp) continue;
				for (const t of kp.targets ?? []) {
					const comp = findInTree(t.component_id, cloned);
					if (comp?.props && t.prop in comp.props) comp.props[t.prop] = value;
				}
			}
			return {
				...definition,
				components: cloned
			};
		}
		const props = __props;
		const fetchedDefinition = ref(null);
		const fetchedKeyProps = ref(null);
		const resolvedDefinition = computed(() => {
			if (!fetchedDefinition.value) return null;
			return applyConfiguration(fetchedDefinition.value, props.configuration, fetchedKeyProps.value);
		});
		async function fetchAndDecode(url) {
			try {
				const response = await fetch(url);
				if (!response.ok) throw new Error(`Failed to fetch: ${response.status}`);
				const data = await response.json();
				const item = data.preset || data.shader;
				if (item?.definition && typeof item.definition === "string") {
					fetchedDefinition.value = decodePreviewDefinition(item.definition, props.obfuscationKey);
					fetchedKeyProps.value = Array.isArray(item.key_props) ? item.key_props : null;
				}
			} catch (e) {
				console.error("[Shaders Preview] Failed to fetch preview:", e);
			}
		}
		watch(() => props.shader, (token) => {
			if (token) {
				fetchedDefinition.value = null;
				fetchedKeyProps.value = null;
				fetchAndDecode(`${props.apiBaseUrl}/api/preview/shader/${token}`);
			} else if (!props.presetId) {
				fetchedDefinition.value = null;
				fetchedKeyProps.value = null;
			}
		}, { immediate: true });
		watch(() => [props.presetId, props.version], ([id, version]) => {
			if (id) {
				fetchedDefinition.value = null;
				fetchedKeyProps.value = null;
				const versionParam = version ? `?version=${encodeURIComponent(version)}` : "";
				fetchAndDecode(`${props.apiBaseUrl}/api/preview/preset/${id}${versionParam}`);
			} else if (!props.shader) {
				fetchedDefinition.value = null;
				fetchedKeyProps.value = null;
			}
		}, { immediate: true });
		onMounted(() => {
			console.warn("[Shaders] Preview component is intended for use with a Shaders license. Visit https://shaders.com for more information.");
		});
		return (_ctx, _cache) => {
			return resolvedDefinition.value ? (openBlock(), createElementBlock("div", _hoisted_1, [createVNode(Shader_default, mergeProps({
				"color-space": resolvedDefinition.value.colorSpace || "p3-linear",
				"enable-performance-tracking": true,
				"is-preview": true,
				style: {
					"width": "100%",
					"height": "100%"
				}
			}, _ctx.$attrs), {
				default: withCtx(() => [(openBlock(true), createElementBlock(Fragment, null, renderList(resolvedDefinition.value.components, (config, index) => {
					return openBlock(), createBlock(RenderComponent, {
						key: `${config.id || "root"}_${index}_v${resolvedDefinition.value.structureVersion || 0}`,
						config: {
							...config,
							renderOrder: config.renderOrder ?? index
						},
						"component-map": componentMap,
						"structure-version": resolvedDefinition.value.structureVersion
					}, null, 8, ["config", "structure-version"]);
				}), 128))]),
				_: 1
			}, 16, ["color-space"]), __props.watermarkText ? (openBlock(), createElementBlock("a", {
				key: 0,
				href: __props.watermarkLink,
				target: "_blank",
				rel: "noopener noreferrer",
				style: {
					"position": "absolute",
					"bottom": "8px",
					"right": "12px",
					"font-size": "11px",
					"font-family": "-apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif",
					"color": "rgba(255, 255, 255, 0.5)",
					"text-decoration": "none",
					"pointer-events": "auto",
					"z-index": "10",
					"text-shadow": "0 1px 3px rgba(0, 0, 0, 0.4)",
					"transition": "color 0.2s ease"
				},
				onMouseenter: _cache[0] || (_cache[0] = ($event) => $event.target.style.color = "rgba(255, 255, 255, 0.8)"),
				onMouseleave: _cache[1] || (_cache[1] = ($event) => $event.target.style.color = "rgba(255, 255, 255, 0.5)")
			}, toDisplayString(__props.watermarkText), 41, _hoisted_2)) : createCommentVNode("", true)])) : createCommentVNode("", true);
		};
	}
});
export { Preview_vue_vue_type_script_setup_true_lang_default as default };
