// shaders/dist/vue/Shader.vue_vue_type_script_setup_true_lang.js
import { computed, createElementBlock, createElementVNode, defineComponent, mergeProps, onBeforeUnmount, onMounted, openBlock, provide, ref, renderSlot, shallowRef, useId, watch } from "vue";
import { shaderRenderer } from "../core/index.js";
import { setColorSpaceMode } from "../core/utilities/transformations/index.js";
import { vec4 } from "three/tsl";
import { isExternalUser, startTelemetry } from "../core/telemetry/index.js";
var Shader_vue_vue_type_script_setup_true_lang_default = /* @__PURE__ */ defineComponent({
	__name: "Shader",
	props: {
		disableTelemetry: {
			type: Boolean,
			default: false
		},
		enablePerformanceTracking: {
			type: Boolean,
			default: true
		},
		colorSpace: { default: "p3-linear" },
		isPreview: {
			type: Boolean,
			default: false
		}
	},
	emits: ["ready"],
	setup(__props, { expose: __expose, emit: __emit }) {
		const props = __props;
		const emit = __emit;
		provide("shaderColorSpace", computed(() => props.colorSpace));
		const containerRef = ref(null);
		const canvasRef = ref(null);
		const rootId = ref("shader-root-" + useId().replace(/[^a-zA-Z0-9_]/g, ""));
		const rendererInstance = shallowRef(shaderRenderer());
		rendererInstance.value.setOnReady(() => emit("ready"));
		const rendererResetSignal = ref(0);
		let telemetryCollector = null;
		let telemetryStartTimeout = null;
		let shouldSendTelemetry = null;
		provide("shaderParentId", rootId.value);
		provide("shaderRendererResetSignal", rendererResetSignal);
		provide("shaderNodeRegister", (id, fragmentNodeFunc, parentId, metadata, uniforms = null, componentDefinition = null, domCanvas = void 0) => {
			if (fragmentNodeFunc === null) rendererInstance.value.removeNode(id);
			else rendererInstance.value.registerNode(id, fragmentNodeFunc, parentId, metadata, uniforms, componentDefinition, domCanvas);
		});
		provide("shaderUniformUpdate", (nodeId, uniformName, value) => {
			rendererInstance.value.updateUniformValue(nodeId, uniformName, value);
		});
		provide("shaderMetadataUpdate", (nodeId, metadata) => {
			rendererInstance.value.updateNodeMetadata(nodeId, metadata);
		});
		const initializeRenderer = async () => {
			if (!canvasRef.value) return;
			if (!rendererInstance.value.isInitialized()) await rendererInstance.value.initialize({
				canvas: canvasRef.value,
				enablePerformanceTracking: props.enablePerformanceTracking,
				colorSpace: props.colorSpace
			});
			rendererInstance.value.registerNode(rootId.value, ({ childNode }) => childNode || vec4(0, 0, 0, 0), null, null, {}, void 0);
			if (shouldSendTelemetry === null) shouldSendTelemetry = isExternalUser();
			if (shouldSendTelemetry && !telemetryCollector) startTelemetryWhenReady();
		};
		const startTelemetryWhenReady = () => {
			const checkRendering = () => {
				if (rendererInstance.value.getPerformanceStats().fps > 0) {
					telemetryCollector = startTelemetry(rendererInstance.value, "2.5.124", props.disableTelemetry, props.isPreview);
					if (telemetryCollector) telemetryCollector.start();
					telemetryStartTimeout = null;
				} else telemetryStartTimeout = setTimeout(checkRendering, 500);
			};
			telemetryStartTimeout = setTimeout(checkRendering, 500);
		};
		let wasVisible = false;
		let visibilityObserver = null;
		const setupVisibilityObserver = () => {
			if (!containerRef.value || visibilityObserver) return;
			visibilityObserver = new IntersectionObserver((entries) => {
				const entry = entries[0];
				if (!entry) return;
				const isCurrentlyVisible = entry.isIntersecting;
				if (isCurrentlyVisible && !wasVisible) {
					if (rendererInstance.value.isInitialized()) {
						rendererInstance.value.startAnimation();
						if (shouldSendTelemetry && !telemetryCollector && !telemetryStartTimeout) startTelemetryWhenReady();
					} else initializeRenderer().then(() => {
						rendererResetSignal.value++;
					}).catch(console.error);
					wasVisible = true;
				} else if (!isCurrentlyVisible && wasVisible) {
					rendererInstance.value.stopAnimation();
					wasVisible = false;
				}
			}, { threshold: 0 });
			visibilityObserver.observe(containerRef.value);
		};
		onMounted(async () => {
			if (canvasRef.value && containerRef.value) {
				const rect = containerRef.value.getBoundingClientRect();
				if (rect.width > 0 && rect.height > 0) {
					await initializeRenderer();
					rendererResetSignal.value++;
					wasVisible = true;
				} else wasVisible = false;
				setupVisibilityObserver();
			}
		});
		let isFirstRun = true;
		watch(() => props.colorSpace, (newColorSpace) => {
			if (isFirstRun) {
				isFirstRun = false;
				return;
			}
			if (rendererInstance.value.isInitialized()) {
				setColorSpaceMode(newColorSpace);
				rendererResetSignal.value++;
			}
		});
		const renderForCapture = async () => {
			try {
				await rendererInstance.value.renderAndWait();
			} catch (error) {
				console.warn("[Shaders] Failed to force render before capture:", error);
			}
		};
		const canvasToBlob = async (source, targetWidth, targetHeight, mimeType, quality) => {
			const sourceBlob = await new Promise((resolve, reject) => {
				source.toBlob((blob) => {
					if (blob) resolve(blob);
					else reject(/* @__PURE__ */ new Error("Failed to read source canvas"));
				}, "image/png");
			});
			if (!(mimeType !== "image/png" || targetWidth !== source.width || targetHeight !== source.height)) return sourceBlob;
			const bitmap = await createImageBitmap(sourceBlob);
			try {
				const tempCanvas = document.createElement("canvas");
				tempCanvas.width = targetWidth;
				tempCanvas.height = targetHeight;
				const ctx = tempCanvas.getContext("2d");
				if (!ctx) throw new Error("Failed to get canvas context");
				ctx.imageSmoothingEnabled = true;
				ctx.imageSmoothingQuality = "high";
				ctx.drawImage(bitmap, 0, 0, targetWidth, targetHeight);
				return await new Promise((resolve, reject) => {
					tempCanvas.toBlob((blob) => {
						if (blob) resolve(blob);
						else reject(/* @__PURE__ */ new Error("Failed to encode image"));
					}, mimeType, quality);
				});
			} finally {
				bitmap.close();
			}
		};
		const captureImage = async (options = {}) => {
			if (!canvasRef.value) throw new Error("Canvas not available");
			await renderForCapture();
			const source = canvasRef.value;
			const format = options.format ?? "png";
			const scale = Math.max(.1, options.scale ?? 1);
			return canvasToBlob(source, Math.round(source.width * scale), Math.round(source.height * scale), format === "png" ? "image/png" : format === "webp" ? "image/webp" : "image/jpeg", format === "png" ? void 0 : options.quality ?? .95);
		};
		const captureScreenshot = async (maxWidth = 1200) => {
			if (!canvasRef.value) throw new Error("Canvas not available");
			await renderForCapture();
			const source = canvasRef.value;
			const sourceWidth = source.width;
			const sourceHeight = source.height;
			let targetWidth = sourceWidth;
			let targetHeight = sourceHeight;
			if (sourceWidth > maxWidth) {
				targetWidth = maxWidth;
				targetHeight = Math.round(sourceHeight / sourceWidth * maxWidth);
			}
			return canvasToBlob(source, targetWidth, targetHeight, "image/jpeg", .85);
		};
		const getPerformanceStats = () => {
			return rendererInstance.value.getPerformanceStats();
		};
		const beginRecordingResolution = async (pixelRatio) => {
			const internal = rendererInstance.value.getInternalRenderer();
			if (!internal || !canvasRef.value) return async () => {};
			const previousRatio = typeof internal.getPixelRatio === "function" ? internal.getPixelRatio() : Math.min(window.devicePixelRatio, 2);
			const cssWidth = canvasRef.value.clientWidth;
			const cssHeight = canvasRef.value.clientHeight;
			const applyRatio = async (ratio) => {
				internal.setPixelRatio(ratio);
				if (typeof internal.setSize === "function") internal.setSize(cssWidth, cssHeight, false);
				await renderForCapture();
			};
			const previousForce = rendererInstance.value.setForceFullFrameRate?.(true) ?? false;
			await applyRatio(pixelRatio);
			return async () => {
				await applyRatio(previousRatio);
				rendererInstance.value.setForceFullFrameRate?.(previousForce);
			};
		};
		__expose({
			captureImage,
			captureScreenshot,
			getPerformanceStats,
			beginRecordingResolution,
			getCanvas: () => canvasRef.value
		});
		onBeforeUnmount(() => {
			if (telemetryCollector) {
				telemetryCollector.stop();
				telemetryCollector = null;
			}
			if (telemetryStartTimeout !== null) {
				clearTimeout(telemetryStartTimeout);
				telemetryStartTimeout = null;
			}
			if (visibilityObserver) {
				visibilityObserver.disconnect();
				visibilityObserver = null;
			}
			rendererInstance.value.cleanup();
		});
		return (_ctx, _cache) => {
			return openBlock(), createElementBlock("div", mergeProps({
				class: "shader",
				ref_key: "containerRef",
				ref: containerRef
			}, _ctx.$attrs), [createElementVNode("canvas", {
				"data-renderer": "shaders",
				ref_key: "canvasRef",
				ref: canvasRef,
				style: {
					"width": "100%",
					"height": "100%",
					"display": "block"
				}
			}, [renderSlot(_ctx.$slots, "default")], 512)], 16);
		};
	}
});
export { Shader_vue_vue_type_script_setup_true_lang_default as default };
