// shaders/dist/vue/WorleyNoise.vue_vue_type_script_setup_true_lang.js
import { Fragment, Teleport, computed, createBlock, createCommentVNode, createElementBlock, createElementVNode, defineComponent, effectScope, inject, mergeDefaults, mergeProps, onBeforeUnmount, onMounted, openBlock, provide, ref, renderSlot, toValue, unref, useId, watch } from "vue";
import { createUniformsMap } from "../core/index.js";
import { setColorSpaceMode } from "../core/utilities/transformations/index.js";
import { componentDefinition } from "../core/shaders/WorleyNoise/index.js";
var _hoisted_1 = ["data-shader-id"];
var _hoisted_2 = ["width", "height"];
var WorleyNoise_vue_vue_type_script_setup_true_lang_default = /* @__PURE__ */ defineComponent({
	__name: "WorleyNoise",
	props: /* @__PURE__ */ mergeDefaults({
		scale: {},
		lacunarity: {},
		persistence: {},
		jitter: {},
		contrast: {},
		balance: {},
		seed: {},
		blendMode: {},
		opacity: {},
		visible: { type: Boolean },
		id: {},
		maskSource: {},
		maskType: {},
		renderOrder: {},
		transform: {},
		colorA: {},
		colorB: {},
		colorSpace: {},
		mode: {},
		distance: {},
		octaves: {},
		speed: {}
	}, {
		blendMode: "normal",
		visible: true,
		...Object.entries(componentDefinition.props).reduce((acc, [key, config]) => {
			acc[key] = config.default;
			return acc;
		}, {})
	}),
	setup(__props) {
		let _experimentalWarnedOnce = false;
		if (componentDefinition.experimental && !_experimentalWarnedOnce) {
			_experimentalWarnedOnce = true;
			const _e = componentDefinition.experimental;
			console.info(`%c⚠ [Shaders] ${componentDefinition.name} is experimental: ${_e.message}`, "color: #f59e0b; font-weight: bold");
		}
		function isPropDriver(value) {
			return typeof value === "object" && value !== null && "type" in value && (value.type === "map" || value.type === "mouse" || value.type === "mouse-position" || value.type === "auto-animate");
		}
		const props = __props;
		const instanceId = (props.id ? props.id.replace(/[^a-zA-Z0-9_]/g, "_") : null) || useId().replace(/[^a-zA-Z0-9_]/g, "");
		provide("shaderParentId", instanceId);
		const DEFAULT_TRANSFORM = {
			offsetX: 0,
			offsetY: 0,
			rotation: 0,
			scale: 1,
			anchorX: .5,
			anchorY: .5,
			edges: "transparent"
		};
		const effectiveTransform = computed(() => ({
			...DEFAULT_TRANSFORM,
			...props.transform
		}));
		const shaderReadyProps = computed(() => {
			let baseProps = { ...Object.entries(componentDefinition.props).reduce((acc, [key, config]) => {
				acc[key] = config.default;
				return acc;
			}, {}) };
			for (const key in props) if (key !== "blendMode" && key !== "opacity" && key !== "visible" && key !== "id" && key !== "maskSource" && key !== "maskType" && key !== "renderOrder" && key !== "transform") {
				const val = toValue(props[key]);
				if (isPropDriver(val)) continue;
				if (val !== Object.entries(componentDefinition.props).reduce((acc, [key$1, config]) => {
					acc[key$1] = config.default;
					return acc;
				}, {})[key]) baseProps[key] = val;
			}
			return baseProps;
		});
		const mapsFromProps = computed(() => {
			const maps = {};
			for (const key of Object.keys(componentDefinition.props)) {
				const val = props[key];
				if (isPropDriver(val)) maps[key] = val;
			}
			return Object.keys(maps).length > 0 ? maps : void 0;
		});
		const shaderColorSpace = inject("shaderColorSpace");
		if (shaderColorSpace) setColorSpaceMode(shaderColorSpace.value);
		const uniforms = createUniformsMap(componentDefinition, shaderReadyProps.value, instanceId);
		const parentId = inject("shaderParentId");
		if (parentId === void 0) throw new Error("Shader components must be used inside an <Shader> component or another shader component");
		const parentRegister = inject("shaderNodeRegister");
		if (parentRegister === void 0) throw new Error("Shader components must be used inside an <Shader> component or another shader component");
		const parentUniformUpdate = inject("shaderUniformUpdate");
		if (parentUniformUpdate === void 0) throw new Error("Shader components require shaderUniformUpdate from parent");
		const parentMetadataUpdate = inject("shaderMetadataUpdate");
		if (parentMetadataUpdate === void 0) throw new Error("Shader components require shaderMetadataUpdate from parent");
		const rendererResetSignal = inject("shaderRendererResetSignal");
		const orderMarker = ref(null);
		const isCapturesDOM = !!componentDefinition.capturesDOM;
		const captureCanvas = ref(null);
		const captureW = ref(typeof window !== "undefined" ? Math.round(window.innerWidth * Math.min(window.devicePixelRatio, 2)) : 0);
		const captureH = ref(typeof window !== "undefined" ? Math.round(window.innerHeight * Math.min(window.devicePixelRatio, 2)) : 0);
		const captureStyle = {
			position: "fixed",
			inset: "0",
			width: "100vw",
			height: "100vh",
			zIndex: "-9999"
		};
		if (isCapturesDOM) {
			const onWinResize = () => {
				const d = Math.min(window.devicePixelRatio, 2);
				captureW.value = Math.round(window.innerWidth * d);
				captureH.value = Math.round(window.innerHeight * d);
			};
			onMounted(() => {
				const d = Math.min(window.devicePixelRatio, 2);
				captureW.value = Math.round(window.innerWidth * d);
				captureH.value = Math.round(window.innerHeight * d);
				window.addEventListener("resize", onWinResize);
			});
			onBeforeUnmount(() => {
				window.removeEventListener("resize", onWinResize);
			});
		}
		let detectedRenderOrder = void 0;
		const registerWithRenderer = () => {
			parentRegister(instanceId, componentDefinition.fragmentNode, parentId, {
				blendMode: props.blendMode,
				opacity: props.opacity,
				visible: props.visible === false ? false : true,
				id: props.id,
				mask: props.maskSource ? {
					source: props.maskSource,
					type: props.maskType || "alpha"
				} : void 0,
				maps: mapsFromProps.value,
				renderOrder: props.renderOrder ?? detectedRenderOrder,
				transform: effectiveTransform.value
			}, uniforms, componentDefinition, isCapturesDOM ? captureCanvas.value ?? void 0 : void 0);
		};
		onMounted(() => {
			if (props.renderOrder === void 0 && orderMarker.value) {
				const parent = orderMarker.value.parentElement;
				if (parent) {
					const siblings = parent.querySelectorAll(":scope > [data-shader-id]");
					const position = Array.from(siblings).indexOf(orderMarker.value);
					if (position >= 0) {
						detectedRenderOrder = position;
						parentMetadataUpdate(instanceId, { renderOrder: position });
					}
				}
			}
		});
		const setupUniformWatchers = () => {
			const scope = effectScope();
			scope.run(() => {
				Object.entries(uniforms).forEach(([propName, { uniform, transform }]) => {
					watch(() => props[propName], (newValue) => {
						if (uniform && uniform.value !== void 0 && !isPropDriver(newValue)) parentUniformUpdate(instanceId, propName, newValue);
					}, { deep: true });
				});
				watch(() => [
					props.blendMode,
					props.opacity,
					props.visible,
					props.id,
					props.maskSource,
					props.maskType,
					props.renderOrder,
					props.transform,
					mapsFromProps.value
				], ([blendMode, opacity, visible]) => {
					parentMetadataUpdate(instanceId, {
						blendMode,
						opacity,
						visible: visible === false ? false : true,
						id: props.id,
						mask: props.maskSource ? {
							source: props.maskSource,
							type: props.maskType || "alpha"
						} : void 0,
						maps: mapsFromProps.value,
						renderOrder: props.renderOrder ?? detectedRenderOrder,
						transform: effectiveTransform.value
					});
				}, { deep: true });
				if (rendererResetSignal) watch(() => rendererResetSignal.value, (newValue) => {
					if (newValue > 0) registerWithRenderer();
				}, { immediate: true });
			});
			onBeforeUnmount(() => {
				scope.stop();
			});
		};
		if (isCapturesDOM) onMounted(() => {
			setupUniformWatchers();
		});
		else setupUniformWatchers();
		onBeforeUnmount(() => {
			parentRegister(instanceId, null, null, null, null, null);
		});
		return (_ctx, _cache) => {
			return openBlock(), createElementBlock(Fragment, null, [createElementVNode("span", {
				ref_key: "orderMarker",
				ref: orderMarker,
				style: { "display": "contents" },
				"data-shader-id": unref(instanceId)
			}, [!isCapturesDOM ? renderSlot(_ctx.$slots, "default", { key: 0 }) : createCommentVNode("", true)], 8, _hoisted_1), isCapturesDOM ? (openBlock(), createBlock(Teleport, {
				key: 0,
				to: "body"
			}, [createElementVNode("canvas", mergeProps({
				ref_key: "captureCanvas",
				ref: captureCanvas
			}, { "layoutsubtree": "" }, {
				width: captureW.value,
				height: captureH.value,
				style: captureStyle
			}), [renderSlot(_ctx.$slots, "default")], 16, _hoisted_2)])) : createCommentVNode("", true)], 64);
		};
	}
});
export { WorleyNoise_vue_vue_type_script_setup_true_lang_default as default };
