// shaders/dist/solid/utils/generatePresetCode.js
const DEFAULT_TRANSFORM = {
  offsetX: 0,
  offsetY: 0,
  rotation: 0,
  scale: 1,
  anchorX: 0.5,
  anchorY: 0.5,
  edges: "transparent"
};
function toObjectLiteral(obj) {
  const entries = Object.entries(obj).map(([key, value]) => {
    const valueStr = typeof value === "string" ? `"${value}"` : value;
    return `${key}: ${valueStr}`;
  });
  return `{ ${entries.join(", ")} }`;
}
const shaderMetadata = {
  "AngularBlur": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 20,
    "center": {
      "x": 0.5,
      "y": 0.5
    }
  },
  "Ascii": {
    "opacity": 1,
    "blendMode": "normal",
    "characters": "@%#*+=-:.",
    "cellSize": 30,
    "fontFamily": "JetBrains Mono",
    "spacing": 1,
    "gamma": 1,
    "alphaThreshold": 0,
    "preserveAlpha": true
  },
  "Aurora": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#a533f8",
    "colorB": "#22ee88",
    "colorC": "#1694e8",
    "colorSpace": "linear",
    "balance": 50,
    "intensity": 80,
    "curtainCount": 4,
    "speed": 5,
    "waviness": 50,
    "rayDensity": 20,
    "height": 120,
    "center": {
      "x": 0.5,
      "y": 0
    },
    "seed": 0
  },
  "BarShift": {
    "opacity": 1,
    "blendMode": "normal",
    "count": 6,
    "angle": 0,
    "intensity": 0.15,
    "seed": 0,
    "speed": 0,
    "edges": "mirror"
  },
  "Beam": {
    "opacity": 1,
    "blendMode": "normal",
    "startPosition": {
      "x": 0.2,
      "y": 0.5
    },
    "endPosition": {
      "x": 0.8,
      "y": 0.5
    },
    "startThickness": 0.2,
    "endThickness": 0.2,
    "startSoftness": 0.5,
    "endSoftness": 0.5,
    "insideColor": "#FF0000",
    "outsideColor": "#0000FF",
    "colorSpace": "linear"
  },
  "Blob": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#ff6b35",
    "colorB": "#e91e63",
    "size": 0.5,
    "deformation": 0.5,
    "softness": 0.5,
    "highlightIntensity": 0.5,
    "highlightX": 0.3,
    "highlightY": -0.3,
    "highlightZ": 0.4,
    "highlightColor": "#ffe11a",
    "speed": 0.5,
    "seed": 1,
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "colorSpace": "linear"
  },
  "Blur": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 50
  },
  "BrickPattern": {
    "opacity": 1,
    "blendMode": "normal",
    "colorBrick": "#000000",
    "colorMortar": "#ffffff",
    "cellsX": 8,
    "cellsY": 10,
    "mortar": 0.05,
    "angle": 0,
    "speed": 0,
    "offset": 0,
    "speedVariance": 0,
    "seed": 0,
    "colorSpace": "linear"
  },
  "BrightnessContrast": {
    "opacity": 1,
    "blendMode": "normal",
    "brightness": 0,
    "contrast": 0
  },
  "Bulge": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "strength": 1,
    "radius": 1,
    "falloff": 0.5,
    "edges": "stretch"
  },
  "ChannelBlur": {
    "opacity": 1,
    "blendMode": "normal",
    "redIntensity": 0,
    "greenIntensity": 20,
    "blueIntensity": 40
  },
  "Checkerboard": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#cccccc",
    "colorB": "#999999",
    "cells": 8,
    "softness": 0,
    "colorSpace": "linear"
  },
  "Chevron": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#000000",
    "colorB": "#ffffff",
    "count": 5,
    "angle": 0,
    "balance": 0.5,
    "softness": 0,
    "speed": 0,
    "offset": 0,
    "colorSpace": "linear"
  },
  "ChromaFlow": {
    "opacity": 1,
    "blendMode": "normal",
    "baseColor": "#0066ff",
    "upColor": "#00ff00",
    "downColor": "#ff0000",
    "leftColor": "#0000ff",
    "rightColor": "#ffff00",
    "intensity": 1,
    "radius": 3,
    "momentum": 30
  },
  "ChromaticAberration": {
    "opacity": 1,
    "blendMode": "normal",
    "strength": 0.2,
    "angle": 0,
    "redOffset": -1,
    "greenOffset": 0,
    "blueOffset": 1
  },
  "Circle": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "radius": 1,
    "softness": 0,
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "strokeThickness": 0,
    "strokeColor": "#000000",
    "strokePosition": "center",
    "colorSpace": "linear"
  },
  "ColorWheel": {
    "opacity": 1,
    "blendMode": "normal",
    "mode": "rainbow",
    "colorA": "#ff0000",
    "colorB": "#00ff88",
    "colorC": "#0066ff",
    "scale": 1,
    "angle": 0,
    "speed": 0.05,
    "colorSpace": "oklch"
  },
  "ConcentricSpin": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 20,
    "rings": 8,
    "smoothness": 0.03,
    "seed": 0,
    "speed": 0.1,
    "speedRandomness": 0.5,
    "edges": "mirror",
    "center": {
      "x": 0.5,
      "y": 0.5
    }
  },
  "ConicGradient": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#FF0080",
    "colorB": "#00BFFF",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "rotation": 0,
    "repeat": 1,
    "colorSpace": "oklch"
  },
  "ContourLines": {
    "opacity": 1,
    "blendMode": "normal",
    "levels": 5,
    "lineWidth": 2,
    "softness": 0,
    "gamma": 0.5,
    "invert": false,
    "source": "luminance",
    "colorMode": "source",
    "lineColor": "#000000",
    "backgroundColor": "transparent"
  },
  "Crescent": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "radius": 0.3,
    "innerRatio": 0.8,
    "offset": 0.2,
    "rotation": 0,
    "softness": 0,
    "strokeThickness": 0,
    "strokeColor": "#000000",
    "strokePosition": "center",
    "colorSpace": "linear"
  },
  "Cross": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "radius": 0.35,
    "thickness": 0.08,
    "rounding": 0,
    "rotation": 0,
    "softness": 0,
    "strokeThickness": 0,
    "strokeColor": "#000000",
    "strokePosition": "center",
    "colorSpace": "linear"
  },
  "CRTScreen": {
    "opacity": 1,
    "blendMode": "normal",
    "pixelSize": 128,
    "colorShift": 1,
    "scanlineIntensity": 0.3,
    "scanlineFrequency": 200,
    "brightness": 1,
    "contrast": 1,
    "vignetteIntensity": 1,
    "vignetteRadius": 0.5
  },
  "Crystal": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "scale": 1,
    "cutout": false,
    "refraction": 0.5,
    "dispersion": 0.5,
    "facets": 5,
    "fresnel": 0.05,
    "fresnelSoftness": 1,
    "fresnelColor": "#ffffff",
    "edgeSoftness": 0,
    "innerZoom": 1.5,
    "lightAngle": 270,
    "highlights": 0.5,
    "shadows": 0.3,
    "brightness": 1.2,
    "tintColor": "#e8e0ff",
    "tintIntensity": 0,
    "tintPreserveLuminosity": true,
    "shape": '{"type":"polygonSDF","radius":0.35,"sides":10}'
  },
  "CursorRipples": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 10,
    "decay": 10,
    "radius": 0.5,
    "chromaticSplit": 1,
    "edges": "stretch"
  },
  "CursorTrail": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#00aaff",
    "colorB": "#ff00aa",
    "radius": 0.5,
    "length": 0.5,
    "shrink": 1,
    "colorSpace": "linear"
  },
  "DiamondGradient": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#4ffb4a",
    "colorB": "#4f1238",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "size": 0.7,
    "rotation": 0,
    "repeat": 1,
    "roundness": 0,
    "colorSpace": "oklch"
  },
  "DiffuseBlur": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 30,
    "edges": "stretch"
  },
  "Dither": {
    "opacity": 1,
    "blendMode": "normal",
    "pattern": "bayer4",
    "pixelSize": 4,
    "threshold": 0.5,
    "spread": 1,
    "colorMode": "custom",
    "colorA": "transparent",
    "colorB": "#ffffff"
  },
  "DOMTexture": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "DotGrid": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "density": 30,
    "dotSize": 0.3,
    "twinkle": 0
  },
  "DropShadow": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#000000",
    "distance": 0.1,
    "angle": 135,
    "blur": 5,
    "intensity": 0.5,
    "cutout": false
  },
  "Duotone": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#ff0000",
    "colorB": "#023af4",
    "blend": 0.5,
    "colorSpace": "linear"
  },
  "Ellipse": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "radiusX": 0.35,
    "radiusY": 0.2,
    "rotation": 0,
    "softness": 0,
    "strokeThickness": 0,
    "strokeColor": "#000000",
    "strokePosition": "center",
    "colorSpace": "linear"
  },
  "Emboss": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "scale": 1,
    "depth": -0.5,
    "lightAngle": 260,
    "lightIntensity": 0.6,
    "shadowIntensity": 0.3,
    "shape": '{"type":"circleSDF","radius":0.35}'
  },
  "FallingLines": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#ffffff",
    "colorB": "#ffffff00",
    "colorSpace": "linear",
    "angle": 90,
    "speed": 0.5,
    "speedVariance": 0.3,
    "density": 15,
    "trailLength": 0.35,
    "balance": 0.5,
    "strokeWidth": 0.15,
    "rounding": 1
  },
  "FilmGrain": {
    "opacity": 1,
    "blendMode": "normal",
    "strength": 0.5,
    "bias": 2,
    "animated": false
  },
  "FloatingParticles": {
    "opacity": 1,
    "blendMode": "normal",
    "randomness": 0.25,
    "speed": 0.25,
    "angle": 90,
    "particleSize": 2,
    "particleSoftness": 0,
    "twinkle": 0.5,
    "count": 5,
    "particleColor": "#ffffff",
    "speedVariance": 0.3,
    "angleVariance": 30,
    "particleDensity": 3
  },
  "Flower": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "radius": 0.4,
    "sides": 5,
    "innerRatio": 0.4,
    "rotation": 0,
    "softness": 0,
    "strokeThickness": 0,
    "strokeColor": "#000000",
    "strokePosition": "center",
    "colorSpace": "linear"
  },
  "FlowField": {
    "opacity": 1,
    "blendMode": "normal",
    "strength": 0.15,
    "detail": 2,
    "speed": 0,
    "evolutionSpeed": 0,
    "edges": "mirror"
  },
  "FlowingGradient": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#0a0015",
    "colorB": "#6b17e6",
    "colorC": "#ff4d6a",
    "colorD": "#ff6b35",
    "colorSpace": "oklch",
    "speed": 1,
    "distortion": 0.5,
    "seed": 0
  },
  "FlutedGlass": {
    "opacity": 1,
    "blendMode": "normal",
    "shape": "bars",
    "angle": 0,
    "frequency": 10,
    "softness": 0.5,
    "waveAmplitude": 0.06,
    "waveFrequency": 1.5,
    "speed": 0,
    "refraction": 1.5,
    "aberration": 0.2,
    "lightAngle": 30,
    "highlight": 0.2,
    "highlightSoftness": 0.3,
    "highlightColor": "#ffffff",
    "edges": "mirror"
  },
  "Fog": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#e0e0e0",
    "colorB": "#888888",
    "seed": 0,
    "speed": 1,
    "turbulence": 1,
    "detail": 15,
    "blending": 0.3,
    "mouseInfluence": 0.1,
    "mouseRadius": 0.1,
    "colorSpace": "linear"
  },
  "Form3D": {
    "opacity": 1,
    "blendMode": "normal",
    "shape3d": '{"type":"ribbon","angle":0,"twist":50,"width":40,"thickness":20,"seed":0}',
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "zoom": 50,
    "glossiness": 50,
    "lighting": 50,
    "uvMode": "stretch",
    "speed": 1
  },
  "FractalNoise": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#000000",
    "colorB": "#ffffff",
    "octaves": 4,
    "detail": 2,
    "contrast": 0.5,
    "speed": 0.15,
    "angle": 0,
    "seed": 0,
    "colorSpace": "linear"
  },
  "Glass": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "scale": 1,
    "cutout": false,
    "refraction": 1,
    "edgeSoftness": 0.1,
    "blur": 0,
    "thickness": 0.2,
    "aberration": 0.5,
    "innerZoom": 1,
    "lightAngle": 300,
    "highlight": 0.05,
    "highlightColor": "#ffffff",
    "highlightSoftness": 0.5,
    "fresnel": 0.1,
    "fresnelSoftness": 0.1,
    "fresnelColor": "#ffffff",
    "tintColor": "#ffffff",
    "tintIntensity": 0,
    "tintPreserveLuminosity": true,
    "shape": '{"type":"circleSDF","radius":0.35}'
  },
  "GlassTiles": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 2,
    "tileCount": 20,
    "rotation": 0,
    "roundness": 0
  },
  "Glitch": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 0.5,
    "speed": 1,
    "rgbShift": 5,
    "blockDensity": 10,
    "colorBarIntensity": 0.2,
    "mirrorAmount": 0.3,
    "scanlineIntensity": 0.2
  },
  "Glow": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 1,
    "threshold": 0.5,
    "size": 25
  },
  "Godrays": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0,
      "y": 0
    },
    "density": 0.3,
    "intensity": 0.8,
    "spotty": 1,
    "speed": 0.5,
    "rayColor": "#4283fb",
    "backgroundColor": "transparent"
  },
  "Grayscale": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Grid": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "cells": 10,
    "thickness": 1,
    "rotation": 0
  },
  "GridDistortion": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 1,
    "decay": 3,
    "radius": 1,
    "gridSize": 20,
    "edges": "stretch"
  },
  "Group": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Halftone": {
    "opacity": 1,
    "blendMode": "normal",
    "style": "classic",
    "frequency": 100,
    "angle": 45,
    "cyanAngle": 15,
    "magentaAngle": 75,
    "yellowAngle": 0,
    "blackAngle": 45,
    "misprint": 0,
    "misprintAngle": 0,
    "paperColor": "#ffffff",
    "cyanColor": "#00ffff",
    "magentaColor": "#ff00ff",
    "yellowColor": "#ffff00",
    "blackColor": "#000000"
  },
  "HexGrid": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#000000",
    "colorB": "#ffffff",
    "cells": 8,
    "thickness": 1,
    "colorSpace": "linear"
  },
  "HueShift": {
    "opacity": 1,
    "blendMode": "normal",
    "shift": 0
  },
  "ImageTexture": {
    "opacity": 1,
    "blendMode": "normal",
    "url": "https://shaders.com/sample.jpg",
    "objectFit": "cover"
  },
  "Invert": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Kaleidoscope": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "segments": 6,
    "angle": 0,
    "edges": "mirror"
  },
  "LensFlare": {
    "opacity": 1,
    "blendMode": "normal",
    "lightPosition": {
      "x": 0.3,
      "y": 0.3
    },
    "intensity": 0.5,
    "ghostIntensity": 0.4,
    "ghostSpread": 0.7,
    "ghostChroma": 0.3,
    "haloIntensity": 0.4,
    "haloRadius": 0.6,
    "haloChroma": 0.6,
    "haloSoftness": 0.8,
    "starburstIntensity": 0.3,
    "starburstPoints": 6,
    "streakIntensity": 0.15,
    "streakLength": 0.5,
    "glareIntensity": 0.2,
    "glareSize": 0.5,
    "edgeFade": 0.2,
    "speed": 0.5
  },
  "LinearBlur": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 30,
    "angle": 0
  },
  "LinearGradient": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#1aff00",
    "colorB": "#0000ff",
    "start": {
      "x": 0,
      "y": 0.5
    },
    "end": {
      "x": 1,
      "y": 0.5
    },
    "angle": 0,
    "edges": "stretch",
    "colorSpace": "linear"
  },
  "Liquify": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 10,
    "stiffness": 3,
    "damping": 3,
    "radius": 1,
    "edges": "stretch"
  },
  "Marble": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#ffffff",
    "colorB": "#3a2d54",
    "colorC": "#0f0f0f",
    "scale": 2,
    "turbulence": 10,
    "speed": 0.05,
    "seed": 0,
    "colorSpace": "linear"
  },
  "Mirror": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "angle": 0,
    "edges": "mirror"
  },
  "MultiPointGradient": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#4776E6",
    "positionA": {
      "x": 0.2,
      "y": 0.2
    },
    "colorB": "#C44DFF",
    "positionB": {
      "x": 0.8,
      "y": 0.2
    },
    "colorC": "#1ABC9C",
    "positionC": {
      "x": 0.2,
      "y": 0.8
    },
    "colorD": "#F8BBD9",
    "positionD": {
      "x": 0.8,
      "y": 0.8
    },
    "colorE": "#FF8C42",
    "positionE": {
      "x": 0.5,
      "y": 0.5
    },
    "smoothness": 2
  },
  "Neon": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "scale": 1,
    "color": "#00ddff",
    "secondaryColor": "#ff00aa",
    "secondaryBlend": 0.5,
    "glowColor": "#00ddff",
    "tubeThickness": 0.2,
    "intensity": 1.5,
    "hotCoreIntensity": 0.6,
    "glowIntensity": 0.6,
    "glowRadius": 0.25,
    "lightAngle": 300,
    "specularIntensity": 0.5,
    "specularSize": 0.5,
    "cornerSmoothing": 0.15,
    "flickerSpeed": 0,
    "flickerAmount": 0.2,
    "flowSpeed": 0,
    "flowAmount": 0.3,
    "shape": '{"type":"circleSDF","radius":0.35}'
  },
  "Paper": {
    "opacity": 1,
    "blendMode": "normal",
    "roughness": 0.3,
    "grainScale": 1,
    "displacement": 0.15,
    "seed": 0
  },
  "Perspective": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "pan": 0,
    "tilt": 0,
    "fov": 60,
    "zoom": 1,
    "offset": {
      "x": 0.5,
      "y": 0.5
    },
    "edges": "transparent"
  },
  "Pixelate": {
    "opacity": 1,
    "blendMode": "normal",
    "scale": 50,
    "gap": 0,
    "roundness": 0
  },
  "Plasma": {
    "opacity": 1,
    "blendMode": "normal",
    "density": 2,
    "speed": 2,
    "intensity": 1.5,
    "warp": 0.4,
    "contrast": 1,
    "balance": 50,
    "colorA": "#7018be",
    "colorB": "#000000",
    "colorSpace": "linear"
  },
  "PolarCoordinates": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "wrap": 1,
    "radius": 1,
    "intensity": 1,
    "edges": "transparent"
  },
  "Polygon": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "radius": 0.4,
    "sides": 6,
    "rounding": 0,
    "rotation": 0,
    "softness": 0,
    "strokeThickness": 0,
    "strokeColor": "#000000",
    "strokePosition": "center",
    "colorSpace": "linear"
  },
  "Posterize": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 5
  },
  "ProgressiveBlur": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 50,
    "angle": 0,
    "center": {
      "x": 0,
      "y": 0.5
    },
    "falloff": 1
  },
  "RadialGradient": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#ff0000",
    "colorB": "#0000ff",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "radius": 1,
    "repeat": 1,
    "aspect": 1,
    "skewAngle": 0,
    "colorSpace": "linear"
  },
  "RectangularCoordinates": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "scale": 1,
    "intensity": 1,
    "edges": "transparent"
  },
  "ReflectivePlane": {
    "opacity": 1,
    "blendMode": "normal",
    "height": 0.7,
    "distance": 0.5,
    "falloff": 0.5,
    "blur": 3,
    "blurDistance": 0.3,
    "edges": "stretch"
  },
  "Ring": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "radius": 0.3,
    "thickness": 0.07,
    "softness": 0,
    "strokeThickness": 0,
    "strokeColor": "#000000",
    "strokePosition": "center",
    "colorSpace": "linear"
  },
  "Ripples": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "colorA": "#ffffff",
    "colorB": "#000000",
    "speed": 1,
    "frequency": 20,
    "softness": 0,
    "thickness": 0.5,
    "phase": 0
  },
  "RoundedRect": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "width": 0.35,
    "height": 0.25,
    "rounding": 0.05,
    "rotation": 0,
    "softness": 0,
    "strokeThickness": 0,
    "strokeColor": "#000000",
    "strokePosition": "center",
    "colorSpace": "linear"
  },
  "Saturation": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 1
  },
  "Sharpness": {
    "opacity": 1,
    "blendMode": "normal",
    "sharpness": 0
  },
  "Shatter": {
    "opacity": 1,
    "blendMode": "normal",
    "crackWidth": 1,
    "intensity": 4,
    "radius": 0.4,
    "decay": 1,
    "seed": 2,
    "chromaticSplit": 1,
    "refractionStrength": 5,
    "shardLighting": 0.1,
    "edges": "mirror"
  },
  "SimplexNoise": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#ffffff",
    "colorB": "#000000",
    "colorSpace": "linear",
    "scale": 2,
    "balance": 0,
    "contrast": 0,
    "seed": 0,
    "speed": 1
  },
  "SineWave": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "amplitude": 0.15,
    "frequency": 1,
    "speed": 1,
    "angle": 0,
    "position": {
      "x": 0.5,
      "y": 0.5
    },
    "thickness": 0.2,
    "softness": 0.4
  },
  "Smoke": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#fc83f9",
    "colorB": "#c21c79",
    "emitFrom": {
      "x": 0.5,
      "y": 1
    },
    "direction": 0,
    "speed": 20,
    "spread": 60,
    "emitRadius": 0.08,
    "intensity": 1,
    "dissipation": 0.2,
    "detail": 25,
    "gravity": 0.5,
    "colorDecay": 0.4,
    "mouseInfluence": 0.1,
    "mouseRadius": 0.1,
    "colorSpace": "linear"
  },
  "SmokeFill": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#8cf3ff",
    "colorB": "#04a0d6",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "scale": 1,
    "emitFrom": {
      "x": 0.5,
      "y": 0.5
    },
    "direction": 0,
    "speed": 10,
    "spread": 60,
    "emitRadius": 0.03,
    "intensity": 1,
    "dissipation": 0.3,
    "detail": 25,
    "gravity": 0.5,
    "colorDecay": 0.4,
    "mouseInfluence": 0.1,
    "mouseRadius": 0.1,
    "colorSpace": "linear",
    "shape": '{"type":"circleSDF","radius":0.35}'
  },
  "Solarize": {
    "opacity": 1,
    "blendMode": "normal",
    "threshold": 0.5,
    "strength": 1
  },
  "SolidColor": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#5b18ca"
  },
  "Spherize": {
    "opacity": 1,
    "blendMode": "normal",
    "radius": 1,
    "depth": 1,
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "lightPosition": {
      "x": 0.3,
      "y": 0.3
    },
    "lightIntensity": 0.5,
    "lightSoftness": 0.5,
    "lightColor": "#ffffff"
  },
  "Spiral": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#000000",
    "colorB": "#ffffff",
    "strokeWidth": 0.5,
    "strokeFalloff": 0,
    "softness": 0,
    "speed": 1,
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "scale": 1,
    "colorSpace": "linear"
  },
  "Star": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "radius": 0.4,
    "sides": 5,
    "innerRatio": 0.4,
    "rotation": 0,
    "softness": 0,
    "strokeThickness": 0,
    "strokeColor": "#000000",
    "strokePosition": "center",
    "colorSpace": "linear"
  },
  "Strands": {
    "opacity": 1,
    "blendMode": "normal",
    "speed": 0.5,
    "amplitude": 1,
    "frequency": 1,
    "lineCount": 12,
    "lineWidth": 0.1,
    "waveColor": "#f1c907",
    "pinEdges": true,
    "start": {
      "x": 0,
      "y": 0.5
    },
    "end": {
      "x": 1,
      "y": 0.5
    }
  },
  "Stretch": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "strength": 1,
    "angle": 0,
    "falloff": 0,
    "edges": "stretch"
  },
  "Stripes": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#000000",
    "colorB": "#ffffff",
    "angle": 45,
    "density": 5,
    "balance": 0.5,
    "softness": 0,
    "speed": 0.2,
    "offset": 0,
    "colorSpace": "linear"
  },
  "StudioBackground": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#d8dbec",
    "keyColor": "#d5e4ea",
    "keyIntensity": 40,
    "keySoftness": 50,
    "fillColor": "#d5e4ea",
    "fillIntensity": 10,
    "fillSoftness": 70,
    "fillAngle": 70,
    "backColor": "#c8d4e8",
    "backIntensity": 20,
    "backSoftness": 80,
    "brightness": 20,
    "vignette": 0,
    "center": {
      "x": 0.5,
      "y": 0.8
    },
    "lightTarget": 100,
    "wallCurvature": 10,
    "ambientIntensity": 50,
    "ambientSpeed": 2,
    "seed": 0
  },
  "SunBurst": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffdd88",
    "background": "#000000",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "rayCount": 12,
    "softness": 0.3,
    "radius": 0.8,
    "feather": 0.5,
    "speed": 0.2
  },
  "Swirl": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#1275d8",
    "colorB": "#e19136",
    "speed": 1,
    "detail": 1,
    "blend": 50,
    "colorSpace": "linear"
  },
  "TiltShift": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 50,
    "width": 0.3,
    "falloff": 0.3,
    "angle": 0,
    "center": {
      "x": 0.5,
      "y": 0.5
    }
  },
  "Tint": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ff8800",
    "amount": 0.5,
    "preserveLuminosity": true
  },
  "Trapezoid": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "bottomWidth": 0.35,
    "topWidth": 0.2,
    "height": 0.25,
    "rotation": 0,
    "softness": 0,
    "strokeThickness": 0,
    "strokeColor": "#000000",
    "strokePosition": "center",
    "colorSpace": "linear"
  },
  "Tritone": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#ce1bea",
    "colorB": "#2fff00",
    "colorC": "#ffff00",
    "blendMid": 0.5,
    "colorSpace": "linear"
  },
  "Truchet": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#000000",
    "colorB": "#ffffff",
    "cells": 10,
    "thickness": 2,
    "seed": 0,
    "colorSpace": "linear"
  },
  "Twirl": {
    "opacity": 1,
    "blendMode": "normal",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "intensity": 1,
    "edges": "stretch"
  },
  "Vesica": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#ffffff",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "radius": 0.35,
    "spread": 0.5,
    "rotation": 0,
    "softness": 0,
    "strokeThickness": 0,
    "strokeColor": "#000000",
    "strokePosition": "center",
    "colorSpace": "linear"
  },
  "VHS": {
    "opacity": 1,
    "blendMode": "normal",
    "wobble": 1,
    "scanlineNoise": 0.6,
    "smear": 0.2,
    "speed": 1
  },
  "Vibrance": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 0
  },
  "VideoTexture": {
    "opacity": 1,
    "blendMode": "normal",
    "url": "https://shaders.com/sample.mp4",
    "objectFit": "cover",
    "loop": true
  },
  "Vignette": {
    "opacity": 1,
    "blendMode": "normal",
    "color": "#000000",
    "center": {
      "x": 0.5,
      "y": 0.5
    },
    "radius": 0.5,
    "falloff": 0.5,
    "intensity": 1
  },
  "Voronoi": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#3186cf",
    "colorB": "#fc02dd",
    "colorBorder": "#000000",
    "scale": 6,
    "speed": 0.5,
    "seed": 0,
    "edgeIntensity": 0.5,
    "edgeSoftness": 0.05,
    "colorSpace": "oklch"
  },
  "WaveDistortion": {
    "opacity": 1,
    "blendMode": "normal",
    "strength": 0.3,
    "frequency": 1,
    "speed": 1,
    "angle": 0,
    "waveType": "sine",
    "edges": "stretch"
  },
  "Weave": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#c4c4c4",
    "colorB": "#4d4d4d",
    "cells": 10,
    "gap": 0.25,
    "rotation": 0
  },
  "WebcamTexture": {
    "opacity": 1,
    "blendMode": "normal",
    "objectFit": "cover",
    "mirror": true
  },
  "WorleyNoise": {
    "opacity": 1,
    "blendMode": "normal",
    "colorA": "#ffffff",
    "colorB": "#000000",
    "colorSpace": "linear",
    "scale": 6,
    "mode": "f1",
    "distance": "euclidean",
    "octaves": 1,
    "lacunarity": 2,
    "persistence": 0.5,
    "jitter": 1,
    "contrast": 1,
    "balance": 0,
    "seed": 0,
    "speed": 0.5
  },
  "ZoomBlur": {
    "opacity": 1,
    "blendMode": "normal",
    "intensity": 30,
    "center": {
      "x": 0.5,
      "y": 0.5
    }
  }
};
function isPropMapValue(value) {
  return typeof value === "object" && value !== null && "type" in value && value.type === "map";
}
function generatePropString(props, componentType, indent = "  ") {
  return Object.entries(props).sort(([a], [b]) => a.localeCompare(b)).filter(([key, value]) => {
    if (key === "maskType" && value === "alpha") return false;
    if ((key === "shape" || key === "shapeType") && props.shapeSdfUrl) return false;
    if (key === "transform" && typeof value === "object") {
      const allDefaults = Object.keys(DEFAULT_TRANSFORM).every(
        (k) => value[k] === DEFAULT_TRANSFORM[k]
      );
      if (allDefaults) return false;
    }
    const componentDefaults = shaderMetadata[componentType] || {};
    if (componentDefaults.hasOwnProperty(key)) {
      const defaultValue = componentDefaults[key];
      if (value !== null && defaultValue !== null && typeof value === "object" && typeof defaultValue === "object") {
        return JSON.stringify(value) !== JSON.stringify(defaultValue);
      }
      return value !== defaultValue;
    }
    if (key === "opacity" && value === 1) return false;
    if (key === "blendMode" && value === "normal") return false;
    return true;
  }).map(([key, value]) => {
    if (key === "transform" && typeof value === "object") {
      const nonDefaultKeys = {};
      for (const k in value) {
        if (value[k] !== DEFAULT_TRANSFORM[k]) {
          nonDefaultKeys[k] = value[k];
        }
      }
      const keys = Object.keys(nonDefaultKeys).sort();
      if (keys.length > 1) {
        const entries = keys.map((k) => `${k}: ${typeof nonDefaultKeys[k] === "string" ? `"${nonDefaultKeys[k]}"` : nonDefaultKeys[k]}`);
        return `${key}={{
${indent}    ${entries.join(`,
${indent}    `)}
${indent}  }}`;
      } else if (keys.length === 1) {
        const k = keys[0];
        const v = nonDefaultKeys[k];
        return `${key}={{ ${k}: ${typeof v === "string" ? `"${v}"` : v} }}`;
      }
    }
    if (isPropMapValue(value)) {
      const entries = Object.keys(value).map((k) => {
        const v = value[k];
        return `${k}: ${typeof v === "string" ? `"${v}"` : v}`;
      });
      return `${key}={{
${indent}    ${entries.join(`,
${indent}    `)}
${indent}  }}`;
    }
    if (typeof value === "string") {
      if (value.startsWith("{") && value.endsWith("}")) {
        try {
          const parsed = JSON.parse(value);
          if (typeof parsed === "object" && parsed !== null) {
            const keys = Object.keys(parsed);
            if (keys.length > 1) {
              const entries = keys.map((k) => `${k}: ${typeof parsed[k] === "string" ? `"${parsed[k]}"` : parsed[k]}`);
              return `${key}={{
${indent}    ${entries.join(`,
${indent}    `)}
${indent}  }}`;
            }
            return `${key}={${toObjectLiteral(parsed)}}`;
          }
        } catch {
        }
      }
      return `${key}=${JSON.stringify(value)}`;
    } else if (value !== null && typeof value === "object") {
      const roundedValue = { ...value };
      if ("x" in roundedValue && typeof roundedValue.x === "number") {
        roundedValue.x = Math.round(roundedValue.x * 100) / 100;
      }
      if ("y" in roundedValue && typeof roundedValue.y === "number") {
        roundedValue.y = Math.round(roundedValue.y * 100) / 100;
      }
      const keys = Object.keys(roundedValue);
      if (keys.length > 1) {
        const entries = keys.map((k) => `${k}: ${typeof roundedValue[k] === "string" ? `"${roundedValue[k]}"` : roundedValue[k]}`);
        return `${key}={{
${indent}    ${entries.join(`,
${indent}    `)}
${indent}  }}`;
      }
      return `${key}={${toObjectLiteral(roundedValue)}}`;
    } else {
      return `${key}={${value}}`;
    }
  }).join("\n" + indent + "  ");
}
function isIdReferenced(id, allComponents) {
  const flatComponents = [];
  function flattenComponents(components) {
    for (const component of components) {
      flatComponents.push(component);
      if (component.children) {
        flattenComponents(component.children);
      }
    }
  }
  flattenComponents(allComponents);
  return flatComponents.some((component) => {
    var _a;
    if (((_a = component.props) == null ? void 0 : _a.maskSource) === id) return true;
    if (component.props) {
      for (const value of Object.values(component.props)) {
        if (isPropMapValue(value) && value.source === id) return true;
      }
    }
    return false;
  });
}
function generatePresetCode(preset, colorSpace) {
  const componentTypes = /* @__PURE__ */ new Set();
  function collectTypes(components) {
    for (const config of components) {
      componentTypes.add(config.type);
      if (config.children) collectTypes(config.children);
    }
  }
  collectTypes(preset.components);
  const generateComponentString = (config, indent = "      ") => {
    const propString = config.props ? generatePropString(config.props, config.type, indent) : "";
    let idString = "";
    if (config.id) {
      const isUsedAsMask = isIdReferenced(config.id, preset.components);
      if (isUsedAsMask) {
        idString = `id="${config.id}"`;
      }
    }
    const attributes = [idString, propString].filter(Boolean).join("\n" + indent + "  ");
    const sampleChildren = config.type === "DOMTexture" && (!config.children || config.children.length === 0) ? `${indent}  <h1>Your HTML content here</h1>` : null;
    if (sampleChildren) {
      if (attributes) {
        return `${indent}<${config.type}
${indent}  ${attributes}>
${sampleChildren}
${indent}</${config.type}>`;
      } else {
        return `${indent}<${config.type}>
${sampleChildren}
${indent}</${config.type}>`;
      }
    } else if (config.children && config.children.length > 0) {
      const childrenString = config.children.map((child) => generateComponentString(child, indent + "  ")).join("\n");
      if (attributes) {
        return `${indent}<${config.type}
${indent}  ${attributes}>
${childrenString}
${indent}</${config.type}>`;
      } else {
        return `${indent}<${config.type}>
${childrenString}
${indent}</${config.type}>`;
      }
    } else {
      if (attributes) {
        return `${indent}<${config.type}
${indent}  ${attributes} />`;
      } else {
        return `${indent}<${config.type} />`;
      }
    }
  };
  const componentStrings = preset.components.map((config) => generateComponentString(config)).join("\n");
  const shaderProps = colorSpace && colorSpace !== "p3-linear" ? ` colorSpace="${colorSpace}"` : "";
  const sortedTypes = Array.from(componentTypes).sort();
  const allImports = ["Shader", ...sortedTypes];
  const importStatement = allImports.length > 3 ? `import {
  ${allImports.join(",\n  ")},
} from 'shaders/solid'` : `import { ${allImports.join(", ")} } from 'shaders/solid'`;
  return `${importStatement}

export default function ShaderEffect() {
  return (
    <Shader${shaderProps}>
${componentStrings}
    </Shader>
  )
}`;
}
const availableComponents = [
  // @ts-ignore - replaced at build time
  "AngularBlur",
  "Ascii",
  "Aurora",
  "BarShift",
  "Beam",
  "Blob",
  "Blur",
  "BrickPattern",
  "BrightnessContrast",
  "Bulge",
  "CRTScreen",
  "ChannelBlur",
  "Checkerboard",
  "Chevron",
  "ChromaFlow",
  "ChromaticAberration",
  "Circle",
  "ColorWheel",
  "ConcentricSpin",
  "ConicGradient",
  "ContourLines",
  "Crescent",
  "Cross",
  "Crystal",
  "CursorRipples",
  "CursorTrail",
  "DOMTexture",
  "DiamondGradient",
  "DiffuseBlur",
  "Dither",
  "DotGrid",
  "DropShadow",
  "Duotone",
  "Ellipse",
  "Emboss",
  "FallingLines",
  "FilmGrain",
  "FloatingParticles",
  "FlowField",
  "Flower",
  "FlowingGradient",
  "FlutedGlass",
  "Fog",
  "Form3D",
  "FractalNoise",
  "Glass",
  "GlassTiles",
  "Glitch",
  "Glow",
  "Godrays",
  "Grayscale",
  "Grid",
  "GridDistortion",
  "Group",
  "Halftone",
  "HexGrid",
  "HueShift",
  "ImageTexture",
  "Invert",
  "Kaleidoscope",
  "LensFlare",
  "LinearBlur",
  "LinearGradient",
  "Liquify",
  "Marble",
  "Mirror",
  "MultiPointGradient",
  "Neon",
  "Paper",
  "Perspective",
  "Pixelate",
  "Plasma",
  "PolarCoordinates",
  "Polygon",
  "Posterize",
  "ProgressiveBlur",
  "RadialGradient",
  "RectangularCoordinates",
  "ReflectivePlane",
  "Ring",
  "Ripples",
  "RoundedRect",
  "Saturation",
  "Sharpness",
  "Shatter",
  "SimplexNoise",
  "SineWave",
  "Smoke",
  "SmokeFill",
  "Solarize",
  "SolidColor",
  "Spherize",
  "Spiral",
  "Star",
  "Strands",
  "Stretch",
  "Stripes",
  "StudioBackground",
  "SunBurst",
  "Swirl",
  "TiltShift",
  "Tint",
  "Trapezoid",
  "Tritone",
  "Truchet",
  "Twirl",
  "VHS",
  "Vesica",
  "Vibrance",
  "VideoTexture",
  "Vignette",
  "Voronoi",
  "WaveDistortion",
  "Weave",
  "WebcamTexture",
  "WorleyNoise",
  "ZoomBlur"
];
export {
  availableComponents,
  generatePresetCode
};
