// shaders/dist/solid/components/Solarize.js
import { createComponent, memo, template, setAttribute, insert, Portal, spread, mergeProps, use } from "solid-js/web";
import { splitProps, createUniqueId, createMemo, createSignal, onMount, onCleanup, createEffect } from "solid-js";
import { createUniformsMap } from "../../core/index.js";
import { setColorSpaceMode } from "../../core/utilities/transformations/index.js";
import { componentDefinition } from "../../core/shaders/Solarize/index.js";
import { useShaderContext, ShaderContext } from "../engine/Shader.js";
var _tmpl$ = /* @__PURE__ */ template(`<span style=display:contents>`), _tmpl$2 = /* @__PURE__ */ template(`<canvas>`);
let _experimentalWarnedOnce = false;
if (componentDefinition.experimental && !_experimentalWarnedOnce) {
  _experimentalWarnedOnce = true;
  const _e = componentDefinition.experimental;
  console.info(`%c⚠ [Shaders] ${componentDefinition.name} is experimental: ${_e.message}`, "color: #f59e0b; font-weight: bold");
}
function isPropDriver(value) {
  return typeof value === "object" && value !== null && "type" in value && (value.type === "map" || value.type === "mouse" || value.type === "mouse-position" || value.type === "auto-animate");
}
const DEFAULT_TRANSFORM = {
  offsetX: 0,
  offsetY: 0,
  rotation: 0,
  scale: 1,
  anchorX: 0.5,
  anchorY: 0.5,
  edges: "transparent"
};
const defaultProps = {
  blendMode: "normal",
  visible: true
  // opacity intentionally has no default - handled by renderer
  // transform intentionally has no default - handled by effectiveTransform
};
try {
  if (componentDefinition && componentDefinition.props) {
    Object.entries(componentDefinition.props).forEach(([key, config]) => {
      const propConfig = config;
      if (propConfig && typeof propConfig === "object" && "default" in propConfig) {
        defaultProps[key] = propConfig.default;
      }
    });
  }
} catch (e) {
  console.warn("Error extracting default props:", e);
}
function ShaderComponent(props) {
  const [local, otherProps] = splitProps(props, ["children"]);
  const context = useShaderContext();
  const {
    shaderParentId: parentId,
    shaderNodeRegister: parentRegister,
    shaderUniformUpdate: parentUniformUpdate,
    shaderMetadataUpdate: parentMetadataUpdate,
    shaderColorSpace
  } = context;
  const instanceId = (props.id ? props.id.replace(/[^a-zA-Z0-9_]/g, "_") : null) || createUniqueId();
  const effectiveProps = createMemo(() => {
    let baseProps = {
      ...defaultProps
    };
    for (const [key, value] of Object.entries(otherProps)) {
      if (value !== void 0 && !isPropDriver(value)) {
        baseProps[key] = value;
      }
    }
    return baseProps;
  });
  const effectiveTransform = createMemo(() => ({
    ...DEFAULT_TRANSFORM,
    ...props.transform
  }));
  const mapsFromProps = createMemo(() => {
    const maps = {};
    for (const [key, value] of Object.entries(otherProps)) {
      if (isPropDriver(value)) maps[key] = value;
    }
    return Object.keys(maps).length > 0 ? maps : void 0;
  });
  let uniformsMap = null;
  const getUniformsMap = () => {
    if (!uniformsMap) {
      setColorSpaceMode(shaderColorSpace);
      uniformsMap = createUniformsMap(componentDefinition, effectiveProps(), instanceId);
    }
    return uniformsMap;
  };
  const childContextValue = createMemo(() => ({
    ...context,
    shaderParentId: instanceId
  }));
  const isCapturesDOM = !!componentDefinition.capturesDOM;
  let captureCanvasEl;
  const [captureSize, setCaptureSize] = createSignal({
    w: typeof window !== "undefined" ? Math.round(window.innerWidth * Math.min(window.devicePixelRatio, 2)) : 0,
    h: typeof window !== "undefined" ? Math.round(window.innerHeight * Math.min(window.devicePixelRatio, 2)) : 0
  });
  if (isCapturesDOM) {
    const onWinResize = () => {
      const d = Math.min(window.devicePixelRatio, 2);
      setCaptureSize({
        w: Math.round(window.innerWidth * d),
        h: Math.round(window.innerHeight * d)
      });
    };
    onMount(() => window.addEventListener("resize", onWinResize));
    onCleanup(() => window.removeEventListener("resize", onWinResize));
  }
  let markerRef;
  let detectedRenderOrder = void 0;
  let isRegistered = false;
  onMount(() => {
    const uniforms = getUniformsMap();
    if (!uniforms) return;
    try {
      const metadata = {
        blendMode: props.blendMode || "normal",
        opacity: props.opacity,
        visible: props.visible === false ? false : true,
        id: props.id,
        mask: props.maskSource ? {
          source: props.maskSource,
          type: props.maskType || "alpha"
        } : void 0,
        maps: mapsFromProps(),
        renderOrder: props.renderOrder ?? detectedRenderOrder,
        transform: effectiveTransform()
      };
      parentRegister(instanceId, componentDefinition.fragmentNode, parentId, metadata, uniforms, componentDefinition, isCapturesDOM ? captureCanvasEl : void 0);
      isRegistered = true;
      if (props.renderOrder === void 0 && markerRef) {
        const parent = markerRef.parentElement;
        if (parent) {
          const siblings = parent.querySelectorAll(":scope > [data-shader-id]");
          const position = Array.from(siblings).indexOf(markerRef);
          if (position >= 0) {
            detectedRenderOrder = position;
            parentMetadataUpdate(instanceId, {
              renderOrder: position
            });
          }
        }
      }
    } catch (error) {
      console.error("Error registering shader node:", error);
    }
  });
  onCleanup(() => {
    isRegistered = false;
    parentRegister(instanceId, null, null, null, null);
  });
  createEffect(() => {
    if (!isRegistered) return;
    const uniforms = getUniformsMap();
    if (!uniforms) return;
    try {
      const props_snapshot = effectiveProps();
      Object.entries(uniforms).forEach(([propName, uniformData]) => {
        if (!uniformData || typeof uniformData !== "object") return;
        const {
          uniform
        } = uniformData;
        if ((uniform == null ? void 0 : uniform.value) !== void 0 && propName in props_snapshot) {
          const newValue = props_snapshot[propName];
          parentUniformUpdate(instanceId, propName, newValue);
        }
      });
    } catch (error) {
      console.error("Error updating uniforms:", error);
    }
  });
  createEffect(() => {
    if (!isRegistered) return;
    try {
      const metadata = {
        blendMode: props.blendMode || "normal",
        opacity: props.opacity,
        visible: props.visible === false ? false : true,
        id: props.id,
        mask: props.maskSource ? {
          source: props.maskSource,
          type: props.maskType || "alpha"
        } : void 0,
        maps: mapsFromProps(),
        renderOrder: props.renderOrder ?? detectedRenderOrder,
        transform: effectiveTransform()
      };
      parentMetadataUpdate(instanceId, metadata);
    } catch (error) {
      console.error("Error updating metadata:", error);
    }
  });
  const canUseDOM = typeof document !== "undefined" && typeof window !== "undefined";
  return createComponent(ShaderContext.Provider, {
    get value() {
      return childContextValue();
    },
    get children() {
      return [(() => {
        var _el$ = _tmpl$();
        var _ref$ = markerRef;
        typeof _ref$ === "function" ? use(_ref$, _el$) : markerRef = _el$;
        setAttribute(_el$, "data-shader-id", instanceId);
        insert(_el$, () => (!isCapturesDOM || !canUseDOM) && local.children);
        return _el$;
      })(), memo(() => isCapturesDOM && canUseDOM && createComponent(Portal, {
        get children() {
          var _el$2 = _tmpl$2();
          var _ref$2 = captureCanvasEl;
          typeof _ref$2 === "function" ? use(_ref$2, _el$2) : captureCanvasEl = _el$2;
          spread(_el$2, mergeProps({
            "layoutsubtree": ""
          }, {
            get width() {
              return captureSize().w;
            },
            get height() {
              return captureSize().h;
            },
            "style": "position:fixed;inset:0;width:100vw;height:100vh;z-index:-9999"
          }), false, true);
          insert(_el$2, () => local.children);
          return _el$2;
        }
      }))];
    }
  });
}
export {
  ShaderComponent as default
};
