// shaders/dist/react/GridDistortion.js
import { ShaderContext } from "./Shader.js";
import { useContext, useEffect, useId, useMemo, useRef, useState } from "react";
import { createPortal } from "react-dom";
import { createUniformsMap } from "../core/index.js";
import { setColorSpaceMode } from "../core/utilities/transformations/index.js";
import { jsx, jsxs } from "react/jsx-runtime";
import { componentDefinition } from "../core/shaders/GridDistortion/index.js";
function isPropDriver(value) {
	return typeof value === "object" && value !== null && "type" in value && (value.type === "map" || value.type === "mouse" || value.type === "mouse-position" || value.type === "auto-animate");
}
function computeEffectiveProps(props, defaultProps$1) {
	let baseProps = { ...defaultProps$1 };
	for (const [key, value] of Object.entries(props)) if (key !== "children" && key !== "ref" && value !== void 0 && !isPropDriver(value)) baseProps[key] = value;
	return baseProps;
}
var MARKER_STYLE = { display: "contents" };
var DEFAULT_TRANSFORM = {
	offsetX: 0,
	offsetY: 0,
	rotation: 0,
	scale: 1,
	anchorX: .5,
	anchorY: .5,
	edges: "transparent"
};
var defaultProps = {
	blendMode: "normal",
	visible: true
};
try {
	if (componentDefinition && componentDefinition.props) Object.entries(componentDefinition.props).forEach(([key, config]) => {
		const propConfig = config;
		if (propConfig && typeof propConfig === "object" && "default" in propConfig) defaultProps[key] = propConfig.default;
	});
} catch (e) {
	console.warn("Error extracting default props:", e);
}
const ShaderComponent = (props) => {
	const context = useContext(ShaderContext);
	if (!context) throw new Error("Shader components must be used inside an <Shader> component or another shader component");
	const { shaderParentId: parentId, shaderNodeRegister: parentRegister, shaderUniformUpdate: parentUniformUpdate, shaderMetadataUpdate: parentMetadataUpdate, shaderColorSpace } = context;
	const reactId = useId();
	const instanceId = useMemo(() => {
		return (props.id ? props.id.replace(/[^a-zA-Z0-9_]/g, "_") : null) || reactId.replace(/[^a-zA-Z0-9_]/g, "");
	}, [props.id, reactId]);
	const effectiveProps = useMemo(() => {
		return computeEffectiveProps(props, defaultProps);
	}, [props]);
	const effectiveTransform = useMemo(() => ({
		...DEFAULT_TRANSFORM,
		...props.transform
	}), [props.transform]);
	const markerRef = useRef(null);
	const isCapturesDOM = !!componentDefinition.capturesDOM;
	const captureCanvasRef = useRef(null);
	const [domMounted, setDomMounted] = useState(false);
	const [captureSize, setCaptureSize] = useState(() => ({
		w: typeof window !== "undefined" ? Math.round(window.innerWidth * Math.min(window.devicePixelRatio, 2)) : 0,
		h: typeof window !== "undefined" ? Math.round(window.innerHeight * Math.min(window.devicePixelRatio, 2)) : 0
	}));
	useEffect(() => {
		if (!isCapturesDOM) return;
		setDomMounted(true);
		const onResize = () => {
			const d = Math.min(window.devicePixelRatio, 2);
			setCaptureSize({
				w: Math.round(window.innerWidth * d),
				h: Math.round(window.innerHeight * d)
			});
		};
		window.addEventListener("resize", onResize);
		return () => window.removeEventListener("resize", onResize);
	}, [isCapturesDOM]);
	const detectedOrderRef = useRef(void 0);
	const uniformsRef = useRef(null);
	if (uniformsRef.current === null) {
		setColorSpaceMode(shaderColorSpace);
		uniformsRef.current = createUniformsMap(componentDefinition, effectiveProps, instanceId);
	}
	const childContextValue = useMemo(() => {
		return {
			...context,
			shaderParentId: instanceId
		};
	}, [context, instanceId]);
	useEffect(() => {
		if (!uniformsRef.current) return;
		try {
			const metadata = {
				blendMode: props.blendMode || "normal",
				opacity: props.opacity,
				visible: props.visible === false ? false : true,
				id: props.id,
				mask: props.maskSource ? {
					source: props.maskSource,
					type: props.maskType || "alpha"
				} : void 0,
				maps: mapsFromProps,
				renderOrder: props.renderOrder ?? detectedOrderRef.current,
				transform: effectiveTransform
			};
			parentRegister(instanceId, componentDefinition.fragmentNode, parentId, metadata, uniformsRef.current, componentDefinition, isCapturesDOM ? captureCanvasRef.current ?? void 0 : void 0);
			return () => {
				parentRegister(instanceId, null, null, null, null);
			};
		} catch (error) {
			console.error("Error registering shader node:", error);
			return () => {};
		}
	}, [
		instanceId,
		parentId,
		parentRegister,
		domMounted
	]);
	useEffect(() => {
		if (props.renderOrder === void 0 && markerRef.current) {
			const parent = markerRef.current.parentElement;
			if (parent) {
				const siblings = parent.querySelectorAll(":scope > [data-shader-id]");
				const position = Array.from(siblings).indexOf(markerRef.current);
				if (position >= 0) {
					detectedOrderRef.current = position;
					parentMetadataUpdate(instanceId, { renderOrder: position });
				}
			}
		}
	}, []);
	const prevPropsRef = useRef({});
	const mapsFromProps = useMemo(() => {
		const maps = {};
		for (const [key, value] of Object.entries(props)) if (key !== "children" && key !== "ref" && isPropDriver(value)) maps[key] = value;
		return Object.keys(maps).length > 0 ? maps : void 0;
	}, [props]);
	useEffect(() => {
		if (!uniformsRef.current) return;
		try {
			Object.entries(uniformsRef.current).forEach(([propName, uniformData]) => {
				if (!uniformData || typeof uniformData !== "object") return;
				const { uniform } = uniformData;
				if (uniform?.value !== void 0 && propName in effectiveProps) {
					const newValue = effectiveProps[propName];
					if (isPropDriver(newValue)) return;
					if (newValue !== prevPropsRef.current[propName]) {
						parentUniformUpdate(instanceId, propName, newValue);
						prevPropsRef.current[propName] = newValue;
					}
				}
			});
		} catch (error) {
			console.error("Error updating uniforms:", error);
		}
	}, [
		effectiveProps,
		instanceId,
		parentUniformUpdate
	]);
	useEffect(() => {
		try {
			parentMetadataUpdate(instanceId, {
				blendMode: props.blendMode || "normal",
				opacity: props.opacity,
				visible: props.visible === false ? false : true,
				id: props.id,
				mask: props.maskSource ? {
					source: props.maskSource,
					type: props.maskType || "alpha"
				} : void 0,
				maps: mapsFromProps,
				renderOrder: props.renderOrder ?? detectedOrderRef.current,
				transform: effectiveTransform
			});
		} catch (error) {
			console.error("Error updating metadata:", error);
		}
	}, [
		props.blendMode,
		props.opacity,
		props.visible,
		props.maskSource,
		props.maskType,
		props.renderOrder,
		props.id,
		mapsFromProps,
		effectiveTransform,
		instanceId,
		parentMetadataUpdate
	]);
	if (props.ref && typeof props.ref === "function") try {
		props.ref(null);
	} catch (e) {}
	const captureCanvas = isCapturesDOM && domMounted ? createPortal(/* @__PURE__ */ jsx("canvas", {
		ref: captureCanvasRef,
		"layoutsubtree": "",
		width: captureSize.w,
		height: captureSize.h,
		style: {
			position: "fixed",
			inset: 0,
			width: "100vw",
			height: "100vh",
			zIndex: -9999
		},
		children: props.children
	}), document.body) : null;
	return /* @__PURE__ */ jsxs(ShaderContext.Provider, {
		value: childContextValue,
		children: [/* @__PURE__ */ jsx("span", {
			ref: markerRef,
			style: MARKER_STYLE,
			"data-shader-id": instanceId,
			children: !isCapturesDOM && props.children
		}), captureCanvas]
	});
};
var GridDistortion_default = ShaderComponent;
export { ShaderComponent, GridDistortion_default as default };
