// shaders/dist/react/Preview.js
import { Shader } from "./Shader.js";
import { Suspense, lazy, useEffect, useMemo, useRef, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
function xorCrypt(data, key) {
	const keyBytes = new TextEncoder().encode(key);
	const result = new Uint8Array(data.length);
	for (let i = 0; i < data.length; i++) result[i] = data[i] ^ keyBytes[i % keyBytes.length];
	return result;
}
function decodePreviewDefinition(encoded, key) {
	const binaryString = atob(encoded);
	const encrypted = new Uint8Array(binaryString.length);
	for (let i = 0; i < binaryString.length; i++) encrypted[i] = binaryString.charCodeAt(i);
	const decrypted = xorCrypt(encrypted, key);
	const json = new TextDecoder().decode(decrypted);
	try {
		return JSON.parse(json);
	} catch {
		throw new Error("Failed to decode shader definition: invalid data or incorrect key");
	}
}
var RESERVED_IDENTIFIERS = new Set([
	"break",
	"case",
	"catch",
	"class",
	"const",
	"continue",
	"default",
	"delete",
	"do",
	"else",
	"export",
	"extends",
	"finally",
	"for",
	"function",
	"if",
	"import",
	"in",
	"instanceof",
	"new",
	"return",
	"super",
	"switch",
	"this",
	"throw",
	"try",
	"typeof",
	"var",
	"void",
	"while",
	"with",
	"yield"
]);
function slugifyIdentifier(label, fallback) {
	const cleaned = label.replace(/[^A-Za-z0-9]+/g, " ").trim();
	if (!cleaned) return fallback;
	const camel = cleaned.split(/\s+/).map((part, i) => i === 0 ? part.charAt(0).toLowerCase() + part.slice(1) : part.charAt(0).toUpperCase() + part.slice(1)).join("");
	const safe = /^[A-Za-z_$]/.test(camel) ? camel : `_${camel}`;
	return RESERVED_IDENTIFIERS.has(safe) ? `_${safe}` : safe;
}
function buildKeyPropIdentifierMap(keyProps) {
	const used = new Set(["style"]);
	const map = /* @__PURE__ */ new Map();
	if (!keyProps?.length) return map;
	keyProps.forEach((kp, i) => {
		if (!kp.targets?.length) return;
		const base = slugifyIdentifier(kp.label, `keyProp${i}`);
		let ident = base;
		let suffix = 1;
		while (used.has(ident)) {
			suffix++;
			ident = `${base}${suffix}`;
		}
		used.add(ident);
		map.set(ident, kp);
	});
	return map;
}
function findInTree(id, list) {
	for (const c of list) {
		if (c.id === id) return c;
		if (c.children?.length) {
			const hit = findInTree(id, c.children);
			if (hit) return hit;
		}
	}
	return null;
}
function applyConfiguration(definition, configuration, keyProps) {
	if (!configuration || !keyProps?.length) return definition;
	const entries = Object.entries(configuration);
	if (entries.length === 0) return definition;
	const identMap = buildKeyPropIdentifierMap(keyProps);
	const cloned = JSON.parse(JSON.stringify(definition.components));
	for (const [ident, value] of entries) {
		const kp = identMap.get(ident);
		if (!kp) continue;
		for (const t of kp.targets ?? []) {
			const comp = findInTree(t.component_id, cloned);
			if (comp?.props && t.prop in comp.props) comp.props[t.prop] = value;
		}
	}
	return {
		...definition,
		components: cloned
	};
}
var componentMap = {
	AngularBlur: lazy(() => import("./AngularBlur.js")),
	Ascii: lazy(() => import("./Ascii.js")),
	Aurora: lazy(() => import("./Aurora.js")),
	BarShift: lazy(() => import("./BarShift.js")),
	Beam: lazy(() => import("./Beam.js")),
	Blob: lazy(() => import("./Blob.js")),
	Blur: lazy(() => import("./Blur.js")),
	BrickPattern: lazy(() => import("./BrickPattern.js")),
	BrightnessContrast: lazy(() => import("./BrightnessContrast.js")),
	Bulge: lazy(() => import("./Bulge.js")),
	CRTScreen: lazy(() => import("./CRTScreen.js")),
	ChannelBlur: lazy(() => import("./ChannelBlur.js")),
	Checkerboard: lazy(() => import("./Checkerboard.js")),
	Chevron: lazy(() => import("./Chevron.js")),
	ChromaFlow: lazy(() => import("./ChromaFlow.js")),
	ChromaticAberration: lazy(() => import("./ChromaticAberration.js")),
	Circle: lazy(() => import("./Circle.js")),
	ColorWheel: lazy(() => import("./ColorWheel.js")),
	ConcentricSpin: lazy(() => import("./ConcentricSpin.js")),
	ConicGradient: lazy(() => import("./ConicGradient.js")),
	ContourLines: lazy(() => import("./ContourLines.js")),
	Crescent: lazy(() => import("./Crescent.js")),
	Cross: lazy(() => import("./Cross.js")),
	Crystal: lazy(() => import("./Crystal.js")),
	CursorRipples: lazy(() => import("./CursorRipples.js")),
	CursorTrail: lazy(() => import("./CursorTrail.js")),
	DOMTexture: lazy(() => import("./DOMTexture.js")),
	DiamondGradient: lazy(() => import("./DiamondGradient.js")),
	DiffuseBlur: lazy(() => import("./DiffuseBlur.js")),
	Dither: lazy(() => import("./Dither.js")),
	DotGrid: lazy(() => import("./DotGrid.js")),
	DropShadow: lazy(() => import("./DropShadow.js")),
	Duotone: lazy(() => import("./Duotone.js")),
	Ellipse: lazy(() => import("./Ellipse.js")),
	Emboss: lazy(() => import("./Emboss.js")),
	FallingLines: lazy(() => import("./FallingLines.js")),
	FilmGrain: lazy(() => import("./FilmGrain.js")),
	FloatingParticles: lazy(() => import("./FloatingParticles.js")),
	FlowField: lazy(() => import("./FlowField.js")),
	Flower: lazy(() => import("./Flower.js")),
	FlowingGradient: lazy(() => import("./FlowingGradient.js")),
	FlutedGlass: lazy(() => import("./FlutedGlass.js")),
	Fog: lazy(() => import("./Fog.js")),
	Form3D: lazy(() => import("./Form3D.js")),
	FractalNoise: lazy(() => import("./FractalNoise.js")),
	Glass: lazy(() => import("./Glass.js")),
	GlassTiles: lazy(() => import("./GlassTiles.js")),
	Glitch: lazy(() => import("./Glitch.js")),
	Glow: lazy(() => import("./Glow.js")),
	Godrays: lazy(() => import("./Godrays.js")),
	Grayscale: lazy(() => import("./Grayscale.js")),
	Grid: lazy(() => import("./Grid.js")),
	GridDistortion: lazy(() => import("./GridDistortion.js")),
	Group: lazy(() => import("./Group.js")),
	Halftone: lazy(() => import("./Halftone.js")),
	HexGrid: lazy(() => import("./HexGrid.js")),
	HueShift: lazy(() => import("./HueShift.js")),
	ImageTexture: lazy(() => import("./ImageTexture.js")),
	Invert: lazy(() => import("./Invert.js")),
	Kaleidoscope: lazy(() => import("./Kaleidoscope.js")),
	LensFlare: lazy(() => import("./LensFlare.js")),
	LinearBlur: lazy(() => import("./LinearBlur.js")),
	LinearGradient: lazy(() => import("./LinearGradient.js")),
	Liquify: lazy(() => import("./Liquify.js")),
	Marble: lazy(() => import("./Marble.js")),
	Mirror: lazy(() => import("./Mirror.js")),
	MultiPointGradient: lazy(() => import("./MultiPointGradient.js")),
	Neon: lazy(() => import("./Neon.js")),
	Paper: lazy(() => import("./Paper.js")),
	Perspective: lazy(() => import("./Perspective.js")),
	Pixelate: lazy(() => import("./Pixelate.js")),
	Plasma: lazy(() => import("./Plasma.js")),
	PolarCoordinates: lazy(() => import("./PolarCoordinates.js")),
	Polygon: lazy(() => import("./Polygon.js")),
	Posterize: lazy(() => import("./Posterize.js")),
	ProgressiveBlur: lazy(() => import("./ProgressiveBlur.js")),
	RadialGradient: lazy(() => import("./RadialGradient.js")),
	RectangularCoordinates: lazy(() => import("./RectangularCoordinates.js")),
	ReflectivePlane: lazy(() => import("./ReflectivePlane.js")),
	Ring: lazy(() => import("./Ring.js")),
	Ripples: lazy(() => import("./Ripples.js")),
	RoundedRect: lazy(() => import("./RoundedRect.js")),
	Saturation: lazy(() => import("./Saturation.js")),
	Sharpness: lazy(() => import("./Sharpness.js")),
	Shatter: lazy(() => import("./Shatter.js")),
	SimplexNoise: lazy(() => import("./SimplexNoise.js")),
	SineWave: lazy(() => import("./SineWave.js")),
	Smoke: lazy(() => import("./Smoke.js")),
	SmokeFill: lazy(() => import("./SmokeFill.js")),
	Solarize: lazy(() => import("./Solarize.js")),
	SolidColor: lazy(() => import("./SolidColor.js")),
	Spherize: lazy(() => import("./Spherize.js")),
	Spiral: lazy(() => import("./Spiral.js")),
	Star: lazy(() => import("./Star.js")),
	Strands: lazy(() => import("./Strands.js")),
	Stretch: lazy(() => import("./Stretch.js")),
	Stripes: lazy(() => import("./Stripes.js")),
	StudioBackground: lazy(() => import("./StudioBackground.js")),
	SunBurst: lazy(() => import("./SunBurst.js")),
	Swirl: lazy(() => import("./Swirl.js")),
	TiltShift: lazy(() => import("./TiltShift.js")),
	Tint: lazy(() => import("./Tint.js")),
	Trapezoid: lazy(() => import("./Trapezoid.js")),
	Tritone: lazy(() => import("./Tritone.js")),
	Truchet: lazy(() => import("./Truchet.js")),
	Twirl: lazy(() => import("./Twirl.js")),
	VHS: lazy(() => import("./VHS.js")),
	Vesica: lazy(() => import("./Vesica.js")),
	Vibrance: lazy(() => import("./Vibrance.js")),
	VideoTexture: lazy(() => import("./VideoTexture.js")),
	Vignette: lazy(() => import("./Vignette.js")),
	Voronoi: lazy(() => import("./Voronoi.js")),
	WaveDistortion: lazy(() => import("./WaveDistortion.js")),
	Weave: lazy(() => import("./Weave.js")),
	WebcamTexture: lazy(() => import("./WebcamTexture.js")),
	WorleyNoise: lazy(() => import("./WorleyNoise.js")),
	ZoomBlur: lazy(() => import("./ZoomBlur.js"))
};
var DEFAULT_KEY = "shaders-preview-key";
function parseStyleString(style) {
	return style.split(";").reduce((acc, declaration) => {
		const [prop, ...rest] = declaration.split(":");
		if (!prop?.trim()) return acc;
		const value = rest.join(":").trim();
		const trimmed = prop.trim();
		const finalProp = trimmed.startsWith("--") ? trimmed : trimmed.replace(/-([a-z])/g, (_, c) => c.toUpperCase());
		acc[finalProp] = value;
		return acc;
	}, {});
}
function RenderComponent({ component }) {
	const Component = componentMap[component.type];
	if (!Component) {
		console.warn(`[Shaders Preview] Unknown component type: ${component.type}`);
		return null;
	}
	const { type, children, id, props: componentProps,...topLevelProps } = component;
	const mergedProps = {
		...topLevelProps,
		...componentProps
	};
	if (id) mergedProps.id = id;
	return /* @__PURE__ */ jsx(Component, {
		...mergedProps,
		children: children?.map((child, index) => /* @__PURE__ */ jsx(RenderComponent, { component: child }, child.id || index))
	});
}
const Preview = ({ shader, presetId, apiBaseUrl = "https://shaders.com", obfuscationKey = DEFAULT_KEY, watermarkText = "Unlock your Shaders Pro license", watermarkLink = "https://shaders.com/dashboard?pricing=true", style, className, configuration, version,...rest }) => {
	const [fetchedDefinition, setFetchedDefinition] = useState(null);
	const [fetchedKeyProps, setFetchedKeyProps] = useState(null);
	const fetchAndDecode = (url) => {
		setFetchedDefinition(null);
		setFetchedKeyProps(null);
		fetch(url).then((res) => {
			if (!res.ok) throw new Error(`Failed to fetch: ${res.status}`);
			return res.json();
		}).then((data) => {
			const item = data.preset || data.shader;
			if (item?.definition && typeof item.definition === "string") {
				setFetchedDefinition(decodePreviewDefinition(item.definition, obfuscationKey));
				setFetchedKeyProps(Array.isArray(item.key_props) ? item.key_props : null);
			}
		}).catch((err) => {
			console.error("[Shaders Preview] Failed to fetch preview:", err);
		});
	};
	useEffect(() => {
		if (shader) fetchAndDecode(`${apiBaseUrl}/api/preview/shader/${shader}`);
		else if (!presetId) {
			setFetchedDefinition(null);
			setFetchedKeyProps(null);
		}
	}, [
		shader,
		presetId,
		apiBaseUrl,
		obfuscationKey
	]);
	useEffect(() => {
		if (presetId) fetchAndDecode(`${apiBaseUrl}/api/preview/preset/${presetId}${version ? `?version=${encodeURIComponent(version)}` : ""}`);
		else if (!shader) {
			setFetchedDefinition(null);
			setFetchedKeyProps(null);
		}
	}, [
		presetId,
		shader,
		apiBaseUrl,
		obfuscationKey,
		version
	]);
	const definition = useMemo(() => {
		if (!fetchedDefinition) return null;
		return applyConfiguration(fetchedDefinition, configuration, fetchedKeyProps);
	}, [
		fetchedDefinition,
		fetchedKeyProps,
		configuration
	]);
	const hasWarnedRef = useRef(false);
	useEffect(() => {
		if (!hasWarnedRef.current && definition) {
			hasWarnedRef.current = true;
			console.warn("[Shaders] The Preview component is intended for use with a valid Shaders license. Please visit https://shaders.com for more information.");
		}
	}, [definition]);
	if (!definition) return null;
	return /* @__PURE__ */ jsxs("div", {
		style: {
			position: "relative",
			...typeof style === "string" ? parseStyleString(style) : style
		},
		className,
		...rest,
		children: [/* @__PURE__ */ jsx(Suspense, {
			fallback: null,
			children: /* @__PURE__ */ jsx(Shader, {
				colorSpace: definition.colorSpace || "p3-linear",
				isPreview: true,
				style: {
					width: "100%",
					height: "100%"
				},
				children: definition.components.map((component, index) => /* @__PURE__ */ jsx(RenderComponent, { component }, component.id || index))
			})
		}), watermarkText && /* @__PURE__ */ jsx("div", {
			style: {
				position: "absolute",
				bottom: "8px",
				right: "12px",
				fontSize: "11px",
				fontFamily: "-apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, sans-serif",
				color: "rgba(255, 255, 255, 0.5)",
				textShadow: "0 1px 3px rgba(0, 0, 0, 0.4)",
				pointerEvents: watermarkLink ? "auto" : "none",
				zIndex: 10,
				transition: "color 0.2s ease"
			},
			onMouseEnter: (e) => {
				if (watermarkLink) e.currentTarget.style.color = "rgba(255, 255, 255, 0.8)";
			},
			onMouseLeave: (e) => {
				if (watermarkLink) e.currentTarget.style.color = "rgba(255, 255, 255, 0.5)";
			},
			children: watermarkLink ? /* @__PURE__ */ jsx("a", {
				href: watermarkLink,
				target: "_blank",
				rel: "noopener noreferrer",
				style: {
					color: "inherit",
					textDecoration: "none"
				},
				children: watermarkText
			}) : watermarkText
		})]
	});
};
export { Preview };
