// shaders/dist/react/Shader.js
import { createContext, useEffect, useId, useMemo, useRef } from "react";
import { shaderRenderer } from "../core/index.js";
import { setColorSpaceMode } from "../core/utilities/transformations/index.js";
import { vec4 } from "three/tsl";
import { isExternalUser, startTelemetry } from "../core/telemetry/index.js";
import { jsx, jsxs } from "react/jsx-runtime";
const ShaderContext = createContext(null);
const Shader = ({ children, disableTelemetry = false, colorSpace = "p3-linear", isPreview = false, onReady, style = {}, className = "",...rest }) => {
	const containerRef = useRef(null);
	const canvasRef = useRef(null);
	const reactId = useId();
	const rootId = useMemo(() => "shader-root-" + reactId.replace(/[^a-zA-Z0-9_]/g, ""), [reactId]);
	const rendererRef = useRef(null);
	if (rendererRef.current === null) rendererRef.current = shaderRenderer();
	const onReadyRef = useRef(onReady);
	onReadyRef.current = onReady;
	useEffect(() => {
		rendererRef.current.setOnReady(() => onReadyRef.current?.());
		return () => {
			rendererRef.current.setOnReady(null);
		};
	}, []);
	const telemetryCollectorRef = useRef(null);
	const telemetryStartTimeoutRef = useRef(null);
	const shouldSendTelemetryRef = useRef(null);
	const nodeRegister = useMemo(() => {
		return (id, fragmentNodeFunc, parentId, metadata, uniforms = null, componentDefinition = null, domCanvas) => {
			if (fragmentNodeFunc === null) {
				try {
					rendererRef.current.removeNode(id);
				} catch (err) {
					console.warn("Error removing node:", err);
				}
				return;
			}
			try {
				rendererRef.current.registerNode(id, fragmentNodeFunc, parentId, metadata, uniforms, componentDefinition, domCanvas);
			} catch (err) {
				console.error("Error registering node:", err, {
					id,
					parentId,
					metadata
				});
			}
		};
	}, []);
	const uniformUpdate = useMemo(() => {
		return (nodeId, uniformName, value) => {
			try {
				rendererRef.current.updateUniformValue(nodeId, uniformName, value);
			} catch (err) {
				console.warn("Error updating uniform:", err);
			}
		};
	}, []);
	const metadataUpdate = useMemo(() => {
		return (nodeId, metadata) => {
			try {
				rendererRef.current.updateNodeMetadata(nodeId, metadata);
			} catch (err) {
				console.warn("Error updating metadata:", err);
			}
		};
	}, []);
	const contextValue = useMemo(() => {
		return {
			shaderParentId: rootId,
			shaderNodeRegister: nodeRegister,
			shaderUniformUpdate: uniformUpdate,
			shaderMetadataUpdate: metadataUpdate,
			shaderColorSpace: colorSpace
		};
	}, [
		rootId,
		nodeRegister,
		uniformUpdate,
		metadataUpdate,
		colorSpace
	]);
	const isInitializedRef = useRef(false);
	const isInitializingRef = useRef(false);
	const isCleanedUpRef = useRef(false);
	const wasVisibleRef = useRef(false);
	const visibilityObserverRef = useRef(null);
	const startTelemetryWhenReady = () => {
		const checkRendering = () => {
			if (isCleanedUpRef.current || !rendererRef.current) {
				telemetryStartTimeoutRef.current = null;
				return;
			}
			if (rendererRef.current.getPerformanceStats().fps > 0) {
				telemetryCollectorRef.current = startTelemetry(rendererRef.current, "2.5.124", disableTelemetry, isPreview);
				if (telemetryCollectorRef.current) telemetryCollectorRef.current.start();
				telemetryStartTimeoutRef.current = null;
			} else telemetryStartTimeoutRef.current = window.setTimeout(checkRendering, 500);
		};
		telemetryStartTimeoutRef.current = window.setTimeout(checkRendering, 500);
	};
	const initializeRenderer = async () => {
		const canvas = canvasRef.current;
		if (!canvas || isInitializedRef.current || isInitializingRef.current) return;
		isInitializingRef.current = true;
		try {
			const renderer = rendererRef.current;
			if (!renderer.isInitialized()) await renderer.initialize({
				canvas,
				colorSpace
			});
			renderer.registerNode(rootId, ({ childNode }) => childNode || vec4(0, 0, 0, 0), null, null, {});
			isInitializedRef.current = true;
			isInitializingRef.current = false;
			if (shouldSendTelemetryRef.current === null) shouldSendTelemetryRef.current = isExternalUser();
			if (shouldSendTelemetryRef.current && !telemetryCollectorRef.current) startTelemetryWhenReady();
		} catch (err) {
			console.error("[Shader] Initialization failed:", err);
			isInitializingRef.current = false;
			throw err;
		}
	};
	const setupVisibilityObserver = () => {
		const container = containerRef.current;
		if (!container || visibilityObserverRef.current) return;
		visibilityObserverRef.current = new IntersectionObserver((entries) => {
			const entry = entries[0];
			if (!entry) return;
			const rect = container.getBoundingClientRect();
			const isCurrentlyVisible = entry.isIntersecting && rect && rect.width > 0 && rect.height > 0;
			if (isCurrentlyVisible && !wasVisibleRef.current) {
				if (rendererRef.current.isInitialized()) {
					rendererRef.current.startAnimation();
					if (shouldSendTelemetryRef.current && !telemetryCollectorRef.current && !telemetryStartTimeoutRef.current) startTelemetryWhenReady();
				} else initializeRenderer().catch(console.error);
				wasVisibleRef.current = true;
			} else if (!isCurrentlyVisible && wasVisibleRef.current) {
				rendererRef.current.stopAnimation();
				wasVisibleRef.current = false;
			}
		}, { threshold: 0 });
		visibilityObserverRef.current.observe(container);
	};
	useEffect(() => {
		isCleanedUpRef.current = false;
		const container = containerRef.current;
		if (!container) return;
		const rect = container.getBoundingClientRect();
		if (rect.width > 0 && rect.height > 0) {
			initializeRenderer().catch(console.error);
			wasVisibleRef.current = true;
		} else wasVisibleRef.current = false;
		setupVisibilityObserver();
		return () => {
			if (isCleanedUpRef.current) return;
			isCleanedUpRef.current = true;
			if (telemetryCollectorRef.current) {
				telemetryCollectorRef.current.stop();
				telemetryCollectorRef.current = null;
			}
			if (telemetryStartTimeoutRef.current !== null) {
				clearTimeout(telemetryStartTimeoutRef.current);
				telemetryStartTimeoutRef.current = null;
			}
			if (visibilityObserverRef.current) {
				visibilityObserverRef.current.disconnect();
				visibilityObserverRef.current = null;
			}
			if (rendererRef.current && isInitializedRef.current) try {
				rendererRef.current.cleanup();
			} catch (err) {
				console.warn("[Shader] Error during cleanup:", err);
			}
			isInitializedRef.current = false;
			isInitializingRef.current = false;
		};
	}, []);
	const isFirstRenderRef = useRef(true);
	useEffect(() => {
		if (isFirstRenderRef.current) {
			isFirstRenderRef.current = false;
			return;
		}
		if (rendererRef.current && isInitializedRef.current) {
			setColorSpaceMode(colorSpace);
			try {
				rendererRef.current.cleanup();
				isInitializedRef.current = false;
				initializeRenderer().catch(console.error);
			} catch (err) {
				console.error("[Shader] Error updating colorSpace:", err);
			}
		}
	}, [colorSpace]);
	return /* @__PURE__ */ jsx(ShaderContext.Provider, {
		value: contextValue,
		children: /* @__PURE__ */ jsxs("div", {
			ref: containerRef,
			className: "shader" + (className ? " " + className : ""),
			style,
			...rest,
			children: [/* @__PURE__ */ jsx("canvas", {
				"data-renderer": "shaders",
				ref: canvasRef,
				style: {
					width: "100%",
					height: "100%",
					display: "block"
				}
			}), children]
		})
	});
};
var Shader_default = Shader;
export { Shader, ShaderContext, Shader_default as default };
